{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ammend DHS index annotations by doing a down-lift from hg38 to hg19\n",
    "# First collect positions that could not be downlifted\n",
    "import collections\n",
    "blackdict = collections.defaultdict(lambda: False)\n",
    "errfile = \"/cbscratch/franco/datasets/DHSindex/hglft_genome_32b00_ea89f0.err\"\n",
    "with open(errfile) as instream:\n",
    "    for line in instream:\n",
    "        if line[0] == \"#\":\n",
    "            continue\n",
    "        blackdict[line.rstrip()] = True\n",
    "\n",
    "# Go line by line through DHSindex and check if the range was downlifted or not\n",
    "# and keep its annotation\n",
    "liftfile = \"/cbscratch/franco/datasets/DHSindex/hglft_genome_32b00_ea89f0.bed\"\n",
    "dhsindexfile = \"/cbscratch/franco/datasets/DHSindex/DHS_Index_and_Vocabulary_hg38_WM20190703.txt\"\n",
    "hg19_dhs_dict = collections.defaultdict(dict)\n",
    "hglft = open(liftfile)\n",
    "with open(dhsindexfile) as instream:\n",
    "    next(instream)\n",
    "    for line in instream:\n",
    "        arr = line.rstrip().split(\"\\t\")\n",
    "        chrm = arr[0]\n",
    "        start = arr[1]\n",
    "        end   = arr[2]\n",
    "        annot = arr[9]\n",
    "        key = \"{:s}:{:s}-{:s}\".format(chrm, start, end)\n",
    "        if not blackdict[key]:\n",
    "            hg19chrom, hg19range = hglft.readline().rstrip().split(\":\")\n",
    "            hg19_dhs_dict[hg19chrom][hg19range] = annot\n",
    "\n",
    "# Write down the newly annotated dictionary in a file, sorted by genomic position\n",
    "hg19_DHS_outfile = \"/cbscratch/franco/datasets/DHSindex/DHS_Index_downlift_hg19.txt\"\n",
    "with open(hg19_DHS_outfile, 'w') as outstream:\n",
    "    for chrm in range(1, 23):\n",
    "        chrom = \"chr\"+str(chrm)\n",
    "        for key in sorted(hg19_dhs_dict[chrom], key=lambda t: int(t.split(\"-\")[0])):\n",
    "            # print(chrom, key, hg19_dhs_dict[chrom][key])\n",
    "            outstream.write(\"{:s}\\t{:s}\\t{:s}\\t{:s}\\n\".format(chrom, key.split(\"-\")[0], key.split(\"-\")[1], hg19_dhs_dict[chrom][key]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import os\n",
    "# here we annotate the best trans-eQTLs for EQTLgen\n",
    "\n",
    "def annotate_snps(dhsfile, gtfile, annot_outfile, mode=\"tejaas\"):\n",
    "    current_chrm = None\n",
    "    dhs = open(dhsfile)\n",
    "    line = dhs.readline()\n",
    "    f = None\n",
    "    with open(annot_outfile, 'w') as outf:\n",
    "        outf.write(\"chr\\trsid\\tpos\\n\")\n",
    "        while line:\n",
    "            arr = line.rstrip().split()\n",
    "            if arr[0][3:] == \"X\":\n",
    "                break\n",
    "            chrm = int(arr[0][3:])\n",
    "            start = int(arr[1])\n",
    "            end = int(arr[2])\n",
    "            if chrm != current_chrm:\n",
    "                if f:\n",
    "                    f.close()\n",
    "                # close previous GT file and open new one\n",
    "                if current_chrm is not None:\n",
    "                    tend = time.time()\n",
    "                    print(\"CHR{:d} took {:g}s\".format(current_chrm, tend-tstart))\n",
    "\n",
    "                current_chrm = chrm\n",
    "                if not os.path.exists(gtfile.format(current_chrm)):\n",
    "                    continue\n",
    "                print(\"Processing CHRM\", current_chrm, end=\" \")\n",
    "                f = open(gtfile.format(current_chrm), 'r')\n",
    "                next(f)\n",
    "                tstart = time.time()\n",
    "                gtline = f.readline()\n",
    "            if not gtline:\n",
    "                line = dhs.readline()\n",
    "            while gtline:\n",
    "                gtarr = gtline.split()\n",
    "                if mode == \"tejaas\":\n",
    "                    rsid = gtarr[0]\n",
    "                    pos = int(gtarr[1])\n",
    "                if mode == \"meqtl\":\n",
    "                    rsid = gtarr[0]\n",
    "                    pos = int(gtarr[6])\n",
    "                # print(rsid, pos, start, end)\n",
    "                if pos < start:\n",
    "                    gtline = f.readline()\n",
    "                    continue # go to next snp\n",
    "                elif pos > end:\n",
    "                    line = dhs.readline()\n",
    "                    break # go to next DHS line\n",
    "                else:\n",
    "                    # print(\"-->\",chrm, rsid, pos, start, end)\n",
    "                    outf.write(\"{:d}\\t{:s}\\t{:d}\\n\".format(chrm, rsid, pos))\n",
    "                    gtline = f.readline()\n",
    "                    continue\n",
    "        if f:\n",
    "            f.close()\n",
    "            tend = time.time()\n",
    "            print(\"CHR{:d} took {:g}s\".format(current_chrm, tend-tstart))\n",
    "        print(\"Done DHS file\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mpmath\n",
    "import collections\n",
    "import numpy as np\n",
    "\n",
    "mpmath.mp.dps = 500\n",
    "def pval(x): return float(mpmath.log10(1 - 0.5 * (1 + mpmath.erf(x/mpmath.sqrt(2)))))\n",
    "\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'inDHS']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "       \n",
    "def eqtlgen(filepath, chrom, snpannot):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            pos   = int(arr[1])\n",
    "            p     = float(arr[2])\n",
    "            logp  = np.log10(p) if p != 0 else -99\n",
    "            inDHS = snpannot.get(rsid, False)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, inDHS=inDHS))\n",
    "    return res\n",
    "        \n",
    "def tejaas(filepath, chrom, snpannot):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            pos   = int(arr[1])\n",
    "            p     = float(arr[5])\n",
    "            q     = float(arr[2])\n",
    "            mu    = float(arr[3])\n",
    "            sigma = float(arr[4])\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pval( (q - mu) / sigma)\n",
    "            inDHS = snpannot.get(rsid, False)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, inDHS=inDHS))\n",
    "    return res\n",
    "\n",
    "# TODO: first I need to report snp positions in matrixeqtl output..\n",
    "\n",
    "def matrixeqtl(filepath, chrom, snpannot):\n",
    "    res = dict()\n",
    "    res_list = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr  = line.strip().split(\"\\t\")\n",
    "            rsid = arr[0]\n",
    "            pos  = int(arr[6])\n",
    "            pval = float(arr[4])\n",
    "            if not res.get(rsid, False):\n",
    "                res[rsid] = [np.log10(pval), pos]\n",
    "    for key in res.keys():\n",
    "        inDHS = snpannot.get(rsid, False)\n",
    "        res_list.append(SNPRes(rsid=key, chrom=chrom, pos=res[key][1], logp=-res[key][0], inDHS=inDHS))\n",
    "    return res_list\n",
    "\n",
    "def matrixeqtl_fdr(filepath, chrom, snpannot):\n",
    "    res = dict()\n",
    "    res_list = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr  = line.strip().split(\"\\t\")\n",
    "            rsid = arr[0]\n",
    "            pos  = int(arr[6])\n",
    "            pval = float(arr[5])\n",
    "            if not res.get(rsid, False):\n",
    "                res[rsid] = [np.log10(pval), pos]\n",
    "    for key in res.keys():\n",
    "        inDHS = snpannot.get(rsid, False)\n",
    "        res_list.append(SNPRes(rsid=key, chrom=chrom, pos=res[key][1], logp=-res[key][0], inDHS=inDHS))\n",
    "    return res_list\n",
    "\n",
    "def read_annot(infile):\n",
    "    annot_dict = dict()\n",
    "    with open(infile) as fin:\n",
    "        _ = fin.readline()\n",
    "        for line in fin:\n",
    "            arr = line.rstrip().split()\n",
    "            annot_dict[arr[1]] = True\n",
    "    return annot_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Null enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample genotype positions randomly and check enrichment\n",
    "# generate files from bash\n",
    "# for CHRM in `seq 1 22`; do mkdir $CHRM; for i in `seq 1 50`; do grep -P \"^$CHRM \" all_snp_pos| shuf -n 2000 | sort -nk 1 -k 3 | tr \" \" \"\\t\" |cut -f 2- > ${CHRM}/random_snp_sample_${i}; done; done\n",
    "\n",
    "basedir = \"/cbscratch/franco/datasets/DHS_null_enrichment/\"\n",
    "gtfile = \"/cbscratch/franco/datasets/DHS_null_enrichment/{:d}/random_snp_sample_\"\n",
    "dhsfile = \"/cbscratch/franco/datasets/multi-tissue.master.ntypes.simple.hg19.bed\"\n",
    "null_annot_outfile = basedir + \"null_SNPs_annots_\"\n",
    "for N in range(1,51):\n",
    "    if not os.path.exists(null_annot_outfile+str(N)+\".txt\"):\n",
    "        annotate_snps(dhsfile, gtfile+str(N), null_annot_outfile+str(N)+\".txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.1614268181818182 0.0018046845034589253\n"
     ]
    }
   ],
   "source": [
    "def file_len(fname):\n",
    "    with open(fname) as f:\n",
    "        for i, l in enumerate(f):\n",
    "            pass\n",
    "    return i + 1\n",
    "\n",
    "null_frac_annot = list()\n",
    "basedir = \"/cbscratch/franco/datasets/DHS_null_enrichment/\"\n",
    "null_annot_outfile = basedir + \"null_SNPs_annots_\"\n",
    "for N in range(1,51):\n",
    "    nullfile = null_annot_outfile+str(N)+\".txt\"\n",
    "    L = file_len(nullfile)\n",
    "    frac_annot = L / (22*2000)\n",
    "    null_frac_annot.append(frac_annot)\n",
    "    \n",
    "null_frac_annot = np.array(null_frac_annot)\n",
    "mean_null_frac_annot = np.mean(null_frac_annot)\n",
    "std_null_frac_annot = np.std(null_frac_annot)\n",
    "print(mean_null_frac_annot, std_null_frac_annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "dhs_mode = \"multi\"\n",
    "if dhs_mode == \"multi\":\n",
    "    dhsfile = \"/cbscratch/franco/datasets/multi-tissue.master.ntypes.simple.hg19.bed\"\n",
    "    dhs_annot_out = \"SNPs_annots.txt\"\n",
    "if dhs_mode == \"dhsindex\":\n",
    "    # dhsfile = \"/cbscratch/franco/datasets/DHSindex/DHS_Index_downlift_hg19.txt\"\n",
    "    dhsfile = \"/cbscratch/franco/datasets/DHSindex/DHS_Index_and_Vocabulary_hg38_WM20190703.txt\"\n",
    "    dhs_annot_out = \"dhsindex_SNPs_annots.txt\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EQTLgen enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing CHRM 1 CHR1 took 0.756557s\n",
      "Processing CHRM 10 CHR10 took 0.466239s\n",
      "Processing CHRM 11 CHR11 took 0.364025s\n",
      "Processing CHRM 12 CHR12 took 0.338149s\n",
      "Processing CHRM 13 CHR13 took 0.198827s\n",
      "Processing CHRM 14 CHR14 took 0.224862s\n",
      "Processing CHRM 15 CHR15 took 0.231487s\n",
      "Processing CHRM 16 CHR16 took 0.248599s\n",
      "Processing CHRM 17 CHR17 took 0.300454s\n",
      "Processing CHRM 18 CHR18 took 0.180515s\n",
      "Processing CHRM 19 CHR19 took 0.210141s\n",
      "Processing CHRM 2 CHR2 took 0.596454s\n",
      "Processing CHRM 20 CHR20 took 0.222431s\n",
      "Processing CHRM 21 CHR21 took 0.0965607s\n",
      "Processing CHRM 22 CHR22 took 0.150762s\n",
      "Processing CHRM 3 CHR3 took 0.46116s\n",
      "Processing CHRM 4 CHR4 took 0.37988s\n",
      "Processing CHRM 5 CHR5 took 0.409307s\n",
      "Processing CHRM 6 CHR6 took 0.42467s\n",
      "Processing CHRM 7 CHR7 took 0.381395s\n",
      "Processing CHRM 8 CHR8 took 0.342839s\n",
      "Processing CHRM 9 CHR9 took 0.304899s\n",
      "Done DHS file\n"
     ]
    }
   ],
   "source": [
    "\n",
    "basedir = \"/cbscratch/franco/datasets/EQTLgen/\"\n",
    "eqtlgen_gtfile = basedir + \"signif/trans-eQTLs_CHR{:d}\"\n",
    "               \n",
    "eqtlgen_annot_outfile = basedir + dhs_annot_out\n",
    "      \n",
    "annotate_snps(dhsfile, eqtlgen_gtfile, eqtlgen_annot_outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of trans-eQTLs in DHS regions (EQTLgen): 0.30753192598384155 1180 / 3837\n",
      "EQTLgen total enrichment: 1.90509\n"
     ]
    }
   ],
   "source": [
    "eqtlgen_annot_outfile = basedir + dhs_annot_out\n",
    "chrms = np.arange(1,23)\n",
    "\n",
    "eqtlgen_snp_annot = read_annot(eqtlgen_annot_outfile)\n",
    "eqtlgen_snp_res = list()\n",
    "for chrom in chrms:\n",
    "    eqtlgen_snp_res += eqtlgen(eqtlgen_gtfile.format(chrom), chrom, eqtlgen_snp_annot)\n",
    "    \n",
    "eqtlgen_dhs_dict = dict()\n",
    "for snp in eqtlgen_snp_res:\n",
    "    eqtlgen_dhs_dict[snp.rsid] = snp.inDHS\n",
    "    \n",
    "# Number of EQTLgen signif trans-eQTLS in DHS regions\n",
    "\n",
    "total_signif_SNPs = len(list(eqtlgen_dhs_dict.keys()))\n",
    "signif_inDHS      = np.sum(np.array(list(eqtlgen_dhs_dict.values())))\n",
    "frac = signif_inDHS / total_signif_SNPs\n",
    "print(\"Fraction of trans-eQTLs in DHS regions (EQTLgen):\",frac, signif_inDHS,\"/\", total_signif_SNPs)\n",
    "print(\"EQTLgen total enrichment: {:g}\".format(frac / mean_null_frac_annot))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "METHOD: gtex-ms raw 0.1_knn30 tejaas\n",
      "METHOD: gtex-ms raw 0.1_knn30 tejaas_rand\n",
      "METHOD: gtex-wb raw 0.1_knn30 tejaas\n",
      "METHOD: gtex-wb raw 0.1_knn30 tejaas_rand\n",
      "METHOD: gtex-sse raw 0.1_knn30 tejaas\n",
      "METHOD: gtex-sse raw 0.1_knn30 tejaas_rand\n",
      "METHOD: gtex-as raw 0.1_knn30 tejaas\n",
      "METHOD: gtex-as raw 0.1_knn30 tejaas_rand\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "import numpy as np\n",
    "import os\n",
    "# here we only try to annotate the best trans-eQTLs only!\n",
    "\n",
    "methods = [\"tejaas\", \"tejaas_rand\"]\n",
    "tissues = [\"gtex-ms\", \"gtex-wb\", \"gtex-sse\", \"gtex-as\"]\n",
    "sbs = [\"0.1_knn30\"]\n",
    "expressions = [\"raw\"]\n",
    "\n",
    "for tissue in tissues:\n",
    "    for expression in expressions:\n",
    "        for sb in sbs:\n",
    "            for method in methods:\n",
    "                print(\"METHOD:\",tissue, expression, sb, method)\n",
    "                basedir = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v6_gtknn/{:s}/{:s}/{:s}/permnull_sb{:s}/\".format(expression, tissue, method, sb)\n",
    "                gtfile = basedir + \"chr{:d}/rr.txt\" #.ld_prune\"\n",
    "                annot_outfile = basedir + dhs_annot_out\n",
    "                if not os.path.exists(annot_outfile):\n",
    "                    annotate_snps(dhsfile, gtfile, annot_outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gtex-ms tejaas raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-ms tejaas_rand raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-wb tejaas raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-wb tejaas_rand raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-sse tejaas raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-sse tejaas_rand raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-as tejaas raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-as tejaas_rand raw 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "datadir = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v6_gtknn/\"\n",
    "snp_res_dict = collections.defaultdict(dict)\n",
    "for sb in sbs:\n",
    "    for expr in expressions:\n",
    "        snp_res_dict[sb][expr] = collections.defaultdict(dict)\n",
    "        for tissue in tissues:\n",
    "            snp_res_dict[sb][expr][tissue] = collections.defaultdict(dict)\n",
    "            for method in methods:\n",
    "                snp_res = list()\n",
    "                print(tissue, method, expr, end=\" \")\n",
    "                annotfile = os.path.join(datadir,expr,tissue,method,\"permnull_sb{:s}\".format(sb),dhs_annot_out)\n",
    "                snp_annot = read_annot(annotfile)\n",
    "                for chrom in chrms:\n",
    "                    print(chrom, end=\" \")\n",
    "                    inputfile = os.path.join(datadir,expr,tissue,method,\"permnull_sb{:s}\".format(sb),\"chr\"+str(chrom),\"rr.txt\")\n",
    "                    snp_res += tejaas(inputfile, chrom, snp_annot)\n",
    "                print(\"\")\n",
    "                snp_res_dict[sb][expr][tissue][method] = snp_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'tmm_cclm'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-bf177d34dfa2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mmyexpr\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mexpressions\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m                 \u001b[0msnp_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msnp_res_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmysb\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmyexpr\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmytissue\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmymethod\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m                 \u001b[0mtop_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mattrgetter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'logp'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'tmm_cclm'"
     ]
    }
   ],
   "source": [
    "from operator import attrgetter\n",
    "\n",
    "def count_inDHS(top_list, cutoff):\n",
    "    teqtls = list()\n",
    "    for x in top_list:\n",
    "        if x.logp > -np.log10(cutoff):\n",
    "            teqtls.append(x.inDHS)\n",
    "        else:\n",
    "            break\n",
    "    return teqtls\n",
    "\n",
    "tejaas_enrich_dict = dict()\n",
    "\n",
    "for mysb in sbs:\n",
    "    for mytissue in tissues:\n",
    "        for mymethod in methods:\n",
    "            for myexpr in expressions:\n",
    "\n",
    "                snp_list = snp_res_dict[mysb][myexpr][mytissue][mymethod]\n",
    "                top_list = sorted(snp_list, key=attrgetter('logp'), reverse=True)\n",
    "\n",
    "                cutoffs = [1e-11, 1e-10, 1e-9, 1e-8, 5e-8] #, 1e-7, 5e-7, 1e-6, 1e-5, 1e-3, 5e-2]\n",
    "\n",
    "                # rr_teqtls = [x.inDHS for x in top_list if x.logp > -np.log10(cutoff)]\n",
    "                fracs_rr = list()\n",
    "                for cutoff in cutoffs:\n",
    "                    rr_teqtls = count_inDHS(top_list, cutoff)\n",
    "                    total_rr_teqtls = len(rr_teqtls)\n",
    "                    if total_rr_teqtls > 0:\n",
    "                        rr_inDHS = np.sum(np.array(rr_teqtls))\n",
    "                        frac_rr = rr_inDHS/total_rr_teqtls\n",
    "                        fracs_rr.append(frac_rr)\n",
    "                        print(\"Fraction of trans-eQTLS in DHS regions @ {:g} ({:s}): {:g}, {:d}/{:d}\".format(cutoff, mymethod, frac_rr, rr_inDHS, total_rr_teqtls))\n",
    "                    else:\n",
    "                        print(\"No signif trans-eQTLs found at {:g} cutoff\".format(cutoff))\n",
    "                        fracs_rr.append(None)\n",
    "                key = \"{:s}_{:s}_{:s}_{:s}\".format(mysb, myexpr, mytissue, mymethod)\n",
    "                tejaas_enrich_dict[key] = fracs_rr\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.1_knn30_raw_gtex-ms_tejaas total enrichment: [1.54868939 1.27539126]\n",
      "0.1_knn30_raw_gtex-ms_tejaas_rand total enrichment: []\n",
      "0.1_knn30_raw_gtex-wb_tejaas total enrichment: []\n",
      "0.1_knn30_raw_gtex-wb_tejaas_rand total enrichment: []\n",
      "0.1_knn30_raw_gtex-sse_tejaas total enrichment: [1.54868939 0.4955806  0.96988628 0.89793717 0.98232498]\n",
      "0.1_knn30_raw_gtex-sse_tejaas_rand total enrichment: [0.]\n",
      "0.1_knn30_raw_gtex-as_tejaas total enrichment: [0.88496536 1.03245959 0.96253205 1.04383029 1.00538371]\n",
      "0.1_knn30_raw_gtex-as_tejaas_rand total enrichment: []\n"
     ]
    }
   ],
   "source": [
    "# Enrichments\n",
    "for mysb in sbs:\n",
    "    for mytissue in tissues:\n",
    "        for mymethod in methods:\n",
    "            for myexpr in expressions:\n",
    "                key = \"{:s}_{:s}_{:s}_{:s}\".format(mysb, myexpr, mytissue, mymethod)\n",
    "                print(\"{:s} total enrichment: {:s}\".format(key, str( np.array(tejaas_enrich_dict[key]) / mean_null_frac_annot ) ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "METHOD: gtex-ms tmm_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 4.12564s\n",
      "Processing CHRM 10 CHR10 took 2.59979s\n",
      "Processing CHRM 11 CHR11 took 2.59018s\n",
      "Processing CHRM 12 CHR12 took 0.541229s\n",
      "Processing CHRM 13 CHR13 took 1.99152s\n",
      "Processing CHRM 14 CHR14 took 1.66439s\n",
      "Processing CHRM 15 CHR15 took 1.67161s\n",
      "Processing CHRM 16 CHR16 took 1.55531s\n",
      "Processing CHRM 17 CHR17 took 1.6461s\n",
      "Processing CHRM 18 CHR18 took 1.4529s\n",
      "Processing CHRM 19 CHR19 took 1.34868s\n",
      "Processing CHRM 2 CHR2 took 4.49375s\n",
      "Processing CHRM 20 CHR20 took 0.219798s\n",
      "Processing CHRM 21 CHR21 took 0.756137s\n",
      "Processing CHRM 22 CHR22 took 0.899945s\n",
      "Processing CHRM 3 CHR3 took 2.80239s\n",
      "Processing CHRM 4 CHR4 took 3.84319s\n",
      "Processing CHRM 5 CHR5 took 3.39421s\n",
      "Processing CHRM 6 CHR6 took 0.520547s\n",
      "Processing CHRM 7 CHR7 took 3.40756s\n",
      "Processing CHRM 8 CHR8 took 2.85502s\n",
      "Processing CHRM 9 CHR9 took 2.24425s\n",
      "Done DHS file\n",
      "METHOD: gtex-ms tmm_lasso matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.02856s\n",
      "Processing CHRM 10 CHR10 took 3.13906s\n",
      "Processing CHRM 11 CHR11 took 3.17748s\n",
      "Processing CHRM 12 CHR12 took 0.553946s\n",
      "Processing CHRM 13 CHR13 took 2.33865s\n",
      "Processing CHRM 14 CHR14 took 1.98082s\n",
      "Processing CHRM 15 CHR15 took 2.02582s\n",
      "Processing CHRM 16 CHR16 took 1.91746s\n",
      "Processing CHRM 17 CHR17 took 1.94602s\n",
      "Processing CHRM 18 CHR18 took 1.7676s\n",
      "Processing CHRM 19 CHR19 took 1.64797s\n",
      "Processing CHRM 2 CHR2 took 5.35373s\n",
      "Processing CHRM 20 CHR20 took 0.233923s\n",
      "Processing CHRM 21 CHR21 took 0.921095s\n",
      "Processing CHRM 22 CHR22 took 1.08865s\n",
      "Processing CHRM 3 CHR3 took 2.4s\n",
      "Processing CHRM 4 CHR4 took 4.85617s\n",
      "Processing CHRM 5 CHR5 took 4.1813s\n",
      "Processing CHRM 6 CHR6 took 0.488669s\n",
      "Processing CHRM 7 CHR7 took 4.01574s\n",
      "Processing CHRM 8 CHR8 took 3.52082s\n",
      "Processing CHRM 9 CHR9 took 2.71406s\n",
      "Done DHS file\n",
      "METHOD: gtex-ms raw_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 4.63943s\n",
      "Processing CHRM 10 CHR10 took 3.43509s\n",
      "Processing CHRM 11 CHR11 took 2.92432s\n",
      "Processing CHRM 12 CHR12 took 0.519234s\n",
      "Processing CHRM 13 CHR13 took 2.51396s\n",
      "Processing CHRM 14 CHR14 took 1.85756s\n",
      "Processing CHRM 15 CHR15 took 1.76958s\n",
      "Processing CHRM 16 CHR16 took 2.06186s\n",
      "Processing CHRM 17 CHR17 took 1.77392s\n",
      "Processing CHRM 18 CHR18 took 1.83456s\n",
      "Processing CHRM 19 CHR19 took 1.51413s\n",
      "Processing CHRM 2 CHR2 took 5.39177s\n",
      "Processing CHRM 20 CHR20 took 0.228453s\n",
      "Processing CHRM 21 CHR21 took 0.852458s\n",
      "Processing CHRM 22 CHR22 took 1.3822s\n",
      "Processing CHRM 3 CHR3 took 2.83519s\n",
      "Processing CHRM 4 CHR4 took 4.83356s\n",
      "Processing CHRM 5 CHR5 took 4.1956s\n",
      "Processing CHRM 6 CHR6 took 0.89606s\n",
      "Processing CHRM 7 CHR7 took 3.71022s\n",
      "Processing CHRM 8 CHR8 took 3.12608s\n",
      "Processing CHRM 9 CHR9 took 2.75613s\n",
      "Done DHS file\n",
      "METHOD: gtex-sse tmm_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 4.81619s\n",
      "Processing CHRM 10 CHR10 took 3.22466s\n",
      "Processing CHRM 11 CHR11 took 2.99864s\n",
      "Processing CHRM 12 CHR12 took 0.380269s\n",
      "Processing CHRM 13 CHR13 took 2.53352s\n",
      "Processing CHRM 14 CHR14 took 2.05395s\n",
      "Processing CHRM 15 CHR15 took 1.70268s\n",
      "Processing CHRM 16 CHR16 took 1.95935s\n",
      "Processing CHRM 17 CHR17 took 1.8954s\n",
      "Processing CHRM 18 CHR18 took 1.96002s\n",
      "Processing CHRM 19 CHR19 took 1.51649s\n",
      "Processing CHRM 2 CHR2 took 5.45844s\n",
      "Processing CHRM 20 CHR20 took 0.282343s\n",
      "Processing CHRM 21 CHR21 took 0.990104s\n",
      "Processing CHRM 22 CHR22 took 0.970245s\n",
      "Processing CHRM 3 CHR3 took 2.15012s\n",
      "Processing CHRM 4 CHR4 took 4.54292s\n",
      "Processing CHRM 5 CHR5 took 3.89987s\n",
      "Processing CHRM 6 CHR6 took 0.639124s\n",
      "Processing CHRM 7 CHR7 took 3.90614s\n",
      "Processing CHRM 8 CHR8 took 0.516823s\n",
      "Processing CHRM 9 CHR9 took 2.62542s\n",
      "Done DHS file\n",
      "METHOD: gtex-sse tmm_lasso matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.4804s\n",
      "Processing CHRM 10 CHR10 took 3.65111s\n",
      "Processing CHRM 11 CHR11 took 3.56031s\n",
      "Processing CHRM 12 CHR12 took 0.374375s\n",
      "Processing CHRM 13 CHR13 took 2.99586s\n",
      "Processing CHRM 14 CHR14 took 2.52712s\n",
      "Processing CHRM 15 CHR15 took 1.99293s\n",
      "Processing CHRM 16 CHR16 took 2.15884s\n",
      "Processing CHRM 17 CHR17 took 2.07724s\n",
      "Processing CHRM 18 CHR18 took 2.17114s\n",
      "Processing CHRM 19 CHR19 took 1.73486s\n",
      "Processing CHRM 2 CHR2 took 6.01034s\n",
      "Processing CHRM 20 CHR20 took 0.354971s\n",
      "Processing CHRM 21 CHR21 took 1.09728s\n",
      "Processing CHRM 22 CHR22 took 1.16118s\n",
      "Processing CHRM 3 CHR3 took 2.75676s\n",
      "Processing CHRM 4 CHR4 took 5.27135s\n",
      "Processing CHRM 5 CHR5 took 4.35733s\n",
      "Processing CHRM 6 CHR6 took 0.816412s\n",
      "Processing CHRM 7 CHR7 took 4.32785s\n",
      "Processing CHRM 8 CHR8 took 0.488712s\n",
      "Processing CHRM 9 CHR9 took 2.9905s\n",
      "Done DHS file\n",
      "METHOD: gtex-sse raw_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.36424s\n",
      "Processing CHRM 10 CHR10 took 3.51797s\n",
      "Processing CHRM 11 CHR11 took 3.4538s\n",
      "Processing CHRM 12 CHR12 took 0.582781s\n",
      "Processing CHRM 13 CHR13 took 2.64667s\n",
      "Processing CHRM 14 CHR14 took 2.29791s\n",
      "Processing CHRM 15 CHR15 took 1.99534s\n",
      "Processing CHRM 16 CHR16 took 2.03241s\n",
      "Processing CHRM 17 CHR17 took 2.07617s\n",
      "Processing CHRM 18 CHR18 took 2.12231s\n",
      "Processing CHRM 19 CHR19 took 1.6455s\n",
      "Processing CHRM 2 CHR2 took 5.80345s\n",
      "Processing CHRM 20 CHR20 took 0.796885s\n",
      "Processing CHRM 21 CHR21 took 1.09542s\n",
      "Processing CHRM 22 CHR22 took 1.06858s\n",
      "Processing CHRM 3 CHR3 took 4.43665s\n",
      "Processing CHRM 4 CHR4 took 5.11075s\n",
      "Processing CHRM 5 CHR5 took 4.16871s\n",
      "Processing CHRM 6 CHR6 took 1.70408s\n",
      "Processing CHRM 7 CHR7 took 4.1189s\n",
      "Processing CHRM 8 CHR8 took 0.513656s\n",
      "Processing CHRM 9 CHR9 took 2.99939s\n",
      "Done DHS file\n",
      "METHOD: gtex-as tmm_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 4.83309s\n",
      "Processing CHRM 10 CHR10 took 2.93809s\n",
      "Processing CHRM 11 CHR11 took 2.94561s\n",
      "Processing CHRM 12 CHR12 took 0.391165s\n",
      "Processing CHRM 13 CHR13 took 2.31567s\n",
      "Processing CHRM 14 CHR14 took 2.01722s\n",
      "Processing CHRM 15 CHR15 took 1.74454s\n",
      "Processing CHRM 16 CHR16 took 1.92094s\n",
      "Processing CHRM 17 CHR17 took 1.80778s\n",
      "Processing CHRM 18 CHR18 took 1.69726s\n",
      "Processing CHRM 19 CHR19 took 1.45671s\n",
      "Processing CHRM 2 CHR2 took 4.91965s\n",
      "Processing CHRM 20 CHR20 took 0.22811s\n",
      "Processing CHRM 21 CHR21 took 0.899556s\n",
      "Processing CHRM 22 CHR22 took 0.936509s\n",
      "Processing CHRM 3 CHR3 took 2.29668s\n",
      "Processing CHRM 4 CHR4 took 4.05948s\n",
      "Processing CHRM 5 CHR5 took 3.91057s\n",
      "Processing CHRM 6 CHR6 took 0.538873s\n",
      "Processing CHRM 7 CHR7 took 3.56499s\n",
      "Processing CHRM 8 CHR8 took 1.48108s\n",
      "Processing CHRM 9 CHR9 took 2.55079s\n",
      "Done DHS file\n",
      "METHOD: gtex-as tmm_lasso matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.43702s\n",
      "Processing CHRM 10 CHR10 took 3.39378s\n",
      "Processing CHRM 11 CHR11 took 3.30531s\n",
      "Processing CHRM 12 CHR12 took 0.40468s\n",
      "Processing CHRM 13 CHR13 took 2.58981s\n",
      "Processing CHRM 14 CHR14 took 2.23147s\n",
      "Processing CHRM 15 CHR15 took 1.92331s\n",
      "Processing CHRM 16 CHR16 took 2.08774s\n",
      "Processing CHRM 17 CHR17 took 1.93489s\n",
      "Processing CHRM 18 CHR18 took 1.88529s\n",
      "Processing CHRM 19 CHR19 took 1.67485s\n",
      "Processing CHRM 2 CHR2 took 5.54899s\n",
      "Processing CHRM 20 CHR20 took 0.240602s\n",
      "Processing CHRM 21 CHR21 took 0.995444s\n",
      "Processing CHRM 22 CHR22 took 1.02799s\n",
      "Processing CHRM 3 CHR3 took 4.87576s\n",
      "Processing CHRM 4 CHR4 took 4.72741s\n",
      "Processing CHRM 5 CHR5 took 4.53303s\n",
      "Processing CHRM 6 CHR6 took 0.807236s\n",
      "Processing CHRM 7 CHR7 took 3.95526s\n",
      "Processing CHRM 8 CHR8 took 1.29588s\n",
      "Processing CHRM 9 CHR9 took 2.78481s\n",
      "Done DHS file\n",
      "METHOD: gtex-as raw_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.55903s\n",
      "Processing CHRM 10 CHR10 took 3.55723s\n",
      "Processing CHRM 11 CHR11 took 3.37743s\n",
      "Processing CHRM 12 CHR12 took 0.414714s\n",
      "Processing CHRM 13 CHR13 took 2.62658s\n",
      "Processing CHRM 14 CHR14 took 2.25673s\n",
      "Processing CHRM 15 CHR15 took 2.02178s\n",
      "Processing CHRM 16 CHR16 took 2.06773s\n",
      "Processing CHRM 17 CHR17 took 2.07912s\n",
      "Processing CHRM 18 CHR18 took 2.07836s\n",
      "Processing CHRM 19 CHR19 took 1.59912s\n",
      "Processing CHRM 2 CHR2 took 5.47994s\n",
      "Processing CHRM 20 CHR20 took 0.230747s\n",
      "Processing CHRM 21 CHR21 took 0.986259s\n",
      "Processing CHRM 22 CHR22 took 1.10537s\n",
      "Processing CHRM 3 CHR3 took 1.66116s\n",
      "Processing CHRM 4 CHR4 took 4.55328s\n",
      "Processing CHRM 5 CHR5 took 4.5095s\n",
      "Processing CHRM 6 CHR6 took 0.983373s\n",
      "Processing CHRM 7 CHR7 took 4.14934s\n",
      "Processing CHRM 8 CHR8 took 3.36106s\n",
      "Processing CHRM 9 CHR9 took 3.03304s\n",
      "Done DHS file\n",
      "METHOD: gtex-wb tmm_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 4.80353s\n",
      "Processing CHRM 10 CHR10 took 2.46714s\n",
      "Processing CHRM 11 CHR11 took 2.34711s\n",
      "Processing CHRM 12 CHR12 took 0.380438s\n",
      "Processing CHRM 13 CHR13 took 2.20559s\n",
      "Processing CHRM 14 CHR14 took 1.59016s\n",
      "Processing CHRM 15 CHR15 took 1.76473s\n",
      "Processing CHRM 16 CHR16 took 1.61426s\n",
      "Processing CHRM 17 CHR17 took 1.54197s\n",
      "Processing CHRM 18 CHR18 took 1.6049s\n",
      "Processing CHRM 19 CHR19 took 1.2453s\n",
      "Processing CHRM 2 CHR2 took 4.27041s\n",
      "Processing CHRM 20 CHR20 took 0.23316s\n",
      "Processing CHRM 21 CHR21 took 0.664333s\n",
      "Processing CHRM 22 CHR22 took 0.886898s\n",
      "Processing CHRM 3 CHR3 took 0.875339s\n",
      "Processing CHRM 4 CHR4 took 3.78s\n",
      "Processing CHRM 5 CHR5 took 3.29442s\n",
      "Processing CHRM 6 CHR6 took 0.7908s\n",
      "Processing CHRM 7 CHR7 took 3.29135s\n",
      "Processing CHRM 8 CHR8 took 0.572504s\n",
      "Processing CHRM 9 CHR9 took 2.41425s\n",
      "Done DHS file\n",
      "METHOD: gtex-wb tmm_lasso matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.62203s\n",
      "Processing CHRM 10 CHR10 took 2.90771s\n",
      "Processing CHRM 11 CHR11 took 2.68182s\n",
      "Processing CHRM 12 CHR12 took 0.371595s\n",
      "Processing CHRM 13 CHR13 took 2.42571s\n",
      "Processing CHRM 14 CHR14 took 1.71078s\n",
      "Processing CHRM 15 CHR15 took 2.06297s\n",
      "Processing CHRM 16 CHR16 took 1.76305s\n",
      "Processing CHRM 17 CHR17 took 1.7681s\n",
      "Processing CHRM 18 CHR18 took 1.76209s\n",
      "Processing CHRM 19 CHR19 took 1.42768s\n",
      "Processing CHRM 2 CHR2 took 4.93204s\n",
      "Processing CHRM 20 CHR20 took 0.231036s\n",
      "Processing CHRM 21 CHR21 took 0.761206s\n",
      "Processing CHRM 22 CHR22 took 1.00309s\n",
      "Processing CHRM 3 CHR3 took 1.37842s\n",
      "Processing CHRM 4 CHR4 took 4.43486s\n",
      "Processing CHRM 5 CHR5 took 3.63822s\n",
      "Processing CHRM 6 CHR6 took 0.954874s\n",
      "Processing CHRM 7 CHR7 took 3.57001s\n",
      "Processing CHRM 8 CHR8 took 0.51165s\n",
      "Processing CHRM 9 CHR9 took 2.66653s\n",
      "Done DHS file\n",
      "METHOD: gtex-wb raw_cclm matrixeqtl\n",
      "Processing CHRM 1 CHR1 took 5.47411s\n",
      "Processing CHRM 10 CHR10 took 2.87807s\n",
      "Processing CHRM 11 CHR11 took 3.27639s\n",
      "Processing CHRM 12 CHR12 took 0.844382s\n",
      "Processing CHRM 13 CHR13 took 2.34284s\n",
      "Processing CHRM 14 CHR14 took 2.13285s\n",
      "Processing CHRM 15 CHR15 took 1.84168s\n",
      "Processing CHRM 16 CHR16 took 1.93539s\n",
      "Processing CHRM 17 CHR17 took 1.7421s\n",
      "Processing CHRM 18 CHR18 took 1.93459s\n",
      "Processing CHRM 19 CHR19 took 1.71012s\n",
      "Processing CHRM 2 CHR2 took 5.4968s\n",
      "Processing CHRM 20 CHR20 took 0.234933s\n",
      "Processing CHRM 21 CHR21 took 1.0651s\n",
      "Processing CHRM 22 CHR22 took 0.965715s\n",
      "Processing CHRM 3 CHR3 took 4.5836s\n",
      "Processing CHRM 4 CHR4 took 4.86195s\n",
      "Processing CHRM 5 CHR5 took 3.69907s\n",
      "Processing CHRM 6 CHR6 took 1.07366s\n",
      "Processing CHRM 7 CHR7 took 3.12235s\n",
      "Processing CHRM 8 CHR8 took 1.06922s\n",
      "Processing CHRM 9 CHR9 took 3.21334s\n",
      "Done DHS file\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "methods = [\"matrixeqtl\"] #, \"matrixeqtl_rand\"]\n",
    "tissues = [\"gtex-ms\", \"gtex-sse\", \"gtex-as\", \"gtex-wb\"]\n",
    "expressions = [\"tmm_cclm\", \"tmm_lasso\", \"raw_cclm\"]\n",
    "\n",
    "for tissue in tissues:\n",
    "    for expression in expressions:\n",
    "        for method in methods:\n",
    "            print(\"METHOD:\",tissue, expression, method)\n",
    "            basedir = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v6_gtknn/{:s}/{:s}/{:s}/\".format(expression, tissue, method)\n",
    "            gtfile = basedir + \"chr{:d}/trans_eqtl.txt.pos\"\n",
    "            annot_outfile = basedir + dhs_annot_out\n",
    "            if not os.path.exists(annot_outfile):\n",
    "                annotate_snps(dhsfile, gtfile, annot_outfile, mode=\"meqtl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gtex-ms matrixeqtl tmm_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-sse matrixeqtl tmm_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-as matrixeqtl tmm_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-wb matrixeqtl tmm_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-ms matrixeqtl tmm_lasso 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-sse matrixeqtl tmm_lasso 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-as matrixeqtl tmm_lasso 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-wb matrixeqtl tmm_lasso 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-ms matrixeqtl raw_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-sse matrixeqtl raw_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-as matrixeqtl raw_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n",
      "gtex-wb matrixeqtl raw_cclm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 \n"
     ]
    }
   ],
   "source": [
    "\n",
    "datadir = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v6_gtknn/\"\n",
    "chrms = np.arange(1,23)\n",
    "snp_res_dict_meqtl = collections.defaultdict(dict)\n",
    "for expr in expressions:\n",
    "    for tissue in tissues:\n",
    "        snp_res_dict_meqtl[expr][tissue] = collections.defaultdict(dict)\n",
    "        for method in methods:\n",
    "            snp_res = list()\n",
    "            print(tissue, method, expr, end=\" \")\n",
    "            annotfile = os.path.join(datadir,expr,tissue,method,dhs_annot_out)\n",
    "            snp_annot = read_annot(annotfile)\n",
    "            for chrom in chrms:\n",
    "                print(chrom, end=\" \")\n",
    "                inputfile = os.path.join(datadir,expr,tissue,method,\"chr\"+str(chrom),\"trans_eqtl.txt.pos\")\n",
    "                snp_res += matrixeqtl(inputfile, chrom, snp_annot)\n",
    "            print(\"\")\n",
    "            snp_res_dict_meqtl[expr][tissue][method] = snp_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/904\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/1032\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1190\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/1699\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/3077\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/987\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/1114\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1346\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/1999\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/3551\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/584\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/768\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1141\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/2959\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/7880\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/1081\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/1240\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1435\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/2159\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/3853\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/1179\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/1308\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1595\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/2351\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/4618\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/1012\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/1185\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1613\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/3232\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/7304\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/723\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/910\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1077\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/1462\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/2996\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/854\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/983\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1137\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/1648\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/3416\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/726\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/986\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/1561\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/3683\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/8669\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/564\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/689\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/840\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/1171\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/2104\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/648\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/765\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/914\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/1329\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/2369\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-11 (matrixeqtl): 0, 0/146\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-10 (matrixeqtl): 0, 0/275\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-09 (matrixeqtl): 0, 0/630\n",
      "Fraction of trans-eQTLS in DHS regions @ 1e-08 (matrixeqtl): 0, 0/2504\n",
      "Fraction of trans-eQTLS in DHS regions @ 5e-08 (matrixeqtl): 0, 0/6430\n"
     ]
    }
   ],
   "source": [
    "meqtl_enrich_dict = dict()\n",
    "\n",
    "for mytissue in tissues:\n",
    "    for mymethod in methods:\n",
    "        for myexpr in expressions:\n",
    "            snp_list = snp_res_dict_meqtl[myexpr][mytissue][mymethod]\n",
    "            top_list = sorted(snp_list, key=attrgetter('logp'), reverse=True)\n",
    "\n",
    "            cutoffs = [1e-11, 1e-10, 1e-9, 1e-8, 5e-8] #, 1e-7, 5e-7, 1e-6, 1e-5, 1e-3, 5e-2]\n",
    "\n",
    "            fracs_rr = list()\n",
    "            for cutoff in cutoffs:\n",
    "                rr_teqtls = count_inDHS(top_list, cutoff)\n",
    "                total_rr_teqtls = len(rr_teqtls)\n",
    "                if total_rr_teqtls > 0:\n",
    "                    rr_inDHS = np.sum(np.array(rr_teqtls))\n",
    "                    frac_rr = rr_inDHS/total_rr_teqtls\n",
    "                    fracs_rr.append(frac_rr)\n",
    "                    print(\"Fraction of trans-eQTLS in DHS regions @ {:g} ({:s}): {:g}, {:d}/{:d}\".format(cutoff, mymethod, frac_rr, rr_inDHS, total_rr_teqtls))\n",
    "                else:\n",
    "                    print(\"No signif trans-eQTLs found at {:g} cutoff\".format(cutoff))\n",
    "            key = \"{:s}_{:s}_{:s}\".format(myexpr, mytissue, mymethod)\n",
    "            meqtl_enrich_dict[key] = fracs_rr  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'tmm_cclm_gtex-ms_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_lasso_gtex-ms_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'raw_cclm_gtex-ms_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_cclm_gtex-sse_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_lasso_gtex-sse_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'raw_cclm_gtex-sse_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_cclm_gtex-as_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_lasso_gtex-as_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'raw_cclm_gtex-as_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_cclm_gtex-wb_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'tmm_lasso_gtex-wb_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0],\n",
       " 'raw_cclm_gtex-wb_matrixeqtl': [0.0, 0.0, 0.0, 0.0, 0.0]}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "ax.plot(cutoffs, tejaas_enrich_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from operator import attrgetter\n",
    "\n",
    "MAX_TOP = 200\n",
    "TOP_W = 100\n",
    "mytissue = \"gtex-ms\"\n",
    "mysb = \"0.1_knn30\"\n",
    "myexpr = \"raw\"\n",
    "mymethod = \"tejaas\"\n",
    "snp_list = snp_res_dict[mysb][myexpr][mytissue][mymethod]\n",
    "top_list = sorted(snp_list, key=attrgetter('logp'), reverse=True)\n",
    "\n",
    "eqtlgen_top_list = sorted(eqtlgen_snp_res, key=attrgetter('logp'), reverse=True)\n",
    "\n",
    "x = np.array([])\n",
    "y = np.array([])\n",
    "ygen = np.array([])\n",
    "\n",
    "wx = np.array([])\n",
    "wy = np.array([])\n",
    "wygen = np.array([])\n",
    "\n",
    "eqtlgen_true = 0\n",
    "true_eqtls = 0\n",
    "rand_eqtls = 0\n",
    "\n",
    "prev = 0\n",
    "for TOPN in range(TOP_W, MAX_TOP, TOP_W):\n",
    "    \n",
    "    window_true_eqtls = np.sum([x.inDHS for x in top_list[prev:TOPN]])\n",
    "    true_eqtls += window_true_eqtls\n",
    "\n",
    "    rand_distrib = []\n",
    "    # myrandmethods = [\"tejaas_rand\"] # + randmethods\n",
    "    # for mymethod in myrandmethods:\n",
    "    myrandmethod = \"tejaas_rand\"\n",
    "    snp_list_rand = snp_res_dict[mysb][myexpr][mytissue][myrandmethod]\n",
    "    top_list_rand = sorted(snp_list_rand, key=attrgetter('logp'), reverse=True)\n",
    "      \n",
    "    window_eqtlgen_true = np.sum([x.inDHS for x in eqtlgen_top_list[prev:TOPN]])\n",
    "    eqtlgen_true += window_eqtlgen_true\n",
    "    \n",
    "    window_rand_eqtls = np.sum([x.inDHS for x in top_list_rand[prev:TOPN]])\n",
    "    rand_eqtls += window_rand_eqtls\n",
    "\n",
    "    print(true_eqtls, rand_eqtls, true_eqtls/rand_eqtls, \"||\", eqtlgen_true, eqtlgen_true/rand_eqtls, window_eqtlgen_true, window_eqtlgen_true/window_rand_eqtls)\n",
    "    x = np.append(x, TOPN)\n",
    "    y = np.append(y, true_eqtls/rand_eqtls)\n",
    "    ygen = np.append(ygen, eqtlgen_true/rand_eqtls)\n",
    "    \n",
    "    wx = np.append(wx, TOPN)\n",
    "    wy = np.append(wy, window_true_eqtls/window_rand_eqtls)\n",
    "    wygen = np.append(wygen, window_eqtlgen_true/window_rand_eqtls)\n",
    "    prev += TOP_W\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "outplotfile = \"test.dhs.png\"\n",
    "\n",
    "fig = plt.figure(figsize=(16,8))\n",
    "ax1 = fig.add_subplot(121)\n",
    "ax2 = fig.add_subplot(122)\n",
    "ax1.plot(x, y, label=\"RR\")\n",
    "ax1.plot(x, ygen, label=\"EQTLgen\")\n",
    "ax1.set_title(\"Cumulative enrichment\")\n",
    "ax1.axhline(y=1, color='red')\n",
    "\n",
    "ax2.bar(wx, wy, 100, alpha=0.3, label=\"RR\")\n",
    "ax2.bar(wx, wygen, 100, alpha=0.3, label=\"EQTLgen\")\n",
    "ax2.axhline(y=1, color='red')\n",
    "ax2.set_title(\"Windowed enrichment\")\n",
    "\n",
    "ax1.set_xlabel(\"First N SNPS\")\n",
    "ax2.set_xlabel(\"First N SNPS\")\n",
    "ax1.set_ylabel(\"Enrichment\")\n",
    "ax2.set_ylabel(\"Enrichment\")\n",
    "\n",
    "fig.suptitle('Main title')\n",
    "ax1.legend()\n",
    "ax2.legend()\n",
    "\n",
    "# fig.tight_layout()\n",
    "# fig.subplots_adjust(top=0.88)\n",
    "plt.savefig(outplotfile, bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
