{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "import gzip\n",
    "import random\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.interpolate import interp1d\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "#plt.switch_backend('agg')\n",
    "import matplotlib\n",
    "from utils import readgtf\n",
    "\n",
    "# from utils import mpl_stylesheet\n",
    "# mpl_stylesheet.banskt_presentation(fontfamily = 'latex-clearsans', fontsize = 18, colors = 'banskt', dpi = 72)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re\n",
    "from functools import wraps\n",
    "import time\n",
    "\n",
    "SNPINFO_FIELDS = ['chrom', 'varid', 'bp_pos', 'ref_allele', 'alt_allele', 'maf']\n",
    "class SnpInfo(collections.namedtuple('_SnpInfo', SNPINFO_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "def timeit(f):\n",
    "    @wraps(f)\n",
    "    def wrap(*args, **kw):\n",
    "        ts = time.time()\n",
    "        result = f(*args, **kw)\n",
    "        te = time.time()\n",
    "        print('{:s} took: {:.6f} seconds'.format(f.__name__, te-ts))\n",
    "        return result\n",
    "    return wrap\n",
    "def normalize_expr(Y):\n",
    "    if isinstance(Y, pd.DataFrame):\n",
    "        Y_cent = (Y.values - np.mean(Y.values, axis = 1).reshape(-1, 1)) / np.std(Y.values, axis = 1).reshape(-1, 1)\n",
    "        Y_cent = pd.DataFrame(Y_cent, index=Y.index, columns=Y.columns)\n",
    "        Y_cent.index.name = Y.index.name\n",
    "    else:\n",
    "        Y_cent = (Y - np.mean(Y, axis = 1).reshape(-1, 1)) / np.std(Y, axis = 1).reshape(-1, 1)\n",
    "    return Y_cent\n",
    "\n",
    "def normalize_and_center_dosage(dosage, snpinfo):\n",
    "    f = [snp.maf for snp in snpinfo]\n",
    "    f = np.array(f).reshape(-1, 1)\n",
    "    gtnorm = (dosage - (2 * f)) / np.sqrt(2 * f * (1 - f))\n",
    "    gtcent = dosage - np.mean(dosage, axis = 1).reshape(-1, 1)\n",
    "    return gtnorm, gtcent #rr uses gtcent\n",
    "\n",
    "def sample_gt(snpinfo, nsample):\n",
    "    nsnps = len(snpinfo)\n",
    "    dosages = np.zeros(nsnps * nsample)\n",
    "    i = 0\n",
    "    for snp in snpinfo:\n",
    "        gtcent = sample_from_maf(nsample, snp.maf)\n",
    "        dosages[i,:] = gtcent\n",
    "        i += 1\n",
    "    return dosages    \n",
    "\n",
    "def sample_from_maf(nsample, maf):\n",
    "    dosage = np.zeros(nsample)\n",
    "    mafratios = np.array([(1 - maf)**2, 2 * maf * (1 - maf), maf**2])\n",
    "    nfreq  = np.random.multinomial(nsample, mafratios, size=1)[0]\n",
    "    f1 = np.repeat(0, nfreq[0])\n",
    "    f2 = np.repeat(1, nfreq[1])\n",
    "    f3 = np.repeat(2, nfreq[2])\n",
    "    x  = np.concatenate((f1,f2,f3))\n",
    "    dosage = np.random.permutation(x)\n",
    "    # gtnorm = (dosage - (2 * maf2d)) / np.sqrt(2 * maf2d * (1 - maf2d))\n",
    "    gtcent = dosage - np.mean(dosage)\n",
    "\n",
    "    return gtcent\n",
    "\n",
    "def simulate_gt(nsnps, nsample):\n",
    "    mafs = np.linspace(0.1, 0.9, nsnps)\n",
    "    gtcent = np.zeros((nsnps, nsample))\n",
    "    snpinfo = list()\n",
    "    for i in range(nsnps):\n",
    "        gtcent[i,:] = sample_from_maf(nsample, mafs[i])\n",
    "        this_snp = SnpInfo(chrom      = 1,\n",
    "                           bp_pos     = i*100,\n",
    "                           varid      = \"rsid\"+str(i),\n",
    "                           ref_allele = \"A\",\n",
    "                           alt_allele = \"G\",\n",
    "                           maf        = mafs[i])\n",
    "        snpinfo.append(this_snp)\n",
    "    return gtcent, snpinfo\n",
    "def knn_correction(expr, dosage, K, f=1):\n",
    "    pca = PCA(n_components=int(f * min(expr.shape[0], expr.shape[1]) ))\n",
    "#     print(\"Original dimension: {:d} x {:d}\".format(expr.shape[0], expr.shape[1]))\n",
    "    pca.fit(expr) # requires N x G\n",
    "    expr_pca = pca.transform(expr)\n",
    "#     print(\"Reduced dimension: {:d} x {:d}\".format(expr_pca.shape[0], expr_pca.shape[1]))\n",
    "\n",
    "    def gene_distance(a, b):\n",
    "        return np.linalg.norm(a - b)\n",
    "\n",
    "    nsample = expr.shape[0]\n",
    "    distance_matrix = np.zeros((nsample, nsample))\n",
    "    for i in range(nsample):\n",
    "        for j in range(i+1, nsample):\n",
    "            dist = gene_distance(expr_pca[i,:], expr_pca[j,:])\n",
    "            distance_matrix[i, j] = dist\n",
    "            distance_matrix[j, i] = dist\n",
    "\n",
    "    kneighbor = K\n",
    "    gx_knn = np.zeros_like(expr)\n",
    "    gt_knn = np.zeros_like(dosage)\n",
    "    neighbor_list = list()\n",
    "\n",
    "    for i in range(nsample):\n",
    "        neighbors = np.argsort(distance_matrix[i, :])[:kneighbor + 1][1:]\n",
    "        gx_knn[i, :] = expr[i, :] - np.mean(expr[neighbors, :], axis = 0)\n",
    "        # noisy_neighbors = np.random.choice(neighbors, size = int(2 * kneighbor / 3), replace = False)\n",
    "        # noisy_neighbors = np.random.choice(neighbors, size = kneighbor, replace = True )\n",
    "        noisy_neighbors = neighbors\n",
    "        gt_knn[:, i] = dosage[:, i] - np.mean(dosage[:, noisy_neighbors], axis = 1)\n",
    "        neighbor_list.append(neighbors)\n",
    "\n",
    "    return gx_knn, gt_knn\n",
    "\n",
    "# # PCA correction - Won't work: makes some singular values == 0\n",
    "def PCA_correction(gx, nComp = 5):\n",
    "    pca = PCA()\n",
    "    pca.fit(gx.T)\n",
    "    gx_pca = np.dot(pca.transform(gx.T)[:, nComp:], pca.components_[nComp:,:]).T\n",
    "    gx_pca_norm = normalize_expr(gx_pca)\n",
    "    return gx_pca_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15000, 350)\n",
      "350\n"
     ]
    }
   ],
   "source": [
    "# Use artificial genotype\n",
    "nsample = 350\n",
    "gtfull, snp_info = simulate_gt(10000, nsample)\n",
    "\n",
    "# # Simulate some expression\n",
    "ngene = 15000\n",
    "gx_rand = np.random.normal(0, 1, size = nsample * ngene).reshape((ngene, nsample)) \n",
    "print(gx_rand.shape)\n",
    "\n",
    "print(np.linalg.matrix_rank(gx_rand))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "nsample = gx_norm.shape[1]\n",
    "\n",
    "## --- Apply KNN correction\n",
    "f=1\n",
    "gx_knn, gt_knn = knn_correction(gx_norm.T, dosage, K=30, f=f)\n",
    "\n",
    "## --- Normalize GX and GT\n",
    "gt_norm, gt_cent = normalize_and_center_dosage(dosage, snp_info)\n",
    "sigmax2     = np.var(gt_cent, axis = 1)\n",
    "\n",
    "gx_knn_norm = normalize_expr(gx_knn.T)\n",
    "gt_knn_norm, gt_knn_cent = normalize_and_center_dosage(gt_knn, snp_info)\n",
    "sigmax2_knn = np.var(gt_knn_cent, axis = 1)\n",
    "\n",
    "shuffle_mask = np.arange(nsample)\n",
    "np.random.shuffle(shuffle_mask)\n",
    "gx_shuffled_norm = copy.deepcopy(gx_norm[:, shuffle_mask])\n",
    "gx_shuffled_knn_norm = copy.deepcopy(gx_knn_norm[:, shuffle_mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gx_knn60, gt_knn60 = knn_correction(gx_norm.T, dosage, K=60, f=f)\n",
    "# gx_knn15, gt_knn15 = knn_correction(gx_norm.T, dosage, K=15, f=f)\n",
    "\n",
    "# gx_knn_norm60 = normalize_expr(gx_knn60.T)\n",
    "# gt_knn_norm60, gt_knn_cent60 = normalize_and_center_dosage(gt_knn60, snp_info)\n",
    "# sigmax2_knn60 = np.var(gt_knn_cent60, axis = 1)\n",
    "\n",
    "# gx_knn_norm15 = normalize_expr(gx_knn15.T)\n",
    "# gt_knn_norm15, gt_knn_cent15 = normalize_and_center_dosage(gt_knn15, snp_info)\n",
    "# sigmax2_knn15 = np.var(gt_knn_cent15, axis = 1)\n",
    "\n",
    "# gx_shuffled_knn_norm60 = copy.deepcopy(gx_knn_norm60[:, shuffle_mask])\n",
    "# gx_shuffled_knn_norm15 = copy.deepcopy(gx_knn_norm15[:, shuffle_mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## --- Obtain CisMasks\n",
    "\n",
    "cismasklist = get_cismasklist(snp_info, genes, chrm, window=1e6)\n",
    "cismaskcomp = compress_cismasklist(cismasklist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Y = gx_knn_norm # / np.sqrt(nsample)\n",
    "# Yt = Y.T\n",
    "# U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "# s2_median = np.median(np.square(S))\n",
    "# s2_10     = np.percentile(np.square(S), 10)\n",
    "# print(s2_median)\n",
    "# print(s2_10)\n",
    "# SIGMA_BETAS = np.repeat(0.05, gt_cent.shape[0])\n",
    "# DY_SIGMA_BETAS = np.sqrt( sigmax2_knn / s2_median )\n",
    "# DY_SIGMA_BETAS10 = np.sqrt( sigmax2_knn / s2_10 )\n",
    "# plt.hist(DY_SIGMA_BETAS, alpha=0.4, label=\"median\")\n",
    "# plt.hist(DY_SIGMA_BETAS10, alpha=0.4, label=\"percentile\")\n",
    "# plt.legend()\n",
    "# plt.show()\n",
    "\n",
    "# # SIGMA_BETAS = DY_SIGMA_BETAS10\n",
    "# SIGMA_BETAS = DY_SIGMA_BETAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Optimizes sb2 to a given Keff\n",
    "\n",
    "# @timeit\n",
    "# def optimize_sb2(S, sigmasx, target):\n",
    "#     sbetas = list()\n",
    "#     S2 = np.square(S)\n",
    "#     S2_lim = np.percentile(S2, 50)\n",
    "#     for sx2 in sigmasx:\n",
    "#         sb2 =  sx2 / S2_lim       # start parameter at median\n",
    "#         S2mod = S2 + (sx2 / sb2)\n",
    "#         N = len(S2)\n",
    "#         Keff = np.sum(S2/S2mod) / N\n",
    "\n",
    "#         while np.abs(Keff - target) > 0.01:\n",
    "#             diff = Keff - target\n",
    "#             sb2 -= diff*(sb2)\n",
    "#             S2mod = S2 + (sx2 / sb2)\n",
    "#             Keff = np.sum(S2/S2mod) / N\n",
    "#         #print(\"Keff\",Keff)\n",
    "#         #print(\"SB2=\",sb2)\n",
    "#         sbetas.append(sb2)\n",
    "#     return np.array(sbetas)\n",
    "\n",
    "# print(len(sigmax2_knn))\n",
    "# sb2_opt = optimize_sb2(S, sigmax2_knn, 0.6)\n",
    "\n",
    "# Keffs = []\n",
    "# S2 = np.square(S)\n",
    "# N = len(S2)\n",
    "# for i in range(len(sigmax2_knn)):\n",
    "#     S2mod = S2 + (sigmax2_knn[i] / sb2_opt[i])\n",
    "#     Keffs.append(np.sum(S2/S2mod) / N)\n",
    "# print(Keffs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pvals_perm(GT, R, W):\n",
    "    mu2, mu4 = moment_data(GT)\n",
    "    N = GT.shape[1]\n",
    "    q11 = np.sum(W)\n",
    "    q2  = np.sum(np.diag(W))\n",
    "    muQ = mu2 * (N * q2 - q11) / (N - 1)\n",
    "\n",
    "    v31 = - mu4 / (N - 1)\n",
    "    v22 = v31 + (N * mu2 * mu2 / (N - 1)) #(N*(mu2**2) - mu4)/(N-1)\n",
    "    v211 = - (v31 + v22) / (N - 2)\n",
    "    v1111 = - 3 * v211 / (N - 3)\n",
    "\n",
    "    q31 = np.dot(np.diag(W),np.sum(W,axis = 1))\n",
    "    q4 = np.sum(np.square(np.diag(W)))\n",
    "    q22 = np.sum(np.square(W))\n",
    "    q211 = np.sum(np.square(np.sum(W,axis = 1)))\n",
    "\n",
    "    sigma2 = v1111*(q11**2 - 2*q2*q11 - 4*q211 + 8*q31 + 2*q22 + q2**2 - 6*q4) + 2*v211*(q2*q11 + 2*q211 - 6*q31 - 2*q22 - q2**2 + 6*q4) + v22*(q2**2 + 2*q22 - 3*q4) + 4*v31*(q31 - q4) + mu4*q4\n",
    "\n",
    "    sigma2 = sigma2 - muQ**2\n",
    "    sigmaQ = np.sqrt(sigma2)\n",
    "    p = 1 - stats.norm.cdf(R, loc=muQ, scale=sigmaQ)\n",
    "    return p, muQ, sigmaQ\n",
    "\n",
    "def moment_data(GT):   #GT ixN\n",
    "    GT2 = np.square(GT)\n",
    "    GT4 = np.square(GT2)\n",
    "    mu2 = np.mean(GT2)\n",
    "    mu4 = np.mean(GT4)\n",
    "    return mu2, mu4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.optimize import minimize\n",
    "\n",
    "class SBoptimizer:\n",
    "\n",
    "    def __init__(self, GT, GX, sx2):\n",
    "\n",
    "        self._GT  = np.ascontiguousarray(GT)\n",
    "        self._GX  = np.ascontiguousarray(GX)\n",
    "        self._sx2 = np.ascontiguousarray(sx2)\n",
    "        self._nsnps = GT.shape[0]\n",
    "        self._nsample = GX.shape[1]\n",
    "        \n",
    "        U, S, VT = np.linalg.svd(GX.T)\n",
    "        self._S = S\n",
    "        self._U = U\n",
    "        self._S2 = np.square(S)\n",
    "        self._opt_sb2 = np.zeros(self._nsnps)\n",
    "    \n",
    "    @property\n",
    "    def sb2(self):\n",
    "        return self._opt_sb2\n",
    "\n",
    "    def get_ML(self, _sb2, i):\n",
    "        # sb2 = sb * sb\n",
    "        sb2 = np.exp(_sb2)\n",
    "        S2mod = self._S2 + (self._sx2[i] / sb2)\n",
    "        Rscore = np.sum(np.square(np.dot(self._U.T, self._GT[i,:])) * (self._S2 / S2mod)) / self._sx2[i]\n",
    "        MLL = -0.5*np.sum(np.log( self._S2 * (sb2 / self._sx2[i]) + 1 )) + 0.5*Rscore\n",
    "\n",
    "        denom = (self._S2 * sb2 + self._sx2[i])\n",
    "        der = 0.5* np.sum( ( self._S2 / denom ) * ( (np.square(np.dot(self._U.T, self._GT[i,:])) / denom ) - 1 ) )\n",
    "        return -MLL, sb2*np.array([-der])\n",
    "\n",
    "    def fit(self):\n",
    "        st = time.time()\n",
    "        \n",
    "        sb_init = np.exp(0.01)\n",
    "        for i in range(self._nsnps):\n",
    "            res = minimize(   self.get_ML,\n",
    "                              sb_init, \n",
    "                              args = i,\n",
    "                              method='L-BFGS-B',\n",
    "                              jac = True,\n",
    "                              #bounds = [[0,1]],\n",
    "                              options={'maxiter': 200000,\n",
    "                                       'maxfun': 2000000,\n",
    "                                       #'ftol': 1e-9,\n",
    "                                       #'gtol': 1e-9,\n",
    "                                       'disp': True})\n",
    "\n",
    "            # print(res)\n",
    "            self._opt_sb2[i] = np.exp(res.x[0])\n",
    "        et = time.time()\n",
    "        print(\"optimization took in total: \",et-st)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def tejaas_rr(myGT, myGX, sigmax2, sb2):\n",
    "    nsnps  = myGT.shape[0]\n",
    "    Rscore = np.zeros(nsnps)\n",
    "    pvals  = np.zeros(nsnps)\n",
    "    muQ    = np.zeros(nsnps)\n",
    "    sigmaQ = np.zeros(nsnps)\n",
    "    Keff   = np.zeros(nsnps)\n",
    "    EmuQ   = np.zeros(nsnps)\n",
    "    EsigmaQ= np.zeros(nsnps)\n",
    "    \n",
    "    # Yt = (GX / np.sqrt(nsample)).T\n",
    "    Yt = myGX.T\n",
    "    U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "    S2 = np.square(S)\n",
    "    \n",
    "    for j in range(nsnps):       \n",
    "        S2mod = S2 + (sigmax2[j] / sb2[j])\n",
    "        Keff[j] = np.sum(S2/S2mod)\n",
    "        \n",
    "        W = np.dot(U, np.dot(np.diag(S2 / S2mod), U.T)) / sigmax2[j]\n",
    "        Rscore[j] = np.sum(np.square(np.dot(U.T, myGT[j,:])) * (S2 / S2mod)) / sigmax2[j]\n",
    "        \n",
    "        pvals[j], muQ[j], sigmaQ[j] = pvals_perm(myGT[j, :].reshape(1, -1), Rscore[j], W)\n",
    "        \n",
    "#         rand_gt = copy.deepcopy(myGT[j,:])\n",
    "#         emp_q = np.zeros(2000)\n",
    "#         for k in range(2000):\n",
    "#             random.shuffle(rand_gt)\n",
    "#             emp_q[k] = np.sum(np.square(np.dot(U.T, rand_gt)) * (S2 / S2mod)) / sigmax2[j]\n",
    "#         EmuQ[j]     = np.mean(emp_q)\n",
    "#         EsigmaQ[j]  = np.var(emp_q)\n",
    "        \n",
    "        \n",
    "    return S2, pvals, Rscore, muQ, sigmaQ, Keff #, EmuQ, EsigmaQ\n",
    "\n",
    "def compare_basic_props(GX, GT, ax1, ax2, label, sigmax2, sb2 = None, Ucorr = False, cismasks = None):\n",
    "        \n",
    "    nsnps  = GT.shape[0]\n",
    "    Rscore = np.zeros(nsnps)\n",
    "    pvals  = np.zeros(nsnps)\n",
    "    muQ    = np.zeros(nsnps)\n",
    "    sigmaQ = np.zeros(nsnps)\n",
    "    Keff   = np.zeros(nsnps)\n",
    "    EmuQ   = np.zeros(nsnps)\n",
    "    EsigmaQ= np.zeros(nsnps)\n",
    "    myGT = GT\n",
    "    myGX = GX\n",
    "    opt_sb2 = np.zeros(nsnps)\n",
    "    \n",
    "    print(\"====== {:s} ======\".format(label))\n",
    "    Yt = myGX.T\n",
    "    U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "    S2 = np.square(S, dtype=np.float64 )\n",
    "    \n",
    "    ax1.scatter(np.arange(len(S)-1), S[:-1], label = label, alpha=0.4)\n",
    "    ax2.hist(S, bins=30, alpha=0.3, label = label)\n",
    "    \n",
    "    if cismasks is not None:\n",
    "        nmasks = len(cismasks)\n",
    "        print(\"Found {:d} cismasks.\".format(nmasks))\n",
    "        for i,cismask in enumerate(cismasks):\n",
    "            print(\"Processing cismask {:d}: {:d} genes masked for {:d} snps\".format(i, len(cismask.rmv_id), cismask.nsnp))\n",
    "            usegenes = np.ones(GX.shape[0], dtype=bool)\n",
    "            if cismask.rmv_id.shape[0] > 0: usegenes[cismask.rmv_id] = False\n",
    "            myGX = GX[usegenes]\n",
    "            myGT = GT[cismask.apply2,:]\n",
    "            \n",
    "            SBOPT = SBoptimizer(myGT, myGX, sigmax2[cismask.apply2])\n",
    "            SBOPT.fit()\n",
    "            sb2 = SBOPT.sb2\n",
    "            opt_sb2[cismask.apply2] = SBOPT.sb2\n",
    "            \n",
    "#             _s2, _pvals, _qscores, _muq, _sigmaq, _keff, _emuq, _esigq = tejaas_rr(myGT, myGX, sigmax2[cismask.apply2], sb2, Z, ZTZ_inv) #[cismask.apply2])\n",
    "            _s2, _pvals, _qscores, _muq, _sigmaq, _keff = tejaas_rr(myGT, myGX, sigmax2[cismask.apply2], sb2) #[cismask.apply2])\n",
    "            pvals[cismask.apply2]  = _pvals\n",
    "            Rscore[cismask.apply2] = _qscores\n",
    "            muQ[cismask.apply2]    = _muq\n",
    "            sigmaQ[cismask.apply2] = _sigmaq\n",
    "            Keff[cismask.apply2]   = _keff\n",
    "#             EmuQ[cismask.apply2]   = _emuq\n",
    "#             EsigmaQ[cismask.apply2]= _esigq\n",
    "            \n",
    "    else:\n",
    "        S2, pvals, Rscore, muQ, sigmaQ = tejaas_rr(myGT, myGX, sigmax2, sb2)        \n",
    "    return S2, pvals, Rscore, muQ, sigmaQ, opt_sb2, Keff #, EmuQ, EsigmaQ\n",
    "\n",
    "fig = plt.figure(figsize = (14, 6))\n",
    "ax1 = fig.add_subplot(121)\n",
    "ax2 = fig.add_subplot(122)\n",
    "# sb2 = np.square(SIGMA_BETAS)\n",
    "sb2 = None # sb2_opt\n",
    "\n",
    "# - No KNN\n",
    "nS2, npvals, nQ, nmuQ, nscaledQ, nopt_sb2, nKeff = compare_basic_props(gx_norm, \n",
    "                                                                       gt_cent, \n",
    "                                                                       ax1, ax2, \n",
    "                                                                       'Norm - no KNN', \n",
    "                                                                       sigmax2, \n",
    "                                                                       sb2 = sb2, \n",
    "                                                                       cismasks=cismaskcomp)\n",
    "shuf_nS2, shuf_npvals, shuf_nQ, shuf_nmuQ, shuf_nscaledQ, shuf_nopt_sb2, shuf_nKeff = compare_basic_props(gx_shuffled_norm, \n",
    "                                                                                                          gt_cent, \n",
    "                                                                                                          ax1, ax2, \n",
    "                                                                                                          'Norm Shuffled - no KNN', \n",
    "                                                                                                          sigmax2, \n",
    "                                                                                                          sb2 = sb2, \n",
    "                                                                                                          cismasks=cismaskcomp)\n",
    "# nS2, npvals, nQ, nmuQ, nscaledQ, nopt_sb2, nKeff, nEmuQ, nEscaledQ = compare_basic_props(gx_norm, gt_cent, ax1, ax2, 'Norm - no KNN', sigmax2, Z, ZTZ_inv, sb2 = sb2, cismasks=cismaskcomp)\n",
    "# shuf_nS2, shuf_npvals, shuf_nQ, shuf_nmuQ, shuf_nscaledQ, shuf_nopt_sb2, shuf_nKeff, shuf_nEmuQ, shuf_nEscaledQ = compare_basic_props(gx_shuffled_norm, gt_cent, ax1, ax2, 'Norm Shuffled - no KNN', sigmax2, Z, ZTZ_inv, sb2 = sb2, cismasks=cismaskcomp)\n",
    "\n",
    "# - With KNN\n",
    "# S2, pvals, Q, muQ, scaledQ, opt_sb2, Keff, EmuQ, EscaledQ = compare_basic_props(gx_knn_norm, gt_knn_cent, ax1, ax2, 'Norm KNN', sigmax2 = sigmax2_knn, sb2 = sb2, cismasks=cismaskcomp)\n",
    "# shuf_S2, shuf_pvals, shuf_Q, shuf_muQ, shuf_scaledQ, shuf_opt_sb2, shuf_Keff, shuf_EmuQ, shuf_EscaledQ  = compare_basic_props(gx_shuffled_knn_norm, gt_knn_cent, ax1, ax2, 'Shuffled Norm KNN', sigmax2 = sigmax2_knn, sb2 = sb2, cismasks=cismaskcomp)\n",
    "S2, pvals, Q, muQ, scaledQ, opt_sb2, Keff = compare_basic_props(gx_knn_norm, \n",
    "                                                                gt_knn_cent, \n",
    "                                                                ax1, ax2, \n",
    "                                                                'Norm KNN', \n",
    "                                                                sigmax2_knn, \n",
    "                                                                sb2 = sb2, \n",
    "                                                                cismasks=cismaskcomp)\n",
    "shuf_S2, shuf_pvals, shuf_Q, shuf_muQ, shuf_scaledQ, shuf_opt_sb2, shuf_Keff = compare_basic_props(gx_shuffled_knn_norm, \n",
    "                                                                                                   gt_knn_cent, \n",
    "                                                                                                   ax1, ax2, \n",
    "                                                                                                   'Shuffled Norm KNN', \n",
    "                                                                                                   sigmax2_knn, \n",
    "                                                                                                   sb2 = sb2, \n",
    "                                                                                                   cismasks=cismaskcomp)\n",
    "\n",
    "# # # - With KNN\n",
    "# S260, pvals60, Q60, muQ60, scaledQ60 = compare_basic_props(gx_knn_norm60, gt_knn_cent60, ax1, ax2, 'Norm KNN 60', sigmax2 = sigmax2_knn60, sb2 = sb2, cismasks=cismaskcomp)\n",
    "# shuf_S260, shuf_pvals60, shuf_Q60, shuf_muQ60, shuf_scaledQ60 = compare_basic_props(gx_shuffled_knn_norm60, gt_knn_cent60, ax1, ax2, 'Shuffled Norm KNN 60', sigmax2 = sigmax2_knn60, sb2 = sb2, cismasks=cismaskcomp)\n",
    "\n",
    "# # - With KNN\n",
    "# S215, pvals15, Q15, muQ15, scaledQ15 = compare_basic_props(gx_knn_norm15, gt_knn_cent15, ax1, ax2, 'Norm KNN15', sigmax2 = sigmax2_knn15, sb2 = sb2, cismasks=cismaskcomp)\n",
    "# shuf_S215, shuf_pvals15, shuf_Q15, shuf_muQ15, shuf_scaledQ15 = compare_basic_props(gx_shuffled_knn_norm15, gt_knn_cent15, ax1, ax2, 'Shuffled Norm KNN15', sigmax2 = sigmax2_knn15, sb2 = sb2, cismasks=cismaskcomp)\n",
    "\n",
    "ax1.set_xlabel(\"Singular values (rank)\")\n",
    "ax1.set_ylabel(\"Singular values S (value)\")\n",
    "ax1.legend()\n",
    "\n",
    "ax2.legend()\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10,10), dpi=90)\n",
    "ax1 = fig.add_subplot(2,2,1)\n",
    "ax1.hist(shuf_npvals, bins=20, alpha=0.3, label=\"norm shuf - no KNN\")\n",
    "ax1.hist(npvals, bins=20, alpha=0.3, label=\"norm - no KNN\")\n",
    "ax1.legend()\n",
    "\n",
    "ax2 = fig.add_subplot(2,2,2)\n",
    "ax2.hist(shuf_pvals, bins=20, alpha=0.3, label=\"KNN shuf\")\n",
    "ax2.hist(pvals, bins=20, alpha=0.3, label=\"KNN\")\n",
    "ax2.legend()\n",
    "\n",
    "# ax3 = fig.add_subplot(2,2,3)\n",
    "# ax3.hist(shuf_pvals60, bins=20, alpha=0.3, label=\"KNN shuf 60\")\n",
    "# ax3.hist(pvals60, bins=20, alpha=0.3, label=\"KNN 60\")\n",
    "# ax3.legend()\n",
    "\n",
    "# ax4 = fig.add_subplot(2,2,4)\n",
    "# ax4.hist(shuf_pvals15, bins=20, alpha=0.3, label=\"KNN shuf 15\")\n",
    "# ax4.hist(pvals15, bins=20, alpha=0.3, label=\"KNN 15\")\n",
    "# ax4.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(14,7), dpi=90)\n",
    "ax1 = fig.add_subplot(1,2,1)\n",
    "ax1.hist(shuf_npvals, bins=30, alpha=0.3, label=\"norm shuf - no KNN\")\n",
    "ax1.hist(npvals, bins=30, alpha=0.3, label=\"norm - no KNN\")\n",
    "ax1.legend()\n",
    "\n",
    "ax2 = fig.add_subplot(1,2,2)\n",
    "ax2.scatter((shuf_npvals), np.log10(shuf_nopt_sb2), s=2, alpha=0.3, label=\"shuf pvals sb2\")\n",
    "ax2.scatter((npvals), np.log10(nopt_sb2), s=2, alpha=0.3, label=\"pvals sb2\")\n",
    "ax2.set_ylabel(\"log10_sb2\")\n",
    "ax2.set_xlabel(\"pvalues\")\n",
    "ax2.legend()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(np.log10(shuf_nopt_sb2), np.log10(nopt_sb2))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# - no KNN\n",
    "qscaled = (nQ - nmuQ)/ nscaledQ\n",
    "shuf_qscaled = (shuf_nQ - shuf_nmuQ)/ shuf_nscaledQ\n",
    "\n",
    "# - KNN 30\n",
    "qscaled_knn = (Q - muQ)/ scaledQ\n",
    "shuf_qscaled_knn = (shuf_Q - shuf_muQ)/ shuf_scaledQ\n",
    "\n",
    "# - KNN 15\n",
    "# qscaled = (Q15 - muQ15)/ scaledQ15\n",
    "# shuf_qscaled = (shuf_Q15 - shuf_muQ15)/ shuf_scaledQ15\n",
    "\n",
    "# - KNN 60\n",
    "# qscaled2 = (Q60 - muQ60)/ scaledQ60\n",
    "# shuf_qscaled2 = (shuf_Q60 - shuf_muQ60)/ shuf_scaledQ60\n",
    "\n",
    "fig = plt.figure(figsize=(14,7))\n",
    "ax1 = fig.add_subplot(121)\n",
    "ax1.hist(qscaled, bins=30, alpha=0.3, density=True, label=\"Actual\")\n",
    "ax1.hist(shuf_qscaled, bins=30, alpha=0.3, density=True, label=\"Shuffled\")\n",
    "ax1.legend()\n",
    "\n",
    "ax2 = fig.add_subplot(122)\n",
    "ax2.hist(qscaled_knn, bins=30, alpha=0.3, density=True, label=\"Actual KNN\")\n",
    "ax2.hist(shuf_qscaled_knn, bins=30, alpha=0.3, density=True, label=\"Shuffled KNN\")\n",
    "ax2.legend()\n",
    "\n",
    "x = np.linspace(-4, 4, 100)\n",
    "rv = stats.norm(loc = 0, scale = 1)\n",
    "ax1.plot(x, rv.pdf(x), label = 'analytical')\n",
    "ax2.plot(x, rv.pdf(x), label = 'analytical')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([snp_info[x] for x in np.where(pvals < 1e-7)[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "[snp_info[x] for x in np.where(pvals < 1e-7)[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sbeta optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue = \"wb\"\n",
    "df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "ngene, nsample = df.shape\n",
    "gx_donors = list(df.columns)\n",
    "gx = df.values\n",
    "gene_names = list(df.index)\n",
    "\n",
    "vcfmask, exprmask = select_donors(gt_donors, gx_donors)\n",
    "genes, indices = select_genes(geneinfo, gene_names)\n",
    "dosage_masked = gtfull[:, vcfmask]\n",
    "\n",
    "snp_info, dosage = filter_snps(snpinfos, dosage_masked)\n",
    "\n",
    "gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "\n",
    "## --- Apply KNN correction\n",
    "f=1\n",
    "gx_knn, gt_knn = knn_correction(gx_norm.T, dosage, K=30, f=f)\n",
    "\n",
    "## --- Normalize GX and GT\n",
    "gt_norm, gt_cent = normalize_and_center_dosage(dosage, snp_info)\n",
    "sigmax2     = np.var(gt_cent, axis = 1)\n",
    "\n",
    "gx_knn_norm = normalize_expr(gx_knn.T)\n",
    "gt_knn_norm, gt_knn_cent = normalize_and_center_dosage(gt_knn, snp_info)\n",
    "sigmax2_knn = np.var(gt_knn_cent, axis = 1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GT = gt_knn_cent\n",
    "GX = gx_knn_norm\n",
    "Nsample = GX.shape[1]\n",
    "\n",
    "cismask = cismaskcomp[0]\n",
    "print(\"Processing cismask {:d}: {:d} genes masked for {:d} snps\".format(0, len(cismask.rmv_id), cismask.nsnp))\n",
    "usegenes = np.ones(GX.shape[0], dtype=bool)\n",
    "if cismask.rmv_id.shape[0] > 0: usegenes[cismask.rmv_id] = False\n",
    "myGX = GX[usegenes] #/ np.sqrt(Nsample)\n",
    "myGT = GT[cismask.apply2,:]\n",
    "\n",
    "Yt = myGX.T\n",
    "U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "S2  = np.square(S)\n",
    "YTY = np.matmul(myGX, Yt)\n",
    "\n",
    "# Gene number\n",
    "G = myGX.shape[0]\n",
    "\n",
    "j = 0\n",
    "print(\"sx2=\",sigmax2_knn[j])\n",
    "SIGMA_BETAS2 = np.linspace(0.02, 0.034, 10)\n",
    "# SIGMA_BETAS2 = np.linspace(0.00001, 0.0001, 20)\n",
    "\n",
    "gridsize   = SIGMA_BETAS2.shape[0]\n",
    "betax_term = np.zeros(gridsize)\n",
    "der        = np.zeros(gridsize)\n",
    "Keff_term  = np.zeros(gridsize)\n",
    "ML_sb2     = np.zeros(gridsize)\n",
    "Rscore     = np.zeros(gridsize)\n",
    "for i,sb2 in enumerate(SIGMA_BETAS2):\n",
    "    S2mod = S2 + (sigmax2_knn[j] / sb2)\n",
    "    \n",
    "    Rscore[i] = np.sum(np.square(np.dot(U.T, myGT[j,:])) * (S2 / S2mod)) / sigmax2_knn[j]\n",
    "    \n",
    "    Ig = np.identity(G) * (sigmax2_knn[j] / sb2)\n",
    "    \n",
    "##     YTY_Ig = np.log(np.linalg.norm(YTY + Ig, 1))\n",
    "##     a = np.power(2*np.pi*sb2, G/2) * YTY_Ig\n",
    "##     b = np.exp(Rscore[i] / (2*sigmax2_knn[j]))\n",
    "    L_YTY_Ig = np.linalg.cholesky(YTY + Ig)\n",
    "    det_YTY_Ig = 2*np.log(np.trace(L_YTY_Ig))\n",
    "    log_a = -(G/2)*np.log(2*np.pi*sb2) - 0.5*det_YTY_Ig\n",
    "#     log_a = (-G)*np.log(sb2) + G*np.log(sigmax2_knn[j]) - 0.5*YTY_Ig\n",
    "    \n",
    "    \n",
    "    log_b = Rscore[i] / (2*sigmax2_knn[j])\n",
    "    ML_sb2[i] = log_a + log_b\n",
    "    print(ML_sb2[i])\n",
    "\n",
    "    Keff_term[i] = np.sum(S2/S2mod)\n",
    "    betax_term[i] = np.sum(np.square(np.dot(U.T, myGT[j,:])) * (S2 / S2mod)) / sb2\n",
    "    der[i]    = ( -Keff_term[i] + betax_term[i] ) / ( 2 * sb2 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S2mod = S2 + (sigmax2_knn[j] / sb2)\n",
    "Keff_term[i] = np.sum(S2/S2mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "GT = gt_knn_cent\n",
    "GX = gx_knn_norm\n",
    "Nsample = GX.shape[1]\n",
    "\n",
    "cismask = cismaskcomp[0]\n",
    "print(\"Processing cismask {:d}: {:d} genes masked for {:d} snps\".format(0, len(cismask.rmv_id), cismask.nsnp))\n",
    "usegenes = np.ones(GX.shape[0], dtype=bool)\n",
    "if cismask.rmv_id.shape[0] > 0: usegenes[cismask.rmv_id] = False\n",
    "myGX = GX[usegenes] #/ np.sqrt(Nsample)\n",
    "myGT = GT[cismask.apply2,:]\n",
    "\n",
    "Yt = myGX.T\n",
    "U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "S2  = np.square(S)\n",
    "YTY = np.matmul(myGX, Yt)\n",
    "\n",
    "# Gene number\n",
    "G = myGX.shape[0]\n",
    "\n",
    "j = 0\n",
    "print(\"sx2=\",sigmax2_knn[j])\n",
    "SIGMA_BETAS2 = np.linspace(1.47e-07, 1.48e-7, 100)\n",
    "# SIGMA_BETAS2 = np.linspace(0.027, 0.036, 10)\n",
    "# SIGMA_BETAS2 = np.linspace(1, 2*Nsample, 100)\n",
    "\n",
    "gridsize   = SIGMA_BETAS2.shape[0]\n",
    "betax_term = np.zeros(gridsize)\n",
    "der        = np.zeros(gridsize)\n",
    "der_old    = np.zeros(gridsize)\n",
    "Keff_term  = np.zeros(gridsize)\n",
    "ML_sb2     = np.zeros(gridsize)\n",
    "Rscore     = np.zeros(gridsize)\n",
    "I = np.identity(Nsample)\n",
    "for i,sb2 in enumerate(SIGMA_BETAS2):\n",
    "    S2mod = S2 + (sigmax2_knn[j] / sb2)\n",
    "    Rscore[i] = np.sum(np.square(np.dot(U.T, myGT[j,:])) * (S2 / S2mod)) / sigmax2_knn[j]\n",
    "    ML_sb2[i] = -0.5*np.sum(np.log( S2 * (sb2 / sigmax2_knn[j]) + 1 )) + 0.5*Rscore[i]\n",
    "    \n",
    "#     # Gradient check, it matches\n",
    "#     delta = 0.000001\n",
    "#     newsb2 = sb2 + delta\n",
    "#     newS2mod = S2 + (sigmax2_knn[j] / newsb2)\n",
    "#     newR = np.sum(np.square(np.dot(U.T, myGT[j,:])) * (S2 / S2mod)) / sigmax2_knn[j]\n",
    "#     newM = -0.5*np.sum(np.log( S2 * (newsb2 / sigmax2_knn[j]) + 1 )) + 0.5*newR\n",
    "#     M_grad = (newM - ML_sb2[i]) / delta\n",
    "#     print(M_grad)\n",
    "        \n",
    "    denom = (S2 * sb2 + sigmax2_knn[j])\n",
    "    der[i] = 0.5* np.sum( ( S2 / denom ) * ( (np.square(np.dot(U.T, myGT[j,:])) / denom ) - 1 ) )\n",
    "\n",
    "#     ## -- der_old below is equal to der above, good! it matches!\n",
    "#     Keff_term[i]  = np.sum(S2/S2mod)\n",
    "#     ## alternative check for beta term\n",
    "#     innerLinv_ST = (S*I / (S2*I + (sigmax2_knn[j]/sb2))) * I\n",
    "#     UT = np.transpose(U)\n",
    "#     V = np.transpose(Vt)\n",
    "#     innerLinv_STUT = np.matmul(innerLinv_ST, UT)\n",
    "#     A2 = np.matmul(V, innerLinv_STUT)\n",
    "#     Bi = np.matmul(A2, myGT[j,:][np.newaxis].T)\n",
    "#     B2 = np.dot(Bi.T, Bi)\n",
    "\n",
    "#     betax_term[i] = B2 / sb2\n",
    "    # der_old[i]    = ( -Keff_term[i] + betax_term[i] ) / ( 2 * sb2 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.optimize import minimize\n",
    "\n",
    "\n",
    "GT = gt_knn_cent\n",
    "GX = gx_knn_norm\n",
    "Nsample = GX.shape[1]\n",
    "\n",
    "cismask = cismaskcomp[0]\n",
    "print(\"Processing cismask {:d}: {:d} genes masked for {:d} snps\".format(0, len(cismask.rmv_id), cismask.nsnp))\n",
    "usegenes = np.ones(GX.shape[0], dtype=bool)\n",
    "if cismask.rmv_id.shape[0] > 0: usegenes[cismask.rmv_id] = False\n",
    "myGX = GX[usegenes] #/ np.sqrt(Nsample)\n",
    "myGT = GT[cismask.apply2,:]\n",
    "\n",
    "\n",
    "class SBoptimizer:\n",
    "\n",
    "    def __init__(self, GT, GX, sx2):\n",
    "\n",
    "        self._GT  = np.ascontiguousarray(GT)\n",
    "        self._GX  = np.ascontiguousarray(GX)\n",
    "        self._sx2 = np.ascontiguousarray(sx2)\n",
    "        self._nsnps = GT.shape[0]\n",
    "        self._nsample = GX.shape[1]\n",
    "        \n",
    "        U, S, VT = np.linalg.svd(GX.T)\n",
    "        self._S = S\n",
    "        self._U = U\n",
    "        self._S2 = np.square(S)\n",
    "        self._opt_sb2 = np.zeros(self._nsnps)\n",
    "    \n",
    "    @property\n",
    "    def sb2(self):\n",
    "        return self._opt_sb2\n",
    "\n",
    "    def get_ML(self, _sb2, i):\n",
    "        # sb2 = sb * sb\n",
    "        sb2 = np.exp(_sb2)\n",
    "        S2mod = self._S2 + (self._sx2[i] / sb2)\n",
    "        Rscore = np.sum(np.square(np.dot(self._U.T, self._GT[i,:])) * (self._S2 / S2mod)) / self._sx2[i]\n",
    "        MLL = -0.5*np.sum(np.log( self._S2 * (sb2 / self._sx2[i]) + 1 )) + 0.5*Rscore\n",
    "\n",
    "        denom = (self._S2 * sb2 + self._sx2[i])\n",
    "        der = 0.5* np.sum( ( self._S2 / denom ) * ( (np.square(np.dot(self._U.T, self._GT[i,:])) / denom ) - 1 ) )\n",
    "        return -MLL, sb2*np.array([-der])\n",
    "\n",
    "    def fit(self):\n",
    "        st = time.time()\n",
    "        \n",
    "        sb_init = np.exp(0.01)\n",
    "        for i in range(self._nsnps):\n",
    "            res = minimize(   self.get_ML,\n",
    "                              sb_init, \n",
    "                              args = i,\n",
    "                              method='L-BFGS-B',\n",
    "                              jac = True,\n",
    "                              #bounds = [[0,1]],\n",
    "                              options={'maxiter': 200000,\n",
    "                                       'maxfun': 2000000,\n",
    "                                       #'ftol': 1e-9,\n",
    "                                       #'gtol': 1e-9,\n",
    "                                       'disp': True})\n",
    "\n",
    "            print(res)\n",
    "            self._opt_sb2[i] = np.exp(res.x[0])\n",
    "        et = time.time()\n",
    "        print(\"optimization took in total: \",et-st)\n",
    "        \n",
    "# GT was subsetted above\n",
    "SBOPT = SBoptimizer(myGT, myGX, sigmax2_knn[cismask.apply2])\n",
    "        \n",
    "# Yt = myGX.T\n",
    "# U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "# S2  = np.square(S)\n",
    "\n",
    "# def get_ML(sb2, S, S2, U, GT, sx2, i):\n",
    "#     S2mod = S2 + (sx2[i] / sb2)\n",
    "#     Rscore = np.sum(np.square(np.dot(U.T, GT[i,:])) * (S2 / S2mod)) / sx2[i]\n",
    "#     MLL = -0.5*np.sum(np.log( S2 * (sb2 / sx2[i]) + 1 )) + 0.5*Rscore\n",
    "            \n",
    "#     denom = (S2 * sb2 + sx2[i])\n",
    "#     der = 0.5* np.sum( ( S2 / denom ) * ( (np.square(np.dot(U.T, GT[i,:])) / denom ) - 1 ) )\n",
    "#     return -MLL, -der\n",
    "\n",
    "# lml_min = optimize.minimize(self._log_marginal_likelihood,\n",
    "#                                         scaledparams,\n",
    "#                                         method='L-BFGS-B',\n",
    "#                                         jac=True,\n",
    "#                                         bounds=bounds,\n",
    "#                                         callback=self._callback_zstates,\n",
    "#                                         options={'maxiter': 200000,\n",
    "#                                                  'maxfun': 2000000,\n",
    "#                                                  'ftol': 1e-9,\n",
    "#                                                  'gtol': 1e-9,\n",
    "#                                                  'disp': True})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "SBOPT.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SBOPT.sb2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(np.sqrt(SBOPT.sb2), bins=20)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(14,8))\n",
    "ax1  = fig.add_subplot(121)\n",
    "ax2  = fig.add_subplot(122)\n",
    "\n",
    "SIGMA_BETAS = SIGMA_BETAS2\n",
    "ax1.scatter(SIGMA_BETAS,  ML_sb2, label=\"log_ML\", c=\"blue\")\n",
    "ax1.scatter(SIGMA_BETAS,  der, label=\"log_derivative\", c='red')\n",
    "ax1.axhline(0, ls=\"--\", c=\"red\")\n",
    "ax1.set_xlabel(r\"$\\sigma_{\\beta}^2$\")\n",
    "ax1.set_xlim(min(SIGMA_BETAS), max(SIGMA_BETAS))\n",
    "# ax1.set_xlim(0.025,0.03)\n",
    "# ax1.set_ylim(0,10)\n",
    "\n",
    "ax2.scatter(SIGMA_BETAS, Keff_term, label=r\"$K_{eff}$\")\n",
    "ax2.scatter(SIGMA_BETAS, betax_term, label=r\"$ \\beta $\")\n",
    "ax2.axhline(GX.shape[1], c=\"red\")\n",
    "ax2.set_xlabel(r\"$\\sigma_{\\beta}^2$\")\n",
    "ax2.set_xlim(min(SIGMA_BETAS), max(SIGMA_BETAS))\n",
    "ax2.legend()\n",
    "ax2.set_ylim(0,1000)\n",
    "\n",
    "ax1.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(der)\n",
    "print(ML_sb2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Singular values of some tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_tissues = [\"as\", \"ms\", \"sse\", \"ag\", \"wb\"]\n",
    "\n",
    "fig = plt.figure(figsize=(14,10))\n",
    "ax1 = fig.add_subplot(111)\n",
    "\n",
    "for tissue in select_tissues:\n",
    "    # df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "    df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_cclm.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "    ngene, nsample = df.shape\n",
    "    gx_donors = list(df.columns)\n",
    "    gx = df.values\n",
    "    gene_names = list(df.index)\n",
    "    \n",
    "    vcfmask, exprmask = select_donors(gt_donors, gx_donors)\n",
    "    genes, indices = select_genes(geneinfo, gene_names)\n",
    "    dosage_masked = gtfull[:, vcfmask]\n",
    "\n",
    "    snp_info, dosage = filter_snps(snpinfos, dosage_masked)\n",
    "\n",
    "    gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "\n",
    "    ## --- Apply KNN correction\n",
    "    f=1\n",
    "    gx_knn, gt_knn = knn_correction(gx_norm.T, dosage, K=30, f=f)\n",
    "\n",
    "    ## --- Normalize GX and GT\n",
    "    gt_norm, gt_cent = normalize_and_center_dosage(dosage, snp_info)\n",
    "    sigmax2     = np.var(gt_cent, axis = 1)\n",
    "\n",
    "    gx_knn_norm = normalize_expr(gx_knn.T)\n",
    "    gt_knn_norm, gt_knn_cent = normalize_and_center_dosage(gt_knn, snp_info)\n",
    "    sigmax2_knn = np.var(gt_knn_cent, axis = 1)\n",
    "    \n",
    "    Y = gx_knn_norm # / np.sqrt(nsample)\n",
    "    Yt = Y.T\n",
    "    U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "    ax1.scatter(np.arange(len(S)-1), S[:-1], label=tissue, alpha=0.4)\n",
    "    \n",
    "ax1.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Without KNN\n",
    "select_tissues = [\"as\", \"ms\", \"sse\", \"ag\", \"wb\"]\n",
    "\n",
    "fig = plt.figure(figsize=(14,10))\n",
    "ax1 = fig.add_subplot(111)\n",
    "\n",
    "for tissue in select_tissues:\n",
    "    # df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "    df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_cclm.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "    ngene, nsample = df.shape\n",
    "    gx_donors = list(df.columns)\n",
    "    gx = df.values\n",
    "    gene_names = list(df.index)\n",
    "    \n",
    "    vcfmask, exprmask = select_donors(gt_donors, gx_donors)\n",
    "    genes, indices = select_genes(geneinfo, gene_names)\n",
    "    dosage_masked = gtfull[:, vcfmask]\n",
    "\n",
    "    snp_info, dosage = filter_snps(snpinfos, dosage_masked)\n",
    "\n",
    "    gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "\n",
    "    ## --- Normalize GX and GT\n",
    "    gt_norm, gt_cent = normalize_and_center_dosage(dosage, snp_info)\n",
    "    sigmax2     = np.var(gt_cent, axis = 1)\n",
    "   \n",
    "    Y = gx_norm # / np.sqrt(nsample)\n",
    "    Yt = Y.T\n",
    "    U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "    ax1.scatter(np.arange(len(S)-1), S[:-1], label=tissue, alpha=0.4)\n",
    "    \n",
    "ax1.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "## What's the usual Keff in other tissues?\n",
    "select_tissues = [\"as\", \"ms\", \"sse\", \"ag\", \"wb\"]\n",
    "select_tissues = [\"ms\", \"wb\"]\n",
    "\n",
    "fig = plt.figure(figsize=(14,10))\n",
    "ax1 = fig.add_subplot(121)\n",
    "ax2 = fig.add_subplot(122)\n",
    "\n",
    "def get_contribs(tissue, gx_norm, dosage, snp_info, ax1, ax2, keff_target=0.5, K=30, f=1):\n",
    "    ## --- Apply KNN correction\n",
    "    gx_knn, gt_knn = knn_correction(gx_norm.T, dosage, K=K, f=f)\n",
    "\n",
    "    ## --- Normalize GX and GT\n",
    "    gt_norm, gt_cent = normalize_and_center_dosage(dosage, snp_info)\n",
    "    sigmax2     = np.var(gt_cent, axis = 1)\n",
    "\n",
    "    gx_knn_norm = normalize_expr(gx_knn.T)\n",
    "    gt_knn_norm, gt_knn_cent = normalize_and_center_dosage(gt_knn, snp_info)\n",
    "    sigmax2_knn = np.var(gt_knn_cent, axis = 1)\n",
    "    \n",
    "    Y = gx_knn_norm # / np.sqrt(nsample)\n",
    "    Yt = Y.T\n",
    "    U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "    \n",
    "    sb2_opt = optimize_sb2(S, sigmax2_knn, keff_target)\n",
    "    S2 = np.square(S)\n",
    "    sb2 = sb2_opt # np.square(DY_SIGMA_BETAS10)\n",
    "    S2mod = S2 + (sigmax2_knn[0] / sb2[0])\n",
    "    contribs = S2/S2mod\n",
    "    \n",
    "    L=int(1 * len(S2)) \n",
    "    N = len(S2)\n",
    "    print(tissue, np.sum(contribs[:L]), np.sum(contribs[:L]) / N)\n",
    "    ax1.scatter(np.arange(len(S)), contribs, s=3, label=\"{:s}_{:d}_{:f}_{:f}\".format(tissue, K, keff_target, f), alpha=0.4)\n",
    "    ax2.scatter(np.arange(len(S)) / N, np.cumsum(contribs) / N, s=3, label=\"{:s}_{:d}_{:f}_{:f}\".format(tissue, K, keff_target, f), alpha=0.4)\n",
    "\n",
    "for tissue in select_tissues:\n",
    "    df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "    ngene, nsample = df.shape\n",
    "    gx_donors = list(df.columns)\n",
    "    gx = df.values\n",
    "    gene_names = list(df.index)\n",
    "    \n",
    "    vcfmask, exprmask = select_donors(gt_donors, gx_donors)\n",
    "    genes, indices = select_genes(geneinfo, gene_names)\n",
    "    dosage_masked = gtfull[:, vcfmask]\n",
    "\n",
    "    snp_info, dosage = filter_snps(snpinfos, dosage_masked)\n",
    "\n",
    "    gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "    get_contribs(tissue, gx_norm, dosage, snp_info, ax1, ax2, keff_target=0.7,  K=30, f=1)\n",
    "#     get_contribs(tissue, gx_norm, dosage, snp_info, ax1, ax2, keff_target=0.8,  K=30, f=0.1)\n",
    "    \n",
    "## Add Framingham\n",
    "df = pd.read_csv(\"/cbscratch/franco/datasets/FHS/expression/fhs.formatted.expr.one_pedno_rep.txt\", header=0, index_col=0, sep=\"\\t\")\n",
    "ngene, nsample = df.shape\n",
    "gx_donors = list(df.columns)\n",
    "gx = df.values\n",
    "gene_names = list(df.index)\n",
    "print(\"Loaded FHS \", gx.shape)\n",
    "\n",
    "# Use real genotype\n",
    "chrm=12\n",
    "f_vcf = \"/cbscratch/franco/datasets/FHS/genotypes/vcfs/chr{:d}.fhs.vcf.gz\".format(chrm)\n",
    "Fgtfull, Fsnpinfos, Fgt_donors = read_vcf(f_vcf, 0, 2000)\n",
    "\n",
    "vcfmask, exprmask = select_donors(Fgt_donors, gx_donors)\n",
    "genes, indices = select_genes(geneinfo, gene_names)\n",
    "dosage_masked = Fgtfull[:, vcfmask]\n",
    "\n",
    "snp_info, dosage = filter_snps(Fsnpinfos, dosage_masked)\n",
    "\n",
    "gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "get_contribs(\"fhs\", gx_norm, dosage, snp_info, ax1, ax2, keff_target=0.7,  K=30, f=1)\n",
    "\n",
    "    \n",
    "ax1.legend()\n",
    "ax2.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## What's the usual Keff in other tissues?\n",
    "select_tissues = [\"as\", \"ms\", \"sse\", \"ag\", \"wb\"]\n",
    "\n",
    "fig = plt.figure(figsize=(14,10))\n",
    "ax1 = fig.add_subplot(121)\n",
    "ax2 = fig.add_subplot(122)\n",
    "\n",
    "for tissue in select_tissues:\n",
    "    df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/{:s}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\".format(tissue), header=0, index_col=0, sep=\"\\t\")\n",
    "    ngene, nsample = df.shape\n",
    "    gx_donors = list(df.columns)\n",
    "    gx = df.values\n",
    "    gene_names = list(df.index)\n",
    "    \n",
    "    vcfmask, exprmask = select_donors(gt_donors, gx_donors)\n",
    "    genes, indices = select_genes(geneinfo, gene_names)\n",
    "    dosage_masked = gtfull[:, vcfmask]\n",
    "\n",
    "    snp_info, dosage = filter_snps(snpinfos, dosage_masked)\n",
    "\n",
    "    gx_norm = normalize_expr( gx[:, exprmask] )\n",
    "\n",
    "    ## --- Apply KNN correction\n",
    "    f=1\n",
    "    gx_knn, gt_knn = knn_correction(gx_norm.T, dosage, K=30, f=f)\n",
    "\n",
    "    ## --- Normalize GX and GT\n",
    "    gt_norm, gt_cent = normalize_and_center_dosage(dosage, snp_info)\n",
    "    sigmax2     = np.var(gt_cent, axis = 1)\n",
    "\n",
    "    gx_knn_norm = normalize_expr(gx_knn.T)\n",
    "    gt_knn_norm, gt_knn_cent = normalize_and_center_dosage(gt_knn, snp_info)\n",
    "    sigmax2_knn = np.var(gt_knn_cent, axis = 1)\n",
    "    \n",
    "    Y = gx_knn_norm # / np.sqrt(nsample)\n",
    "    Yt = Y.T\n",
    "    U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "    s2_median = np.median(np.square(S))\n",
    "    s2_10     = np.percentile(np.square(S), 60)\n",
    "    \n",
    "    SIGMA_BETAS = np.repeat(0.1, gt_cent.shape[0])\n",
    "    DY_SIGMA_BETAS = np.sqrt( sigmax2 / s2_median )\n",
    "    DY_SIGMA_BETAS10 = np.sqrt( sigmax2 / s2_10 )\n",
    "    avg_sb_m  = np.mean(DY_SIGMA_BETAS)\n",
    "    avg_sb_10 = np.mean(DY_SIGMA_BETAS10)\n",
    "\n",
    "    S2 = np.square(S)\n",
    "    sb2 = np.square(DY_SIGMA_BETAS10)\n",
    "    S2mod = S2 + (sigmax2_knn[0] / sb2[0])\n",
    "    contribs = S2/S2mod\n",
    "    \n",
    "    L=int(0.1 * len(S2)) \n",
    "    print(tissue, np.sum(contribs[:L]), np.sum(contribs[:L]) / len(S2))\n",
    "    ax1.scatter(np.arange(len(S)-1), contribs[:-1], label=tissue, alpha=0.4)\n",
    "    ax2.scatter(np.arange(len(S)-1), np.log10(contribs[:-1]), label=tissue, alpha=0.4)\n",
    "    \n",
    "ax1.legend()\n",
    "ax2.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick and dirty DHS enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dhs_dict = collections.defaultdict(list)\n",
    "pcutoffs = [5e-08, 1e-10, 1e-15, 1e-20, 1e-25, 1e-30, 1e-35]\n",
    "title = \"multi_tissue\"\n",
    "for pcutoff in pcutoffs:\n",
    "    dhsfile = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_opt_sb2_EUR/raw/summary_5e-08/dhs_enrichments/dhs_enrichment_{:s}_{:g}.txt\".format(title, pcutoff)\n",
    "    with open(dhsfile) as instream:\n",
    "        next(instream)\n",
    "        for line in instream:\n",
    "            arr = line.strip().split(\"\\t\")\n",
    "            tissue = arr[0]\n",
    "            enrichment = float(arr[5])\n",
    "            dhs_dict[tissue].append(enrichment)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dhs_dict = collections.defaultdict(list)\n",
    "# keffs = [\"0.5\", \"0.6\", \"0.7\", \"0.8\"]\n",
    "# for keff in keffs:\n",
    "#     dhsfile = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_dysb/summary_5e-08_k{:s}/dhs_enrichments/dhs_enrichment_multi_tissue_keff{:s}.txt\".format(keff, keff)\n",
    "#     with open(dhsfile) as instream:\n",
    "#         next(instream)\n",
    "#         for line in instream:\n",
    "#             arr = line.strip().split(\"\\t\")\n",
    "#             tissue = arr[0]\n",
    "#             enrichment = float(arr[5])\n",
    "#             dhs_dict[tissue].append(enrichment)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(14,8))\n",
    "ax  = fig.add_subplot(111)\n",
    "for t in dhs_dict:\n",
    "    while len(dhs_dict[t]) != len(pcutoffs):\n",
    "        dhs_dict[t] = np.append(dhs_dict[t],0.0)\n",
    "    ax.plot([np.log10(x) for x in pcutoffs], dhs_dict[t], label=t)\n",
    "    ax.set_title(\"ALL SAMPLES\")\n",
    "    ax.set_ylabel(\"enrichment\")\n",
    "    ax.set_xlabel(\"log p-val\")\n",
    "ax.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(14,8))\n",
    "ax  = fig.add_subplot(111)\n",
    "for t in dhs_dict:\n",
    "    while len(dhs_dict[t]) != len(pcutoffs):\n",
    "        dhs_dict[t] = np.append(dhs_dict[t],0.0)\n",
    "    ax.plot([np.log10(x) for x in pcutoffs], dhs_dict[t], label=t)\n",
    "    ax.set_title(\"ONLY EUR\")\n",
    "    ax.set_ylabel(\"enrichment\")\n",
    "    ax.set_xlabel(\"log p-val\")\n",
    "ax.legend()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
