{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import sys, os\n",
    "import numpy as np\n",
    "sys.path.append('../')\n",
    "sys.path.append(\"../../../tejaas/\")\n",
    "from iotools.readRPKM import ReadRPKM\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip\n",
    "from utils import utils\n",
    "# from gtex_normalization import normalize_expression\n",
    "\n",
    "def centerscale_expr(Y):\n",
    "    return (Y - np.mean(Y, axis = 1).reshape(-1, 1)) / np.std(Y, axis = 1).reshape(-1, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### correct covariates BEFORE normalization, directly on the RPKMs\n",
    "\n",
    "def get_donors(path):\n",
    "    donor_ids = list()\n",
    "    with open(path, 'r') as instream:\n",
    "        # skip first two lines\n",
    "        next(instream)\n",
    "        next(instream)\n",
    "        for line in instream:\n",
    "            donor_ids.append(line.strip().split()[0])\n",
    "    return donor_ids\n",
    "\n",
    "def read_gct(gct_file, donor_ids):\n",
    "    \"\"\"\n",
    "    Load GCT as DataFrame\n",
    "    \"\"\"    \n",
    "    df = pd.read_csv(gct_file, sep='\\t', skiprows=2, index_col=0)\n",
    "    df.drop('Description', axis=1, inplace=True)\n",
    "    df.index.name = 'gene_id'\n",
    "    df = df[[i for i in df.columns if '-'.join(i.split('-')[:2]) in donor_ids]]\n",
    "    return df\n",
    "\n",
    "### this is in the gtex_normalization.py file\n",
    "import scipy.stats as stats\n",
    "def normalize_quantiles(df):\n",
    "    \"\"\"\n",
    "    Note: replicates behavior of R function normalize.quantiles from library(\"preprocessCore\")  \n",
    "\n",
    "    Reference:\n",
    "     [1] Bolstad et al., Bioinformatics 19(2), pp. 185-193, 2003\n",
    "    \n",
    "    Adapted from https://github.com/andrewdyates/quantile_normalize\n",
    "    \"\"\"\n",
    "    \n",
    "    M = df.values.copy()\n",
    "    \n",
    "    Q = M.argsort(axis=0)\n",
    "    m,n = M.shape\n",
    "\n",
    "    # compute quantile vector\n",
    "    quantiles = np.zeros(m)\n",
    "    for i in range(n):\n",
    "        quantiles += M[Q[:,i],i]\n",
    "    quantiles = quantiles / n\n",
    "    \n",
    "    for i in range(n):\n",
    "        # Get equivalence classes; unique values == 0\n",
    "        dupes = np.zeros(m, dtype=np.int)\n",
    "        for j in range(m-1):\n",
    "            if M[Q[j,i],i]==M[Q[j+1,i],i]:\n",
    "                dupes[j+1] = dupes[j]+1\n",
    "                \n",
    "        # Replace column with quantile ranks\n",
    "        M[Q[:,i],i] = quantiles\n",
    "\n",
    "        # Average together equivalence classes\n",
    "        j = m-1\n",
    "        while j >= 0:\n",
    "            if dupes[j] == 0:\n",
    "                j -= 1\n",
    "            else:\n",
    "                idxs = Q[j-dupes[j]:j+1,i]\n",
    "                M[idxs,i] = np.median(M[idxs,i])\n",
    "                j -= 1 + dupes[j]\n",
    "        assert j == -1\n",
    "    \n",
    "    return M\n",
    "      \n",
    "    \n",
    "def inverse_quantile_normalization(M):\n",
    "    \"\"\"\n",
    "    After quantile normalization of samples, standardize expression of each gene\n",
    "    \"\"\"\n",
    "    R = stats.mstats.rankdata(M,axis=1)  # ties are averaged\n",
    "    Q = stats.norm.ppf(R/(M.shape[1]+1))\n",
    "    return Q\n",
    "        \n",
    "        \n",
    "# def normalize_expression(expression_df, counts_df, expression_threshold=0.1, count_threshold=5, min_samples=10):\n",
    "#     \"\"\"\n",
    "#     Genes are thresholded based on the following expression rules:\n",
    "#       >=min_samples with >expression_threshold expression values\n",
    "#       >=min_samples with >count_threshold read counts\n",
    "#     \"\"\"\n",
    "#     # donor_ids = ['-'.join(i.split('-')[:2]) for i in expression_df.columns]\n",
    "#     donor_ids = expression_df.columns\n",
    "    \n",
    "#     # expression thresholds\n",
    "#     mask = ((np.sum(expression_df>expression_threshold,axis=1)>=min_samples) & (np.sum(counts_df>count_threshold,axis=1)>=min_samples)).values\n",
    "    \n",
    "#     # apply normalization\n",
    "#     M = normalize_quantiles(expression_df.loc[mask])\n",
    "#     R = inverse_quantile_normalization(M)\n",
    "\n",
    "#     quant_std_df = pd.DataFrame(data=R, columns=donor_ids, index=expression_df.loc[mask].index)    \n",
    "#     quant_df = pd.DataFrame(data=M, columns=donor_ids, index=expression_df.loc[mask].index)\n",
    "#     return quant_std_df, quant_df\n",
    "\n",
    "\n",
    "### Apply LM correction (python)\n",
    "from sklearn import linear_model\n",
    "\n",
    "def lmcorrect(expression_df, cov_df):   \n",
    "    donor_ids = expression_df.columns\n",
    "\n",
    "    #sort donors\n",
    "    cov_df = cov_df[expression_df.columns]\n",
    "\n",
    "    reg = linear_model.LinearRegression()\n",
    "    reg.fit(cov_df.T, expression_df.T)\n",
    "\n",
    "    # reg.score(df_cov.T, crop_expression_df.T)\n",
    "    # print(reg.coef_)\n",
    "    residuals = expression_df - reg.predict(cov_df.T).T\n",
    "    return residuals, reg.coef_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/main/tissues.txt\"\n",
    "tissuenames, descriptions = utils.read_tissues(tissue_file)\n",
    "# tissues = [\"gtex-\"+t for t in tissuenames]\n",
    "\n",
    "# tissue=\"sse\"\n",
    "for tissue in tissuenames:\n",
    "    print(tissue)\n",
    "    rpkmpath=\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/rpkms2/{:s}_rpkm.gct\".format(tissue)\n",
    "    countspath=\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/rpkms2/{:s}_counts.gct\".format(tissue)\n",
    "    donorspath =\"/cbscratch/franco/datasets/gtex/gtex.sample\"\n",
    "\n",
    "    expression_threshold=0.1    # 'Selects genes with > expression_threshold expression in at least min_samples')\n",
    "    count_threshold=5,          # 'Selects genes with > count_threshold reads in at least min_samples')\n",
    "    min_samples=10              # 'Minimum number of samples that must satisfy thresholds')\n",
    "\n",
    "    donor_ids = get_donors(donorspath)\n",
    "    expression_df = read_gct(rpkmpath, donor_ids)\n",
    "    counts_df = read_gct(countspath, donor_ids)\n",
    "\n",
    "    if expression_df.shape[1] < min_samples:\n",
    "        raise ValueError(\"tissue has less samples than threshold\")\n",
    "\n",
    "    expr_ids = list(expression_df.columns)\n",
    "    tissue_counts_df = counts_df.loc[:,expr_ids]\n",
    "\n",
    "    # load covariates\n",
    "    covfile = \"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/covariates/{:s}_nopeer_covariates_w_age.txt\".format(tissue)\n",
    "    df_cov = pd.read_table(covfile, header=0, index_col=0)\n",
    "\n",
    "    expr_corr_rpkms, coefs = lmcorrect_rpkms(expression_df, counts_df, df_cov)\n",
    "\n",
    "    expr_precov_corr_norm = centerscale_expr(inverse_quantile_normalization(expr_corr_rpkms))\n",
    "    expr_precov_corr_norm_df = pd.DataFrame(expr_precov_corr_norm, columns=expr_corr_rpkms.columns, index=expr_corr_rpkms.index)\n",
    "    expr_precov_corr_norm_df.index.name = \"gene_id\"\n",
    "\n",
    "    expr_precov_corr_norm_df.to_csv(\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/lmcorrected/{:s}_age_lmcorrected.precorr.txt\".format(tissue), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "\n",
    "def prepare_expression(counts_df, tpm_df, sample_frac_threshold=0.2, count_threshold=6, tpm_threshold=0.1, mode='tmm'):\n",
    "    \"\"\"\n",
    "    Genes are thresholded based on the following expression rules:\n",
    "      TPM >= tpm_threshold in >= sample_frac_threshold*samples\n",
    "      read counts >= count_threshold in sample_frac_threshold*samples\n",
    "    \n",
    "    vcf_lookup: lookup table mapping sample IDs to VCF IDs\n",
    "    \n",
    "    Between-sample normalization modes:\n",
    "      tmm: TMM from edgeR\n",
    "      qn:  quantile normalization\n",
    "    \"\"\"\n",
    "\n",
    "    ns = tpm_df.shape[1]\n",
    "\n",
    "    # expression thresholds\n",
    "    if mode.lower()=='tmm':\n",
    "        mask = (\n",
    "            (np.sum(tpm_df>=tpm_threshold,axis=1)>=sample_frac_threshold*ns) &\n",
    "            (np.sum(counts_df>=count_threshold,axis=1)>=sample_frac_threshold*ns)\n",
    "        ).values\n",
    "    else:\n",
    "        mask = (\n",
    "            (np.sum(tpm_df>=tpm_threshold,axis=1)>=10) &\n",
    "            (np.sum(counts_df>=count_threshold,axis=1)>=10)\n",
    "        ).values\n",
    "\n",
    "    # apply normalization\n",
    "    if mode.lower()=='tmm':\n",
    "        tmm_counts_df = edgeR_cpm(counts_df, normalized_lib_sizes=True)\n",
    "        norm_df = inverse_quantile_normalization(tmm_counts_df[mask])\n",
    "    elif mode.lower()=='qn':\n",
    "        qn_df = normalize_quantiles(tpm_df.loc[mask])\n",
    "        norm_df = inverse_quantile_normalization(qn_df)       \n",
    "    else:\n",
    "        raise ValueError('Unsupported mode {}'.format(mode))\n",
    "\n",
    "    tmm_norm_df = pd.DataFrame(norm_df, columns=expression_df.columns, index=expression_df.index[mask])\n",
    "    tmm_norm_df.index.name = \"gene_id\"\n",
    "    \n",
    "    return tmm_norm_df\n",
    "\n",
    "def edgeR_calcNormFactors(counts_df, ref=None, logratio_trim=0.3, sum_trim=0.05, acutoff=-1e10, verbose=False):\n",
    "    \"\"\"\n",
    "    Calculate TMM (Trimmed Mean of M values) normalization.\n",
    "    Reproduces edgeR::calcNormFactors.default\n",
    "    Scaling factors for the library sizes that minimize the log-fold changes between the samples for most genes.\n",
    "    Effective library size: TMM scaling factor * library size\n",
    "    References:\n",
    "     [1] Robinson & Oshlack, 2010\n",
    "     [2] R functions:\n",
    "          edgeR::calcNormFactors.default\n",
    "          edgeR:::.calcFactorWeighted\n",
    "          edgeR:::.calcFactorQuantile\n",
    "    \"\"\"\n",
    "\n",
    "    # discard genes with all-zero counts\n",
    "    Y = counts_df.values.copy()\n",
    "    allzero = np.sum(Y>0,axis=1)==0\n",
    "    if np.any(allzero):\n",
    "        Y = Y[~allzero,:]\n",
    "\n",
    "    # select reference sample\n",
    "    if ref is None:  # reference sample index\n",
    "        f75 = np.percentile(Y/np.sum(Y,axis=0), 75, axis=0)\n",
    "        ref = np.argmin(np.abs(f75-np.mean(f75)))\n",
    "        if verbose:\n",
    "            print('Reference sample index: '+str(ref))\n",
    "\n",
    "    N = np.sum(Y, axis=0)  # total reads in each library\n",
    "\n",
    "    # with np.errstate(divide='ignore'):\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter('ignore')\n",
    "        logR = np.log2((Y/N).T / (Y[:,ref]/N[ref])).T  # log fold change; Mg in [1]\n",
    "        absE = 0.5*(np.log2(Y/N).T + np.log2(Y[:,ref]/N[ref])).T  # average log relative expression; Ag in [1]\n",
    "        v = (N-Y)/N/Y\n",
    "        v = (v.T + v[:,ref]).T  # w in [1]\n",
    "\n",
    "    ns = Y.shape[1]\n",
    "    tmm = np.zeros(ns)\n",
    "    for i in range(ns):\n",
    "        fin = np.isfinite(logR[:,i]) & np.isfinite(absE[:,i]) & (absE[:,i] > acutoff)\n",
    "        n = np.sum(fin)\n",
    "\n",
    "        loL = np.floor(n*logratio_trim)+1\n",
    "        hiL = n + 1 - loL\n",
    "        loS = np.floor(n*sum_trim)+1\n",
    "        hiS = n + 1 - loS\n",
    "        rankR = stats.rankdata(logR[fin,i])\n",
    "        rankE = stats.rankdata(absE[fin,i])\n",
    "        keep = (rankR >= loL) & (rankR <= hiL) & (rankE >= loS) & (rankE <= hiS)\n",
    "        # in [1], w erroneously defined as 1/v ?\n",
    "        tmm[i] = 2**(np.nansum(logR[fin,i][keep]/v[fin,i][keep]) / np.nansum(1/v[fin,i][keep]))\n",
    "\n",
    "    tmm = tmm / np.exp(np.mean(np.log(tmm)))\n",
    "    return tmm\n",
    "\n",
    "\n",
    "def edgeR_cpm(counts_df, tmm=None, normalized_lib_sizes=True):\n",
    "    \"\"\"\n",
    "    Return edgeR normalized/rescaled CPM (counts per million)\n",
    "    Reproduces edgeR::cpm.DGEList\n",
    "    \"\"\"\n",
    "    lib_size = counts_df.sum(axis=0)\n",
    "    if normalized_lib_sizes:\n",
    "        if tmm is None:\n",
    "            tmm = edgeR_calcNormFactors(counts_df)\n",
    "        lib_size = lib_size * tmm\n",
    "    return counts_df / lib_size * 1e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading counts\n",
      "reading expr\n",
      "  * Converting to TPM\n"
     ]
    }
   ],
   "source": [
    "tissue = \"ms\"\n",
    "rpkmpath=\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/rpkms2/{:s}_rpkm.gct\".format(tissue)\n",
    "countspath=\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/rpkms2/{:s}_counts.gct\".format(tissue)\n",
    "donorspath =\"/cbscratch/franco/datasets/gtex/gtex.sample\"\n",
    "\n",
    "donor_ids = get_donors(donorspath)\n",
    "print(\"reading counts\")\n",
    "counts_df = read_gct(countspath, donor_ids)\n",
    "\n",
    "print(\"reading expr\")\n",
    "# tpm_df = read_gct(args.tpm_gct, sample_ids) # from gtex-pipeline\n",
    "expression_df = read_gct(rpkmpath, donor_ids) # my pipeline\n",
    "\n",
    "convert_tpm = True\n",
    "if convert_tpm:\n",
    "    print('  * Converting to TPM')\n",
    "    tpm_expr_df = expression_df/expression_df.sum(0)*1e6\n",
    "\n",
    "# Generate normalized expression BED files for eQTL analyses\n",
    "convert_tpm=True    #, help='Convert to TPM (in case input is in RPKM/FPKM)')\n",
    "tpm_threshold=0.1   #, help='Selects genes with > expression_threshold expression in at least sample_frac_threshold')\n",
    "count_threshold=6   #, help='Selects genes with >= count_threshold reads in at least sample_frac_threshold samples')\n",
    "sample_frac_threshold=0.2  #, help='Minimum fraction of samples that must satisfy thresholds')\n",
    "normalization_method='tmm' #, help='Normalization method: TMM or quantile normalization (qn)')\n",
    "\n",
    "norm_tmm_df = prepare_expression(counts_df, tpm_expr_df, \n",
    "                             sample_frac_threshold=sample_frac_threshold,\n",
    "                             count_threshold=count_threshold, tpm_threshold=tpm_threshold, \n",
    "                             mode=normalization_method)\n",
    "\n",
    "normalization_method='qn'\n",
    "norm_qn_df = prepare_expression(counts_df, expression_df, \n",
    "                             sample_frac_threshold=sample_frac_threshold,\n",
    "                             count_threshold=count_threshold, tpm_threshold=tpm_threshold, \n",
    "                             mode=normalization_method)\n",
    "\n",
    "newcolumns = [\"-\".join(i.split(\"-\")[:2]) for i in norm_tmm_df.columns]\n",
    "norm_tmm_df.columns = newcolumns\n",
    "\n",
    "newcolumns = [\"-\".join(i.split(\"-\")[:2]) for i in norm_qn_df.columns]\n",
    "norm_qn_df.columns = newcolumns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111VG</th>\n",
       "      <th>GTEX-111YS</th>\n",
       "      <th>GTEX-1122O</th>\n",
       "      <th>GTEX-1128S</th>\n",
       "      <th>GTEX-113JC</th>\n",
       "      <th>GTEX-117XS</th>\n",
       "      <th>GTEX-117YW</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZY6K</th>\n",
       "      <th>GTEX-ZYFC</th>\n",
       "      <th>GTEX-ZYFD</th>\n",
       "      <th>GTEX-ZYFG</th>\n",
       "      <th>GTEX-ZYT6</th>\n",
       "      <th>GTEX-ZYWO</th>\n",
       "      <th>GTEX-ZYY3</th>\n",
       "      <th>GTEX-ZZ64</th>\n",
       "      <th>GTEX-ZZPT</th>\n",
       "      <th>GTEX-ZZPU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>2.130272</td>\n",
       "      <td>-0.971114</td>\n",
       "      <td>0.413066</td>\n",
       "      <td>-1.548313</td>\n",
       "      <td>-2.396011</td>\n",
       "      <td>-0.287825</td>\n",
       "      <td>0.420619</td>\n",
       "      <td>-1.647538</td>\n",
       "      <td>-0.497574</td>\n",
       "      <td>-0.627416</td>\n",
       "      <td>...</td>\n",
       "      <td>1.167288</td>\n",
       "      <td>0.705237</td>\n",
       "      <td>-0.353442</td>\n",
       "      <td>-0.594039</td>\n",
       "      <td>-1.127175</td>\n",
       "      <td>0.302290</td>\n",
       "      <td>0.405537</td>\n",
       "      <td>-1.153711</td>\n",
       "      <td>1.404329</td>\n",
       "      <td>-0.932983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009.2</th>\n",
       "      <td>1.462346</td>\n",
       "      <td>2.130272</td>\n",
       "      <td>-1.571692</td>\n",
       "      <td>-0.732131</td>\n",
       "      <td>0.521225</td>\n",
       "      <td>0.787585</td>\n",
       "      <td>-2.774724</td>\n",
       "      <td>-0.696385</td>\n",
       "      <td>-2.396011</td>\n",
       "      <td>-1.734255</td>\n",
       "      <td>...</td>\n",
       "      <td>0.159940</td>\n",
       "      <td>-0.251917</td>\n",
       "      <td>-0.585798</td>\n",
       "      <td>-1.039994</td>\n",
       "      <td>0.284218</td>\n",
       "      <td>2.067638</td>\n",
       "      <td>2.012200</td>\n",
       "      <td>0.309546</td>\n",
       "      <td>0.390546</td>\n",
       "      <td>1.039994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237683.5</th>\n",
       "      <td>0.309546</td>\n",
       "      <td>0.316819</td>\n",
       "      <td>0.083188</td>\n",
       "      <td>-1.647538</td>\n",
       "      <td>-1.101410</td>\n",
       "      <td>0.181018</td>\n",
       "      <td>0.287825</td>\n",
       "      <td>0.145928</td>\n",
       "      <td>-0.513310</td>\n",
       "      <td>-1.028175</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.529174</td>\n",
       "      <td>-2.130272</td>\n",
       "      <td>1.269059</td>\n",
       "      <td>0.159940</td>\n",
       "      <td>0.013849</td>\n",
       "      <td>1.916916</td>\n",
       "      <td>0.428196</td>\n",
       "      <td>-0.635869</td>\n",
       "      <td>-0.166958</td>\n",
       "      <td>0.138933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239906.1</th>\n",
       "      <td>0.759566</td>\n",
       "      <td>0.266239</td>\n",
       "      <td>-0.181018</td>\n",
       "      <td>-1.472507</td>\n",
       "      <td>0.845575</td>\n",
       "      <td>-0.768839</td>\n",
       "      <td>1.253717</td>\n",
       "      <td>-0.855517</td>\n",
       "      <td>-0.787585</td>\n",
       "      <td>-0.806613</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.774724</td>\n",
       "      <td>0.497574</td>\n",
       "      <td>1.703911</td>\n",
       "      <td>-1.300674</td>\n",
       "      <td>-0.537155</td>\n",
       "      <td>1.734255</td>\n",
       "      <td>-0.230531</td>\n",
       "      <td>0.906557</td>\n",
       "      <td>0.797063</td>\n",
       "      <td>-0.117988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463.4</th>\n",
       "      <td>-0.714144</td>\n",
       "      <td>0.602320</td>\n",
       "      <td>1.482823</td>\n",
       "      <td>-1.962333</td>\n",
       "      <td>-0.619007</td>\n",
       "      <td>-1.386003</td>\n",
       "      <td>-2.202581</td>\n",
       "      <td>0.714144</td>\n",
       "      <td>-0.244776</td>\n",
       "      <td>-0.723108</td>\n",
       "      <td>...</td>\n",
       "      <td>0.885864</td>\n",
       "      <td>0.251917</td>\n",
       "      <td>0.394286</td>\n",
       "      <td>0.181018</td>\n",
       "      <td>-1.016498</td>\n",
       "      <td>-0.309546</td>\n",
       "      <td>-0.891002</td>\n",
       "      <td>-0.561311</td>\n",
       "      <td>-1.101410</td>\n",
       "      <td>0.383084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 361 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GTEX-1117F  GTEX-111CU  GTEX-111FC  GTEX-111VG  GTEX-111YS  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000227232.4    2.130272   -0.971114    0.413066   -1.548313   -2.396011   \n",
       "ENSG00000238009.2    1.462346    2.130272   -1.571692   -0.732131    0.521225   \n",
       "ENSG00000237683.5    0.309546    0.316819    0.083188   -1.647538   -1.101410   \n",
       "ENSG00000239906.1    0.759566    0.266239   -0.181018   -1.472507    0.845575   \n",
       "ENSG00000228463.4   -0.714144    0.602320    1.482823   -1.962333   -0.619007   \n",
       "\n",
       "                   GTEX-1122O  GTEX-1128S  GTEX-113JC  GTEX-117XS  GTEX-117YW  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000227232.4   -0.287825    0.420619   -1.647538   -0.497574   -0.627416   \n",
       "ENSG00000238009.2    0.787585   -2.774724   -0.696385   -2.396011   -1.734255   \n",
       "ENSG00000237683.5    0.181018    0.287825    0.145928   -0.513310   -1.028175   \n",
       "ENSG00000239906.1   -0.768839    1.253717   -0.855517   -0.787585   -0.806613   \n",
       "ENSG00000228463.4   -1.386003   -2.202581    0.714144   -0.244776   -0.723108   \n",
       "\n",
       "                     ...      GTEX-ZY6K  GTEX-ZYFC  GTEX-ZYFD  GTEX-ZYFG  \\\n",
       "gene_id              ...                                                   \n",
       "ENSG00000227232.4    ...       1.167288   0.705237  -0.353442  -0.594039   \n",
       "ENSG00000238009.2    ...       0.159940  -0.251917  -0.585798  -1.039994   \n",
       "ENSG00000237683.5    ...      -0.529174  -2.130272   1.269059   0.159940   \n",
       "ENSG00000239906.1    ...      -2.774724   0.497574   1.703911  -1.300674   \n",
       "ENSG00000228463.4    ...       0.885864   0.251917   0.394286   0.181018   \n",
       "\n",
       "                   GTEX-ZYT6  GTEX-ZYWO  GTEX-ZYY3  GTEX-ZZ64  GTEX-ZZPT  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000227232.4  -1.127175   0.302290   0.405537  -1.153711   1.404329   \n",
       "ENSG00000238009.2   0.284218   2.067638   2.012200   0.309546   0.390546   \n",
       "ENSG00000237683.5   0.013849   1.916916   0.428196  -0.635869  -0.166958   \n",
       "ENSG00000239906.1  -0.537155   1.734255  -0.230531   0.906557   0.797063   \n",
       "ENSG00000228463.4  -1.016498  -0.309546  -0.891002  -0.561311  -1.101410   \n",
       "\n",
       "                   GTEX-ZZPU  \n",
       "gene_id                       \n",
       "ENSG00000227232.4  -0.932983  \n",
       "ENSG00000238009.2   1.039994  \n",
       "ENSG00000237683.5   0.138933  \n",
       "ENSG00000239906.1  -0.117988  \n",
       "ENSG00000228463.4   0.383084  \n",
       "\n",
       "[5 rows x 361 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_qn_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111VG</th>\n",
       "      <th>GTEX-111YS</th>\n",
       "      <th>GTEX-1122O</th>\n",
       "      <th>GTEX-1128S</th>\n",
       "      <th>GTEX-113JC</th>\n",
       "      <th>GTEX-117XS</th>\n",
       "      <th>GTEX-117YW</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZY6K</th>\n",
       "      <th>GTEX-ZYFC</th>\n",
       "      <th>GTEX-ZYFD</th>\n",
       "      <th>GTEX-ZYFG</th>\n",
       "      <th>GTEX-ZYT6</th>\n",
       "      <th>GTEX-ZYWO</th>\n",
       "      <th>GTEX-ZYY3</th>\n",
       "      <th>GTEX-ZZ64</th>\n",
       "      <th>GTEX-ZZPT</th>\n",
       "      <th>GTEX-ZZPU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>1.962333</td>\n",
       "      <td>-0.855517</td>\n",
       "      <td>0.173983</td>\n",
       "      <td>-1.571692</td>\n",
       "      <td>-2.288654</td>\n",
       "      <td>-0.173983</td>\n",
       "      <td>0.513310</td>\n",
       "      <td>-1.766285</td>\n",
       "      <td>-0.635869</td>\n",
       "      <td>-0.610642</td>\n",
       "      <td>...</td>\n",
       "      <td>1.181085</td>\n",
       "      <td>0.353442</td>\n",
       "      <td>-0.383084</td>\n",
       "      <td>-0.529174</td>\n",
       "      <td>-1.140342</td>\n",
       "      <td>0.295050</td>\n",
       "      <td>0.346081</td>\n",
       "      <td>-1.088798</td>\n",
       "      <td>1.088798</td>\n",
       "      <td>-0.835716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009.2</th>\n",
       "      <td>1.766285</td>\n",
       "      <td>1.734255</td>\n",
       "      <td>-1.253717</td>\n",
       "      <td>-0.545172</td>\n",
       "      <td>0.360823</td>\n",
       "      <td>0.577597</td>\n",
       "      <td>-2.243542</td>\n",
       "      <td>-0.497574</td>\n",
       "      <td>-2.243542</td>\n",
       "      <td>-1.916916</td>\n",
       "      <td>...</td>\n",
       "      <td>0.428196</td>\n",
       "      <td>-0.055423</td>\n",
       "      <td>-0.778178</td>\n",
       "      <td>-1.088798</td>\n",
       "      <td>0.055423</td>\n",
       "      <td>1.800237</td>\n",
       "      <td>2.396011</td>\n",
       "      <td>0.273420</td>\n",
       "      <td>0.732131</td>\n",
       "      <td>0.825937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237683.5</th>\n",
       "      <td>0.360823</td>\n",
       "      <td>0.048490</td>\n",
       "      <td>0.368223</td>\n",
       "      <td>-1.571692</td>\n",
       "      <td>-1.181085</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166958</td>\n",
       "      <td>0.375643</td>\n",
       "      <td>-0.561311</td>\n",
       "      <td>-1.114200</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.383084</td>\n",
       "      <td>-1.766285</td>\n",
       "      <td>1.223888</td>\n",
       "      <td>0.055423</td>\n",
       "      <td>-0.048490</td>\n",
       "      <td>1.675059</td>\n",
       "      <td>0.458757</td>\n",
       "      <td>-0.652913</td>\n",
       "      <td>0.104053</td>\n",
       "      <td>0.041558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463.4</th>\n",
       "      <td>-0.561311</td>\n",
       "      <td>0.420619</td>\n",
       "      <td>1.525751</td>\n",
       "      <td>-1.875140</td>\n",
       "      <td>-0.652913</td>\n",
       "      <td>-1.503941</td>\n",
       "      <td>-2.067638</td>\n",
       "      <td>0.714144</td>\n",
       "      <td>-0.237647</td>\n",
       "      <td>-0.806613</td>\n",
       "      <td>...</td>\n",
       "      <td>0.938347</td>\n",
       "      <td>0.466463</td>\n",
       "      <td>0.390546</td>\n",
       "      <td>0.159940</td>\n",
       "      <td>-1.039994</td>\n",
       "      <td>-0.458757</td>\n",
       "      <td>-0.816237</td>\n",
       "      <td>-0.521225</td>\n",
       "      <td>-0.960077</td>\n",
       "      <td>0.273420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237094.7</th>\n",
       "      <td>2.774724</td>\n",
       "      <td>-1.253717</td>\n",
       "      <td>0.111018</td>\n",
       "      <td>-0.678843</td>\n",
       "      <td>-0.383084</td>\n",
       "      <td>0.405537</td>\n",
       "      <td>0.166958</td>\n",
       "      <td>0.398030</td>\n",
       "      <td>0.353442</td>\n",
       "      <td>-0.723108</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.734255</td>\n",
       "      <td>0.076242</td>\n",
       "      <td>1.333644</td>\n",
       "      <td>-0.237647</td>\n",
       "      <td>-0.896163</td>\n",
       "      <td>0.062360</td>\n",
       "      <td>1.016498</td>\n",
       "      <td>0.152931</td>\n",
       "      <td>-0.443425</td>\n",
       "      <td>-1.916916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 361 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GTEX-1117F  GTEX-111CU  GTEX-111FC  GTEX-111VG  GTEX-111YS  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000227232.4    1.962333   -0.855517    0.173983   -1.571692   -2.288654   \n",
       "ENSG00000238009.2    1.766285    1.734255   -1.253717   -0.545172    0.360823   \n",
       "ENSG00000237683.5    0.360823    0.048490    0.368223   -1.571692   -1.181085   \n",
       "ENSG00000228463.4   -0.561311    0.420619    1.525751   -1.875140   -0.652913   \n",
       "ENSG00000237094.7    2.774724   -1.253717    0.111018   -0.678843   -0.383084   \n",
       "\n",
       "                   GTEX-1122O  GTEX-1128S  GTEX-113JC  GTEX-117XS  GTEX-117YW  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000227232.4   -0.173983    0.513310   -1.766285   -0.635869   -0.610642   \n",
       "ENSG00000238009.2    0.577597   -2.243542   -0.497574   -2.243542   -1.916916   \n",
       "ENSG00000237683.5    0.000000    0.166958    0.375643   -0.561311   -1.114200   \n",
       "ENSG00000228463.4   -1.503941   -2.067638    0.714144   -0.237647   -0.806613   \n",
       "ENSG00000237094.7    0.405537    0.166958    0.398030    0.353442   -0.723108   \n",
       "\n",
       "                     ...      GTEX-ZY6K  GTEX-ZYFC  GTEX-ZYFD  GTEX-ZYFG  \\\n",
       "gene_id              ...                                                   \n",
       "ENSG00000227232.4    ...       1.181085   0.353442  -0.383084  -0.529174   \n",
       "ENSG00000238009.2    ...       0.428196  -0.055423  -0.778178  -1.088798   \n",
       "ENSG00000237683.5    ...      -0.383084  -1.766285   1.223888   0.055423   \n",
       "ENSG00000228463.4    ...       0.938347   0.466463   0.390546   0.159940   \n",
       "ENSG00000237094.7    ...      -1.734255   0.076242   1.333644  -0.237647   \n",
       "\n",
       "                   GTEX-ZYT6  GTEX-ZYWO  GTEX-ZYY3  GTEX-ZZ64  GTEX-ZZPT  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000227232.4  -1.140342   0.295050   0.346081  -1.088798   1.088798   \n",
       "ENSG00000238009.2   0.055423   1.800237   2.396011   0.273420   0.732131   \n",
       "ENSG00000237683.5  -0.048490   1.675059   0.458757  -0.652913   0.104053   \n",
       "ENSG00000228463.4  -1.039994  -0.458757  -0.816237  -0.521225  -0.960077   \n",
       "ENSG00000237094.7  -0.896163   0.062360   1.016498   0.152931  -0.443425   \n",
       "\n",
       "                   GTEX-ZZPU  \n",
       "gene_id                       \n",
       "ENSG00000227232.4  -0.835716  \n",
       "ENSG00000238009.2   0.825937  \n",
       "ENSG00000237683.5   0.041558  \n",
       "ENSG00000228463.4   0.273420  \n",
       "ENSG00000237094.7  -1.916916  \n",
       "\n",
       "[5 rows x 361 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_tmm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ms\n",
      "before cs 361\n",
      "after cs 361\n",
      "360\n",
      "353\n",
      "360\n"
     ]
    }
   ],
   "source": [
    "\n",
    "tissue=\"ms\"\n",
    "print(tissue)\n",
    "rpkmpath=\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/rpkms2/{:s}_rpkm.gct\".format(tissue)\n",
    "countspath=\"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/rpkms2/{:s}_counts.gct\".format(tissue)\n",
    "donorspath =\"/cbscratch/franco/datasets/gtex/gtex.sample\"\n",
    "\n",
    "expression_threshold=0.1    # 'Selects genes with > expression_threshold expression in at least min_samples')\n",
    "count_threshold=5,          # 'Selects genes with > count_threshold reads in at least min_samples')\n",
    "min_samples=10              # 'Minimum number of samples that must satisfy thresholds')\n",
    "\n",
    "donor_ids = get_donors(donorspath)\n",
    "expression_df = read_gct(rpkmpath, donor_ids)\n",
    "counts_df = read_gct(countspath, donor_ids)\n",
    "\n",
    "if expression_df.shape[1] < min_samples:\n",
    "    raise ValueError(\"tissue has less samples than threshold\")\n",
    "\n",
    "expr_ids = list(expression_df.columns)\n",
    "tissue_counts_df = counts_df.loc[:,expr_ids]\n",
    "\n",
    "newcolumns = [\"-\".join(i.split(\"-\")[:2]) for i in expr_ids]\n",
    "tissue_counts_df.columns = newcolumns\n",
    "expression_df.columns = newcolumns\n",
    "\n",
    "# QC Filter\n",
    "expr_threshold = 0.1\n",
    "count_threshold = 10\n",
    "mask = (\n",
    "            (np.sum(expression_df>=expr_threshold,axis=1)>=10) &\n",
    "            (np.sum(tissue_counts_df>=count_threshold,axis=1)>=10)\n",
    "        ).values\n",
    "expression_df = expression_df.loc[mask]\n",
    "\n",
    "print(\"before cs\", np.linalg.matrix_rank(expression_df))\n",
    "\n",
    "# load covariates\n",
    "covfile = \"/cbscratch/franco/trans-eqtl/preprocess/gtex/expression/covariates/{:s}_nopeer_covariates_w_age.txt\".format(tissue)\n",
    "df_cov = pd.read_table(covfile, header=0, index_col=0)\n",
    "\n",
    "expr_qn = centerscale_expr(inverse_quantile_normalization(normalize_quantiles(expression_df)))\n",
    "expr_qn_df = pd.DataFrame(expr_qn, columns=expression_df.columns, index=expression_df.index)\n",
    "\n",
    "expr_qn_corr, coefs = lmcorrect(expr_qn_df, df_cov)\n",
    "\n",
    "_expr_cs = centerscale_expr(expression_df.values)\n",
    "print(\"after cs\", np.linalg.matrix_rank(expression_df))\n",
    "expression_df = pd.DataFrame(_expr_cs, columns=expression_df.columns, index=expression_df.index)\n",
    "expr_corr, coefs = lmcorrect(expression_df, df_cov)\n",
    "expr_corr_qn = inverse_quantile_normalization(expr_corr)\n",
    "\n",
    "print(np.linalg.matrix_rank(expr_qn))\n",
    "print(np.linalg.matrix_rank(expr_qn_corr))\n",
    "print(np.linalg.matrix_rank(expr_corr_qn))\n",
    "\n",
    "# expr_corr_rpkms, coefs = lmcorrect_rpkms(expression_df, counts_df, df_cov)\n",
    "\n",
    "# expr_precov_corr_norm = centerscale_expr(inverse_quantile_normalization(expr_corr_rpkms))\n",
    "# expr_precov_corr_norm_df = pd.DataFrame(expr_precov_corr_norm, columns=expr_corr_rpkms.columns, index=expr_corr_rpkms.index)\n",
    "# expr_precov_corr_norm_df.index.name = \"gene_id\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "360\n"
     ]
    }
   ],
   "source": [
    "expr_qn_corr_qn = inverse_quantile_normalization(expr_qn_corr)\n",
    "\n",
    "print(np.linalg.matrix_rank(expr_qn_corr_qn))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 302)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cov.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21536, 361)\n",
      "360\n"
     ]
    }
   ],
   "source": [
    "print(expression_df.shape)\n",
    "print(np.linalg.matrix_rank(expression_df))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "361\n",
      "361\n",
      "361\n"
     ]
    }
   ],
   "source": [
    "print(np.linalg.matrix_rank(norm_tmm_df.values))\n",
    "print(np.linalg.matrix_rank(norm_qn_df.values))\n",
    "print(np.linalg.matrix_rank(expression_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD8CAYAAABn919SAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAEIxJREFUeJzt3X2MVXV+x/HPBzplQGl0mdkJCjiEGCI67SijNKEaV9GKKSokG2s3u5pdQzeuqZptsmRJqnZjYmtXJaZuygYVjQ/ZRHDVaqsoRk1UHHSUhxEfNlpHCE86EVNIHf32j7mw4zAz9+nce+b+fL+Syb333HPu+R7m8smP3/3dL44IAQAa34S8CwAAZINAB4BEEOgAkAgCHQASQaADQCIIdABIBIEOAIkg0AEgEQQ6ACTiT+p5spaWlmhvb6/nKQGg4W3evHlfRLQW26+ugd7e3q7u7u56nhIAGp7tj0rZjykXAEgEgQ4AiSDQASARdZ1DB4BKfPnll+rr69OhQ4fyLqWmmpubNWPGDDU1NVV0PIEOYNzr6+vT1KlT1d7eLtt5l1MTEaH9+/err69Ps2fPrug1mHIBMO4dOnRI06ZNSzbMJcm2pk2bVtW/Qgh0AA0h5TA/rNprJNABIBHMoQNoOBu278709RbNa8v09fJCoAMqPSBS+YuPNBHoAFCCW265Rffff79mzpyp1tZWzZ8/X08++aQWLFigjRs3qr+/X2vWrNHZZ5+dW43MoQNAEZs3b9YjjzyiN998U+vWrdPrr79+5LmBgQFt2rRJd955p26++eYcq2SEDgBFvfTSS1q6dKmmTJkiSbrkkkuOPLds2TJJ0vz58/Xhhx/mUd4RjNABoASjLSmcNGmSJGnixIkaGBioZ0lHIdABoIhzzjlH69ev18GDB3XgwAE98cQTeZc0IqZcADSceq82OuOMM3T55Zers7NTJ510Uq4ffI6FQEdDynodci3OyxLHtKxcuVIrV66UJN10002SpBdeeOHI8y0tLcyhAwCywQgdAMp0eIQ+3jBCB4BEFA102zNtb7Tda3ub7esK22+y/YntnsLPxbUvFwAwmlKmXAYk/Twi3rA9VdJm288WnrsjIv6tduUBAEpVNNAjYpekXYX7B2z3Sjqx1oUBAMpT1oeittslnS7pNUkLJV1r+0eSujU4iv8s6wIB4Cg7ns729eYuHvPp/v5+PfTQQ7rmmmuyPW/GSv5Q1Paxkh6VdH1EfC7pN5LmSOrU4Aj+16Mct9x2t+3uvXv3ZlAyANRXf3+/7r777rzLKKqkEbrtJg2G+YMRsU6SImL3kOd/K+nJkY6NiNWSVktSV1dXVFswANTbihUr9MEHH6izs1NNTU2aMmWK2tra1NPTo2XLlqmjo0OrVq3SwYMH9dhjj2nOnDm66qqrNHnyZL3zzjv66KOPdO+992rt2rV65ZVXtGDBAt13332Z11nKKhdLWiOpNyJuH7J9+pDdlkramnl1ADAO3HrrrZozZ456enp022236a233tKqVau0ZcsWPfDAA3r33Xe1adMmXX311brrrruOHPfZZ5/p+eef1x133KElS5bohhtu0LZt27Rlyxb19PRkXmcpUy4LJf1Q0nnDlij+q+0ttt+W9D1JN2ReHQCMQ2eeeaamT5+uSZMmac6cObrwwgslSR0dHd/4+v+SJUtkWx0dHWpra1NHR4cmTJigU089tSZtAkpZ5fKypJH6Rj6VeTUA0AAOt8yVpAkTJhx5PGHChG+00B26ffgxtWi1yzdFAaCIqVOn6sCBA3mXURS9XAA0niLLDLM2bdo0LVy4UKeddpomT56strbx2UnTEfVbeNLV1RXd3d11Ox/SlVf73HLQPjc7vb29OuWUU/Iuoy5GulbbmyOiq9ixjNAxrmQR1C07n8+gkkH7Tjiv4mNLvRaCH1lhDh0AEkGgA2gI9Zwezku110igAxj3mpubtX///qRDPSK0f/9+NTc3V/wazKEDGPdmzJihvr4+pd4Pqrm5WTNmzKj4eAIdwLjX1NSk2bNn513GuMeUCwAkghE6GkqWSxLreb5qlj8CpWKEDgCJINABIBEEOgAkgkAHgEQQ6ACQCAIdABJBoANAIliHjvFjx9Nq2dmfdxVAw2KEDgCJINABIBEEOgAkgkAHgEQQ6ACQCAIdABJBoANAIliHjrrYsH130X1SXoM+Vl/1np2Dt8V6pi+a15ZlSUgQI3QASASBDgCJINABIBFFA932TNsbbffa3mb7usL279h+1vZ7hdvja18uAGA0pYzQByT9PCJOkfSXkn5me56kFZKei4iTJT1XeAwAyEnRQI+IXRHxRuH+AUm9kk6UdKmktYXd1kq6rFZFAgCKK2sO3Xa7pNMlvSapLSJ2SYOhL+m7oxyz3Ha37e69e/dWVy0AYFQlB7rtYyU9Kun6iPi81OMiYnVEdEVEV2trayU1AgBKUFKg227SYJg/GBHrCpt3255eeH66pD21KREAUIpSVrlY0hpJvRFx+5CnHpd0ZeH+lZJ+n315AIBSlfLV/4WSfihpi+2ewrZfSrpV0u9s/0TS/0j6fm1KBACUomigR8TLkjzK0+dnWw4AoFJ8UxQAEkGgA0AiaJ+LqpTSFlcau30sBhX7M+rZWbzFrkSb3W8zRugAkAgCHQASQaADQCIIdABIBIEOAIkg0AEgEQQ6ACSCQAeARBDoAJAIAh0AEkGgA0AiCHQASASBDgCJINABIBG0z0XVaI1bPyX9WU88Tpq7uPbFYNxhhA4AiSDQASARBDoAJIJAB4BEEOgAkAgCHQASQaADQCJYh46jbNi+O+8SUIWej/u176viv8NF89rqUA3qiRE6ACSCQAeARBDoAJCIooFu+x7be2xvHbLtJtuf2O4p/Fxc2zIBAMWUMkK/T9JFI2y/IyI6Cz9PZVsWAKBcRQM9Il6U9GkdagEAVKGaOfRrbb9dmJI5PrOKAAAVqXQd+m8k/UpSFG5/LenHI+1oe7mk5ZI0a9asCk+HvNDrvDGV9Hubd0XtC0FdVTRCj4jdEfFVRHwt6beSzhpj39UR0RURXa2trZXWCQAooqJAtz19yMOlkraOti8AoD6KTrnYfljSuZJabPdJulHSubY7NTjl8qGkv69hjQCAEhQN9IgYaaJtTQ1qAQBUgW+KAkAiCHQASASBDgCJINABIBEEOgAkgkAHgEQQ6ACQCAIdABJBoANAIgh0AEhEpe1z0YA2bN995P5Y7VVb6lEMcjf0/TCWRfPaalwJssIIHQASQaADQCIIdABIBIEOAIkg0AEgEQQ6ACSCQAeARLAOHfiWGuu7CIftO+G8OlSCrDBCB4BEEOgAkAgCHQASQaADQCIIdABIBIEOAIkg0AEgEQQ6ACSCQAeARBDoAJCIooFu+x7be2xvHbLtO7aftf1e4fb42pYJACimlBH6fZIuGrZthaTnIuJkSc8VHgMAclQ00CPiRUmfDtt8qaS1hftrJV2WcV0AgDJVOofeFhG7JKlw+93sSgIAVKLm7XNtL5e0XJJmzZpV69N9K23Yvvsbj0dri9pSj2KQnOHvr7EsmtdWw0pQTKUj9N22p0tS4XbPaDtGxOqI6IqIrtbW1gpPBwAoptJAf1zSlYX7V0r6fTblAAAqVcqyxYclvSJpru0+2z+RdKukC2y/J+mCwmMAQI6KzqFHxBWjPHV+xrUAAKrAN0UBIBEEOgAkgkAHgETUfB06gMY12ncahtp3wnl1qASlYIQOAIkg0AEgEQQ6ACSCQAeARBDoAJAIAh0AEsGyxbHseLr4PnMX1+z0I7UtHWkZGW1xkaeh78menSPvM3xpI212a4MROgAkgkAHgEQQ6ACQCAIdABJBoANAIgh0AEgEgQ4AiSDQASARBDoAJIJAB4BEEOgAkAgCHQASQaADQCIIdABIBIEOAImgH3oORupzLh3d65w+50jF8Pf2SH3Th/dMl+ibXi5G6ACQCAIdABJBoANAIqqaQ7f9oaQDkr6SNBARXVkUBQAoXxYfin4vIvZl8DoAgCow5QIAiag20EPSM7Y3214+0g62l9vutt29d+/eKk8HABhNtYG+MCLOkLRY0s9snzN8h4hYHRFdEdHV2tpa5ekAAKOpKtAjYmfhdo+k9ZLOyqIoAED5Kg5028fYnnr4vqQLJW3NqjAAQHmqWeXSJmm97cOv81BE/FcmVQEAylZxoEfEHyT9RYa1AACqwLJFAEgEgQ4AiaB9boZGa4srfbN9KG1xgaMNb7Er0Wa3XIzQASARBDoAJIJAB4BEEOgAkAgCHQASQaADQCIIdABIBOvQS9Dzcf+oz+37avS15wCyV8p69ZHWqkvpr1dnhA4AiSDQASARBDoAJIJAB4BEEOgAkAgCHQAS8a1dtjhWq9vDWnaOvlzxj/scvYQKQL5G+3s5UjvekXTOPG7wztzFGVVUH4zQASARBDoAJIJAB4BEEOgAkAgCHQASQaADQCIIdABIRMOsQy9l3TgAZOFIy+yPH87sNTsXXZHZa42GEToAJIJAB4BEEOgAkIiqAt32RbZ32H7f9oqsigIAlK/iQLc9UdK/S1osaZ6kK2zPy6owAEB5qhmhnyXp/Yj4Q0T8n6RHJF2aTVkAgHJVE+gnSvp4yOO+wjYAQA6qWYfuEbbFUTvZyyUtLzz8wvaOwv0WSfuqOP94xrU1nlSvS+Laxom/K/eAodd2UikHVBPofZJmDnk8Q9JR7eMjYrWk1cO32+6OiK4qzj9ucW2NJ9Xrkri2RlXJtVUz5fK6pJNtz7b9p5L+VtLjVbweAKAKFY/QI2LA9rWS/lvSREn3RMS2zCoDAJSlql4uEfGUpKcqPPyoaZiEcG2NJ9Xrkri2RlX2tTniqM8xAQANiK/+A0Aicg1027+y/bbtHtvP2D4hz3qyYvs22+8Urm297ePyrikrtr9ve5vtr20nsbog1RYWtu+xvcf21rxryZrtmbY32u4tvB+vy7umLNhutr3J9luF67q5rOPznHKx/WcR8Xnh/j9ImhcRP82toIzYvlDS84UPjv9FkiLiFzmXlQnbp0j6WtJ/SPrHiOjOuaSqFFpYvCvpAg0uxX1d0hURsT3XwjJg+xxJX0i6PyJOy7ueLNmeLml6RLxhe6qkzZIua/Tfm21LOiYivrDdJOllSddFxKulHJ/rCP1wmBccoxG+mNSIIuKZiBgoPHxVg2v0kxARvRGxo/ieDSPZFhYR8aKkT/OuoxYiYldEvFG4f0BSrxL4pnoM+qLwsKnwU3Iu5j6HbvsW2x9L+oGkf8q7nhr4saSn8y4Co6KFRYOz3S7pdEmv5VtJNmxPtN0jaY+kZyOi5OuqeaDb3mB76wg/l0pSRKyMiJmSHpR0ba3ryUqx6yrss1LSgAavrWGUcm0JKamFBcYn28dKelTS9cP+xd+wIuKriOjU4L/sz7Jd8nRZzf9P0YhYVOKuD0n6T0k31rCczBS7LttXSvobSedHg60NLeN3loKSWlhg/CnMMT8q6cGIWJd3PVmLiH7bL0i6SFJJH2znvcrl5CEPL5H0Tl61ZMn2RZJ+IemSiPjfvOvBmGhh0YAKHx6ukdQbEbfnXU9WbLceXhVne7KkRSojF/Ne5fKopLkaXDXxkaSfRsQnuRWUEdvvS5okaX9h06sprN6RJNtLJd0lqVVSv6SeiPjrfKuqju2LJd2pP7awuCXnkjJh+2FJ52qwa99uSTdGxJpci8qI7b+S9JKkLRrMD0n6ZeHb6w3L9p9LWqvB9+IESb+LiH8u+fgGmw0AAIwi91UuAIBsEOgAkAgCHQASQaADQCIIdABIBIEOAIkg0AEgEQQ6ACTi/wG6qUl4nacbnwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "i=1\n",
    "plt.hist(norm_qn_df.iloc[i], bins=30, alpha=0.3, label=\"qn\")\n",
    "plt.hist(norm_tmm_df.iloc[i], bins=30, alpha=0.3, label=\"tmm\")\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ngene = 15000\n",
    "nsample = 350\n",
    "gx_rand = np.random.normal(0, 1, size = nsample * ngene).reshape((nsample, ngene))\n",
    "gx_donors = random.sample(gt_donors, nsample)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
