{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import collections\n",
    "import gzip\n",
    "import random\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.interpolate import interp1d\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "#plt.switch_backend('agg')\n",
    "import matplotlib\n",
    "import sys\n",
    "sys.path.append('../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "# from utils import mpl_stylesheet\n",
    "# mpl_stylesheet.banskt_presentation(fontfamily = 'latex-clearsans', fontsize = 18, colors = 'banskt', dpi = 72)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "f_vcf = \"/cbscratch/franco/datasets/gtex_v8/genotypes/vcfs_0.01/GTEX_v8_2019-07-29_WGS_838Indiv_Freeze_NoMissingGT_SNPfilter_MAF0.01_chr17.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "\n",
    "SNPINFO_FIELDS = ['chrom', 'varid', 'bp_pos', 'ref_allele', 'alt_allele', 'maf']\n",
    "class SnpInfo(collections.namedtuple('_SnpInfo', SNPINFO_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "from utils.containers import GeneInfo, CisMask\n",
    "\n",
    "SNP_COMPLEMENT = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "\n",
    "def filter_snps(snpinfo, dosage):\n",
    "    # Predixcan style filtering of snps\n",
    "    newsnps = list()\n",
    "    newdosage = list()\n",
    "    npoly = 0\n",
    "    nambi = 0\n",
    "    nunkn = 0\n",
    "    nlowf = 0\n",
    "    nhwep = 0\n",
    "    for i, snp in enumerate(snpinfo):\n",
    "        pos = snp.bp_pos\n",
    "        refAllele = snp.ref_allele\n",
    "        effectAllele = snp.alt_allele\n",
    "        rsid = snp.varid\n",
    "        maf = round(snp.maf, 3)\n",
    "        # Skip non-single letter polymorphisms\n",
    "        if len(refAllele) > 1 or len(effectAllele) > 1:\n",
    "            npoly += 1\n",
    "            continue\n",
    "        # Skip ambiguous strands\n",
    "        if SNP_COMPLEMENT[refAllele] == effectAllele:\n",
    "            nambi += 1\n",
    "            continue\n",
    "        # Skip unknown RSIDs\n",
    "        if rsid == '.':\n",
    "            nunkn += 1\n",
    "            continue\n",
    "        # Skip low MAF\n",
    "        if not (maf >= 0.01 and maf <=0.99):\n",
    "            nlowf += 1\n",
    "            continue\n",
    "        newsnps.append(snp)\n",
    "        newdosage.append(dosage[i])\n",
    "        # newdosage.append(intdosage)\n",
    "    print(\"Removed {:d} SNPs because of non-single letter polymorphisms\".format(npoly))\n",
    "    print(\"Removed {:d} SNPs because of ambiguous strands\".format(nambi))\n",
    "    print(\"Removed {:d} SNPs because of unknown RSIDs\".format(nunkn))\n",
    "    print(\"Removed {:d} SNPs because of low MAF < 0.01\".format(nlowf))\n",
    "    return newsnps, np.array(newdosage)    \n",
    "\n",
    "def read_vcf(filename, startsnp, endsnp):\n",
    "    dosage = list()\n",
    "    snpinfo = list()\n",
    "    linenum = 0\n",
    "    with gzip.open(filename, 'r') as vcf:\n",
    "        for line in vcf:\n",
    "            linestrip = line.decode().strip()\n",
    "            if linestrip[:2] == '##': continue\n",
    "            if linestrip[:6] == '#CHROM':\n",
    "                linesplit = linestrip.split(\"\\t\")\n",
    "                donor_ids = linesplit[9:]\n",
    "            else:\n",
    "                if linenum >= startsnp and linenum < endsnp:\n",
    "                    linesplit = linestrip.split(\"\\t\")\n",
    "                    chrom = linesplit[0]\n",
    "                    if chrom.startswith(\"chr\"):\n",
    "                        chrom = int(chrom[3:])\n",
    "                    else:\n",
    "                        chrom = int(linesplit[0])\n",
    "                    pos   = int(linesplit[1])\n",
    "                    varid = linesplit[2]\n",
    "                    ref   = linesplit[3]\n",
    "                    alt   = linesplit[4]\n",
    "\n",
    "                    gtindx = linesplit[8].split(':').index(\"GT\")\n",
    "                    gt = [x.split(':')[gtindx] for x in linesplit[9:]]\n",
    "                    ds = [ float(int(x[0]) + int(x[2])) if len(x) == 3 and x[0] != \".\" and x[2] != \".\" else \".\" for x in gt ]\n",
    "\n",
    "                    ds_notna = [float(x) for x in ds if x != \".\"]\n",
    "                    freq = sum(ds_notna) / 2 / len(ds_notna)\n",
    "                    maf = freq\n",
    "                    snpdosage = [float(x) if x != '.' else 2 * freq for x in ds]\n",
    "\n",
    "                    this_snp = SnpInfo(chrom      = chrom,\n",
    "                                       bp_pos     = pos,\n",
    "                                       varid      = varid,\n",
    "                                       ref_allele = ref,\n",
    "                                       alt_allele = alt,\n",
    "                                       maf        = maf)\n",
    "\n",
    "                    dosage.append(snpdosage)\n",
    "                    snpinfo.append(this_snp)\n",
    "                linenum += 1\n",
    "                \n",
    "                if linenum > endsnp:\n",
    "                    break\n",
    "    return np.array(dosage), snpinfo, donor_ids\n",
    "\n",
    "\n",
    "def normalize_expr(Y):\n",
    "    newY = (Y - np.mean(Y, axis = 1).reshape(-1, 1)) / np.std(Y, axis = 1).reshape(-1, 1)\n",
    "    return newY\n",
    "\n",
    "def select_donors(vcf_donors, expr_donors):\n",
    "    ''' Make sure that donors are in the same order for both expression and genotype\n",
    "    '''\n",
    "    common_donors = [x for x in vcf_donors if x in expr_donors]\n",
    "    vcfmask = np.array([vcf_donors.index(x) for x in common_donors])\n",
    "    exprmask = np.array([expr_donors.index(x) for x in common_donors])\n",
    "    return vcfmask, exprmask\n",
    "\n",
    "def select_genes(info, names):\n",
    "    ''' Select genes which would be analyzed. \n",
    "        Make sure the indices are not mixed up\n",
    "    '''\n",
    "    allowed = [x.ensembl_id for x in info]\n",
    "    common  = [x for x in names if x in allowed]\n",
    "    genes = [x for x in info if x.ensembl_id in common]\n",
    "    indices = [names.index(x.ensembl_id) for x in genes]\n",
    "    return genes, np.array(indices)\n",
    "\n",
    "def normalize_and_center_dosage(dosage, snpinfo):\n",
    "    f = [snp.maf for snp in snpinfo]\n",
    "    f = np.array(f).reshape(-1, 1)\n",
    "    gtcent = dosage - np.mean(dosage, axis = 1).reshape(-1, 1)\n",
    "    gtnorm = (dosage - (2 * f)) / np.sqrt(2 * f * (1 - f))\n",
    "    return gtcent, gtnorm\n",
    "\n",
    "def sample_gt(snpinfo, nsample):\n",
    "    nsnps = len(snpinfo)\n",
    "    dosages = np.zeros(nsnps * nsample)\n",
    "    i = 0\n",
    "    for snp in snpinfo:\n",
    "        gtcent = sample_from_maf(nsample, snp.maf)\n",
    "        dosages[i,:] = gtcent\n",
    "        i += 1\n",
    "    return dosages    \n",
    "\n",
    "def sample_from_maf(nsample, maf):\n",
    "    dosage = np.zeros(nsample)\n",
    "    mafratios = np.array([(1 - maf)**2, 2 * maf * (1 - maf), maf**2])\n",
    "    nfreq  = np.random.multinomial(nsample, mafratios, size=1)[0]\n",
    "    f1 = np.repeat(0, nfreq[0])\n",
    "    f2 = np.repeat(1, nfreq[1])\n",
    "    f3 = np.repeat(2, nfreq[2])\n",
    "    x  = np.concatenate((f1,f2,f3))\n",
    "    dosage = np.random.permutation(x)\n",
    "    # gtnorm = (dosage - (2 * maf2d)) / np.sqrt(2 * maf2d * (1 - maf2d))\n",
    "    gtcent = dosage - np.mean(dosage)\n",
    "\n",
    "    return gtcent\n",
    "\n",
    "def simulate_gt(nsnps, nsample):\n",
    "    mafs = np.linspace(0.1, 0.9, nsnps)\n",
    "    gtcent = np.zeros((nsnps, nsample))\n",
    "    snpinfo = list()\n",
    "    for i in range(nsnps):\n",
    "        gtcent[i,:] = sample_from_maf(nsample, mafs[i])\n",
    "        this_snp = SnpInfo(chrom      = 1,\n",
    "                           bp_pos     = i*100,\n",
    "                           varid      = \"rsid\"+str(i),\n",
    "                           ref_allele = \"A\",\n",
    "                           alt_allele = \"G\",\n",
    "                           maf        = mafs[i])\n",
    "        snpinfo.append(this_snp)\n",
    "    return gtcent, snpinfo\n",
    "\n",
    "def knn_correction(expr, dosage, K=30):\n",
    "    pca = PCA(n_components=min(expr.shape[0], expr.shape[1]))\n",
    "    print(\"Original dimension: {:d} x {:d}\".format(expr.shape[0], expr.shape[1]))\n",
    "    pca.fit(expr) # requires N x G\n",
    "    expr_pca = pca.transform(expr)\n",
    "    print(\"Reduced dimension: {:d} x {:d}\".format(expr_pca.shape[0], expr_pca.shape[1]))\n",
    "\n",
    "    def gene_distance(a, b):\n",
    "        return np.linalg.norm(a - b)\n",
    "\n",
    "    nsample = expr.shape[0]\n",
    "    distance_matrix = np.zeros((nsample, nsample))\n",
    "    for i in range(nsample):\n",
    "        for j in range(i+1, nsample):\n",
    "            dist = gene_distance(expr_pca[i,:], expr_pca[j,:])\n",
    "            distance_matrix[i, j] = dist\n",
    "            distance_matrix[j, i] = dist\n",
    "\n",
    "    kneighbor = K\n",
    "    gx_knn = np.zeros_like(expr)\n",
    "    gt_knn = np.zeros_like(dosage)\n",
    "    neighbor_list = list()\n",
    "\n",
    "    for i in range(nsample):\n",
    "        neighbors = np.argsort(distance_matrix[i, :])[:kneighbor + 1][1:]\n",
    "        gx_knn[i, :] = expr[i, :] - np.mean(expr[neighbors, :], axis = 0)\n",
    "        # noisy_neighbors = np.random.choice(neighbors, size = int(2 * kneighbor / 3), replace = False)\n",
    "        # noisy_neighbors = np.random.choice(neighbors, size = kneighbor, replace = True )\n",
    "        noisy_neighbors = neighbors\n",
    "        gt_knn[:, i] = dosage[:, i] - np.mean(dosage[:, noisy_neighbors], axis = 1)\n",
    "        neighbor_list.append(neighbors)\n",
    "\n",
    "    return gx_knn, gt_knn\n",
    "\n",
    "def get_cismasklist(snpinfo, geneinfo, chrom, window=1e6):\n",
    "    chr_genes_ix = [[] for ichrm in range(22)] \n",
    "    chr_genes = [[] for ichrm in range(22)]\n",
    "    if chrom is not None:\n",
    "        chr_genes_ix[chrom - 1] = np.array([i for i, g in enumerate(geneinfo) if g.chrom == chrom])\n",
    "        chr_genes[chrom - 1] = [geneinfo[ix] for ix in chr_genes_ix[chrom - 1]]\n",
    "    else:\n",
    "        for ichrm in range(22):\n",
    "            chr_genes_ix[ichrm] = np.array([i for i, g in enumerate(geneinfo) if g.chrom == ichrm + 1])\n",
    "            chr_genes[ichrm] = [geneinfo[ix] for ix in chr_genes_ix[ichrm]]\n",
    "    genemasks = list()\n",
    "    iprev = 0\n",
    "    ichrmprev = 0\n",
    "    for snp in snpinfo:\n",
    "        pos = snp.bp_pos\n",
    "        left = pos - window\n",
    "        right = pos + window\n",
    "        ichrm = chrom - 1 if chrom is not None else snp.chrom - 1\n",
    "        iprev_started = False\n",
    "        if ichrm != ichrmprev:\n",
    "            iprev = 0\n",
    "            ichrmprev = ichrm\n",
    "        thismask = list()\n",
    "        for i, g in enumerate(chr_genes[ichrm][iprev:]):\n",
    "            gstart = g.start\n",
    "            gend = g.end\n",
    "            if gstart >= left and gstart <= right:\n",
    "                # thismask.append(iprev + i)\n",
    "                thismask.append(chr_genes_ix[ichrm][iprev + i])\n",
    "                if not iprev_started:\n",
    "                    new_start_iloc = iprev\n",
    "                    iprev_started = True\n",
    "            elif gend >= left and gend <= right:\n",
    "                # thismask.append(iprev + i)\n",
    "                thismask.append(chr_genes_ix[ichrm][iprev + i])\n",
    "                if not iprev_started:\n",
    "                    new_start_iloc = iprev\n",
    "                    iprev_started = True\n",
    "            if gstart > right:\n",
    "                break\n",
    "        if len(thismask) > 0:\n",
    "            #genemasks.append(chr_genes_ix[np.array(thismask)])\n",
    "            #iprev = thismask[0]\n",
    "            genemasks.append(np.array(thismask))\n",
    "            iprev = new_start_iloc\n",
    "        else:\n",
    "            genemasks.append(np.array([]))\n",
    "    return genemasks\n",
    "\n",
    "def compress_cismasklist(genemasks):\n",
    "    cismasks = list()\n",
    "    appendmask = False\n",
    "    endmask = False\n",
    "    setprev = False\n",
    "    snplist = list()\n",
    "    for i, mask in enumerate(genemasks):\n",
    "        if not setprev:\n",
    "            prev_mask = mask\n",
    "            setprev = True\n",
    "        if np.all(np.array_equal(mask, prev_mask)):\n",
    "            snplist.append(i)\n",
    "        else:\n",
    "            appendmask = True\n",
    "\n",
    "        if i == len(genemasks) - 1: endmask = True # no more masks to process\n",
    "\n",
    "        if appendmask:\n",
    "            thismask = CisMask(rmv_id = prev_mask, apply2 = snplist)\n",
    "            cismasks.append(thismask)\n",
    "            snplist = list([i])\n",
    "            prev_mask = mask\n",
    "            if not endmask:\n",
    "                appendmask = False\n",
    "\n",
    "        if endmask:\n",
    "            # if not appendmask:\n",
    "            #     snplist.append(i)\n",
    "            thismask = CisMask(rmv_id = mask, apply2 = snplist)\n",
    "            cismasks.append(thismask)\n",
    "\n",
    "    return cismasks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removed 0 SNPs because of non-single letter polymorphisms\n",
      "Removed 0 SNPs because of ambiguous strands\n",
      "Removed 0 SNPs because of unknown RSIDs\n",
      "Removed 0 SNPs because of low MAF < 0.01\n"
     ]
    }
   ],
   "source": [
    "# Use real genotype\n",
    "# gtfull, snpinfo, gt_donors = read_vcf(f_vcf, 9080, 9105)\n",
    "gtfull, snpinfo, gt_donors = read_vcf(f_vcf, 0, 1)\n",
    "snp_info, gtfull = filter_snps(snpinfo, gtfull)\n",
    "\n",
    "# Use artificial genotype\n",
    "# gtfull, snp_info = simulate_gt(10000, len(gt_donors))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13538, 605)\n",
      "604\n"
     ]
    }
   ],
   "source": [
    "# # Simulate some expression\n",
    "# ngene = 15000\n",
    "# nsample = 350\n",
    "# gx_rand = np.random.normal(0, 1, size = nsample * ngene).reshape((ngene, nsample)) \n",
    "# gx_donors = random.sample(gt_donors, nsample)\n",
    "# print(gx_rand.shape)\n",
    "\n",
    "# Load real expression\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/sse_tpms_qcfilter.txt.protein_coding_filtered\", header=0, index_col=0, sep=\"\\t\")\n",
    "ngene, nsample = df.shape\n",
    "gx_donors = list(df.columns)\n",
    "gene_names = list(df.index)\n",
    "gx = df.values\n",
    "print(gx.shape)\n",
    "print(np.linalg.matrix_rank(gx))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from iotools import readgtf\n",
    "\n",
    "gtf_file = \"/cbscratch/franco/datasets/GENCODE/gencode.v26.annotation.gtf.gz\"\n",
    "geneinfo = readgtf.gencode_v12(gtf_file, trim=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "vcfmask, exprmask = select_donors(gt_donors, gx_donors)\n",
    "genes, indices = select_genes(geneinfo, gene_names)       \n",
    "expr = gx[:, exprmask][indices, :]\n",
    "dosage_filtered_selected = gtfull[:, vcfmask]\n",
    "geneinfo = genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom=17\n",
    "window=1e6\n",
    "cismasklist = get_cismasklist(snp_info, geneinfo, chrom, window=window)\n",
    "cismaskcomp = compress_cismasklist(cismasklist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original dimension: 605 x 13538\n",
      "Reduced dimension: 605 x 605\n"
     ]
    }
   ],
   "source": [
    "gt_cent, gt_norm = normalize_and_center_dosage(dosage_filtered_selected, snp_info)\n",
    "\n",
    "gx_norm = normalize_expr(expr)\n",
    "gx_corr, gt_corr = knn_correction(gx_norm.T, dosage_filtered_selected, 30)\n",
    "gx_knn_norm = normalize_expr(gx_corr.T)\n",
    "gt_knn_cent, gt_knn_norm = normalize_and_center_dosage(gt_corr, snp_info)\n",
    "\n",
    "sigmax2 = np.var(gt_cent, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "SIGMA_BETAS = np.repeat( 0.1 ** 2, gt_knn_cent.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pvals_perm(GT, R, W):\n",
    "    mu2, mu4 = moment_data(GT)\n",
    "    N = GT.shape[1]\n",
    "    q11 = np.sum(W)\n",
    "    q2  = np.sum(np.diag(W))\n",
    "    muQ = mu2 * (N * q2 - q11) / (N - 1)\n",
    "\n",
    "    v31 = - mu4 / (N - 1)\n",
    "    v22 = v31 + (N * mu2 * mu2 / (N - 1)) #(N*(mu2**2) - mu4)/(N-1)\n",
    "    v211 = - (v31 + v22) / (N - 2)\n",
    "    v1111 = - 3 * v211 / (N - 3)\n",
    "\n",
    "    q31 = np.dot(np.diag(W),np.sum(W,axis = 1))\n",
    "    q4 = np.sum(np.square(np.diag(W)))\n",
    "    q22 = np.sum(np.square(W))\n",
    "    q211 = np.sum(np.square(np.sum(W,axis = 1)))\n",
    "\n",
    "    sigma2 = v1111*(q11**2 - 2*q2*q11 - 4*q211 + 8*q31 + 2*q22 + q2**2 - 6*q4) + 2*v211*(q2*q11 + 2*q211 - 6*q31 - 2*q22 - q2**2 + 6*q4) + v22*(q2**2 + 2*q22 - 3*q4) + 4*v31*(q31 - q4) + mu4*q4\n",
    "\n",
    "    sigma2 = sigma2 - muQ**2\n",
    "    sigmaQ = np.sqrt(sigma2)\n",
    "    p = 1 - stats.norm.cdf(R, loc=muQ, scale=sigmaQ)\n",
    "    return p, muQ, sigmaQ\n",
    "\n",
    "def moment_data(GT):   #GT ixN\n",
    "    GT2 = np.square(GT)\n",
    "    GT4 = np.square(GT2)\n",
    "    mu2 = np.mean(GT2)\n",
    "    mu4 = np.mean(GT4)\n",
    "    return mu2, mu4\n",
    "\n",
    "def get_betas_force(gt, expr, sigma_x2, sb2):\n",
    "    ngenes, nsamples = expr.shape\n",
    "    nsnps = gt.shape[0]\n",
    "    print(\"Samples:\", nsamples)\n",
    "    print(\"Genes:\", ngenes)\n",
    "    print(\"N SNPs:\", nsnps)\n",
    "    k = ngenes\n",
    "    I = np.identity(k)\n",
    "    \n",
    "    B = np.zeros((nsnps, ngenes))\n",
    "    YTY = np.matmul(expr, expr.T)\n",
    "    for snpi in range(nsnps):\n",
    "        Is  = I * sigma_x2[snpi]/sb2[snpi]\n",
    "        W   = YTY + Is\n",
    "        Winv= np.linalg.inv(W)\n",
    "        A   = np.matmul(Winv, expr)\n",
    "        Bi  = np.matmul(A, gt[snpi,:][np.newaxis].T)\n",
    "        B[snpi,:] = Bi.reshape(-1)\n",
    "    return B\n",
    "\n",
    "def get_betas(gt, expr, sigma_x2, sb2):\n",
    "    ngenes, nsamples = expr.shape\n",
    "    nsnps = gt.shape[0]\n",
    "    print(\"Samples:\", nsamples)\n",
    "    print(\"Genes:\", ngenes)\n",
    "    print(\"N SNPs:\", nsnps)\n",
    "    k = min(nsamples, ngenes)\n",
    "    I = np.identity(k)\n",
    "    \n",
    "    U, s, VT = np.linalg.svd(expr.T, full_matrices=False)\n",
    "    S = I*s  \n",
    "\n",
    "    B = np.zeros((nsnps, ngenes))\n",
    "    for snpi in range(nsnps):\n",
    "        innerLinv_ST = (s / (s*s + (sigma_x2[snpi]/sb2[snpi]))) * I\n",
    "        # print(\"innerLinv_ST:\", innerLinv_ST.shape)\n",
    "        UT = np.transpose(U)\n",
    "        V = np.transpose(VT)\n",
    "        # print(\"U:\",U.shape)\n",
    "\n",
    "        innerLinv_STUT = np.matmul(innerLinv_ST, UT)\n",
    "        # print(\"innerLinv_STUT:\", innerLinv_STUT.shape)  # if G < N, then UT is k x N and then innerL_STUT is k x N\n",
    "        A2 = np.matmul(V, innerLinv_STUT)\n",
    "        # print(\"A:\", A2.shape)\n",
    "\n",
    "        Bi = np.matmul(A2, gt[snpi,:][np.newaxis].T)\n",
    "        B[snpi,:] = Bi.reshape(-1)\n",
    "    return B\n",
    "\n",
    "def reshape_masked_betas(b, mask, ngenes):\n",
    "    print(\"reshaping {:d} betas into ({:d},{:d}) with {:d} masked genes out of {:d}\".format(len(b), len(mask.apply2), (ngenes - len(mask.rmv_id)), len(mask.rmv_id), ngenes)) \n",
    "    _b = b.reshape(len(mask.apply2), ngenes-len(mask.rmv_id))\n",
    "    paddedBeta = np.zeros( (len(mask.apply2), ngenes) )\n",
    "    inv_ind = np.delete(np.arange(ngenes), mask.rmv_id)\n",
    "    paddedBeta[:, inv_ind] = _b\n",
    "    return paddedBeta.reshape(-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13538, 605)\n",
      "Masked expression dims: (13526, 605)\n",
      "Samples: 605\n",
      "Genes: 13526\n",
      "N SNPs: 1\n",
      "[[2.74844614 0.         0.         ... 0.         0.         0.        ]\n",
      " [0.         2.74844614 0.         ... 0.         0.         0.        ]\n",
      " [0.         0.         2.74844614 ... 0.         0.         0.        ]\n",
      " ...\n",
      " [0.         0.         0.         ... 2.74844614 0.         0.        ]\n",
      " [0.         0.         0.         ... 0.         2.74844614 0.        ]\n",
      " [0.         0.         0.         ... 0.         0.         2.74844614]]\n",
      "(13526, 13526)\n",
      "reshaping 1 betas into (1,13526) with 12 masked genes out of 13538\n",
      "Samples: 605\n",
      "Genes: 13526\n",
      "N SNPs: 1\n",
      "reshaping 1 betas into (1,13526) with 12 masked genes out of 13538\n"
     ]
    }
   ],
   "source": [
    "def compare_basic_props(GX, GT, sigmax2 = [0.1], sb2 = None, i = 0):\n",
    "    # Yt = (GX / np.sqrt(nsample)).T\n",
    "#     K = np.linalg.matrix_rank(GX)    \n",
    "    nsnps  = GT.shape[0]\n",
    "    ngenes = GX.shape[0]\n",
    "    Rscore = np.zeros(nsnps)\n",
    "    pvals  = np.zeros(nsnps)\n",
    "    muQ    = np.zeros(nsnps)\n",
    "    sigmaQ = np.zeros(nsnps)\n",
    "    mu2    = np.zeros(nsnps)\n",
    "    mu4    = np.zeros(nsnps)\n",
    "    Keffs  = np.zeros(nsnps)\n",
    "    for mask in cismaskcomp:\n",
    "        applysnps = mask.apply2\n",
    "        \n",
    "        usegenes = np.ones(GX.shape[0], dtype=bool)\n",
    "        if mask.rmv_id.shape[0] > 0: usegenes[mask.rmv_id] = False\n",
    "        masked_gx = np.ascontiguousarray(GX[usegenes])\n",
    "        \n",
    "        print(\"Masked expression dims:\", masked_gx.shape)\n",
    "        Yt = masked_gx.T\n",
    "        U, S, Vt = np.linalg.svd(Yt, full_matrices=False)\n",
    "        S2 = np.square(S)\n",
    "        \n",
    "        for j in applysnps:       \n",
    "            S2mod = S2 + sigmax2[j] / sb2[j]\n",
    "\n",
    "            Keffs[j] = np.sum(S2 / S2mod)\n",
    "            W = np.dot(U, np.dot(np.diag(S2 / S2mod), U.T)) / sigmax2[j]\n",
    "            Rscore[j] = np.sum(np.square(np.dot(U.T, GT[j,:])) * (S2 / S2mod)) / sigmax2[j]\n",
    "\n",
    "            pvals[j], muQ[j], sigmaQ[j] = pvals_perm(GT[j, :].reshape(1, -1), Rscore[j], W)\n",
    "            mu2[j], mu4[j] = moment_data(GT[j, :].reshape(1, -1))\n",
    "    \n",
    "        betas_force = get_betas_force(GT, masked_gx, sigmax2, sb2)\n",
    "        betas_force = reshape_masked_betas(betas_force, mask, ngenes)\n",
    "        betas = get_betas(GT, masked_gx, sigmax2, sb2)\n",
    "        betas = reshape_masked_betas(betas, mask, ngenes)\n",
    "#     print(\"Qscore:\", Rscore[:5])\n",
    "#     print(\"muQ:\", muQ[:5])\n",
    "#     print(\"sigmaQ:\", sigmaQ[:5])\n",
    "#     print(\"pvals:\", pvals[:5])\n",
    "    return pvals, Rscore, muQ, sigmaQ, betas, betas_force\n",
    "\n",
    "i = 0\n",
    "\n",
    "print(gx_knn_norm.shape)\n",
    "pvals, Qscores, muQ, sigmaQ, betas, betas_force = compare_basic_props(gx_knn_norm, gt_knn_cent, sigmax2 = sigmax2, sb2 = SIGMA_BETAS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-3.94930681e-05, -2.01902428e-04,  2.74877836e-04, ...,\n",
       "       -5.61347486e-05, -2.14336100e-05,  5.78134080e-04])"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "betas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-3.94930681e-05, -2.01902428e-04,  2.74877836e-04, ...,\n",
       "       -5.61347486e-05, -2.14336100e-05,  5.78134080e-04])"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "betas_old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
