{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('/usr/users/fsimone/trans-eqtl-pipeline/preprocess/gtex_v8/scripts')\n",
    "import os\n",
    "import numpy as np\n",
    "import gzip\n",
    "# from containers import GeneInfo\n",
    "import re\n",
    "import collections\n",
    "\n",
    "GENEINFO_FIELDS = ['name', 'ensembl_id', 'chrom', 'start', 'end', 'typ']\n",
    "class GeneInfo(collections.namedtuple('_GeneInfo', GENEINFO_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "\n",
    "def gencode_v12(filepath, feature = 'gene', trim=False, biotype=['protein_coding'], include_chrom = 0, include_chroms=['{:d}'.format(x + 1) for x in range(22)]):    \n",
    "    annotfile = os.path.realpath(filepath)\n",
    "    geneinfo = list()\n",
    "    lncRNA_list = [\"macro_lncRNA\",\"non_coding\",\"bidirectional_promoter_lncRNA\",\"3prime_overlapping_ncRNA\",\"sense_overlapping\",\"processed_transcript\",\"sense_intronic\",\"TEC\",\"antisense\",\"lincRNA\"]\n",
    "    mode = \"v19\"\n",
    "    if re.search(\"v26\", filepath):\n",
    "        mode = \"v26\"\n",
    "        \n",
    "    if \"lncRNA\" in biotype:\n",
    "        mybiotype = biotype + lncRNA_list\n",
    "    else:\n",
    "        mybiotype = biotype\n",
    "        \n",
    "    if annotfile.endswith(\"affy\"):\n",
    "        geneinfo = affy_exon_chip(annotfile, include_chrom = include_chrom, include_chroms=include_chroms)\n",
    "        return geneinfo\n",
    "    else:\n",
    "        try:\n",
    "            with gzip.open(annotfile, 'r') as mfile:\n",
    "                for line in mfile:\n",
    "                    linesplit = line.decode().strip().split('\\t')\n",
    "                    if linesplit[0][0] == '#' or linesplit[2] != feature: continue # skip header\n",
    "\n",
    "                    chrom = linesplit[0][3:]\n",
    "                    if include_chrom > 0:\n",
    "                        include_chroms = ['{:d}'.format(include_chrom)]\n",
    "                    if chrom not in include_chroms: continue\n",
    "\n",
    "                    # Any particular biotype selected?\n",
    "                    infolist = linesplit[8].split(';')\n",
    "\n",
    "                    if mode == \"v19\":\n",
    "                        if len(mybiotype) > 0:\n",
    "                            rowtype = infolist[2].strip().split(' ')[1].replace('\"','')\n",
    "                            if rowtype not in mybiotype: continue\n",
    "                        gene_name = infolist[4].strip().split(' ')[1].replace('\"','')\n",
    "\n",
    "                    if mode == \"v26\":\n",
    "                        if len(mybiotype) > 0:\n",
    "                            rowtype = infolist[1].strip().split(' ')[1].replace('\"','')\n",
    "                            if rowtype not in mybiotype: continue\n",
    "                        gene_name = infolist[2].strip().split(' ')[1].replace('\"','')\n",
    "\n",
    "                    # TSS: gene start (0-based coordinates for BED)\n",
    "                    if linesplit[6] == '+':\n",
    "                        start = np.int64(linesplit[3]) - 1\n",
    "                        end   = np.int64(linesplit[4])\n",
    "                    elif linesplit[6] == '-':\n",
    "                        start = np.int64(linesplit[3])  # last base of gene\n",
    "                        end   = np.int64(linesplit[4]) - 1\n",
    "                    else:\n",
    "                        raise ValueError('Strand not specified.')\n",
    "\n",
    "                    # For simulation\n",
    "                    if linesplit[1] == 'SIMULATION':\n",
    "                        start = np.int64(linesplit[3])\n",
    "                        end   = np.int64(linesplit[4])\n",
    "\n",
    "                    gene_id = infolist[0].strip().split(' ')[1].replace('\"','')\n",
    "                    if trim:\n",
    "                        gene_id = gene_id.split(\".\")[0]\n",
    "                    this_gene = GeneInfo(name       = gene_name,\n",
    "                                         ensembl_id = gene_id,\n",
    "                                         chrom      = int(chrom),\n",
    "                                         start      = start,\n",
    "                                         end        = end,\n",
    "                                         typ        = rowtype)\n",
    "\n",
    "                    geneinfo.append(this_gene)\n",
    "        except IOError as err:\n",
    "            raise IOError('{:s}: {:s}'.format(annotfile, err.strerror))\n",
    "\n",
    "        return geneinfo\n",
    "\n",
    "def affy_exon_chip(filepath, include_chrom = 0, include_chroms=['{:d}'.format(x + 1) for x in range(22)]):\n",
    "    geneinfo = list()\n",
    "    try:\n",
    "        with open(filepath, 'r') as mfile:\n",
    "            next(mfile) # skip header\n",
    "            for line in mfile:\n",
    "                linesplit = line.strip().split('\\t')\n",
    "                if linesplit[0][0] == '#' : continue \n",
    "\n",
    "                chrom = linesplit[2][3:]\n",
    "                if include_chrom > 0:\n",
    "                    include_chroms = ['{:d}'.format(include_chrom)]\n",
    "                if chrom not in include_chroms: continue\n",
    "\n",
    "                # TSS: gene start (0-based coordinates for BED)\n",
    "                if linesplit[3] == '+':\n",
    "                    start = np.int64(linesplit[4]) - 1\n",
    "                    end   = np.int64(linesplit[5])\n",
    "                elif linesplit[3] == '-':\n",
    "                    start = np.int64(linesplit[4])  # last base of gene\n",
    "                    end   = np.int64(linesplit[5]) - 1\n",
    "                else:\n",
    "                    raise ValueError('Strand not specified.')\n",
    "\n",
    "                gene_name = linesplit[7].split(\"//\")[0].rstrip()\n",
    "                transcript_cluster_id = linesplit[0]\n",
    "                this_gene = GeneInfo(name       = gene_name,\n",
    "                                     ensembl_id = transcript_cluster_id,\n",
    "                                     chrom      = int(chrom),\n",
    "                                     start      = start,\n",
    "                                     end        = end)\n",
    "\n",
    "                geneinfo.append(this_gene)\n",
    "    except IOError as err:\n",
    "        raise IOError('{:s}: {:s}'.format(annotfile, err.strerror))\n",
    "\n",
    "    return geneinfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = \"/cbscratch/franco/datasets/GENCODE/gencode.v26.annotation.gtf.gz\"\n",
    "gene_info = gencode_v12(infile, biotype=['protein_coding', 'lncRNA'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_samples(donorfile):    \n",
    "    with open(donorfile, 'r') as samfile:\n",
    "        sample = 0\n",
    "        samplenames = list()\n",
    "        # skip first two lines\n",
    "        next(samfile)\n",
    "        next(samfile)\n",
    "        for line in samfile:\n",
    "            if re.search('^#', line):\n",
    "                continue\n",
    "            sample += 1\n",
    "            samplenames.append(line.strip().split()[0])\n",
    "    return samplenames\n",
    "\n",
    "def filter_donors(df, donors):\n",
    "    donor_list = df.columns\n",
    "    common  = [x for x in donors if x in donor_list]\n",
    "    print(\"{:d} donors remained from {:d}\".format(len(common), len(donor_list)))\n",
    "    return df[common]\n",
    "\n",
    "def filter_rows(df, genedict, gene_dict_type):\n",
    "    gx_gene_list = df.index\n",
    "    common  = [genedict[x] for x in gx_gene_list]\n",
    "    current_types = [gene_dict_type[x] for x in gx_gene_list if genedict[x] == True]\n",
    "    print(\"{:d} genes remained from {:d}\".format(sum(common), len(gx_gene_list)))\n",
    "    return df[common], current_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12470, 670)\n",
      "12046 genes remained from 12470\n",
      "670 donors remained from 670\n",
      "(12046, 670)\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "donor_file = \"/cbscratch/franco/datasets/gtex_v8/genotypes/gtex_v8.sample\"\n",
    "gx_file    = \"/cbscratch/franco/trans-eqtl/new_preprocess_aug2019/gtex_v8/expression/tpms/wb_tpms_qcfilter.txt\"\n",
    "donors = read_samples(donor_file)\n",
    "gene_dict = defaultdict(lambda: False)\n",
    "gene_dict_type = dict()\n",
    "for g in gene_info:\n",
    "    gene_dict[g.ensembl_id] = True\n",
    "    gene_dict_type[g.ensembl_id] = g.typ\n",
    "\n",
    "gx_df = pd.read_table(gx_file, sep=\"\\t\", header=0, index_col=0)\n",
    "print(gx_df.shape)\n",
    "new_gx_df, current_types = filter_rows(gx_df, gene_dict, gene_dict_type)\n",
    "sorted_gx_df = filter_donors(new_gx_df, donors)\n",
    "\n",
    "print(sorted_gx_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         non_coding\t3\t0\n",
      "                       macro_lncRNA\t1\t0\n",
      "                                TEC\t1053\t58\n",
      "                     sense_intronic\t888\t57\n",
      "                     protein_coding\t18901\t10414\n",
      "                  sense_overlapping\t184\t40\n",
      "                          antisense\t5419\t945\n",
      "      bidirectional_promoter_lncRNA\t8\t0\n",
      "               processed_transcript\t520\t103\n",
      "                            lincRNA\t7327\t427\n",
      "           3prime_overlapping_ncRNA\t30\t2\n",
      "                              Total\t34334\t12046\n"
     ]
    }
   ],
   "source": [
    "types = [x.typ for x in gene_info]\n",
    "types_uniq = list(set(types))\n",
    "\n",
    "\n",
    "for t in types_uniq:\n",
    "    print(\"{:>35}\\t{:>d}\\t{:>d}\".format(t, types.count(t), current_types.count(t) ))\n",
    "print(\"{:>35}\\t{:>d}\\t{:>d}\".format(\"Total\", len(types), len(current_types) ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
