{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import os, sys\n",
    "sys.path.append('../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import operator\n",
    "import json\n",
    "from utils import utils\n",
    "json_file = \"../gtex_v8_metadata.json\"\n",
    "tissue_file = \"../plots/tissues.txt\"\n",
    "tshorts, tfulls_plain = utils.read_tissues(tissue_file, plain=True)\n",
    "_, tfulls = utils.read_tissues(tissue_file)\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors  = dict()\n",
    "tissue_names   = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tshorts, tfulls_plain):\n",
    "    tissue_names[tshort] = tfull\n",
    "for tshort, tfull in zip(tshorts, tfulls):\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]\n",
    "\n",
    "sorted_tissues = [x[0] for x in sorted(tissue_samples.items(), key=operator.itemgetter(1))]\n",
    "select_tissues = ['ms', 'wb', 'sse', 'as', 'hlv', 'colt', 'bco', 'aa', 'ag' , 'ac', 'bca', 'ac', 'fib', 'bceh', 'bhi', 'tes', 'nt', 'thy']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA = pd.read_csv(\"multi_r2_ALL.txt\", header=0, index_col=0, sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import fastcluster\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "def calculate_linkage(matrix):\n",
    "    data = matrix\n",
    "    row_linkage = fastcluster.linkage_vector(data.fillna(0).values,\n",
    "                               method='single',\n",
    "                               metric=\"euclidean\")\n",
    "#     row_linkage = hierarchy.linkage(data.fillna(0), method=\"average\", metric='euclidean')\n",
    "\n",
    "    row_dendrogram = hierarchy.dendrogram(row_linkage, no_plot=True, color_threshold=-np.inf)\n",
    "    row_reordered_ind = row_dendrogram['leaves']\n",
    "\n",
    "    data = matrix.T\n",
    "    col_linkage = fastcluster.linkage_vector(data.fillna(0).values,\n",
    "                               method='single',\n",
    "                               metric=\"euclidean\")\n",
    "#     col_linkage = hierarchy.linkage(data.fillna(0), method=\"average\", metric='euclidean')\n",
    "    \n",
    "    col_dendrogram = hierarchy.dendrogram(col_linkage, no_plot=True, color_threshold=-np.inf)\n",
    "    col_reordered_ind = col_dendrogram['leaves']\n",
    "    return matrix.iloc[row_reordered_ind, col_reordered_ind]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'calculate_linkage' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-713a9909d470>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mseaborn\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m;\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0mclusterDATA\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcalculate_linkage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDATA\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# fig = plt.figure(figsize=(14,17))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'calculate_linkage' is not defined"
     ]
    }
   ],
   "source": [
    "import matplotlib\n",
    "# matplotlib.use(\"Agg\")\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import seaborn as sns; sns.set()\n",
    "\n",
    "clusterDATA = calculate_linkage(DATA).T\n",
    "\n",
    "# fig = plt.figure(figsize=(14,17))  \n",
    "fig = plt.figure(figsize=(20,8), dpi=300)  \n",
    "ax  = fig.add_subplot(111)\n",
    "# cbar_ax = fig.add_axes([.905, .3, .05, .3])\n",
    "cmap = plt.get_cmap(\"Blues\")\n",
    "# cmap = plt.get_cmap(\"Spectral_r\")\n",
    "\n",
    "mymask = pd.isna(clusterDATA)\n",
    "# cmap.set_bad(\"gray\")\n",
    "\n",
    "h = sns.heatmap(clusterDATA,\n",
    "                cmap=cmap, #\"RdBu_r\",\n",
    "                robust=True,\n",
    "                linewidths=.5,\n",
    "                cbar_kws={'label': '$R^2$', 'shrink':0.35,\n",
    "                          #'ticks':[-2.0,-1.0, 0.0, 1.0, 2.0, 3.0, 4.0],\n",
    "                          'extend':'both',\n",
    "                          'pad':0.01,\n",
    "                          'aspect':15},\n",
    "                ax=ax,\n",
    "                mask=mymask)\n",
    "#                 cbar_ax = cbar_ax)\n",
    "# h.set_facecolor('white')\n",
    "\n",
    "\n",
    "# cbar=plt.colorbar(ax)\n",
    "# cbar.set_yticklabels([\"a\",\"a\",\"a\",\"a\",\"a\",\"a\"])\n",
    "# h.figure.axes[1].set_yticklabels([\"-2\",\"-1\",\"0\",\"1\",\"2\",\"3\",\"> 4\"])\n",
    "\n",
    "# ylabels = [\"3' overlapping ncRNA\", \"Transcription Factors\", \"Processed Transcript\", \"Sense Intronic\", \"Sense Overlapping\", \"Protein Coding\", \"Antisense\", \"lincRNA\", \"\", \"Cis-eQTL Enrichment\", \"GTEx cis-eQTLs\", \"GTEx cis-sQTLs\"]\n",
    "# for l in DATA.index:\n",
    "#     if l == \"tmp\":\n",
    "#         ylabels.append(\"\")\n",
    "#     else:\n",
    "#         ylabels.append(l)\n",
    "# ax.set_yticklabels(ylabels, rotation=\"horizontal\")\n",
    "ax.set_yticklabels(clusterDATA.index, rotation=90) #, ha='right')\n",
    "\n",
    "# cbar = ax.figure.colorbar(im, ax=ax)\n",
    "# cbar.ax.set_ylabel(\"AUC\", rotation=-90, va=\"bottom\")\n",
    "\n",
    "my_colors=[tissue_colors[t] for t in clusterDATA.index]\n",
    "ax.set_aspect('equal')\n",
    "for t, X in zip(clusterDATA.index, ax.get_yticks()):\n",
    "    ax.add_patch(patches.Circle((-0.5, X), radius=.3,\n",
    "                                 fill=True,clip_on=False, color=tissue_colors[t]))\n",
    "\n",
    "labels = ax.get_yticklabels()\n",
    "ax.set_yticklabels([tissue_names[l.get_text()] for l in labels], rotation='horizontal', ha='right')    \n",
    "    \n",
    "# labels = ax.get_yticklabels()\n",
    "# for label in labels:\n",
    "#     print(label.get_position())\n",
    "#     label.set_y(label.get_position()[1] + 0.3)\n",
    "#     print(label.get_position())\n",
    "\n",
    "plt.show()\n",
    "# plt.savefig(\"heatmap.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_index = list()\n",
    "cov_color = list()\n",
    "for i in knn_cov_r2.index:\n",
    "    if i in sample_cov_dict:\n",
    "        new_index.append(sample_cov_dict[i])\n",
    "        cov_color.append(\"#ff9f19\")\n",
    "    elif i in cov_dict:\n",
    "        new_index.append(cov_dict[i])\n",
    "        cov_color.append(\"#26d18a\")\n",
    "    else:\n",
    "        new_index.append(i)\n",
    "        cov_color.append(\"#9b26d1\")\n",
    "new_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns; sns.set()\n",
    "\n",
    "tmp_df = knn_cov_r2.T.copy(deep=True)\n",
    "tmp_df.index = [tissue_names[x] for x in tmp_df.index]\n",
    "tmp_df.columns = new_index\n",
    "# tmp_df.columns = [tissue_names[x] for x in tmp_df.columns]\n",
    "# tmp_df.index = new_index\n",
    "\n",
    "# fig, (cax, ax) = plt.subplots(nrows=2, figsize=(28,14), gridspec_kw={\"height_ratios\":[0.025,1]})\n",
    "\n",
    "sns.set_context(\"paper\", font_scale=1.8)\n",
    "g = sns.clustermap(tmp_df.fillna(0),\n",
    "                   row_colors = [tissue_colors[x] for x in knn_cov_r2.columns], \n",
    "                   col_colors = cov_color,\n",
    "                   mask = pd.isna(tmp_df),\n",
    "                   figsize=(30,8),\n",
    "                   cmap=\"Blues\",\n",
    "                   linewidths=.95,\n",
    "                   cbar_kws={\"orientation\": \"vertical\", \"label\":\"RÂ²\"},\n",
    "                   robust=True ) #col_cluster=False)\n",
    "g.ax_row_dendrogram.set_visible(False)\n",
    "#g.gs.set_height_ratios([0.1,0.1,0.5])\n",
    "g.cax.set_visible(False)\n",
    "#g.ax_row_dendrogram.set_axis_off()\n",
    "#g.cax.set_ax(g.ax_row_dendrogram.axes)\n",
    "#cg.ax_row_dendrogram.set_xlim([0,0])\n",
    "#g.ax_row_colors.set_alpha(0.9)\n",
    "s = g.ax_heatmap.set_xticklabels([i for i in g.ax_heatmap.get_xticklabels()], rotation=45, ha='right')\n",
    "# g.savefig(\"knn_clustermap_rotated_labels.png\")\n",
    "#g.gs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib \n",
    "\n",
    "def fixedWidthClusterMap(dataFrame, cellSizePixels=50):\n",
    "    # Calulate the figure size, this gets us close, but not quite to the right place\n",
    "    dpi = matplotlib.rcParams['figure.dpi']\n",
    "    marginWidth = matplotlib.rcParams['figure.subplot.right']-matplotlib.rcParams['figure.subplot.left']\n",
    "    marginHeight = matplotlib.rcParams['figure.subplot.top']-matplotlib.rcParams['figure.subplot.bottom']\n",
    "    Ny,Nx = dataFrame.shape\n",
    "    figWidth = (Nx*cellSizePixels/dpi)/0.8/marginWidth\n",
    "    figHeigh = (Ny*cellSizePixels/dpi)/0.8/marginHeight\n",
    "\n",
    "    # sns.set_context(\"paper\", font_scale=1.8)\n",
    "    grid = sns.clustermap(dataFrame.fillna(0),\n",
    "                       row_colors = [tissue_colors[x] for x in knn_cov_r2.columns], \n",
    "                       col_colors = cov_color,\n",
    "                       mask = pd.isna(dataFrame),\n",
    "                       figsize=(figWidth, figHeigh),\n",
    "                       cmap=\"Blues\",\n",
    "                       linewidths=.95,\n",
    "                       cbar_kws={\"orientation\": \"vertical\", \"label\":\"RÂ²\"},\n",
    "                       robust=True ) #col_cluster=False)\n",
    "    grid.ax_row_dendrogram.set_visible(False)\n",
    "    grid.cax.set_visible(False)\n",
    "#     grid.ax_col_dendrogram.set_visible(False)\n",
    "    s = grid.ax_heatmap.set_xticklabels([i for i in g.ax_heatmap.get_xticklabels()], rotation=45, ha='right')\n",
    "\n",
    "    # calculate the size of the heatmap axes\n",
    "    axWidth = (Nx*cellSizePixels)/(figWidth*dpi)\n",
    "    axHeight = (Ny*cellSizePixels)/(figHeigh*dpi)\n",
    "    \n",
    "    print(axWidth, axHeight)\n",
    "    # resize dendrograms to match\n",
    "    ax_row_orig_pos = grid.ax_row_dendrogram.get_position()\n",
    "#     grid.ax_row_dendrogram.set_position([ax_row_orig_pos.x0, ax_row_orig_pos.y0, \n",
    "#                                          ax_row_orig_pos.width, axHeight])\n",
    "    ax_col_orig_pos = grid.ax_col_dendrogram.get_position()\n",
    "#     grid.ax_col_dendrogram.set_position([ax_col_orig_pos.x0, ax_heatmap_orig_pos.y0+axHeight,\n",
    "#                                          axWidth, ax_col_orig_pos.height])\n",
    "    \n",
    "#     ax_colcolors_orig_pos = grid.ax_col_colors.get_position()\n",
    "#     grid.ax_col_colors.set_position(ax_colcolors_orig_pos.x0, ax_colcolors_orig_pos.y0, \n",
    "#                                     axWidth, ax_colcolors_orig_pos+axHeight)\n",
    "    \n",
    "#     print(grid.ax_col_colors.get_position())\n",
    "#     # resize heatmap\n",
    "#     ax_heatmap_orig_pos = grid.ax_heatmap.get_position()\n",
    "# #     grid.ax_heatmap.set_position([ax_heatmap_orig_pos.x0, ax_heatmap_orig_pos.y0, \n",
    "# #                                   axWidth, axHeight])\n",
    "#     grid.ax_heatmap.set_position([ax_col_orig_pos.x0, ax_heatmap_orig_pos.y0, \n",
    "#                                   axWidth, axHeight])\n",
    "    return grid\n",
    "    \n",
    "grid = fixedWidthClusterMap(tmp_df, cellSizePixels=100)\n",
    "grid.savefig(\"clustermap_squared_knn.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns; sns.set()\n",
    "\n",
    "tmp_df = knn_cov_r2.T.copy(deep=True)\n",
    "tmp_df.index = [tissue_names[x] for x in tmp_df.index]\n",
    "tmp_df.columns = new_index\n",
    "\n",
    "vmin = np.percentile(tmp_df.values, 2)\n",
    "vmax = np.percentile(tmp_df.values, 98)\n",
    "\n",
    "sns.set_context(\"paper\", font_scale=1.8)\n",
    "g = sns.heatmap(tmp_df,\n",
    "                cmap=\"Blues\",\n",
    "                linewidths=.75,\n",
    "                cbar_kws={\"orientation\": \"vertical\", \"label\":\"RÂ²\"},\n",
    "                robust=True ) #col_cluster=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
