{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# shapeit_file = \"/cbscratch/franco/from_saikat/gtex_v8_202003/all_variants_pvalues_tejaas.txt\"\n",
    "# shapeit_df = pd.read_csv(shapeit_file, sep=\"\\t\", header=0, index_col=0)\n",
    "# snps_list = list(shapeit_df.index)\n",
    "\n",
    "raqtl_snp_file = \"/cbscratch/franco/datasets/raQTLs/SuRE_SNP_table_LP190708.txt.gz\"\n",
    "sure_df = pd.read_csv(raqtl_snp_file, sep=\"\\t\", header=0, usecols=[1])\n",
    "snps_list = list(sure_df.values.reshape(-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip \n",
    "\n",
    "dbsnp_dict = dict()\n",
    "with gzip.open(\"/cbscratch/franco/datasets/gtex_v8/genotypes/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz\") as instream:\n",
    "    next(instream)\n",
    "    for line in instream:\n",
    "        arr = line.decode().rstrip().split()\n",
    "        #if len(arr[3]) == 1 and len(arr[4]) == 1:\n",
    "        dbsnp_dict[arr[2]] = arr[6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('../../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "import json\n",
    "from utils import utils\n",
    "\n",
    "#from utils import mpl_stylesheet\n",
    "#mpl_stylesheet.banskt_presentation(fontfamily = 'latex-clearsans', fontsize = 22, colors = 'banskt', dpi = 300)\n",
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/analysis/plots/tissue_table.txt\"\n",
    "tissuenames, descriptions, tstring = utils.read_tissues_str(tissue_file)\n",
    "\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors = dict()\n",
    "tissue_names = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tissuenames, descriptions):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading  as\n",
      "Loading  av\n",
      "Loading  ag\n",
      "Loading  aa\n",
      "Loading  ac\n",
      "Loading  at\n",
      "Loading  bam\n",
      "Loading  ban\n",
      "Loading  bca\n",
      "Loading  bceh\n",
      "Loading  bce\n",
      "Loading  bco\n",
      "Loading  bfr\n",
      "Loading  bhi\n",
      "Loading  bhy\n",
      "Loading  bnu\n",
      "Loading  bpu\n",
      "Loading  bsp\n",
      "Loading  bsu\n",
      "Loading  br\n",
      "Loading  ebv\n",
      "Loading  fib\n",
      "Loading  cols\n",
      "Loading  colt\n",
      "Loading  esog\n",
      "Loading  esom\n",
      "Loading  esomu\n",
      "Loading  haa\n",
      "haa has 0 trans-eqtls\n",
      "Loading  hlv\n",
      "Loading  kc\n",
      "Loading  liv\n",
      "Loading  lu\n",
      "Loading  msg\n",
      "Loading  ms\n",
      "Loading  nt\n",
      "Loading  ov\n",
      "Loading  pan\n",
      "pan has 0 trans-eqtls\n",
      "Loading  pit\n",
      "Loading  pro\n",
      "Loading  snse\n",
      "Loading  sse\n",
      "Loading  si\n",
      "Loading  spl\n",
      "Loading  sto\n",
      "Loading  tes\n",
      "Loading  thy\n",
      "Loading  ut\n",
      "Loading  va\n",
      "va has 0 trans-eqtls\n",
      "Loading  wb\n",
      "wb has 0 trans-eqtls\n",
      "uniq trans-eqtls: 7481\n"
     ]
    }
   ],
   "source": [
    "import collections\n",
    "import re\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'dbSNP', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "def tejaas(filepath, dbsnp_dict):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = rsid.split(\"_\")[0][3:]\n",
    "            pos   = int(arr[2])\n",
    "            p     = float(arr[7])\n",
    "            logp  = np.log10(p) if p!=0 else np.log10(10e-30)\n",
    "            maf   = float(arr[3])\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, dbSNP=dbsnp_dict[rsid], maf=maf))\n",
    "    return res\n",
    "\n",
    "def smart_LD_filter(trans_eqtls, trans_eqtls_ld_regions_file):\n",
    "    pass_snps = list()\n",
    "    with open(trans_eqtls_ld_regions_file) as instream:\n",
    "        for line in instream:\n",
    "            arr = line.strip(\"\\n\").split(\"\\t\")\n",
    "            if len(arr[4]) > 0:\n",
    "                pass_snps.append(arr[0])\n",
    "    return [x for x in trans_eqtls if x.rsid in pass_snps]\n",
    "\n",
    "paths = dict()\n",
    "paths[\"gamma01\"] = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma01_knn30_cut5e-8/\"\n",
    "paths[\"optim\"] = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_optim_gamma_knn30_cut5e-8/\"\n",
    "\n",
    "PATH=\"gamma01\"\n",
    "# PATH=\"optim\" \n",
    "set_bg=\"old\"\n",
    "basepath = paths[PATH]\n",
    "trans_dict = dict()\n",
    "trans_filename = \"trans_eqtls_ldpruned.txt\"\n",
    "for tissue in tissuenames:\n",
    "    tejaas_file = os.path.join(basepath, tissue, trans_filename)\n",
    "    if not os.path.exists(tejaas_file):\n",
    "        print(\"{:s} has no trans-eqtl results\".format(tissue))\n",
    "        continue\n",
    "    print(\"Loading \", tissue)\n",
    "    transeqtls = tejaas(tejaas_file, dbsnp_dict)\n",
    "    if len(transeqtls) > 0:\n",
    "        trans_dict[tissue] = transeqtls\n",
    "        if re.search(\"ldpruned\",trans_filename):\n",
    "            regions_file = os.path.join(basepath, tissue, \"ld_regions.txt\")\n",
    "            trans_dict[tissue] = smart_LD_filter(transeqtls, regions_file)\n",
    "    else:\n",
    "        trans_dict[tissue] = []\n",
    "        print(\"{:s} has 0 trans-eqtls\".format(tissue))\n",
    "\n",
    "all_transeqtls = list()\n",
    "for tissue in tissuenames:\n",
    "    all_transeqtls += [x.rsid for x in trans_dict[tissue]]\n",
    "    \n",
    "all_uniq_teqtls = list(set(all_transeqtls))\n",
    "print(\"uniq trans-eqtls:\", len(all_uniq_teqtls))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "k562_file_control = \"/cbscratch/franco/datasets/raQTLs/k562.matched.control.LP190708.txt.gz\"\n",
    "k562_file = \"/cbscratch/franco/datasets/raQTLs/k562.sign.id.LP190708.txt.gz\"\n",
    "hepg2_file_control = \"/cbscratch/franco/datasets/raQTLs/hepg2.matched.control.LP190708.txt.gz\"\n",
    "hepg2_file = \"/cbscratch/franco/datasets/raQTLs/hepg2.sign.id.LP190708.txt.gz\"\n",
    "\n",
    "def load_raqtls(filename):\n",
    "    raqtl_dict = collections.defaultdict(lambda: False)\n",
    "    with gzip.open(filename) as instream:\n",
    "        header = next(instream).decode().strip().split(\"\\t\")\n",
    "        for line in instream:\n",
    "            arr = line.decode().strip().split(\"\\t\")\n",
    "            raqtl_dict[arr[1]] = True\n",
    "    return raqtl_dict\n",
    "        \n",
    "# similar to Blood\n",
    "raqtl_k562 = load_raqtls(k562_file)\n",
    "raqtl_k562_control = load_raqtls(k562_file_control)\n",
    "\n",
    "# similar to Liver\n",
    "raqtl_hepg2 = load_raqtls(hepg2_file)\n",
    "raqtl_hepg2_control = load_raqtls(hepg2_file_control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_rand_bg(snps_list, raqtl_dict, dbsnp_dict, nchoose = 20000, niter = 50):\n",
    "    counters = list()\n",
    "    print(f'Iteration', end=\"\")\n",
    "    for k in range(niter):\n",
    "        counter = 0\n",
    "        chooseidx = np.sort(np.random.choice(len(snps_list), nchoose, replace = False))\n",
    "        print(f' {k}', end=\"\")\n",
    "        for idx in chooseidx:\n",
    "            var_id = snps_list[idx]\n",
    "            if raqtl_dict[dbsnp_dict[var_id]]:\n",
    "                counter += 1\n",
    "        counters.append(counter)\n",
    "    print(\"\")\n",
    "    return np.mean(np.array(counters)) / nchoose\n",
    "\n",
    "def sample_rand_bg_dbSNP(snps_list, raqtl_dict, dbsnp_dict, nchoose = 20000, niter = 50):\n",
    "    counters = list()\n",
    "    print(f'Iteration', end=\"\")\n",
    "    for k in range(niter):\n",
    "        counter = 0\n",
    "        chooseidx = np.sort(np.random.choice(len(snps_list), nchoose, replace = False))\n",
    "        print(f' {k}', end=\"\")\n",
    "        for idx in chooseidx:\n",
    "            var_id = snps_list[idx]\n",
    "            if raqtl_dict[var_id]:\n",
    "                counter += 1\n",
    "        counters.append(counter)\n",
    "    print(\"\")\n",
    "    return np.mean(np.array(counters)) / nchoose\n",
    "\n",
    "# # K562 background\n",
    "# bg_k562         = sample_rand_bg_dbSNP(snps_list, raqtl_k562, dbsnp_dict)\n",
    "# bg_control_k562 = sample_rand_bg_dbSNP(snps_list, raqtl_k562_control, dbsnp_dict)\n",
    "\n",
    "# # HEPG2 background\n",
    "# bg_hepg2         = sample_rand_bg_dbSNP(snps_list, raqtl_hepg2, dbsnp_dict)\n",
    "# bg_control_hepg2 = sample_rand_bg_dbSNP(snps_list, raqtl_hepg2_control, dbsnp_dict)\n",
    "\n",
    "# import json\n",
    "# bg_data = dict(zip(['k562', 'k562_control', 'hpeg2', 'hpeg2_control'], [bg_k562, bg_control_k562, bg_hepg2, bg_control_hepg2]))\n",
    "# bg_json = json.dumps(bg_data)\n",
    "# with open(\"/cbscratch/franco/datasets/raQTLs/raqtls_background_orig_paper.txt\", 'w') as outstream:\n",
    "#     outstream.write(bg_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k562 as 1 0.0038461538461538464 2.607561929595828 0.31872082598228424\n",
      "k562 av 0 0.0 0.0 1.0\n",
      "k562 ag 0 0.0 0.0 1.0\n",
      "k562 aa 1 0.001402524544179523 0.9508640977488293 0.6509190285626489\n",
      "k562 ac 3 0.0035587188612099642 2.412690753362688 0.129938964368404\n",
      "k562 at 0 0.0 0.0 1.0\n",
      "k562 bam 3 0.005136986301369863 3.482702577199907 0.05652066421435423\n",
      "k562 ban 2 0.0045045045045045045 3.0539013589861046 0.14020258483987413\n",
      "k562 bca 0 0.0 0.0 1.0\n",
      "k562 bceh 0 0.0 0.0 1.0\n",
      "k562 bce 0 0.0 0.0 1.0\n",
      "k562 bco 0 0.0 0.0 1.0\n",
      "k562 bfr 0 0.0 0.0 1.0\n",
      "k562 bhi 0 0.0 0.0 1.0\n",
      "k562 bhy 0 0.0 0.0 1.0\n",
      "k562 bnu 0 0.0 0.0 1.0\n",
      "k562 bpu 0 0.0 0.0 1.0\n",
      "k562 bsp 0 0.0 0.0 1.0\n",
      "k562 bsu 0 0.0 0.0 1.0\n",
      "k562 br 1 0.00398406374501992 2.7010601661152003 0.30960978325230903\n",
      "k562 ebv 4 0.025157232704402517 17.055750986035605 0.0001011256392547195\n",
      "k562 fib 0 0.0 0.0 1.0\n",
      "k562 cols 0 0.0 0.0 1.0\n",
      "k562 colt 3 0.004511278195488722 3.0584936918567607 0.07668164067682359\n",
      "k562 esog 0 0.0 0.0 1.0\n",
      "k562 esom 0 0.0 0.0 1.0\n",
      "k562 esomu 0 0.0 0.0 1.0\n",
      "k562 hlv 0 0.0 0.0 1.0\n",
      "k562 kc 0 0.0 0.0 1.0\n",
      "k562 liv 0 0.0 0.0 1.0\n",
      "k562 lu 1 0.003968253968253968 2.690341673392521 0.3106281088220119\n",
      "k562 msg 0 0.0 0.0 1.0\n",
      "k562 ms 11 0.008409785932721712 5.701549784896076 5.792613934193455e-06\n",
      "k562 nt 1 0.0030959752321981426 2.0989662591173848 0.3792188715454711\n",
      "k562 ov 0 0.0 0.0 1.0\n",
      "k562 pit 0 0.0 0.0 1.0\n",
      "k562 pro 0 0.0 0.0 1.0\n",
      "k562 snse 2 0.005089058524173028 3.4502091689308667 0.11515419887768188\n",
      "k562 sse 1 0.004901960784313725 3.3233632436025258 0.26001260906555823\n",
      "k562 si 0 0.0 0.0 1.0\n",
      "k562 spl 0 0.0 0.0 1.0\n",
      "k562 sto 1 0.007874015748031496 5.338315761377285 0.17094044609947723\n",
      "k562 tes 2 0.002932551319648094 1.9881703862020976 0.26643153711521605\n",
      "k562 thy 1 0.0019569471624266144 1.326743838933298 0.5296515233842213\n",
      "k562 ut 3 0.0031982942430703624 2.1683350800477035 0.16243177533857284\n",
      "k562 all 29 0.0038764871006549927 2.62812684790169 4.861026261337149e-06\n",
      "k562_control as 0 0.0 0.0 1.0\n",
      "k562_control av 0 0.0 0.0 1.0\n",
      "k562_control ag 0 0.0 0.0 1.0\n",
      "k562_control aa 1 0.001402524544179523 0.9095489910373042 0.6672269474713644\n",
      "k562_control ac 3 0.0035587188612099642 2.3078591836640494 0.14274136191643233\n",
      "k562_control at 1 0.005405405405405406 3.505450976268097 0.24835426982972464\n",
      "k562_control bam 3 0.005136986301369863 3.331378924364373 0.0628040475574096\n",
      "k562_control ban 1 0.0022522522522522522 1.4606045734450404 0.4959986875668503\n",
      "k562_control bca 0 0.0 0.0 1.0\n",
      "k562_control bceh 0 0.0 0.0 1.0\n",
      "k562_control bce 0 0.0 0.0 1.0\n",
      "k562_control bco 0 0.0 0.0 1.0\n",
      "k562_control bfr 0 0.0 0.0 1.0\n",
      "k562_control bhi 0 0.0 0.0 1.0\n",
      "k562_control bhy 0 0.0 0.0 1.0\n",
      "k562_control bnu 0 0.0 0.0 1.0\n",
      "k562_control bpu 0 0.0 0.0 1.0\n",
      "k562_control bsp 0 0.0 0.0 1.0\n",
      "k562_control bsu 0 0.0 0.0 1.0\n",
      "k562_control br 0 0.0 0.0 1.0\n",
      "k562_control ebv 0 0.0 0.0 1.0\n",
      "k562_control fib 1 0.00909090909090909 5.895531187359981 0.15612501012172744\n",
      "k562_control cols 0 0.0 0.0 1.0\n",
      "k562_control colt 0 0.0 0.0 1.0\n",
      "k562_control esog 0 0.0 0.0 1.0\n",
      "k562_control esom 0 0.0 0.0 1.0\n",
      "k562_control esomu 0 0.0 0.0 1.0\n",
      "k562_control hlv 0 0.0 0.0 1.0\n",
      "k562_control kc 0 0.0 0.0 1.0\n",
      "k562_control liv 0 0.0 0.0 1.0\n",
      "k562_control lu 0 0.0 0.0 1.0\n",
      "k562_control msg 0 0.0 0.0 1.0\n",
      "k562_control ms 2 0.0015290519877675841 0.9916031049076421 0.5987694047379096\n",
      "k562_control nt 0 0.0 0.0 1.0\n",
      "k562_control ov 0 0.0 0.0 1.0\n",
      "k562_control pit 0 0.0 0.0 1.0\n",
      "k562_control pro 1 0.002178649237472767 1.4128723978422615 0.5075312432552108\n",
      "k562_control snse 2 0.005089058524173028 3.300297356791847 0.12378236486710052\n",
      "k562_control sse 2 0.00980392156862745 6.357925790290175 0.04010674588135592\n",
      "k562_control si 0 0.0 0.0 1.0\n",
      "k562_control spl 0 0.0 0.0 1.0\n",
      "k562_control sto 1 0.007874015748031496 5.106365595351165 0.17797555293186343\n",
      "k562_control tes 1 0.001466275659824047 0.9508921269935454 0.6509205219868038\n",
      "k562_control thy 4 0.007827788649706457 5.076386932364758 0.008562681564744693\n",
      "k562_control ut 2 0.0021321961620469083 1.3827471868008485 0.4241935629156986\n",
      "k562_control all 19 0.0025397674107739605 1.6470605776744232 0.02677601962058642\n",
      "hpeg2 as 0 0.0 0.0 1.0\n",
      "hpeg2 av 0 0.0 0.0 1.0\n",
      "hpeg2 ag 0 0.0 0.0 1.0\n",
      "hpeg2 aa 1 0.001402524544179523 1.1926229117172815 0.567849519897512\n",
      "hpeg2 ac 3 0.0035587188612099642 3.026121480620718 0.07861224139996902\n",
      "hpeg2 at 0 0.0 0.0 1.0\n",
      "hpeg2 bam 1 0.0017123287671232876 1.4560618768055167 0.4970115874730212\n",
      "hpeg2 ban 1 0.0022522522522522522 1.915180486609058 0.4069349927284164\n",
      "hpeg2 bca 0 0.0 0.0 1.0\n",
      "hpeg2 bceh 0 0.0 0.0 1.0\n",
      "hpeg2 bce 0 0.0 0.0 1.0\n",
      "hpeg2 bco 0 0.0 0.0 1.0\n",
      "hpeg2 bfr 0 0.0 0.0 1.0\n",
      "hpeg2 bhi 0 0.0 0.0 1.0\n",
      "hpeg2 bhy 0 0.0 0.0 1.0\n",
      "hpeg2 bnu 0 0.0 0.0 1.0\n",
      "hpeg2 bpu 1 0.5 425.1700680272109 0.0023506170240000005\n",
      "hpeg2 bsp 0 0.0 0.0 1.0\n",
      "hpeg2 bsu 0 0.0 0.0 1.0\n",
      "hpeg2 br 0 0.0 0.0 1.0\n",
      "hpeg2 ebv 1 0.006289308176100629 5.348051170153597 0.17063425491957157\n",
      "hpeg2 fib 0 0.0 0.0 1.0\n",
      "hpeg2 cols 0 0.0 0.0 1.0\n",
      "hpeg2 colt 2 0.0030075187969924814 2.5574139430208174 0.18472098001737505\n",
      "hpeg2 esog 0 0.0 0.0 1.0\n",
      "hpeg2 esom 0 0.0 0.0 1.0\n",
      "hpeg2 esomu 0 0.0 0.0 1.0\n",
      "hpeg2 hlv 0 0.0 0.0 1.0\n",
      "hpeg2 kc 0 0.0 0.0 1.0\n",
      "hpeg2 liv 1 0.043478260869565216 36.97131026323573 0.02670097052972858\n",
      "hpeg2 lu 1 0.003968253968253968 3.3743656192635783 0.2566039924885052\n",
      "hpeg2 msg 0 0.0 0.0 1.0\n",
      "hpeg2 ms 3 0.0022935779816513763 1.9503214129688575 0.20073437587513251\n",
      "hpeg2 nt 1 0.0030959752321981426 2.632632000168489 0.3161875028312235\n",
      "hpeg2 ov 0 0.0 0.0 1.0\n",
      "hpeg2 pit 0 0.0 0.0 1.0\n",
      "hpeg2 pro 1 0.002178649237472767 1.852592889007455 0.4173109763416103\n",
      "hpeg2 snse 0 0.0 0.0 1.0\n",
      "hpeg2 sse 0 0.0 0.0 1.0\n",
      "hpeg2 si 0 0.0 0.0 1.0\n",
      "hpeg2 spl 0 0.0 0.0 1.0\n",
      "hpeg2 sto 0 0.0 0.0 1.0\n",
      "hpeg2 tes 0 0.0 0.0 1.0\n",
      "hpeg2 thy 0 0.0 0.0 1.0\n",
      "hpeg2 ut 1 0.0010660980810234541 0.906545987264842 0.6683705754905686\n",
      "hpeg2 all 14 0.0018714075658334448 1.5913329641440859 0.06399954013576524\n",
      "hpeg2_control as 1 0.0038461538461538464 3.1786395422759064 0.27005820921870416\n",
      "hpeg2_control av 0 0.0 0.0 1.0\n",
      "hpeg2_control ag 0 0.0 0.0 1.0\n",
      "hpeg2_control aa 2 0.002805049088359046 2.318222387073592 0.21388193046119555\n",
      "hpeg2_control ac 4 0.004744958481613286 3.921453290589493 0.020164519579242438\n",
      "hpeg2_control at 0 0.0 0.0 1.0\n",
      "hpeg2_control bam 0 0.0 0.0 1.0\n",
      "hpeg2_control ban 0 0.0 0.0 1.0\n",
      "hpeg2_control bca 0 0.0 0.0 1.0\n",
      "hpeg2_control bceh 0 0.0 0.0 1.0\n",
      "hpeg2_control bce 0 0.0 0.0 1.0\n",
      "hpeg2_control bco 0 0.0 0.0 1.0\n",
      "hpeg2_control bfr 0 0.0 0.0 1.0\n",
      "hpeg2_control bhi 0 0.0 0.0 1.0\n",
      "hpeg2_control bhy 0 0.0 0.0 1.0\n",
      "hpeg2_control bnu 0 0.0 0.0 1.0\n",
      "hpeg2_control bpu 0 0.0 0.0 1.0\n",
      "hpeg2_control bsp 0 0.0 0.0 1.0\n",
      "hpeg2_control bsu 0 0.0 0.0 1.0\n",
      "hpeg2_control br 1 0.00398406374501992 3.2926146653057193 0.26206083712270606\n",
      "hpeg2_control ebv 1 0.006289308176100629 5.19777535214928 0.17511103528377664\n",
      "hpeg2_control fib 0 0.0 0.0 1.0\n",
      "hpeg2_control cols 0 0.0 0.0 1.0\n",
      "hpeg2_control colt 1 0.0015037593984962407 1.2427763623935875 0.5529734125138959\n",
      "hpeg2_control esog 0 0.0 0.0 1.0\n",
      "hpeg2_control esom 0 0.0 0.0 1.0\n",
      "hpeg2_control esomu 0 0.0 0.0 1.0\n",
      "hpeg2_control hlv 0 0.0 0.0 1.0\n",
      "hpeg2_control kc 0 0.0 0.0 1.0\n",
      "hpeg2_control liv 0 0.0 0.0 1.0\n",
      "hpeg2_control lu 1 0.003968253968253968 3.2795487340941887 0.26295374350978756\n",
      "hpeg2_control msg 0 0.0 0.0 1.0\n",
      "hpeg2_control ms 3 0.0022935779816513763 1.8955189931003111 0.21211034069801846\n",
      "hpeg2_control nt 1 0.0030959752321981426 2.55865721669268 0.3236649185029628\n",
      "hpeg2_control ov 0 0.0 0.0 1.0\n",
      "hpeg2_control pit 0 0.0 0.0 1.0\n",
      "hpeg2_control pro 1 0.002178649237472767 1.800536559894849 0.42634452592338223\n",
      "hpeg2_control snse 5 0.01272264631043257 10.514583727630223 0.00013384222301746743\n",
      "hpeg2_control sse 0 0.0 0.0 1.0\n",
      "hpeg2_control si 0 0.0 0.0 1.0\n",
      "hpeg2_control spl 0 0.0 0.0 1.0\n",
      "hpeg2_control sto 0 0.0 0.0 1.0\n",
      "hpeg2_control tes 0 0.0 0.0 1.0\n",
      "hpeg2_control thy 2 0.003913894324853229 3.2346234089696106 0.12788957187870648\n",
      "hpeg2_control ut 4 0.0042643923240938165 3.5242911769370386 0.028314674697762568\n",
      "hpeg2_control all 18 0.0024060954417858575 1.988508629575089 0.0055984533592419025\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as ss\n",
    "\n",
    "if set_bg == \"old\":\n",
    "    with open(\"/cbscratch/franco/datasets/raQTLs/raqtls_background_shapeit2.txt\") as ins:\n",
    "        bg_data = json.load(ins)\n",
    "if set_bg == \"orig\":\n",
    "    with open(\"/cbscratch/franco/datasets/raQTLs/raqtls_background_orig_paper.txt\") as ins:\n",
    "        bg_data = json.load(ins)\n",
    "\n",
    "data = dict(zip(['k562', 'k562_control', 'hpeg2', 'hpeg2_control'],[raqtl_k562, raqtl_k562_control, raqtl_hepg2, raqtl_hepg2_control]))\n",
    "results = collections.defaultdict(dict)\n",
    "\n",
    "for dataset in data:\n",
    "    raqtl_dict = data[dataset]\n",
    "    this_res = collections.defaultdict(dict)\n",
    "    for tissue in tissuenames:\n",
    "        if len(trans_dict[tissue]) > 0:    \n",
    "            counter = 0\n",
    "            for teqtl in trans_dict[tissue]:\n",
    "                if raqtl_dict[teqtl.dbSNP]:\n",
    "                    counter += 1\n",
    "            frac = counter / len(trans_dict[tissue])\n",
    "            enrichment = frac/bg_data[dataset]\n",
    "            pval = ss.binom_test(counter, len(trans_dict[tissue]), bg_data[dataset], alternative='greater')\n",
    "            print(dataset, tissue, counter, frac, enrichment, pval)\n",
    "            this_res[tissue]['count'] = counter\n",
    "            this_res[tissue]['frac'] = frac\n",
    "            this_res[tissue]['enrichment'] = enrichment\n",
    "            this_res[tissue]['pval'] = pval\n",
    "    counter = 0\n",
    "    for teqtl in all_uniq_teqtls:\n",
    "        if raqtl_dict[dbsnp_dict[teqtl]]:\n",
    "            counter += 1\n",
    "    frac = counter / len(all_uniq_teqtls)\n",
    "    enrichment = frac/bg_data[dataset]\n",
    "    pval = ss.binom_test(counter, len(all_uniq_teqtls), bg_data[dataset], alternative='greater')\n",
    "    this_res['all']['count'] = counter\n",
    "    this_res['all']['frac'] = frac\n",
    "    this_res['all']['enrichment'] = enrichment\n",
    "    this_res['all']['pval'] = pval\n",
    "    print(dataset, 'all', counter, frac, enrichment, pval)\n",
    "    results[dataset] = this_res\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "json_dict = json.dumps(results)\n",
    "with open(f\"raqtl_enrichment_{set_bg}bg_{PATH}_ldpruned.txt\", 'w') as outstream:\n",
    "    outstream.write(json_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gamma01'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PATH"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
