{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# shapeit_file = \"/cbscratch/franco/from_saikat/gtex_v8_202003/all_variants_pvalues_tejaas.txt\"\n",
    "# shapeit_df = pd.read_csv(shapeit_file, sep=\"\\t\", header=0, index_col=0)\n",
    "# snps_list = list(shapeit_df.index)\n",
    "\n",
    "raqtl_snp_file = \"/cbscratch/franco/datasets/raQTLs/SuRE_SNP_table_LP190708.txt.gz\"\n",
    "sure_df = pd.read_csv(raqtl_snp_file, sep=\"\\t\", header=0, usecols=[1])\n",
    "snps_list = list(sure_df.values.reshape(-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip \n",
    "\n",
    "dbsnp_dict = dict()\n",
    "with gzip.open(\"/cbscratch/franco/datasets/gtex_v8/genotypes/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz\") as instream:\n",
    "    next(instream)\n",
    "    for line in instream:\n",
    "        arr = line.decode().rstrip().split()\n",
    "        #if len(arr[3]) == 1 and len(arr[4]) == 1:\n",
    "        dbsnp_dict[arr[2]] = arr[6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('../../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "import json\n",
    "from utils import utils\n",
    "\n",
    "#from utils import mpl_stylesheet\n",
    "#mpl_stylesheet.banskt_presentation(fontfamily = 'latex-clearsans', fontsize = 22, colors = 'banskt', dpi = 300)\n",
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/analysis/plots/tissue_table.txt\"\n",
    "tissuenames, descriptions, tstring = utils.read_tissues_str(tissue_file)\n",
    "\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors = dict()\n",
    "tissue_names = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tissuenames, descriptions):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]\n",
    "    \n",
    "# special_tissues = ['ag', 'haa', 'liv', 'msg', 'pan', 'pit', 'si', 'spl', 'va', 'wb']\n",
    "# special_tissues = ['haa', 'pan', 'spl', 'wb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading  as\n",
      "Loading  av\n",
      "Loading  ag\n",
      "Loading  aa\n",
      "Loading  ac\n",
      "Loading  at\n",
      "Loading  bam\n",
      "Loading  ban\n",
      "Loading  bca\n",
      "Loading  bceh\n",
      "Loading  bce\n",
      "Loading  bco\n",
      "Loading  bfr\n",
      "Loading  bhi\n",
      "Loading  bhy\n",
      "Loading  bnu\n",
      "Loading  bpu\n",
      "Loading  bsp\n",
      "Loading  bsu\n",
      "Loading  br\n",
      "Loading  ebv\n",
      "Loading  fib\n",
      "Loading  cols\n",
      "Loading  colt\n",
      "Loading  esog\n",
      "Loading  esom\n",
      "Loading  esomu\n",
      "Loading  haa\n",
      "Loading  hlv\n",
      "Loading  kc\n",
      "Loading  liv\n",
      "Loading  lu\n",
      "Loading  msg\n",
      "Loading  ms\n",
      "Loading  nt\n",
      "Loading  ov\n",
      "Loading  pan\n",
      "Loading  pit\n",
      "Loading  pro\n",
      "Loading  snse\n",
      "Loading  sse\n",
      "Loading  si\n",
      "Loading  spl\n",
      "Loading  sto\n",
      "Loading  tes\n",
      "Loading  thy\n",
      "Loading  ut\n",
      "Loading  va\n",
      "va has 0 trans-eqtls\n",
      "Loading  wb\n",
      "uniq trans-eqtls: 18851\n"
     ]
    }
   ],
   "source": [
    "import collections\n",
    "import re\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'dbSNP', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "def tejaas(filepath, dbsnp_dict):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = rsid.split(\"_\")[0][3:]\n",
    "            pos   = int(arr[2])\n",
    "            p     = float(arr[7])\n",
    "            logp  = np.log10(p) if p!=0 else np.log10(10e-30)\n",
    "            maf   = float(arr[3])\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, dbSNP=dbsnp_dict[rsid], maf=maf))\n",
    "    return res\n",
    "\n",
    "def smart_LD_filter(trans_eqtls, trans_eqtls_ld_regions_file):\n",
    "    pass_snps = list()\n",
    "    with open(trans_eqtls_ld_regions_file) as instream:\n",
    "        for line in instream:\n",
    "            arr = line.strip(\"\\n\").split(\"\\t\")\n",
    "            if len(arr[4]) > 0:\n",
    "                pass_snps.append(arr[0])\n",
    "    return [x for x in trans_eqtls if x.rsid in pass_snps]\n",
    "\n",
    "paths = dict()\n",
    "paths[\"gamma01\"] = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma01_knn30_cut5e-8/\"\n",
    "#paths[\"optim\"] = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_optim_gamma_knn30_cut5e-8/\"\n",
    "paths[\"gamma0006\"] = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma0006_knn30_cut5e-8/\"\n",
    "\n",
    "\n",
    "set_bg   = \"old\"\n",
    "set_path = \"gamma0006\"\n",
    "\n",
    "# PATH=f\"{set_path}_LDregions\" # with smartLD filter\n",
    "PATH=f\"{set_path}_noregions\" # without smartLD filter\n",
    "PATH=\"combined_noregions\" # without smartLD filter\n",
    "\n",
    "\n",
    "trans_dict = dict()\n",
    "trans_filename = \"trans_eqtls_ldpruned.txt\"\n",
    "for tissue in tissuenames:\n",
    "    if tissue in special_tissues:\n",
    "        basepath = paths['gamma0006']\n",
    "    else:\n",
    "        basepath = paths['gamma01']\n",
    "#     basepath = paths[set_path]\n",
    "    tejaas_file = os.path.join(basepath , tissue, trans_filename)\n",
    "    if not os.path.exists(tejaas_file):\n",
    "        print(\"{:s} has no trans-eqtl results\".format(tissue))\n",
    "        trans_dict[tissue] = []\n",
    "        continue\n",
    "    print(\"Loading \", tissue)\n",
    "    transeqtls = tejaas(tejaas_file, dbsnp_dict)\n",
    "    if len(transeqtls) > 0:\n",
    "        trans_dict[tissue] = transeqtls\n",
    "        if re.search(\"ldpruned\",trans_filename) and re.search(\"LDregions\", PATH):\n",
    "            regions_file = os.path.join(basepath, tissue, \"ld_regions.txt\")\n",
    "            trans_dict[tissue] = smart_LD_filter(transeqtls, regions_file)\n",
    "    else:\n",
    "        trans_dict[tissue] = []\n",
    "        print(\"{:s} has 0 trans-eqtls\".format(tissue))\n",
    "\n",
    "all_transeqtls = list()\n",
    "for tissue in tissuenames:\n",
    "    all_transeqtls += [x.rsid for x in trans_dict[tissue]]\n",
    "    \n",
    "all_uniq_teqtls = list(set(all_transeqtls))\n",
    "print(\"uniq trans-eqtls:\", len(all_uniq_teqtls))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "k562_file_control = \"/cbscratch/franco/datasets/raQTLs/k562.matched.control.LP190708.txt.gz\"\n",
    "k562_file = \"/cbscratch/franco/datasets/raQTLs/k562.sign.id.LP190708.txt.gz\"\n",
    "hepg2_file_control = \"/cbscratch/franco/datasets/raQTLs/hepg2.matched.control.LP190708.txt.gz\"\n",
    "hepg2_file = \"/cbscratch/franco/datasets/raQTLs/hepg2.sign.id.LP190708.txt.gz\"\n",
    "\n",
    "def load_raqtls(filename):\n",
    "    raqtl_dict = collections.defaultdict(lambda: False)\n",
    "    with gzip.open(filename) as instream:\n",
    "        header = next(instream).decode().strip().split(\"\\t\")\n",
    "        for line in instream:\n",
    "            arr = line.decode().strip().split(\"\\t\")\n",
    "            raqtl_dict[arr[1]] = True\n",
    "    return raqtl_dict\n",
    "        \n",
    "# similar to Blood\n",
    "raqtl_k562 = load_raqtls(k562_file)\n",
    "raqtl_k562_control = load_raqtls(k562_file_control)\n",
    "\n",
    "# similar to Liver\n",
    "raqtl_hepg2 = load_raqtls(hepg2_file)\n",
    "raqtl_hepg2_control = load_raqtls(hepg2_file_control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_rand_bg(snps_list, raqtl_dict, dbsnp_dict, nchoose = 20000, niter = 50):\n",
    "    counters = list()\n",
    "    print(f'Iteration', end=\"\")\n",
    "    for k in range(niter):\n",
    "        counter = 0\n",
    "        chooseidx = np.sort(np.random.choice(len(snps_list), nchoose, replace = False))\n",
    "        print(f' {k}', end=\"\")\n",
    "        for idx in chooseidx:\n",
    "            var_id = snps_list[idx]\n",
    "            if raqtl_dict[dbsnp_dict[var_id]]:\n",
    "                counter += 1\n",
    "        counters.append(counter)\n",
    "    print(\"\")\n",
    "    return np.mean(np.array(counters)) / nchoose\n",
    "\n",
    "def sample_rand_bg_dbSNP(snps_list, raqtl_dict, dbsnp_dict, nchoose = 20000, niter = 50):\n",
    "    counters = list()\n",
    "    print(f'Iteration', end=\"\")\n",
    "    for k in range(niter):\n",
    "        counter = 0\n",
    "        chooseidx = np.sort(np.random.choice(len(snps_list), nchoose, replace = False))\n",
    "        print(f' {k}', end=\"\")\n",
    "        for idx in chooseidx:\n",
    "            var_id = snps_list[idx]\n",
    "            if raqtl_dict[var_id]:\n",
    "                counter += 1\n",
    "        counters.append(counter)\n",
    "    print(\"\")\n",
    "    return np.mean(np.array(counters)) / nchoose\n",
    "\n",
    "# # K562 background\n",
    "# bg_k562         = sample_rand_bg_dbSNP(snps_list, raqtl_k562, dbsnp_dict)\n",
    "# bg_control_k562 = sample_rand_bg_dbSNP(snps_list, raqtl_k562_control, dbsnp_dict)\n",
    "\n",
    "# # HEPG2 background\n",
    "# bg_hepg2         = sample_rand_bg_dbSNP(snps_list, raqtl_hepg2, dbsnp_dict)\n",
    "# bg_control_hepg2 = sample_rand_bg_dbSNP(snps_list, raqtl_hepg2_control, dbsnp_dict)\n",
    "\n",
    "# import json\n",
    "# bg_data = dict(zip(['k562', 'k562_control', 'hepg2', 'hepg2_control'], [bg_k562, bg_control_k562, bg_hepg2, bg_control_hepg2]))\n",
    "# bg_json = json.dumps(bg_data)\n",
    "# with open(\"/cbscratch/franco/datasets/raQTLs/raqtls_background_orig_paper.txt\", 'w') as outstream:\n",
    "#     outstream.write(bg_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k562 as 3 0.005119453924914676 3.470816220281136 0.05698388226276416\n",
      "k562 av 0 0.0 0.0 1.0\n",
      "k562 ag 1 0.005434782608695652 3.6845983787767134 0.23784120646472884\n",
      "k562 aa 4 0.0030816640986132513 2.089263795670001 0.12765780702202212\n",
      "k562 ac 4 0.002599090318388564 1.7620951311108908 0.19447842196869689\n",
      "k562 at 0 0.0 0.0 1.0\n",
      "k562 bam 8 0.00676818950930626 4.588603057156787 0.0004503264147058246\n",
      "k562 ban 4 0.0041928721174004195 2.8426251643392675 0.05440121344573258\n",
      "k562 bca 0 0.0 0.0 1.0\n",
      "k562 bceh 0 0.0 0.0 1.0\n",
      "k562 bce 0 0.0 0.0 1.0\n",
      "k562 bco 0 0.0 0.0 1.0\n",
      "k562 bfr 0 0.0 0.0 1.0\n",
      "k562 bhi 0 0.0 0.0 1.0\n",
      "k562 bhy 0 0.0 0.0 1.0\n",
      "k562 bnu 1 0.002638522427440633 1.788828764366531 0.4284696474284174\n",
      "k562 bpu 0 0.0 0.0 1.0\n",
      "k562 bsp 0 0.0 0.0 1.0\n",
      "k562 bsu 0 0.0 0.0 1.0\n",
      "k562 br 3 0.005940594059405941 4.027521396207417 0.03970542416872813\n",
      "k562 ebv 5 0.016835016835016835 11.413570735604633 9.093065307333595e-05\n",
      "k562 fib 0 0.0 0.0 1.0\n",
      "k562 cols 0 0.0 0.0 1.0\n",
      "k562 colt 5 0.004258943781942078 2.88741951318107 0.03155919129470737\n",
      "k562 esog 0 0.0 0.0 1.0\n",
      "k562 esom 0 0.0 0.0 1.0\n",
      "k562 esomu 0 0.0 0.0 1.0\n",
      "k562 haa 1 0.015625 10.593220338983052 0.09014468674490916\n",
      "k562 hlv 2 0.008097165991902834 5.489604062307007 0.05213175647276049\n",
      "k562 kc 0 0.0 0.0 1.0\n",
      "k562 liv 0 0.0 0.0 1.0\n",
      "k562 lu 2 0.0041928721174004195 2.8426251643392675 0.1569723530840485\n",
      "k562 msg 0 0.0 0.0 1.0\n",
      "k562 ms 14 0.006213936972924989 4.212838625711857 1.0341406665627367e-05\n",
      "k562 nt 2 0.002777777777777778 1.8832391713747647 0.28705222096858285\n",
      "k562 ov 0 0.0 0.0 1.0\n",
      "k562 pan 1 0.00558659217877095 3.787520121200644 0.23219532693021178\n",
      "k562 pit 0 0.0 0.0 1.0\n",
      "k562 pro 2 0.002257336343115124 1.5303975207560163 0.3756819274052572\n",
      "k562 snse 3 0.004137931034482759 2.805376972530684 0.0933595720988503\n",
      "k562 sse 2 0.005 3.3898305084745766 0.11852051631926984\n",
      "k562 si 0 0.0 0.0 1.0\n",
      "k562 spl 1 0.0007385524372230429 0.500713516761385 0.8644784144068002\n",
      "k562 sto 1 0.003968253968253968 2.690341673392521 0.3106281088220119\n",
      "k562 tes 6 0.004665629860031105 3.163138888156681 0.013051968363884602\n",
      "k562 thy 4 0.004008016032064128 2.7172990047892394 0.06204600591407431\n",
      "k562 ut 5 0.0029708853238265003 2.014159541577288 0.1063382331419289\n",
      "k562 wb 18 0.002871271335141171 1.946624633994014 0.006871783518400274\n",
      "k562 all 64 0.003395045355684049 2.3017256648705415 2.9443640714002825e-09\n",
      "k562_control as 0 0.0 0.0 1.0\n",
      "k562_control av 2 0.004319654427645789 2.801332313648371 0.1605896497222466\n",
      "k562_control ag 0 0.0 0.0 1.0\n",
      "k562_control aa 1 0.0007704160246533128 0.49962128706440523 0.8650783119057199\n",
      "k562_control ac 4 0.002599090318388564 1.6855319833907678 0.21558122352747977\n",
      "k562_control at 1 0.0025575447570332483 1.6585893365974373 0.4530442257460322\n",
      "k562_control bam 3 0.0025380710659898475 1.6459604837807054 0.27541313067871254\n",
      "k562_control ban 1 0.0010482180293501049 0.679778229150522 0.7705805658866549\n",
      "k562_control bca 0 0.0 0.0 1.0\n",
      "k562_control bceh 0 0.0 0.0 1.0\n",
      "k562_control bce 0 0.0 0.0 1.0\n",
      "k562_control bco 0 0.0 0.0 1.0\n",
      "k562_control bfr 0 0.0 0.0 1.0\n",
      "k562_control bhi 0 0.0 0.0 1.0\n",
      "k562_control bhy 0 0.0 0.0 1.0\n",
      "k562_control bnu 0 0.0 0.0 1.0\n",
      "k562_control bpu 0 0.0 0.0 1.0\n",
      "k562_control bsp 0 0.0 0.0 1.0\n",
      "k562_control bsu 0 0.0 0.0 1.0\n",
      "k562_control br 1 0.0019801980198019802 1.2841751101180157 0.5412780396438697\n",
      "k562_control ebv 0 0.0 0.0 1.0\n",
      "k562_control fib 1 0.004784688995215311 3.1029111512420955 0.2756834309312652\n",
      "k562_control cols 0 0.0 0.0 1.0\n",
      "k562_control colt 2 0.0017035775127768314 1.1047843792327052 0.5404082818526906\n",
      "k562_control esog 0 0.0 0.0 1.0\n",
      "k562_control esom 0 0.0 0.0 1.0\n",
      "k562_control esomu 0 0.0 0.0 1.0\n",
      "k562_control haa 0 0.0 0.0 1.0\n",
      "k562_control hlv 0 0.0 0.0 1.0\n",
      "k562_control kc 1 0.07142857142857142 46.32203075782842 0.021372952444397974\n",
      "k562_control liv 0 0.0 0.0 1.0\n",
      "k562_control lu 1 0.0020964360587002098 1.359556458301044 0.5210225118929439\n",
      "k562_control msg 0 0.0 0.0 1.0\n",
      "k562_control ms 5 0.0022192632046160675 1.439210897935193 0.2696025201005083\n",
      "k562_control nt 1 0.001388888888888889 0.9007061536244416 0.6708023260671704\n",
      "k562_control ov 0 0.0 0.0 1.0\n",
      "k562_control pan 0 0.0 0.0 1.0\n",
      "k562_control pit 0 0.0 0.0 1.0\n",
      "k562_control pro 1 0.001128668171557562 0.731950824615799 0.7451974719293712\n",
      "k562_control snse 6 0.008275862068965517 5.366966322286328 0.0010364434862317473\n",
      "k562_control sse 2 0.005 3.2425421530479897 0.12736521537518392\n",
      "k562_control si 0 0.0 0.0 1.0\n",
      "k562_control spl 4 0.0029542097488921715 1.9158299279456366 0.15893763202110633\n",
      "k562_control sto 2 0.007936507936507936 5.14689230642538 0.05839300521305654\n",
      "k562_control tes 3 0.0023328149300155523 1.512850149167025 0.31872194008777616\n",
      "k562_control thy 4 0.004008016032064128 2.5992321868120154 0.07047443223048586\n",
      "k562_control ut 2 0.0011883541295306002 0.7706576715503244 0.731922464331931\n",
      "k562_control wb 17 0.0027117562609666615 1.758596796995241 0.020421158251719366\n",
      "k562_control all 48 0.0025462840167630365 1.6512866515973 0.0007888492418545743\n",
      "hepg2 as 0 0.0 0.0 1.0\n",
      "hepg2 av 0 0.0 0.0 1.0\n",
      "hepg2 ag 0 0.0 0.0 1.0\n",
      "hepg2 aa 2 0.0015408320493066256 1.3102313344444096 0.45108999926393906\n",
      "hepg2 ac 6 0.003898635477582846 3.3151662224343927 0.010587377942380785\n",
      "hepg2 at 1 0.0025575447570332483 2.1747829566609256 0.3687710111689422\n",
      "hepg2 bam 2 0.001692047377326565 1.4388157970463988 0.40480053712071024\n",
      "hepg2 ban 4 0.0041928721174004195 3.565367446769064 0.027305236015468456\n",
      "hepg2 bca 0 0.0 0.0 1.0\n",
      "hepg2 bceh 0 0.0 0.0 1.0\n",
      "hepg2 bce 0 0.0 0.0 1.0\n",
      "hepg2 bco 0 0.0 0.0 1.0\n",
      "hepg2 bfr 0 0.0 0.0 1.0\n",
      "hepg2 bhi 0 0.0 0.0 1.0\n",
      "hepg2 bhy 0 0.0 0.0 1.0\n",
      "hepg2 bnu 1 0.002638522427440633 2.243641519932511 0.3597946403211126\n",
      "hepg2 bpu 1 0.1 85.03401360544218 0.011697960844488623\n",
      "hepg2 bsp 0 0.0 0.0 1.0\n",
      "hepg2 bsu 0 0.0 0.0 1.0\n",
      "hepg2 br 1 0.0019801980198019802 1.6838418535731126 0.44801227786177467\n",
      "hepg2 ebv 2 0.006734006734006734 5.726196202386679 0.04839725361588491\n",
      "hepg2 fib 0 0.0 0.0 1.0\n",
      "hepg2 cols 1 0.013333333333333334 11.337868480725625 0.08446977649129242\n",
      "hepg2 colt 6 0.005110732538330494 4.345861001981713 0.0029699487560573153\n",
      "hepg2 esog 0 0.0 0.0 1.0\n",
      "hepg2 esom 0 0.0 0.0 1.0\n",
      "hepg2 esomu 0 0.0 0.0 1.0\n",
      "hepg2 haa 0 0.0 0.0 1.0\n",
      "hepg2 hlv 0 0.0 0.0 1.0\n",
      "hepg2 kc 0 0.0 0.0 1.0\n",
      "hepg2 liv 2 0.05405405405405406 45.9643316786174 0.0008961631405632952\n",
      "hepg2 lu 2 0.0041928721174004195 3.565367446769064 0.10913571912181891\n",
      "hepg2 msg 0 0.0 0.0 1.0\n",
      "hepg2 ms 4 0.001775410563692854 1.5097028602830391 0.27476825268924726\n",
      "hepg2 nt 2 0.002777777777777778 2.362055933484505 0.2080582873573183\n",
      "hepg2 ov 0 0.0 0.0 1.0\n",
      "hepg2 pan 1 0.00558659217877095 4.750503553376658 0.1899245386369149\n",
      "hepg2 pit 0 0.0 0.0 1.0\n",
      "hepg2 pro 3 0.003386004514672686 2.8792555396876582 0.08806655690424009\n",
      "hepg2 snse 0 0.0 0.0 1.0\n",
      "hepg2 sse 1 0.0025 2.1258503401360547 0.3754205977871917\n",
      "hepg2 si 0 0.0 0.0 1.0\n",
      "hepg2 spl 1 0.0007385524372230429 0.6280207799515671 0.7967342432226876\n",
      "hepg2 sto 0 0.0 0.0 1.0\n",
      "hepg2 tes 2 0.0015552099533437014 1.322457443319474 0.4463953424710894\n",
      "hepg2 thy 5 0.00501002004008016 4.260221122517143 0.007040293253248134\n",
      "hepg2 ut 1 0.0005941770647653001 0.5052526060929423 0.8619822097817278\n",
      "hepg2 wb 18 0.002871271335141171 2.4415572577731046 0.000646759463822181\n",
      "hepg2 all 45 0.0023871412657153466 2.0298820286695123 1.334288923269296e-05\n",
      "hepg2_control as 3 0.005119453924914676 4.230953656954278 0.035156421068090506\n",
      "hepg2_control av 0 0.0 0.0 1.0\n",
      "hepg2_control ag 1 0.005434782608695652 4.491555874955084 0.19970494931624375\n",
      "hepg2_control aa 6 0.004622496147919877 3.820244750347006 0.005502972770277337\n",
      "hepg2_control ac 7 0.004548408057179987 3.7590149232892456 0.003068944670944688\n",
      "hepg2_control at 0 0.0 0.0 1.0\n",
      "hepg2_control bam 3 0.0025380710659898475 2.0975793933800393 0.17376727443248355\n",
      "hepg2_control ban 2 0.0020964360587002098 1.7325917840497602 0.32084480230289175\n",
      "hepg2_control bca 0 0.0 0.0 1.0\n",
      "hepg2_control bceh 0 0.0 0.0 1.0\n",
      "hepg2_control bce 0 0.0 0.0 1.0\n",
      "hepg2_control bco 0 0.0 0.0 1.0\n",
      "hepg2_control bfr 0 0.0 0.0 1.0\n",
      "hepg2_control bhi 0 0.0 0.0 1.0\n",
      "hepg2_control bhy 0 0.0 0.0 1.0\n",
      "hepg2_control bnu 1 0.002638522427440633 2.180597047471598 0.36800112883013636\n",
      "hepg2_control bpu 0 0.0 0.0 1.0\n",
      "hepg2_control bsp 0 0.0 0.0 1.0\n",
      "hepg2_control bsu 0 0.0 0.0 1.0\n",
      "hepg2_control br 1 0.0019801980198019802 1.6365272890925457 0.4574201327866535\n",
      "hepg2_control ebv 1 0.003367003367003367 2.782647410746584 0.3020358889628345\n",
      "hepg2_control fib 0 0.0 0.0 1.0\n",
      "hepg2_control cols 1 0.013333333333333334 11.019283746556475 0.0868041858734266\n",
      "hepg2_control colt 3 0.002555366269165247 2.111872949723345 0.17140204853744065\n",
      "hepg2_control esog 0 0.0 0.0 1.0\n",
      "hepg2_control esom 0 0.0 0.0 1.0\n",
      "hepg2_control esomu 0 0.0 0.0 1.0\n",
      "hepg2_control haa 0 0.0 0.0 1.0\n",
      "hepg2_control hlv 0 0.0 0.0 1.0\n",
      "hepg2_control kc 0 0.0 0.0 1.0\n",
      "hepg2_control liv 0 0.0 0.0 1.0\n",
      "hepg2_control lu 3 0.006289308176100629 5.19777535214928 0.020839531451406993\n",
      "hepg2_control msg 0 0.0 0.0 1.0\n",
      "hepg2_control ms 5 0.0022192632046160675 1.8341018219967502 0.14088877295597674\n",
      "hepg2_control nt 3 0.004166666666666667 3.4435261707988984 0.058090243007686886\n",
      "hepg2_control ov 0 0.0 0.0 1.0\n",
      "hepg2_control pan 0 0.0 0.0 1.0\n",
      "hepg2_control pit 0 0.0 0.0 1.0\n",
      "hepg2_control pro 2 0.002257336343115124 1.8655672257149787 0.29074434798576326\n",
      "hepg2_control snse 6 0.008275862068965517 6.839555428897122 0.0002958097952982175\n",
      "hepg2_control sse 0 0.0 0.0 1.0\n",
      "hepg2_control si 0 0.0 0.0 1.0\n",
      "hepg2_control spl 2 0.0014771048744460858 1.2207478301207322 0.48748743887066803\n",
      "hepg2_control sto 2 0.007936507936507936 6.559097468188377 0.037941333965238994\n",
      "hepg2_control tes 5 0.0038880248833592537 3.213243705255582 0.02126288632958195\n",
      "hepg2_control thy 4 0.004008016032064128 3.3124099438546515 0.03433503931222295\n",
      "hepg2_control ut 8 0.004753416518122401 3.928443403406943 0.0012164937269546025\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hepg2_control wb 15 0.002392726112617642 1.9774595972046631 0.011216047266037218\n",
      "hepg2_control all 50 0.002652379184128163 2.192048912502614 5.764635462947197e-07\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as ss\n",
    "\n",
    "if set_bg == \"old\":\n",
    "    with open(\"/cbscratch/franco/datasets/raQTLs/raqtls_background_shapeit2.txt\") as ins:\n",
    "        bg_data = json.load(ins)\n",
    "if set_bg == \"orig\":\n",
    "    with open(\"/cbscratch/franco/datasets/raQTLs/raqtls_background_orig_paper.txt\") as ins:\n",
    "        bg_data = json.load(ins)\n",
    "\n",
    "data = dict(zip(['k562', 'k562_control', 'hepg2', 'hepg2_control'],[raqtl_k562, raqtl_k562_control, raqtl_hepg2, raqtl_hepg2_control]))\n",
    "results = collections.defaultdict(dict)\n",
    "\n",
    "variant_in_dataset = collections.defaultdict(list)\n",
    "for dataset in data:\n",
    "    raqtl_dict = data[dataset]\n",
    "    this_res = collections.defaultdict(dict)\n",
    "    for tissue in tissuenames:\n",
    "        if len(trans_dict[tissue]) > 0:    \n",
    "            counter = 0\n",
    "            for teqtl in trans_dict[tissue]:\n",
    "                if raqtl_dict[teqtl.dbSNP]:\n",
    "                    counter += 1\n",
    "            frac = counter / len(trans_dict[tissue])\n",
    "            enrichment = frac/bg_data[dataset]\n",
    "            pval = ss.binom_test(counter, len(trans_dict[tissue]), bg_data[dataset], alternative='greater')\n",
    "            print(dataset, tissue, counter, frac, enrichment, pval)\n",
    "            this_res[tissue]['count'] = counter\n",
    "            this_res[tissue]['frac'] = frac\n",
    "            this_res[tissue]['enrichment'] = enrichment\n",
    "            this_res[tissue]['pval'] = pval\n",
    "    counter = 0\n",
    "    for teqtl in all_uniq_teqtls:\n",
    "        if raqtl_dict[dbsnp_dict[teqtl]]:\n",
    "            variant_in_dataset[dataset].append(teqtl)\n",
    "            counter += 1\n",
    "    frac = counter / len(all_uniq_teqtls)\n",
    "    enrichment = frac/bg_data[dataset]\n",
    "    pval = ss.binom_test(counter, len(all_uniq_teqtls), bg_data[dataset], alternative='greater')\n",
    "    this_res['all']['count'] = counter\n",
    "    this_res['all']['frac'] = frac\n",
    "    this_res['all']['enrichment'] = enrichment\n",
    "    this_res['all']['pval'] = pval\n",
    "    print(dataset, 'all', counter, frac, enrichment, pval)\n",
    "    results[dataset] = this_res\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "json_dict = json.dumps(results)\n",
    "with open(f\"raqtl_enrichment_{set_bg}bg_{PATH}_ldpruned.txt\", 'w') as outstream:\n",
    "    outstream.write(json_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k562 64\n",
      "k562_control 48\n",
      "hepg2 45\n",
      "hepg2_control 50\n",
      "intersection: 8\n",
      "Total raQTLs: 101\n",
      "Total fraction of trans-eQTLs that are raQTLs: 0.005357805951938889\n"
     ]
    }
   ],
   "source": [
    "for d in variant_in_dataset:\n",
    "    print(d, len(variant_in_dataset[d]))\n",
    "    \n",
    "int_n = len(set.intersection(set(variant_in_dataset['k562']), set(variant_in_dataset['hepg2'])))\n",
    "print(\"intersection:\", int_n)\n",
    "total_raqtl = 64 + 45 - int_n\n",
    "print(f\"Total raQTLs: {total_raqtl}\")\n",
    "print(f\"Total fraction of trans-eQTLs that are raQTLs: {total_raqtl/len(all_uniq_teqtls)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
