{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "shapeit_file = \"/cbscratch/franco/from_saikat/gtex_v8_202003/all_variants_pvalues_tejaas.txt\"\n",
    "shapeit_df = pd.read_csv(shapeit_file, sep=\"\\t\", header=0, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip \n",
    "\n",
    "dbsnp_dict = dict()\n",
    "with gzip.open(\"/cbscratch/franco/datasets/gtex_v8/genotypes/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.txt.gz\") as instream:\n",
    "    next(instream)\n",
    "    for line in instream:\n",
    "        arr = line.decode().rstrip().split()\n",
    "        #if len(arr[3]) == 1 and len(arr[4]) == 1:\n",
    "        dbsnp_dict[arr[2]] = arr[6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('../../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "import json\n",
    "from utils import utils\n",
    "\n",
    "#from utils import mpl_stylesheet\n",
    "#mpl_stylesheet.banskt_presentation(fontfamily = 'latex-clearsans', fontsize = 22, colors = 'banskt', dpi = 300)\n",
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/analysis/plots/tissue_table.txt\"\n",
    "tissuenames, descriptions, tstring = utils.read_tissues_str(tissue_file)\n",
    "\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors = dict()\n",
    "tissue_names = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tissuenames, descriptions):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading  as\n",
      "Loading  av\n",
      "Loading  ag\n",
      "Loading  aa\n",
      "Loading  ac\n",
      "Loading  at\n",
      "Loading  bam\n",
      "Loading  ban\n",
      "Loading  bca\n",
      "Loading  bceh\n",
      "Loading  bce\n",
      "Loading  bco\n",
      "Loading  bfr\n",
      "Loading  bhi\n",
      "Loading  bhy\n",
      "Loading  bnu\n",
      "Loading  bpu\n",
      "Loading  bsp\n",
      "Loading  bsu\n",
      "Loading  br\n",
      "Loading  ebv\n",
      "Loading  fib\n",
      "Loading  cols\n",
      "Loading  colt\n",
      "Loading  esog\n",
      "Loading  esom\n",
      "Loading  esomu\n",
      "Loading  haa\n",
      "haa has 0 trans-eqtls\n",
      "Loading  hlv\n",
      "Loading  kc\n",
      "Loading  liv\n",
      "Loading  lu\n",
      "Loading  msg\n",
      "Loading  ms\n",
      "Loading  nt\n",
      "Loading  ov\n",
      "Loading  pan\n",
      "pan has 0 trans-eqtls\n",
      "Loading  pit\n",
      "Loading  pro\n",
      "Loading  snse\n",
      "Loading  sse\n",
      "Loading  si\n",
      "Loading  spl\n",
      "Loading  sto\n",
      "Loading  tes\n",
      "Loading  thy\n",
      "Loading  ut\n",
      "Loading  va\n",
      "va has 0 trans-eqtls\n",
      "Loading  wb\n",
      "wb has 0 trans-eqtls\n"
     ]
    }
   ],
   "source": [
    "import collections\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'dbSNP', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "def tejaas(filepath, dbsnp_dict):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = rsid.split(\"_\")[0][3:]\n",
    "            pos   = int(arr[2])\n",
    "            p     = float(arr[7])\n",
    "            logp  = np.log10(p) if p!=0 else np.log10(10e-30)\n",
    "            maf   = float(arr[3])\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, dbSNP=dbsnp_dict[rsid], maf=maf))\n",
    "    return res\n",
    "\n",
    "def smart_LD_filter(trans_eqtls, trans_eqtls_ld_regions_file):\n",
    "    pass_snps = list()\n",
    "    with open(trans_eqtls_ld_regions_file) as instream:\n",
    "        for line in instream:\n",
    "            arr = line.strip(\"\\n\").split(\"\\t\")\n",
    "            if len(arr[4]) > 0:\n",
    "                pass_snps.append(arr[0])\n",
    "    return [x for x in trans_eqtls if x.rsid in pass_snps]\n",
    "\n",
    "basepath = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma01_knn30_cut5e-8/\"\n",
    "trans_dict = dict()\n",
    "trans_filename = \"trans_eqtls.txt\" # \"trans_eqtls_ldpruned.txt\"\n",
    "for tissue in tissuenames:\n",
    "    tejaas_file = os.path.join(basepath, tissue, trans_filename)\n",
    "    if not os.path.exists(tejaas_file):\n",
    "        print(\"{:s} has no trans-eqtl results\".format(tissue))\n",
    "        continue\n",
    "    print(\"Loading \", tissue)\n",
    "    transeqtls = tejaas(tejaas_file, dbsnp_dict)\n",
    "    if len(transeqtls) > 0:\n",
    "        trans_dict[tissue] = transeqtls\n",
    "    else:\n",
    "        trans_dict[tissue] = []\n",
    "        print(\"{:s} has 0 trans-eqtls\".format(tissue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "raqtl_file_control = \"/cbscratch/franco/datasets/raQTLs/k562.matched.control.LP190708.txt.gz\"\n",
    "raqtl_file = \"/cbscratch/franco/datasets/raQTLs/k562.sign.id.LP190708.txt.gz\"\n",
    "\n",
    "raqtl_dict = collections.defaultdict(lambda: False)\n",
    "with gzip.open(raqtl_file) as instream:\n",
    "    header = next(instream).decode().strip().split(\"\\t\")\n",
    "    for line in instream:\n",
    "        arr = line.decode().strip().split(\"\\t\")\n",
    "        raqtl_dict[arr[1]] = True\n",
    "        \n",
    "raqtl_control_dict = collections.defaultdict(lambda: False)\n",
    "with gzip.open(raqtl_file_control) as instream:\n",
    "    header = next(instream).decode().strip().split(\"\\t\")\n",
    "    for line in instream:\n",
    "        arr = line.decode().strip().split(\"\\t\")\n",
    "        raqtl_control_dict[arr[1]] = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as 5 3\n",
      "av 4 7\n",
      "ag 1 5\n",
      "aa 26 13\n",
      "ac 29 23\n",
      "at 5 4\n",
      "bam 19 9\n",
      "ban 18 12\n",
      "bca 0 0\n",
      "bceh 1 2\n",
      "bce 0 0\n",
      "bco 0 1\n",
      "bfr 0 1\n",
      "bhi 0 0\n",
      "bhy 1 1\n",
      "bnu 3 4\n",
      "bpu 0 0\n",
      "bsp 0 0\n",
      "bsu 0 2\n",
      "br 9 3\n",
      "ebv 6 1\n",
      "fib 1 1\n",
      "cols 2 1\n",
      "colt 19 18\n",
      "esog 1 0\n",
      "esom 0 0\n",
      "esomu 0 0\n",
      "hlv 4 1\n",
      "kc 0 1\n",
      "liv 1 0\n",
      "lu 7 5\n",
      "msg 0 0\n",
      "ms 48 28\n",
      "nt 7 4\n",
      "ov 0 0\n",
      "pit 0 0\n",
      "pro 15 17\n",
      "snse 11 11\n",
      "sse 7 4\n",
      "si 0 0\n",
      "spl 0 0\n",
      "sto 4 3\n",
      "tes 20 14\n",
      "thy 14 9\n",
      "ut 27 21\n"
     ]
    }
   ],
   "source": [
    "for tissue in tissuenames:\n",
    "    if len(trans_dict[tissue]) > 0:\n",
    "        counter = 0\n",
    "        counter_control = 0\n",
    "        for teqtl in trans_dict[tissue]:\n",
    "            if raqtl_dict[teqtl.dbSNP]:\n",
    "                counter += 1\n",
    "            if raqtl_control_dict[teqtl.dbSNP]:\n",
    "                counter_control += 1\n",
    "        print(tissue, counter, counter_control)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
