{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import collections\n",
    "import sys\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "from iotools import readgtf\n",
    "from utils import cismasking\n",
    "\n",
    "sys.path.append('../../')\n",
    "import operator\n",
    "import json\n",
    "from utils import utils\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/analysis/plots/tissue_table.txt\"\n",
    "tshorts, tfulls_plain = utils.read_tissues(tissue_file, plain=True)\n",
    "_, tfulls = utils.read_tissues(tissue_file)\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors  = dict()\n",
    "tissue_names   = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tshorts, tfulls_plain):\n",
    "    tissue_names[tshort] = tfull\n",
    "for tshort, tfull in zip(tshorts, tfulls):\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]\n",
    "\n",
    "brain_tissues = ['bam','ban','bca','bceh','bce','bco','bfr','bhi','bhy','bnu','bpu','bsp','bsu']   \n",
    "optim_tissues = ['haa', 'pan', 'spl', 'wb']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = \"/cbscratch/franco/datasets\"\n",
    "gtffile = os.path.join(base_dir, \"GENCODE/gencode.v26.annotation.gtf.gz\")\n",
    "genemapfile  = os.path.join(base_dir, \"crossmappability/hg38_gene_mappability.txt.gz\")\n",
    "crossmapfile = os.path.join(base_dir, \"crossmappability/hg38_cross_mappability_strength.txt.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNPINFO_FIELDS = ['chrom', 'varid', 'bp_pos', 'ref_allele', 'alt_allele', 'maf']\n",
    "class SnpInfo(collections.namedtuple('_SnpInfo', SNPINFO_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "    \n",
    "def Snp2info(snps_list):\n",
    "    snpinfo = list()\n",
    "    for snpid in snps_list:\n",
    "        arr = snpid.split(\"_\")\n",
    "        chrom = int(arr[0][3:])\n",
    "        pos = int(arr[1])\n",
    "        this_snp = SnpInfo(chrom      = chrom,\n",
    "                           bp_pos     = pos,\n",
    "                           varid      = snpid,\n",
    "                           ref_allele = \"X\",\n",
    "                           alt_allele = \"Y\",\n",
    "                           maf        = None)\n",
    "        snpinfo.append(this_snp)\n",
    "    return snpinfo\n",
    "\n",
    "def knn_correction(expr, dosage, K, f=1):\n",
    "    assert (expr.shape[0] == dosage.shape[1])\n",
    "    pca = PCA(n_components=int(f * min(expr.shape[0], expr.shape[1]) ))\n",
    "#     print(\"Original dimension: {:d} x {:d}\".format(expr.shape[0], expr.shape[1]))\n",
    "    pca.fit(expr) # requires N x G\n",
    "    expr_pca = pca.transform(expr)\n",
    "#     print(\"Reduced dimension: {:d} x {:d}\".format(expr_pca.shape[0], expr_pca.shape[1]))\n",
    "\n",
    "    def gene_distance(a, b):\n",
    "        return np.linalg.norm(a - b)\n",
    "\n",
    "    nsample = expr.shape[0]\n",
    "    distance_matrix = np.zeros((nsample, nsample))\n",
    "    for i in range(nsample):\n",
    "        for j in range(i+1, nsample):\n",
    "            dist = gene_distance(expr_pca[i,:], expr_pca[j,:])\n",
    "            distance_matrix[i, j] = dist\n",
    "            distance_matrix[j, i] = dist\n",
    "\n",
    "    kneighbor = K\n",
    "    gx_knn = np.zeros_like(expr)\n",
    "    gt_knn = np.zeros_like(dosage)\n",
    "    neighbor_list = list()\n",
    "\n",
    "    for i in range(nsample):\n",
    "        neighbors = np.argsort(distance_matrix[i, :])[:kneighbor + 1][1:]\n",
    "        gx_knn[i, :] = expr[i, :] - np.mean(expr[neighbors, :], axis = 0)\n",
    "        # noisy_neighbors = np.random.choice(neighbors, size = int(2 * kneighbor / 3), replace = False)\n",
    "        # noisy_neighbors = np.random.choice(neighbors, size = kneighbor, replace = True )\n",
    "        noisy_neighbors = neighbors\n",
    "        gt_knn[:, i] = dosage[:, i] - np.mean(dosage[:, noisy_neighbors], axis = 1)\n",
    "        neighbor_list.append(neighbors)\n",
    "\n",
    "    return gx_knn, gt_knn\n",
    "    \n",
    "def extend_cismask(genes, cismaskcomp, cross_gene_dict):\n",
    "    # cross_gene_dict = read_crossmap(crossmapfile)\n",
    "    genes_ix_dict = dict(zip([x.ensembl_id for x in genes], np.arange(len(genes))))\n",
    "    crossmap_cismaskcomp = list()\n",
    "    for cismask in cismaskcomp:\n",
    "        gene_list = list()  # list of crossmapped genes for the cismask\n",
    "        for gene in [genes[i] for i in cismask.rmv_id]:\n",
    "            gene_list += cross_gene_dict[gene.ensembl_id]\n",
    "        uniq_gene_list = list(set(gene_list))\n",
    "        ugene_dict = collections.defaultdict(lambda: False) # significant speed-up by making a dict\n",
    "        for g in uniq_gene_list:\n",
    "            ugene_dict[g] = True\n",
    "        cm_gene2rmv_ix = [genes_ix_dict[x.ensembl_id] for x in genes if ugene_dict[x.ensembl_id]]\n",
    "        if len(cm_gene2rmv_ix) > 0:\n",
    "            new_mask_genes = sorted(list(set(list(cismask.rmv_id) + cm_gene2rmv_ix)))\n",
    "            new_cismask = cismask._replace(rmv_id = np.array(new_mask_genes))\n",
    "            crossmap_cismaskcomp.append(new_cismask)\n",
    "        else:\n",
    "            crossmap_cismaskcomp.append(cismask)\n",
    "    return crossmap_cismaskcomp\n",
    "\n",
    "def read_tissue_genes(gx_file, geneinfo_dict):\n",
    "    genes = list()\n",
    "    with open(gx_file) as inst:\n",
    "        next(inst)\n",
    "        for line in inst:\n",
    "            ensembl_id = line.split()[0]\n",
    "            genes.append(geneinfo_dict[ensembl_id])\n",
    "    return genes\n",
    "\n",
    "def read_crossmap(crossmapfile):\n",
    "    gene_pairs = collections.defaultdict(list)\n",
    "    with gzip.open(crossmapfile, 'r') as instream:\n",
    "        for line in instream:\n",
    "            arr = line.rstrip().split()\n",
    "            gene_pairs[arr[0].decode('utf-8')].append(arr[1].decode('utf-8'))\n",
    "    return gene_pairs\n",
    "\n",
    "cross_gene_dict = read_crossmap(crossmapfile) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneinfo = readgtf.gencode(gtffile, trim=False)\n",
    "\n",
    "### reformat gene info\n",
    "geneinfo_dict = dict()\n",
    "for g in geneinfo:\n",
    "    geneinfo_dict[g.ensembl_id] = g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mpmath\n",
    "mpmath.mp.dps = 50\n",
    "def pvalue(x): return float(mpmath.log10(1 - 0.5 * (1 + mpmath.erf(x/mpmath.sqrt(2)))))\n",
    "\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "def tejaas(filepath):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = int(arr[1])\n",
    "            pos   = int(arr[2])\n",
    "            maf   = float(arr[3])\n",
    "            q     = float(arr[4])\n",
    "            mu    = float(arr[5])\n",
    "            sigma = float(arr[6])\n",
    "            p     = float(arr[7])\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pvalue( (q - mu) / sigma)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf))\n",
    "    return res\n",
    "\n",
    "base_resdir = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_SHAPEIT2/raw/gtex_v8-aa/tejaas/permnull_sb0.1_knn30_crossmap/chr1/rr.txt\"\n",
    "\n",
    "target_file    = \"trans_eqtls.txt.ld_prune\"\n",
    "\n",
    "## SHAPEIT 2 data\n",
    "path_sb01    = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma01_knn30_cut5e-8/\"\n",
    "path_sb0006  = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma0006_knn30_cut5e-8/\"\n",
    "path_raw     = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_SHAPEIT2/raw/summary_5e-08/\"\n",
    "\n",
    "\n",
    "teqtls        = os.path.join(path_sb01, \"{:s}/trans_eqtls_ldpruned.txt\") \n",
    "teqtls_cm     = os.path.join(path_raw, \"{:s}/tejaas\", \"permnull_sb0.1_knn30_crossmap\", target_file)\n",
    "alt_teqtls    = os.path.join(path_sb0006, \"{:s}/trans_eqtls_ldpruned.txt\") \n",
    "alt_teqtls_cm = os.path.join(path_raw, \"{:s}/tejaas\", \"permnull_sb0.006_knn30_crossmap\", target_file)\n",
    "\n",
    "teqtl_dict = collections.defaultdict(dict)\n",
    "for ts in tshorts:\n",
    "    if ts not in brain_tissues:\n",
    "        if ts in optim_tissues:\n",
    "            teqtl_dict[\"sb\"][ts] = tejaas(alt_teqtls.format(ts))\n",
    "            teqtl_dict[\"cm\"][ts] = tejaas(alt_teqtls_cm.format(ts))\n",
    "        else:\n",
    "            teqtl_dict[\"sb\"][ts] = tejaas(teqtls.format(ts))\n",
    "            teqtl_dict[\"cm\"][ts] = tejaas(teqtls_cm.format(ts))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "cm_allpval_file = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_SHAPEIT2/all_variants_pvalues_tejaas_4optims_crossmap.txt\"\n",
    "pval_df_cm = pd.read_csv(cm_allpval_file, header=0, index_col=0, sep=\"\\t\")\n",
    "\n",
    "allpval_file = \"/cbscratch/franco/from_saikat/gtex_v8_202003/all_variants_pvalues_tejaas_4optims.txt\"\n",
    "pval_df = pd.read_csv(allpval_file, header=0, index_col=0, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8048655, 38)\n",
      "(8048655, 49)\n",
      "(8048655, 36)\n",
      "(8048655, 36)\n"
     ]
    }
   ],
   "source": [
    "## Subset tissue dataframes of pvalues\n",
    "print(pval_df_cm.shape)\n",
    "print(pval_df.shape)\n",
    "\n",
    "tissues_set = [x for x in tshorts if x not in brain_tissues]\n",
    "pval_df_cm_subset = pval_df_cm[tissues_set]\n",
    "pval_df_subset = pval_df[tissues_set]\n",
    "\n",
    "print(pval_df_cm_subset.shape)\n",
    "print(pval_df_subset.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tissue: haa\n",
      "SNPs before cm: 64\n",
      "SNPs after cm:  236\n",
      "SNPs brought in 173\n",
      "masked crossmap-genes: 4472\n",
      "masked cis-genes: 8\n",
      "Tissue: pan\n",
      "SNPs before cm: 179\n",
      "SNPs after cm:  356\n",
      "SNPs brought in 178\n",
      "masked crossmap-genes: 3842\n",
      "masked cis-genes: 1\n",
      "Tissue: spl\n",
      "SNPs before cm: 1354\n",
      "SNPs after cm:  1666\n",
      "SNPs brought in 371\n",
      "masked crossmap-genes: 5944\n",
      "masked cis-genes: 4\n",
      "Tissue: wb\n",
      "SNPs before cm: 6269\n",
      "SNPs after cm:  11482\n",
      "SNPs brought in 5264\n",
      "masked crossmap-genes: 5688\n",
      "masked cis-genes: 13\n"
     ]
    }
   ],
   "source": [
    "global_cutoff = -np.log10(5e-8)\n",
    "\n",
    "for ts in optim_tissues:\n",
    "    print(f\"Tissue: {ts}\")\n",
    "\n",
    "    # gx_file = f\"/cbscratch/franco/trans-eqtl/new_preprocess_feb2020_freeze/gtex_v8/expression/tpms/{ts}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\"\n",
    "    gx_file = f\"/cbscratch/franco/trans-eqtl/new_preprocess_feb2020_freeze/gtex_v8/expression/tpms/{ts}_knn30_norm.txt.protein_coding_lncRNA_filtered\"\n",
    "    df = pd.read_csv(gx_file, header=0, index_col=0, sep=\"\\t\")\n",
    "\n",
    "    snpids    = [x.rsid for x in teqtl_dict[\"sb\"][ts]]\n",
    "    snpids_cm = [x.rsid for x in teqtl_dict[\"cm\"][ts]]\n",
    "\n",
    "    both_sets = list(set(snpids + snpids_cm))  \n",
    "\n",
    "    logpvals    = pval_df_subset[ts][both_sets]\n",
    "    logpvals_cm = pval_df_cm_subset[ts][both_sets]\n",
    "\n",
    "    # Find snps that become significant after crossmap filter\n",
    "    sig_after_cm_ix      = logpvals < global_cutoff\n",
    "    snpinfo_tissue       = Snp2info(both_sets)\n",
    "    snpinfo_sig_after_cm = [snp for i, snp in enumerate(snpinfo_tissue) if sig_after_cm_ix[i]]\n",
    "\n",
    "    print(f\"SNPs before cm: {len(snpids)}\")\n",
    "    print(f\"SNPs after cm:  {len(snpids_cm)}\")\n",
    "    print(f\"SNPs brought in {np.sum(sig_after_cm_ix)}\")\n",
    "\n",
    "    ts_genes = read_tissue_genes(gx_file, geneinfo_dict)\n",
    "\n",
    "    #### Obtain cismasks and add crossmappable genes\n",
    "    cismasklist = cismasking.get_cismasklist(snpinfo_sig_after_cm, ts_genes, None, window=1e6)\n",
    "    cismaskcomp = cismasking.compress_cismasklist(cismasklist)\n",
    "    crossmap_cismaskcomp = extend_cismask(ts_genes, cismaskcomp, cross_gene_dict)\n",
    "\n",
    "    print(f\"masked crossmap-genes: {len(crossmap_cismaskcomp[0].rmv_id)}\")\n",
    "    print(f\"masked cis-genes: {len(cismaskcomp[0].rmv_id)}\")\n",
    "\n",
    "    usegenes_cm = np.ones(df.shape[0], dtype=bool)\n",
    "    thismask_cm = crossmap_cismaskcomp[1]\n",
    "    if thismask_cm.rmv_id.shape[0] > 0: usegenes_cm[thismask_cm.rmv_id] = False      \n",
    "    df_cm = df[usegenes_cm]\n",
    "\n",
    "    usegenes = np.ones(df.shape[0], dtype=bool)\n",
    "    thismask = cismaskcomp[1]\n",
    "    if thismask.rmv_id.shape[0] > 0: usegenes[thismask.rmv_id] = False   \n",
    "    df_cis = df[usegenes]\n",
    "\n",
    "    ## Write down these expressions to find out if there are differences in the sigma beta optimization\n",
    "\n",
    "    # df_cm.to_csv(f\"{ts}_tpms_crossmap_filtered.txt\", sep=\"\\t\", header=True, index=True)\n",
    "    # df_cis.to_csv(f\"{ts}_tpms_cismask_filtered.txt\", sep=\"\\t\", header=True, index=True)\n",
    "    df_cm.to_csv(f\"{ts}_knn30_crossmap_filtered.txt\", sep=\"\\t\", header=True, index=True)\n",
    "    df_cis.to_csv(f\"{ts}_knn30_cismask_filtered.txt\", sep=\"\\t\", header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tissue: haa\n",
      "Tissue: pan\n",
      "Tissue: spl\n",
      "Tissue: wb\n"
     ]
    }
   ],
   "source": [
    "## Copy original expressions to facilitate calculations\n",
    "import shutil\n",
    "\n",
    "for ts in optim_tissues:\n",
    "    print(f\"Tissue: {ts}\")\n",
    "\n",
    "    # gx_file = f\"/cbscratch/franco/trans-eqtl/new_preprocess_feb2020_freeze/gtex_v8/expression/tpms/{ts}_tpms_qcfilter.txt.protein_coding_lncRNA_filtered\"\n",
    "    # dest_file = f\"{ts}_tpms_qcfilter.txt\"\n",
    "    gx_file = f\"/cbscratch/franco/trans-eqtl/new_preprocess_feb2020_freeze/gtex_v8/expression/tpms/{ts}_knn30_norm.txt.protein_coding_lncRNA_filtered\"\n",
    "    dest_file = f\"{ts}_knn30_norm.txt\"\n",
    "    shutil.copy(gx_file, dest_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyenv37",
   "language": "python",
   "name": "pyenv37"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
