{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "import gzip\n",
    "import random\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import os, re\n",
    "from functools import wraps\n",
    "import time\n",
    "\n",
    "\n",
    "def timeit(f):\n",
    "    @wraps(f)\n",
    "    def wrap(*args, **kw):\n",
    "        ts = time.time()\n",
    "        result = f(*args, **kw)\n",
    "        te = time.time()\n",
    "        print('{:s} took: {:.6f} seconds'.format(f.__name__, te-ts))\n",
    "        return result\n",
    "    return wrap\n",
    "\n",
    "SNPINFO_FIELDS = ['chrom', 'varid', 'bp_pos', 'ref_allele', 'alt_allele', 'maf']\n",
    "class SnpInfo(collections.namedtuple('_SnpInfo', SNPINFO_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "SNP_COMPLEMENT = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "\n",
    "def HWEcheck(x):\n",
    "    gt = x.tolist()\n",
    "    f = np.array([0] * 3)\n",
    "    f[0] = gt.count(0)\n",
    "    f[1] = gt.count(1)\n",
    "    f[2] = gt.count(2)\n",
    "    n = sum(f)\n",
    "    #p_A = (2 * f[0] + f[1]) / (2 * n)\n",
    "    #p_a = (2 * f[2] + f[1]) / (2 * n)\n",
    "    X2 = n * ( (4 * f[0] * f[2] - f[1] ** 2) / ((2 * f[0] + f[1]) * (2 * f[2] + f[1])) )**2\n",
    "    pval = 1 - stats.chi2.cdf(X2, 1)\n",
    "    return pval\n",
    "\n",
    "\n",
    "def filter_snps(snpinfo, dosage):\n",
    "        # Predixcan style filtering of snps\n",
    "        newsnps = list()\n",
    "        newdosage = list()\n",
    "        npoly = 0\n",
    "        nambi = 0\n",
    "        nunkn = 0\n",
    "        nlowf = 0\n",
    "        nlowf_actual = 0\n",
    "        nhwep = 0\n",
    "        maf_limit = 0.01\n",
    "        for i, snp in enumerate(snpinfo):\n",
    "            pos = snp.bp_pos\n",
    "            refAllele = snp.ref_allele\n",
    "            effectAllele = snp.alt_allele\n",
    "            rsid = snp.varid\n",
    "            maf = round(snp.maf, 3)\n",
    "            maf_actual = sum(dosage[i]) / 2 / len(dosage[i])\n",
    "            # Skip non-single letter polymorphisms\n",
    "            if len(refAllele) > 1 or len(effectAllele) > 1:\n",
    "                npoly += 1\n",
    "                continue\n",
    "            # Skip ambiguous strands\n",
    "            if SNP_COMPLEMENT[refAllele] == effectAllele:\n",
    "                nambi += 1\n",
    "                continue\n",
    "            # Skip unknown RSIDs\n",
    "            if rsid == '.':\n",
    "                nunkn += 1\n",
    "                continue\n",
    "            # Skip low MAF\n",
    "            if not (maf >= maf_limit and maf <= (1 - maf_limit)):\n",
    "                nlowf += 1\n",
    "                continue\n",
    "            if not (maf_actual >= maf_limit and maf_actual <= (1 - maf_limit)):\n",
    "                nlowf_actual += 1\n",
    "                continue\n",
    "            # if np.all(dosage[i] == dosage[i][0:]):\n",
    "            #     nsame += 1\n",
    "            #     continue\n",
    "            # # Convert to integers 0, 1 or 2\n",
    "            # bins = [0.66, 1.33]\n",
    "            # intdosage = np.digitize(dosage[i], bins)\n",
    "            # # Remove SNPs out of HWE\n",
    "            # hwep = self.HWEcheck(intdosage)\n",
    "            # if(hwep < 0.000001):\n",
    "            #    nhwep += 1\n",
    "            #    # self.logger.debug(\"SNP {:s} has a HWE p-value of {:g}\".format(rsid, hwep))\n",
    "            #    continue\n",
    "            new_snp = snp._replace(maf = maf_actual)\n",
    "            newsnps.append(new_snp)\n",
    "            newdosage.append(dosage[i])\n",
    "            # newdosage.append(intdosage)\n",
    "#         print(\"Removed {:d} SNPs because of non-single letter polymorphisms\".format(npoly))\n",
    "#         print(\"Removed {:d} SNPs because of ambiguous strands\".format(nambi))\n",
    "#         print(\"Removed {:d} SNPs because of unknown RSIDs\".format(nunkn))\n",
    "#         print(\"Removed {:d} SNPs because of low MAF < {:g}\".format(nlowf, maf_limit))\n",
    "#         print(\"Removed {:d} SNPs because of low MAF (current)\".format(nlowf_actual))\n",
    "        return newsnps, np.array(newdosage)   \n",
    "\n",
    "def read_vcf(filename, startsnp, endsnp, mode=\"DS\", samplefile=None):\n",
    "    dosage = list()\n",
    "    snpinfo = list()\n",
    "    linenum = 0\n",
    "    with gzip.open(filename, 'r') as vcf:\n",
    "        for line in vcf:\n",
    "            linestrip = line.decode().strip()\n",
    "            if linestrip[:2] == '##': continue\n",
    "            if linestrip[:6] == '#CHROM':\n",
    "                linesplit = linestrip.split(\"\\t\")\n",
    "                donor_ids = linesplit[9:]\n",
    "            else:\n",
    "                if linenum >= startsnp and linenum < endsnp:\n",
    "                    linesplit = linestrip.split(\"\\t\")\n",
    "                    if linesplit[0].startswith(\"chr\"):\n",
    "                        chrom = int(linesplit[0][3:])\n",
    "                    else:\n",
    "                        chrom = int(linesplit[0])\n",
    "                    pos   = int(linesplit[1])\n",
    "                    varid = linesplit[2]\n",
    "                    ref   = linesplit[3]\n",
    "                    alt   = linesplit[4]\n",
    "\n",
    "                    if mode == \"DS\":\n",
    "                        if \"DS\" not in linesplit[8].split(':'):\n",
    "                            mode = \"GT\"\n",
    "                        else:\n",
    "                            dsindx = linesplit[8].split(':').index(\"DS\")\n",
    "                            ds = [x.split(':')[dsindx] for x in linesplit[9:]]\n",
    "                            gtindx = linesplit[8].split(':').index(\"GT\")\n",
    "                            for i, x in enumerate(ds):\n",
    "                                if x == \".\":\n",
    "                                    gt = linesplit[9+i].split(':')[gtindx]\n",
    "                                    if len(gt) == 3 and gt[0] != \".\" and gt[2] != \".\":\n",
    "                                        ds[i] = float(int(gt[0]) + int(gt[2]))\n",
    "\n",
    "                    if mode == \"GT\":\n",
    "                        if \"GT\" not in linesplit[8].split(':'):\n",
    "                            print(\"ERROR: no GT field in VCF file\")\n",
    "                            raise\n",
    "                        gtindx = linesplit[8].split(':').index(\"GT\")\n",
    "                        gt = [x.split(':')[gtindx] for x in linesplit[9:]]\n",
    "                        ds = [ float(int(x[0]) + int(x[2])) if len(x) == 3 and x[0] != \".\" and x[2] != \".\" else \".\" for x in gt ]\n",
    "\n",
    "                    ds_notna = [float(x) for x in ds if x != \".\"]\n",
    "                    freq = sum(ds_notna) / 2 / len(ds_notna)\n",
    "                    maf = freq\n",
    "                    snpdosage = [float(x) if x != '.' else 2 * freq for x in ds]\n",
    "\n",
    "                    this_snp = SnpInfo(chrom      = chrom,\n",
    "                                       bp_pos     = pos,\n",
    "                                       varid      = varid,\n",
    "                                       ref_allele = ref,\n",
    "                                       alt_allele = alt,\n",
    "                                       maf        = maf)\n",
    "\n",
    "                    dosage.append(snpdosage)\n",
    "                    snpinfo.append(this_snp)\n",
    "                linenum += 1\n",
    "                \n",
    "                if linenum > endsnp:\n",
    "                    break\n",
    "                    \n",
    "    if samplefile is not None:\n",
    "        if os.path.exists(samplefile):\n",
    "            with open(samplefile, 'r') as samfile:\n",
    "                sample = 0\n",
    "                samplenames = list()\n",
    "                next(samfile)\n",
    "                next(samfile)\n",
    "                for line in samfile:\n",
    "                    if re.search('^#', line):\n",
    "                        continue\n",
    "                    samplenames.append(line.strip().split()[0])\n",
    "            common_ids = [x for x in samplenames if x in donor_ids]\n",
    "            print(\"GT Sample selection {:d} samples were retained from a total of {:d} samples\".format(len(common_ids), len(donor_ids)))\n",
    "            ix = [donor_ids.index(x) for x in common_ids]\n",
    "            return np.array(dosage)[:,ix], snpinfo, common_ids\n",
    "        else:\n",
    "            print(\"samplefile does not exist\")\n",
    "            raise\n",
    "    else:\n",
    "        return np.array(dosage), snpinfo, donor_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GT Sample selection 8453 samples were retained from a total of 8453 samples\n"
     ]
    }
   ],
   "source": [
    "chrm=12\n",
    "f_vcf = \"/cbscratch/franco/datasets/FHS/genotypes/vcfs/chr{:d}.fhs.vcf.gz\".format(chrm)\n",
    "samplefile = \"/cbscratch/franco/datasets/FHS/genotypes/merged_dosages/chr12.fhs.dosages.sample\"\n",
    "gtfull, snpinfos, gt_donors = read_vcf(f_vcf, 0, 20000, samplefile=samplefile)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mafs = [x.maf for x in snpinfos]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.347, 1.907, 1.897, ..., 1.857, 1.382, 1.348],\n",
       "       [0.979, 1.929, 1.884, ..., 1.853, 0.981, 0.547],\n",
       "       [2.   , 2.   , 2.   , ..., 1.999, 2.   , 1.995],\n",
       "       ...,\n",
       "       [1.926, 1.999, 1.963, ..., 1.998, 1.995, 1.11 ],\n",
       "       [1.926, 1.97 , 1.972, ..., 1.998, 1.995, 1.111],\n",
       "       [1.889, 1.995, 1.966, ..., 1.998, 1.993, 0.994]])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfull"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
