{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import argparse\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import collections\n",
    "from sklearn import linear_model\n",
    "\n",
    "import rnaseqnorm\n",
    "import readgtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissueoutdir=\"/scratch/sbanerj/trans-eqtl/input/gtex_v8/preprocess_gx/Cervix_Ectocervix\"\n",
    "tpmfile = os.path.join(tissueoutdir, \"all_genes_tpm.gct\")\n",
    "countsfile = os.path.join(tissueoutdir, \"all_genes_counts.gct\")\n",
    "covfile = os.path.join(tissueoutdir, \"covariates.txt\")\n",
    "outfile = os.path.join(tissueoutdir, \"gtexv8_ect.txt\")\n",
    "vcf_sample_list = \"/scratch/sbanerj/trans-eqtl/input/gtex_v8/preprocess_gx/vcf_samples.list\"\n",
    "gtf_file = \"/scratch/sbanerj/data/GENCODE/gencode.v26.annotation.gtf.gz\"\n",
    "tpm_threshold = 0.1\n",
    "count_threshold = 6\n",
    "sample_frac_threshold = 0.2\n",
    "methods = ['tmm_std_cclm']\n",
    "biotype = ['protein_coding']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gct(gct_file):\n",
    "    \"\"\"\n",
    "    Load GCT as DataFrame\n",
    "    Returns G x N dataframe\n",
    "    Drops the description, converts full sample names to sample IDs\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(gct_file, sep='\\t', skiprows=2, index_col=0)\n",
    "    df.drop('Description', axis=1, inplace=True)\n",
    "    df.index.name = 'gene_id'\n",
    "    colnames = ['-'.join(x.split('-')[:2]) for x in df.columns]\n",
    "    df.columns = colnames\n",
    "    return df\n",
    "\n",
    "def read_covariates(cov_file, col_ix):\n",
    "    \"\"\"\n",
    "    Load covariates as DataFrams\n",
    "    Returns C x N covariates (C = number of covariates, N = number of samples)\n",
    "    Orders samples according to 'col_ix'\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(cov_file, sep='\\t', index_col=0)\n",
    "    df.index.name = 'covname'\n",
    "    df = df[col_ix]\n",
    "    \n",
    "    ## drop covariate with equal entries (e.g. gender in tissues like ovary, testis, etc.)\n",
    "    #eq_mask = np.array([np.all(np.isclose(X, X[0])) for X in np.array(df)])\n",
    "    #if np.any(eq_mask):\n",
    "    #    print( f'Dropping covariates for equal elements: {\", \".join(df.loc[eq_mask].index)}' )\n",
    "    #    df = df.loc[~eq_mask]\n",
    "              \n",
    "    ## identify and drop collinear covariates\n",
    "    C = df.astype(np.float64).T\n",
    "    Q, R = np.linalg.qr(C - np.mean(C, axis=0))\n",
    "    collinear_mask = np.abs(np.diag(R)) < np.finfo(np.float64).eps * C.shape[1]\n",
    "    if np.any(collinear_mask):\n",
    "        print( f'Dropping collinear covariates: {\", \".join(df.loc[collinear_mask].index)}' )\n",
    "        df = df.loc[~collinear_mask]\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def read_samples(sample_file):\n",
    "    with open(sample_file, 'r') as infile:\n",
    "        content = infile.readlines()\n",
    "    samples = [x.strip() for x in content]\n",
    "    return samples\n",
    "\n",
    "def center_and_scale(Y):\n",
    "    \"\"\"\n",
    "    requires  G x N matrix, where G is the number of genes, and Y is the number of samples\n",
    "    \"\"\"\n",
    "    newY = (Y - np.mean(Y, axis = 1).reshape(-1, 1)) / np.std(Y, axis = 1).reshape(-1, 1)\n",
    "    return newY\n",
    "\n",
    "\n",
    "def center_and_scale_df(M):\n",
    "    return pd.DataFrame(center_and_scale(np.array(M)), index=M.index, columns=M.columns)\n",
    "\n",
    "\n",
    "def covcorrlm(M, W):\n",
    "    \"\"\"\n",
    "    inputs: M is pd.DataFrame, W is np.array\n",
    "            M of shape G x N (G = number of genes, N = number of samples)\n",
    "            W of shape C x N (C = number of covariates)\n",
    "    \"\"\"\n",
    "    X = np.array(M)\n",
    "    Wnorm = center_and_scale(np.array(W))\n",
    "    linreg = linear_model.LinearRegression()\n",
    "    linreg.fit(Wnorm.T, X.T)\n",
    "    Xcorr = X - linreg.predict(Wnorm.T).T\n",
    "    #return Xcorr\n",
    "    return pd.DataFrame(Xcorr, index=M.index, columns=M.columns)\n",
    "\n",
    "\n",
    "def covcorrlasso(M, W, alpha = 0.05):\n",
    "    \"\"\"\n",
    "    inputs: M is pd.DataFrame, W is np.array\n",
    "            M of shape G x N (G = number of genes, N = number of samples)\n",
    "            W of shape C x N (C = number of covariates)\n",
    "    \"\"\"\n",
    "    X = np.array(M)\n",
    "    Wnorm = center_and_scale(np.array(W))\n",
    "    Xcorr = np.zeros(X.shape)\n",
    "    for i in range(X.shape[0]):\n",
    "        lassoreg = linear_model.Lasso(alpha = alpha)\n",
    "        lassoreg.fit(Wnorm.T, X[i, :])\n",
    "        Xcorr[i] = X[i, :] - lassoreg.predict(Wnorm.T).T\n",
    "    #return Xcorr\n",
    "    return pd.DataFrame(Xcorr, index=M.index, columns=M.columns)\n",
    "\n",
    "def tmm_normalization(counts_df, mask):\n",
    "    tmm_counts_df = rnaseqnorm.edgeR_cpm(counts_df, normalized_lib_sizes = True)\n",
    "    norm_df = rnaseqnorm.inverse_normal_transform(tmm_counts_df[mask])\n",
    "    return norm_df\n",
    "\n",
    "\n",
    "def qn_normalization(tpm_df, mask):\n",
    "    qn_df = rnaseqnorm.normalize_quantiles(tpm_df.loc[mask])\n",
    "    norm_df = rnaseqnorm.inverse_normal_transform(qn_df)\n",
    "    return norm_df\n",
    "\n",
    "\n",
    "def raw_gx(tpm_df, mask):\n",
    "    R = tpm_df.loc[mask]\n",
    "    norm_df = pd.DataFrame(np.array(R), index=R.index, columns=R.columns)\n",
    "    return norm_df\n",
    "\n",
    "\n",
    "def filter_genes(df, gene_info):\n",
    "    gene_list = df.index\n",
    "    keep = [g.ensembl_id for g in gene_info if g.ensembl_id in gene_list]\n",
    "    return df.loc[keep]\n",
    "\n",
    "\n",
    "def pp_options(method, gx_df, mask, cov = None):\n",
    "    # a pythonic dictionary runs through all functions and creates the full dictionary\n",
    "    # hence if-else is used.\n",
    "    if method == 'raw':\n",
    "        print('   > Raw. Only filtering genes')\n",
    "        res_df = raw_gx(gx_df, mask)\n",
    "    if method == 'std':\n",
    "        print('   > Center and scale')\n",
    "        res_df = center_and_scale_df(gx_df)\n",
    "    elif method == 'qn':\n",
    "        print('   > QN')\n",
    "        res_df = qn_normalization(gx_df, mask)\n",
    "    elif method == 'tmm':\n",
    "        print('   > TMM')\n",
    "        res_df = tmm_normalization(gx_df, mask)\n",
    "    elif method == 'cclm':\n",
    "        print('   > Covariate correction with linear model')\n",
    "        res_df = covcorrlm(gx_df, cov)\n",
    "    elif method == 'cclasso':\n",
    "        print('   > Adaptive covariate correction using LASSO')\n",
    "        res_df = covcorrlasso(gx_df, cov)\n",
    "    return res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gene expression TPM file.\n",
      "Reading gene expression Counts file.\n",
      "Reading covariates.\n",
      "Dropping collinear covariates: SEX\n"
     ]
    }
   ],
   "source": [
    "# Read gene expression\n",
    "print (\"Reading gene expression TPM file.\")\n",
    "tpm_df = read_gct(tpmfile)\n",
    "gxdonors = list(tpm_df.columns)\n",
    "genes = list(tpm_df.index)\n",
    "\n",
    "print (\"Reading gene expression Counts file.\")\n",
    "counts_df = read_gct(countsfile)\n",
    "\n",
    "\"\"\"\n",
    "Sample lookup. \n",
    "Only include samples which are also in VCF.\n",
    "\"\"\"\n",
    "gtdonors = read_samples(vcf_sample_list)\n",
    "common_ix = [x for x in gxdonors if x in gtdonors]\n",
    "tpm_df = tpm_df[common_ix]\n",
    "counts_df = counts_df[common_ix]\n",
    "\n",
    "\"\"\"\n",
    "Expression thresholds.\n",
    "Genes are thresholded based on the following expression rules:\n",
    "  TPM >= tpm_threshold in >= sample_frac_threshold * samples\n",
    "  read counts >= count_threshold in sample_frac_threshold * samples\n",
    "\"\"\"\n",
    "ns = tpm_df.shape[1]\n",
    "mask = (\n",
    "    (np.sum(tpm_df >= tpm_threshold, axis = 1) >= sample_frac_threshold * ns) &\n",
    "    (np.sum(counts_df >= count_threshold, axis = 1) >= sample_frac_threshold * ns)\n",
    ").values\n",
    "\n",
    "# Read covariates\n",
    "if any(['cc' in x for x in methods]):\n",
    "    print (\"Reading covariates.\")\n",
    "    cov_df = read_covariates(covfile, tpm_df.columns)\n",
    "else:\n",
    "    cov_df = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-U3ZN</th>\n",
       "      <th>GTEX-Q2AG</th>\n",
       "      <th>GTEX-TSE9</th>\n",
       "      <th>GTEX-S4UY</th>\n",
       "      <th>GTEX-S341</th>\n",
       "      <th>GTEX-P78B</th>\n",
       "      <th>GTEX-T5JW</th>\n",
       "      <th>GTEX-S32W</th>\n",
       "      <th>GTEX-N7MT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>covname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGE</th>\n",
       "      <td>33.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>68.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRISCHD</th>\n",
       "      <td>29.0</td>\n",
       "      <td>694.0</td>\n",
       "      <td>1021.0</td>\n",
       "      <td>331.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>473.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>440.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         GTEX-U3ZN  GTEX-Q2AG  GTEX-TSE9  GTEX-S4UY  GTEX-S341  GTEX-P78B  \\\n",
       "covname                                                                     \n",
       "AGE           33.0       42.0       60.0       41.0       48.0       48.0   \n",
       "TRISCHD       29.0      694.0     1021.0      331.0       60.0      473.0   \n",
       "\n",
       "         GTEX-T5JW  GTEX-S32W  GTEX-N7MT  \n",
       "covname                                   \n",
       "AGE           27.0       58.0       68.0  \n",
       "TRISCHD       45.0       34.0      440.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying tmm_std_cclm\n",
      "  Starting with TPM\n",
      "  No, wait! Starting with TMM here.\n",
      "  Step: tmm\n",
      "   > TMM\n",
      "  Step: tmm_std\n",
      "   > Center and scale\n",
      "  Step: tmm_std_cclm\n",
      "   > Covariate correction with linear model\n"
     ]
    }
   ],
   "source": [
    "# Perform the QC preprocessing\n",
    "gxpp = collections.defaultdict(lambda: None)\n",
    "\n",
    "for method in methods:\n",
    "    print(f'Applying {method}')\n",
    "    msteps = method.split('_')\n",
    "    prevstep = 'none'\n",
    "    curstep = ''\n",
    "\n",
    "    gxpp['none'] = tpm_df\n",
    "    print(\"  Starting with TPM\")\n",
    "    if method.startswith('tmm'):\n",
    "        print(\"  No, wait! Starting with TMM here.\")\n",
    "        gxpp['none'] = counts_df\n",
    "\n",
    "    for mstep in msteps:\n",
    "        curstep = f'{mstep}' if curstep == '' else f'{curstep}_{mstep}'\n",
    "        print(f'  Step: {curstep}')\n",
    "        if gxpp[curstep] is None:\n",
    "            gxpp[curstep] = pp_options(mstep, gxpp[prevstep], mask, cov = cov_df)\n",
    "        prevstep = curstep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading GENCODE file to keep protein_coding genes\n"
     ]
    }
   ],
   "source": [
    "# Read GTF file for filtering genes\n",
    "print (f'Reading GENCODE file to keep {\", \".join(biotype)} genes')\n",
    "gene_info = readgtf.gencode_v12(gtf_file, trim=False, biotype=biotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing all outputs.\n",
      " - tmm_std_cclm: /scratch/sbanerj/trans-eqtl/input/gtex_v8/preprocess_gx/Cervix_Ectocervix/gtexv8_ect_tmm_std_cclm_protein_coding.txt\n"
     ]
    }
   ],
   "source": [
    "# Filter genes and write output\n",
    "print (\"Writing all outputs.\")\n",
    "oprefix = os.path.splitext(outfile)[0]\n",
    "osuffix = os.path.splitext(outfile)[1]\n",
    "biotype_str = '_'.join(biotype)\n",
    "for method in methods:\n",
    "    ofilename = f'{oprefix}_{method}_{biotype_str}{osuffix}'\n",
    "    print(f' - {method}: {ofilename}')\n",
    "    QCdf = filter_genes(gxpp[method], gene_info)\n",
    "    QCdf.to_csv(ofilename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
