{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "# ammend DHS index annotations by doing a down-lift from hg38 to hg19\n",
    "# First collect positions that could not be downlifted\n",
    "import collections\n",
    "blackdict = collections.defaultdict(lambda: False)\n",
    "errfile = \"/cbscratch/franco/datasets/DHSindex/hglft_genome_32b00_ea89f0.err\"\n",
    "with open(errfile) as instream:\n",
    "    for line in instream:\n",
    "        if line[0] == \"#\":\n",
    "            continue\n",
    "        blackdict[line.rstrip()] = True\n",
    "\n",
    "# Go line by line through DHSindex and check if the range was downlifted or not\n",
    "# and keep its annotation\n",
    "liftfile = \"/cbscratch/franco/datasets/DHSindex/hglft_genome_32b00_ea89f0.bed\"\n",
    "dhsindexfile = \"/cbscratch/franco/datasets/DHSindex/DHS_Index_and_Vocabulary_hg38_WM20190703.txt\"\n",
    "hg19_dhs_dict = collections.defaultdict(dict)\n",
    "hglft = open(liftfile)\n",
    "with open(dhsindexfile) as instream:\n",
    "    next(instream)\n",
    "    for line in instream:\n",
    "        arr = line.rstrip().split(\"\\t\")\n",
    "        chrm = arr[0]\n",
    "        start = arr[1]\n",
    "        end   = arr[2]\n",
    "        annot = arr[9]\n",
    "        key = \"{:s}:{:s}-{:s}\".format(chrm, start, end)\n",
    "        if not blackdict[key]:\n",
    "            hg19chrom, hg19range = hglft.readline().rstrip().split(\":\")\n",
    "            hg19_dhs_dict[hg19chrom][hg19range] = annot\n",
    "\n",
    "# Write down the newly annotated dictionary in a file, sorted by genomic position\n",
    "hg19_DHS_outfile = \"/cbscratch/franco/datasets/DHSindex/DHS_Index_downlift_hg19.txt\"\n",
    "with open(hg19_DHS_outfile, 'w') as outstream:\n",
    "    for chrm in range(1, 23):\n",
    "        chrom = \"chr\"+str(chrm)\n",
    "        for key in sorted(hg19_dhs_dict[chrom], key=lambda t: int(t.split(\"-\")[0])):\n",
    "            # print(chrom, key, hg19_dhs_dict[chrom][key])\n",
    "            outstream.write(\"{:s}\\t{:s}\\t{:s}\\t{:s}\\n\".format(chrom, key.split(\"-\")[0], key.split(\"-\")[1], hg19_dhs_dict[chrom][key]))\n",
    "'''            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../../')\n",
    "import numpy as np\n",
    "import collections\n",
    "import os\n",
    "import json\n",
    "import mpmath\n",
    "from utils import utils\n",
    "import scipy.stats as ss\n",
    "\n",
    "mpmath.mp.dps = 50\n",
    "def pvalue(x): return float(mpmath.log10(1 - 0.5 * (1 + mpmath.erf(x/mpmath.sqrt(2)))))\n",
    "\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "type_master_list = [\"Cancer / epithelial\",\"Cardiac\",\"Digestive\",\"Lymphoid\",\"Musculoskeletal\",\\\n",
    "                    \"Myeloid / erythroid\",\"Neural\",\"Organ devel. / renal\",\"Placental / trophoblast\",\\\n",
    "                    \"Primitive / embryonic\",\"Pulmonary devel.\", \\\n",
    "                    \"Renal / cancer\",\"Stromal A\",\"Stromal B\",\"Tissue invariant\",\"Vascular / endothelial\"]    \n",
    "\n",
    "def tejaas(filepath):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = int(arr[1])\n",
    "            pos   = int(arr[2])\n",
    "            maf   = float(arr[3])\n",
    "            q     = float(arr[4])\n",
    "            mu    = float(arr[5])\n",
    "            sigma = float(arr[6])\n",
    "            p     = float(arr[7])\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pvalue( (q - mu) / sigma)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf))\n",
    "    return res\n",
    "\n",
    "def tejaas_old(filepath):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            pos   = int(arr[1])\n",
    "            p     = float(arr[5])\n",
    "            chrom = int(arr[6])\n",
    "            maf   = float(arr[7])\n",
    "            q     = float(arr[2])\n",
    "            mu    = float(arr[3])\n",
    "            sigma = float(arr[4])\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pvalue( (q - mu) / sigma)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf))\n",
    "    return res\n",
    "\n",
    "def read_snplist(filename, mafcutoff=0.01):\n",
    "    rsidlist = list()\n",
    "    maflist  = list()\n",
    "    with open(filename) as instream:\n",
    "        next(instream)\n",
    "        for line in instream:\n",
    "            arr  = line.strip().split(\"\\t\")\n",
    "            rsid = arr[0]\n",
    "            maf  = float(arr[1])\n",
    "            if maf >= mafcutoff and maf <= (1 - mafcutoff) :\n",
    "                rsidlist.append(rsid)\n",
    "                maflist.append(maf)\n",
    "    return rsidlist, maflist\n",
    "\n",
    "def find_annotated(res_dict, dhs_file, isannotated=False):\n",
    "    dhs = open(dhs_file)\n",
    "    line = dhs.readline()\n",
    "    prev_chrm = 0\n",
    "    nannot = 0\n",
    "    nannot_type = collections.defaultdict(int)\n",
    "\n",
    "    sorted_res_dict = dict()\n",
    "    for chrm in range(1,23):\n",
    "        sorted_res_dict[chrm] = sorted(res_dict[chrm])\n",
    "\n",
    "    while line:\n",
    "        arr = line.rstrip().split(\"\\t\")\n",
    "        if arr[0][3:] == \"X\" or arr[0][3:] == \"Y\":\n",
    "            line = dhs.readline()\n",
    "            continue\n",
    "        chrm = int(arr[0][3:])\n",
    "        start = int(arr[1])\n",
    "        end = int(arr[2])\n",
    "        if isannotated:\n",
    "            atype = arr[9]\n",
    "        if chrm != prev_chrm:\n",
    "            remaining = sorted_res_dict[chrm]\n",
    "            checked = 0\n",
    "        if len(remaining) == 0:\n",
    "            ## No more SNPs in this chromosome, just continue reading the DHS file\n",
    "            line = dhs.readline()\n",
    "        else:\n",
    "            for pos in remaining:\n",
    "                if pos < start:\n",
    "                    checked += 1\n",
    "                    remaining = sorted_res_dict[chrm][checked:]\n",
    "                    continue # go to next SNP\n",
    "                elif pos > end:\n",
    "                    line = dhs.readline()\n",
    "                    break # go to next DHS line, keep checking the remaining results\n",
    "                else:\n",
    "                    # this is an annotated SNP\n",
    "                    checked += 1\n",
    "                    remaining = sorted_res_dict[chrm][checked:]\n",
    "                    nannot += 1\n",
    "                    if isannotated:\n",
    "                        nannot_type[atype] += 1\n",
    "                    continue # go to next SNP\n",
    "        prev_chrm = chrm\n",
    "    dhs.close()\n",
    "    return nannot, nannot_type\n",
    "\n",
    "def annotated_random_all(gwrsids, dhs_file, isannotated=False):\n",
    "    res_dict = dict()\n",
    "    for chrm in range(1, 23):\n",
    "        res_dict[chrm] = list()\n",
    "    for var_id in gwrsids:\n",
    "        info = var_id.split('_')\n",
    "        chrm = int(info[0][3:])\n",
    "        bppos = int(info[1])\n",
    "        res_dict[chrm].append(bppos)\n",
    "    nannot, nannot_type = find_annotated(res_dict, dhs_file, isannotated)\n",
    "    return nannot, nannot_type\n",
    "\n",
    "def annotated_random(gwrsids, nchoose, dhs_file, isannotated=False):\n",
    "    chooseidx = np.sort(np.random.choice(len(gwrsids), nchoose, replace = False))\n",
    "    res_dict = dict()\n",
    "    for chrm in range(1, 23):\n",
    "        res_dict[chrm] = list()\n",
    "    for idx in chooseidx:\n",
    "        var_id = gwrsids[idx]\n",
    "        info = var_id.split('_')\n",
    "        chrm = int(info[0][3:])\n",
    "        bppos = int(info[1])\n",
    "        res_dict[chrm].append(bppos)\n",
    "    nannot, nannot_type = find_annotated(res_dict, dhs_file, isannotated)\n",
    "    return nannot, nannot_type\n",
    "\n",
    "def sample_binomial(n, p, NTIMES):\n",
    "    array_n = list()\n",
    "    for i in range(NTIMES):\n",
    "        n_success = np.random.binomial(n, p)\n",
    "        array_n.append(n_success)\n",
    "    return array_n\n",
    "\n",
    "def enrichment_pval(ntrans, DHS_RANDOM_BG, enrichment):\n",
    "    randtrans = sample_binomial(ntrans, DHS_RANDOM_BG, 10000000)\n",
    "    num_null = np.array(randtrans) /  ntrans\n",
    "\n",
    "    null_enrichments = num_null / DHS_RANDOM_BG\n",
    "    ecdf = ECDF(null_enrichments)\n",
    "    pval = 1 - ecdf(enrichment)\n",
    "    return pval\n",
    "\n",
    "def sample_rand_bg(snpsfilefmt, dhs_file, nchoose = 20000, niter = 20, isannotated=False):\n",
    "    if not os.path.exists(snpsfilefmt):\n",
    "        print(\"File does not exist\", snpsfilefmt)\n",
    "        raise\n",
    "    snps_list, maf_list = read_snplist(snpsfilefmt)\n",
    "    nannot_rand = list()\n",
    "    nannot_type_list = list()\n",
    "    nannot_type_array = list()\n",
    "    print(f'Iteration', end=\"\")\n",
    "    for k in range(niter):\n",
    "        nannot_k, nannot_type_k = annotated_random(snps_list, nchoose, dhs_file, isannotated)\n",
    "        print(f' {k}', end=\"\")\n",
    "        #print(f'Iteration {k}: {nannot_k}')\n",
    "        nannot_rand.append(nannot_k)\n",
    "        nannot_type_array.append([nannot_type_k[k] if k in nannot_type_k else 0 for k in type_master_list ])\n",
    "    print(\"\")\n",
    "    frac_rand = np.mean(nannot_rand) / nchoose\n",
    "    frac_rand_type = np.mean(np.array(nannot_type_array), axis=0) / nchoose\n",
    "    return frac_rand, len(snps_list), dict(zip(type_master_list, frac_rand_type))\n",
    "\n",
    "def full_rand_bg(snpsfilefmt, dhs_file, isannotated=False):\n",
    "    if not os.path.exists(snpsfilefmt):\n",
    "        print(\"File does not exist\", snpsfilefmt)\n",
    "        raise\n",
    "    snps_list, maf_list = read_snplist(snpsfilefmt)\n",
    "    nannot_rand, nannot_type_array_tmp = annotated_random_all(snps_list, dhs_file, isannotated)\n",
    "    nannot_type_array = [nannot_type_array_tmp[k] if k in nannot_type_array_tmp else 0 for k in type_master_list]\n",
    "    frac_rand = nannot_rand / len(snps_list)\n",
    "    frac_rand_type = np.array(nannot_type_array) / len(snps_list)\n",
    "    return frac_rand, len(snps_list), dict(zip(type_master_list, frac_rand_type))\n",
    "\n",
    "def smart_LD_filter(trans_eqtls, trans_eqtls_ld_regions_file):\n",
    "    pass_snps = list()\n",
    "    with open(trans_eqtls_ld_regions_file) as instream:\n",
    "        for line in instream:\n",
    "            arr = line.strip(\"\\n\").split(\"\\t\")\n",
    "            if len(arr[4]) > 0:\n",
    "                pass_snps.append(arr[0])\n",
    "    return [x for x in trans_eqtls if x.rsid in pass_snps]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 0\n"
     ]
    }
   ],
   "source": [
    "type_master_list = [\"Cancer / epithelial\",\"Cardiac\",\"Digestive\",\"Lymphoid\",\"Musculoskeletal\",\"Myeloid / erythroid\",\"Neural\",\"Organ devel. / renal\",\"Placental / trophoblast\",\"Primitive / embryonic\",\"Pulmonary devel.\",\"Renal / cancer\",\"Stromal A\",\"Stromal B\",\"Tissue invariant\",\"Vascular / endothelial\"]\n",
    "snpsfilefmt = f\"{basedir}/raw/{sumdir}/aa/tejaas/{preprocs[0]}/snps_list.txt\"\n",
    "if not os.path.exists(snpsfilefmt):\n",
    "    print(\"File does not exist\", snpsfilefmt)\n",
    "    raise\n",
    "snps_list, maf_list = read_snplist(snpsfilefmt)\n",
    "nchoose = 20000 # len(trans_eqtls_signif) \n",
    "niter = 1\n",
    "nannot_rand = list()\n",
    "nannot_type_list = list()\n",
    "nannot_type_array = list()\n",
    "print(f'Iteration', end=\"\")\n",
    "for k in range(niter):\n",
    "    nannot_k, nannot_type_k = annotated_random(snps_list, nchoose, dhs_file, isannotated=True)\n",
    "    print(f' {k}', end=\"\")\n",
    "    #print(f'Iteration {k}: {nannot_k}')\n",
    "    nannot_rand.append(nannot_k)\n",
    "    if len(nannot_type_list) == 0:\n",
    "        nannot_type_list = list(nannot_type_k.keys())\n",
    "    nannot_type_array.append([nannot_type_k[k] if k in nannot_type_k else 0 for k in type_master_list ])\n",
    "print(\"\")\n",
    "frac_rand = np.mean(nannot_rand) / nchoose\n",
    "frac_rand_type = np.mean(np.array(nannot_type_array), axis=0) / nchoose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use LD results is ON\n"
     ]
    }
   ],
   "source": [
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/main/tissues.txt\"\n",
    "\n",
    "expressions = [\"raw\"]\n",
    "sbtypes = [\"sb\"]\n",
    "sbs     = [\"0.1\"] #[\"0.005\", \"0.007\"]\n",
    "Ks       = [\"30\"]\n",
    "extras  = [\"\"]\n",
    "sumdir = \"summary_5e-08\"\n",
    "preprocs = list()\n",
    "use_LD = True\n",
    "if use_LD:\n",
    "    print(\"Use LD results is ON\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "basedir  = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_lncRNA_freeze\"\n",
    "#master_snp_data_file = basedir+\"/snp_data_{:s}.txt\".format(title)\n",
    "\n",
    "# title    = \"multi_tissue\"\n",
    "# dhs_file = \"/cbscratch/franco/datasets/multi-tissue.master.ntypes.simple.hg19_hglift_hg38_clean_sorted.bed\"\n",
    "# isannotated = False\n",
    "\n",
    "title    = \"DHSindex\"\n",
    "dhs_file = \"/cbscratch/franco/datasets/DHSindex/DHSindex.bed\"\n",
    "isannotated = True\n",
    "\n",
    "master_snp_data_file = \"/cbscratch/franco/datasets/gtex_v8_dhs_{:s}_background.txt\".format(title)\n",
    "min_trans_eqtl = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrmlist = np.arange(1,23)\n",
    "\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "tshorts, tfulls = utils.read_tissues(tissue_file)\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors  = dict()\n",
    "tissue_names   = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tshorts, tfulls):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]\n",
    "#sorted_tissues = [x[0] for x in sorted(tissue_samples.items(), key=itemgetter(1))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sbtype in sbtypes:\n",
    "    if sbtype == \"sbDynamic\":\n",
    "        sb_iter = keffs\n",
    "    if sbtype == \"sb\":\n",
    "        sb_iter = sbs\n",
    "\n",
    "    for K in Ks:\n",
    "        for param in sb_iter:\n",
    "            sb_proc = sbtype+param\n",
    "            for extra in extras:\n",
    "                preproc = \"permnull_{:s}_knn{:s}{:s}\".format(sb_proc, K, extra)\n",
    "                preprocs.append(preproc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snp_data file exists\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "for preproc in preprocs:\n",
    "    for expr in expressions:\n",
    "        resdir = basedir+\"/{:s}/{:s}\".format(expr, sumdir)\n",
    "        pcutoffs = [5e-08, 1e-08, 1e-10, 1e-15] #, 1e-08, 1e-10, 1e-15, 1e-20, 1e-25]# [1e-06, 1e-07, 5e-08, 1e-08, 1e-10, 1e-15, 1e-20, 1e-25]\n",
    "        # trans_eqtls_file = \"trans_eqtls_{:s}.txt\".format(file_pcutoff)\n",
    "        trans_eqtls_file = \"trans_eqtls.txt\"\n",
    "        if use_LD:\n",
    "            trans_eqtls_ld_regions_file = trans_eqtls_file+\".ld_regions\"\n",
    "            trans_eqtls_file = trans_eqtls_file+\".ld_prune\"\n",
    "        snpsfile = \"snps_list.txt\"\n",
    "\n",
    "        dhs_frac_rand = dict()\n",
    "        dhs_frac_type_rand = dict()\n",
    "        outdir = os.path.join(resdir, \"dhs_enrichments\", preproc)\n",
    "        if not os.path.exists(outdir):\n",
    "            os.makedirs(outdir)\n",
    "        \n",
    "#         for tissue in tshorts:\n",
    "#             dhs_frac_rand[tissue], Nsnps, dhs_frac_type_rand[tissue] = sample_rand_bg(snpsfilefmt, dhs_file, nchoose = 20000, niter = 20, isannotated = isannotated)\n",
    "#             raise\n",
    "            \n",
    "        if os.path.exists(master_snp_data_file):\n",
    "            print(\"snp_data file exists\")\n",
    "            dhs_frac_type_rand = collections.defaultdict(dict)\n",
    "            with open(master_snp_data_file) as ifile:\n",
    "                next(ifile)\n",
    "                for line in ifile:\n",
    "                    arr = line.strip().split(\"\\t\")\n",
    "                    if arr[1] == \"all\":\n",
    "                        dhs_frac_rand[arr[0]] = float(arr[3])\n",
    "                    else:\n",
    "                        dhs_frac_type_rand[arr[0]][arr[1]] = float(arr[3])\n",
    "        else:\n",
    "            print(\"Creating snp_data file\")\n",
    "            with open(master_snp_data_file, 'w') as ofile:\n",
    "                ofile.write(\"tissue\\ttissue_name\\tn_snps\\tdhs_frac_rand\\n\")\n",
    "                for tissue in tshorts[:1]:\n",
    "                    snpsfilefmt = f'{resdir}/{tissue}/tejaas/{preproc}/{snpsfile}'\n",
    "                    st = time.time()\n",
    "                    dhs_frac_rand[tissue], Nsnps, dhs_frac_type_rand[tissue] = sample_rand_bg(snpsfilefmt, dhs_file, nchoose = 20000, niter = 20, isannotated = isannotated)\n",
    "                    endt = time.time()\n",
    "                    print(\"Took \", endt-st)\n",
    "                    print (f'Fraction of annotated SNPs: {tissue} - all - {dhs_frac_rand[tissue]:7.4f}')\n",
    "                    ofile.write(f\"{tissue}\\tall\\t{Nsnps}\\t{dhs_frac_rand[tissue]}\\n\")\n",
    "                    for dhs_type in type_master_list:\n",
    "                        print(f\"{tissue}\\t{dhs_type}\\t{Nsnps * dhs_frac_type_rand[tissue][dhs_type]}\\t{dhs_frac_type_rand[tissue][dhs_type]}\")\n",
    "                        ofile.write(f\"{tissue}\\t{dhs_type}\\t-\\t{dhs_frac_type_rand[tissue][dhs_type]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Took  2142.001889228821\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "st = time.time()\n",
    "tissue = \"as\"\n",
    "snpsfilefmt = f'{resdir}/{tissue}/tejaas/{preproc}/{snpsfile}'\n",
    "dhs_frac_rand[tissue], Nsnps, dhs_frac_type_rand[tissue] = full_rand_bg(snpsfilefmt, dhs_file, isannotated = isannotated)\n",
    "endt = time.time()\n",
    "print(\"Took \", endt-st)\n",
    "# Took  2142.001889228821"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Full random\n",
    "# {'Cancer / epithelial': 0.012979240275598257,\n",
    "#  'Cardiac': 0.009184469197369814,\n",
    "#  'Digestive': 0.011347179262189226,\n",
    "#  'Lymphoid': 0.02010506621995873,\n",
    "#  'Musculoskeletal': 0.017838289343329247,\n",
    "#  'Myeloid / erythroid': 0.013527310202994466,\n",
    "#  'Neural': 0.03676996787682157,\n",
    "#  'Organ devel. / renal': 0.014578293313455197,\n",
    "#  'Placental / trophoblast': 0.02294375211743871,\n",
    "#  'Primitive / embryonic': 0.06592449150525996,\n",
    "#  'Pulmonary devel.': 0.0064684170979479585,\n",
    "#  'Renal / cancer': 0.010847704409911379,\n",
    "#  'Stromal A': 0.004080152533522579,\n",
    "#  'Stromal B': 0.0268531342218897,\n",
    "#  'Tissue invariant': 0.010595789374274226,\n",
    "#  'Vascular / endothelial': 0.006124584019280325}\n",
    "\n",
    "### sampled\n",
    "# {'as': {'Cancer / epithelial': 0.0128475,\n",
    "#   'Cardiac': 0.009255,\n",
    "#   'Digestive': 0.011532500000000001,\n",
    "#   'Lymphoid': 0.020069999999999998,\n",
    "#   'Musculoskeletal': 0.018267500000000002,\n",
    "#   'Myeloid / erythroid': 0.0132875,\n",
    "#   'Neural': 0.036545,\n",
    "#   'Organ devel. / renal': 0.015005000000000001,\n",
    "#   'Placental / trophoblast': 0.02304,\n",
    "#   'Primitive / embryonic': 0.0662725,\n",
    "#   'Pulmonary devel.': 0.0065875,\n",
    "#   'Renal / cancer': 0.010595,\n",
    "#   'Stromal A': 0.004035,\n",
    "#   'Stromal B': 0.026660000000000003,\n",
    "#   'Tissue invariant': 0.010459999999999999,\n",
    "#   'Vascular / endothelial': 0.0061475},"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as: 74 annotated out of 199. Enrichment = {'all': 1.2791637449731241, 'Cancer / epithelial': 1.1638970765815178, 'Cardiac': 1.5860470157203694, 'Digestive': 0.8771766315759465, 'Lymphoid': 0.9885898198727562, 'Musculoskeletal': 2.252157144264741, 'Myeloid / erythroid': 1.4853502688019815, 'Neural': 1.3635779467175102, 'Organ devel. / renal': 0.6970869607269919, 'Placental / trophoblast': 1.5206259330805112, 'Primitive / embryonic': 1.3741845312246217, 'Pulmonary devel.': 0.0, 'Renal / cancer': 0.4558970857918534, 'Stromal A': 0.0, 'Stromal B': 1.309966647313469, 'Tissue invariant': 0.48237347042387363, 'Vascular / endothelial': 3.330654931659124} - {'all': 0.008264287792807053, 'Cancer / epithelial': 0.4768690207960834, 'Cardiac': 0.29384174480169895, 'Digestive': 0.6662369993180367, 'Lymphoid': 0.5775736253884594, 'Musculoskeletal': 0.02755145156611675, 'Myeloid / erythroid': 0.2838918447373115, 'Neural': 0.20138819451301598, 'Organ devel. / renal': 0.782631111404567, 'Placental / trophoblast': 0.18023618495123497, 'Primitive / embryonic': 0.10750301976713693, 'Pulmonary devel.': 1.0, 'Renal / cancer': 0.889821761696987, 'Stromal A': 1.0, 'Stromal B': 0.28819195469442854, 'Tissue invariant': 0.8755630668677652, 'Vascular / endothelial': 0.033379928682983735}\n",
      "as: 59 annotated out of 160. Enrichment = {'all': 1.2684680346055281, 'Cancer / epithelial': 0.9650646593321752, 'Cardiac': 1.315097317201473, 'Digestive': 1.0909884355225834, 'Lymphoid': 1.2295585884667406, 'Musculoskeletal': 2.1008403361344534, 'Myeloid / erythroid': 1.3855532976168483, 'Neural': 1.6959500712299032, 'Organ devel. / renal': 0.0, 'Placental / trophoblast': 1.621095860801902, 'Primitive / embryonic': 1.2343803410687835, 'Pulmonary devel.': 0.0, 'Renal / cancer': 0.5670220004536176, 'Stromal A': 0.0, 'Stromal B': 1.3965180150823946, 'Tissue invariant': 0.5999520038396928, 'Vascular / endothelial': 3.1068765534382763} - {'all': 0.020077110642401173, 'Cancer / epithelial': 0.6150596734062763, 'Cardiac': 0.44993823812110373, 'Digestive': 0.548375687937551, 'Lymphoid': 0.4098745914369913, 'Musculoskeletal': 0.06834236099502163, 'Myeloid / erythroid': 0.3682406674587624, 'Neural': 0.07317920008409179, 'Organ devel. / renal': 1.0, 'Placental / trophoblast': 0.16789364205887605, 'Primitive / embryonic': 0.2560013828679087, 'Pulmonary devel.': 1.0, 'Renal / cancer': 0.8302437440567594, 'Stromal A': 1.0, 'Stromal B': 0.2611263030579407, 'Tissue invariant': 0.8127934480215222, 'Vascular / endothelial': 0.07352688592282636}\n",
      "as: 15 annotated out of 47. Enrichment = {'all': 1.0978446747397286, 'Cancer / epithelial': 1.6426632499271068, 'Cardiac': 4.47692703728161, 'Digestive': 0.0, 'Lymphoid': 2.0928656824965794, 'Musculoskeletal': 1.1919661481613921, 'Myeloid / erythroid': 0.0, 'Neural': 1.154689410199083, 'Organ devel. / renal': 0.0, 'Placental / trophoblast': 2.759312103492599, 'Primitive / embryonic': 0.6464839756825052, 'Pulmonary devel.': 0.0, 'Renal / cancer': 0.0, 'Stromal A': 0.0, 'Stromal B': 0.7923506468552594, 'Tissue invariant': 0.0, 'Vascular / endothelial': 3.5255336776604556} - {'all': 0.3861756428246442, 'Cancer / epithelial': 0.4581382633140163, 'Cardiac': 0.07374300216629597, 'Digestive': 1.0, 'Lymphoid': 0.24774096887485772, 'Musculoskeletal': 0.5710984792155777, 'Myeloid / erythroid': 1.0, 'Neural': 0.520860149586369, 'Organ devel. / renal': 1.0, 'Placental / trophoblast': 0.09487344909779123, 'Primitive / embryonic': 0.8242787813010644, 'Pulmonary devel.': 1.0, 'Renal / cancer': 1.0, 'Stromal A': 1.0, 'Stromal B': 0.7217741034880494, 'Tissue invariant': 1.0, 'Vascular / endothelial': 0.24761292345783975}\n",
      "as: 1 annotated out of 3. Enrichment = {'all': 1.1466377713948275, 'Cancer / epithelial': 0.0, 'Cardiac': 35.06926179203928, 'Digestive': 0.0, 'Lymphoid': 0.0, 'Musculoskeletal': 0.0, 'Myeloid / erythroid': 0.0, 'Neural': 0.0, 'Organ devel. / renal': 0.0, 'Placental / trophoblast': 0.0, 'Primitive / embryonic': 0.0, 'Pulmonary devel.': 0.0, 'Renal / cancer': 0.0, 'Stromal A': 0.0, 'Stromal B': 0.0, 'Tissue invariant': 0.0, 'Vascular / endothelial': 0.0} - {'all': 0.6431541131871525, 'Cancer / epithelial': 1.0, 'Cardiac': 0.028244823654462625, 'Digestive': 1.0, 'Lymphoid': 1.0, 'Musculoskeletal': 1.0, 'Myeloid / erythroid': 1.0, 'Neural': 1.0, 'Organ devel. / renal': 1.0, 'Placental / trophoblast': 1.0, 'Primitive / embryonic': 1.0, 'Pulmonary devel.': 1.0, 'Renal / cancer': 1.0, 'Stromal A': 1.0, 'Stromal B': 1.0, 'Tissue invariant': 1.0, 'Vascular / endothelial': 1.0}\n"
     ]
    }
   ],
   "source": [
    "for preproc in preprocs:\n",
    "    for expr in expressions:\n",
    "        resdir = basedir+\"/{:s}/{:s}\".format(expr, sumdir)\n",
    "        pcutoffs = [5e-08, 1e-08, 1e-10, 1e-15] #, 1e-08, 1e-10, 1e-15, 1e-20, 1e-25]# [1e-06, 1e-07, 5e-08, 1e-08, 1e-10, 1e-15, 1e-20, 1e-25]\n",
    "        # trans_eqtls_file = \"trans_eqtls_{:s}.txt\".format(file_pcutoff)\n",
    "        trans_eqtls_file = \"trans_eqtls.txt\"\n",
    "        if use_LD:\n",
    "            trans_eqtls_ld_regions_file = trans_eqtls_file+\".ld_regions\"\n",
    "            trans_eqtls_file = trans_eqtls_file+\".ld_prune\"\n",
    "        snpsfile = \"snps_list.txt\"\n",
    "        \n",
    "        for pcutoff in pcutoffs:\n",
    "            logcutoff = -np.log10(pcutoff)\n",
    "            transeqtls = dict()\n",
    "            dhs_annotated = dict()\n",
    "            dhs_annotated_type = dict()\n",
    "            enrichment = collections.defaultdict(dict)\n",
    "            pval_binom = collections.defaultdict(dict)\n",
    "            for tissue in tshorts[:1]:  \n",
    "                filefmt = f'{resdir}/{tissue}/tejaas/{preproc}/{trans_eqtls_file}'\n",
    "                if not os.path.exists(filefmt):\n",
    "                    print(\"File does not exist\", filefmt)\n",
    "                    continue\n",
    "                trans_eqtls = tejaas(filefmt)\n",
    "                if use_LD:\n",
    "                    # delete the lonely signif SNPs\n",
    "                    filefmt_regions = f'{resdir}/{tissue}/tejaas/{preproc}/{trans_eqtls_ld_regions_file}'\n",
    "                    trans_eqtls = smart_LD_filter(trans_eqtls, filefmt_regions)\n",
    "                trans_eqtls_signif = [x for x in trans_eqtls if x.logp >= logcutoff]\n",
    "                transeqtls[tissue] = trans_eqtls_signif\n",
    "                # print(f'{tissue} @ {pcutoff}: {len(transeqtls[tissue])} trans-eQTLs')\n",
    "\n",
    "                dhs_annotated[tissue] = 0\n",
    "                if len(transeqtls[tissue]) > min_trans_eqtl:\n",
    "                    res_dict = dict()\n",
    "                    for chrm in range(1, 23):\n",
    "                        res_dict[chrm] = list()\n",
    "                    for x in  transeqtls[tissue]:\n",
    "                        res_dict[x.chrom].append(x.pos)\n",
    "                    dhs_annotated[tissue], dhs_annotated_type[tissue] = find_annotated(res_dict, dhs_file, isannotated)\n",
    "                    pval_binom[tissue][\"all\"] = ss.binom_test(dhs_annotated[tissue], len(transeqtls[tissue]), dhs_frac_rand[tissue], alternative='greater')\n",
    "                    enrichment[tissue][\"all\"] = float(dhs_annotated[tissue]) / len(transeqtls[tissue]) / dhs_frac_rand[tissue]\n",
    "                    for dhs_type in type_master_list:\n",
    "                        pval_binom[tissue][dhs_type] = ss.binom_test(dhs_annotated_type[tissue][dhs_type], len(transeqtls[tissue]), dhs_frac_type_rand[tissue][dhs_type], alternative='greater')\n",
    "                        enrichment[tissue][dhs_type] = float(dhs_annotated_type[tissue][dhs_type]) / len(transeqtls[tissue]) / dhs_frac_type_rand[tissue][dhs_type]\n",
    "                    # e_pval[tissue] = enrichment_pval(len(transeqtls[tissue]), dhs_frac_rand[tissue], enrichment[tissue])\n",
    "                    print (f\"{tissue}: {dhs_annotated[tissue]} annotated out of {len(transeqtls[tissue])}. Enrichment = {enrichment[tissue]['all']} - {pval_binom[tissue]['all']}\") #{e_pval[tissue]}\n",
    "\n",
    "            \n",
    "            outfile_dhs = os.path.join(outdir, \"dhs_enrichment_{:s}_{:g}.txt\".format(title, pcutoff))\n",
    "            if use_LD:\n",
    "                outfile_dhs = outfile_dhs + \".ld_prune\"\n",
    "            fileperm = \"w\"\n",
    "            write_header = True\n",
    "#             if opts.overwrite:\n",
    "#                 fileperm = \"w\"\n",
    "            if os.path.exists(outfile_dhs):\n",
    "                write_header = False\n",
    "            with open(outfile_dhs, fileperm) as outstream:\n",
    "                if write_header:\n",
    "                    outstream.write(\"tissue\\tdhs_type\\ttranseqtls\\tinDHS\\tEnrichment\\tpval_binom\\n\")\n",
    "                for tissue in tshorts:\n",
    "                    if tissue in transeqtls:\n",
    "                        nteqtl = len(transeqtls[tissue])\n",
    "                        if nteqtl > 0:\n",
    "                            outstream.write(f\"{tissue}\\tall\\t{nteqtl}\\t{dhs_annotated[tissue]}\\t{enrichment[tissue]['all']}\\t{pval_binom[tissue]['all']}\\n\")\n",
    "                            for dhs_type in type_master_list:\n",
    "                                outstream.write(f\"{tissue}\\t{dhs_type}\\t{nteqtl}\\t{dhs_annotated_type[tissue][dhs_type]}\\t{enrichment[tissue][dhs_type]}\\t{pval_binom[tissue][dhs_type]}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
