{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File exists!\n"
     ]
    }
   ],
   "source": [
    "import os, sys\n",
    "sys.path.append('../../')\n",
    "from utils import utils\n",
    "import json\n",
    "import numpy as np\n",
    "import gzip\n",
    "import collections\n",
    "\n",
    "annot_file = \"/cbscratch/franco/datasets/gtex_v8/expression/WGS_Feature_overlap_collapsed_VEP_short_4torus.MAF01.txt.gz\"\n",
    "snp_file   = \"/cbscratch/franco/datasets/gtex_v8/genotypes/vcfs_SHAPEIT2/0.01/gtex_v8_snpinfo_SHAPEIT2.txt\"\n",
    "# Subset full set of annotations to our genotype\n",
    "out_file = \"/cbscratch/franco/datasets/gtex_v8/expression/GTEx_v8_Features_subset_SHAPEIT2_genotype.MAF01.txt.gz\"\n",
    "\n",
    "if not os.path.exists(out_file):\n",
    "    print(\"Reading Feature annotation\")\n",
    "    snp_dict = collections.defaultdict(lambda: False)\n",
    "    with open(snp_file) as instream:\n",
    "        for line in instream:\n",
    "            arr = line.strip().split()\n",
    "            snp_dict[arr[1]] = True\n",
    "\n",
    "    print(\"Writing custom genotype annotations\")\n",
    "    with gzip.open(out_file, 'wb') as outstream:\n",
    "        with gzip.open(annot_file) as instream:\n",
    "            header = next(instream)\n",
    "            outstream.write(header)\n",
    "            for line in instream:\n",
    "                arr = line.decode().strip().split(\"\\t\")\n",
    "                if snp_dict[arr[0]]:\n",
    "                    outstream.write(line)\n",
    "else:\n",
    "    print(\"File exists!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(out_file, header=0, sep=\"\\t\", index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enhancer_d</th>\n",
       "      <th>promoter_d</th>\n",
       "      <th>open_chromatin_region_d</th>\n",
       "      <th>promoter_flanking_region_d</th>\n",
       "      <th>CTCF_binding_site_d</th>\n",
       "      <th>TF_binding_site_d</th>\n",
       "      <th>3_prime_UTR_variant_d</th>\n",
       "      <th>5_prime_UTR_variant_d</th>\n",
       "      <th>frameshift_variant_d</th>\n",
       "      <th>intron_variant_d</th>\n",
       "      <th>missense_variant_d</th>\n",
       "      <th>non_coding_transcript_exon_variant_d</th>\n",
       "      <th>splice_acceptor_variant_d</th>\n",
       "      <th>splice_donor_variant_d</th>\n",
       "      <th>splice_region_variant_d</th>\n",
       "      <th>stop_gained_d</th>\n",
       "      <th>synonymous_variant_d</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_13550_G_A_b38</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_14677_G_A_b38</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_16841_G_T_b38</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_16856_A_G_b38</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_17005_A_G_b38</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    enhancer_d  promoter_d  open_chromatin_region_d  \\\n",
       "SNP                                                                   \n",
       "chr1_13550_G_A_b38           1           0                        0   \n",
       "chr1_14677_G_A_b38           0           0                        0   \n",
       "chr1_16841_G_T_b38           0           0                        0   \n",
       "chr1_16856_A_G_b38           0           0                        0   \n",
       "chr1_17005_A_G_b38           0           0                        0   \n",
       "\n",
       "                    promoter_flanking_region_d  CTCF_binding_site_d  \\\n",
       "SNP                                                                   \n",
       "chr1_13550_G_A_b38                           0                    0   \n",
       "chr1_14677_G_A_b38                           0                    0   \n",
       "chr1_16841_G_T_b38                           0                    0   \n",
       "chr1_16856_A_G_b38                           0                    0   \n",
       "chr1_17005_A_G_b38                           0                    0   \n",
       "\n",
       "                    TF_binding_site_d  3_prime_UTR_variant_d  \\\n",
       "SNP                                                            \n",
       "chr1_13550_G_A_b38                  0                      0   \n",
       "chr1_14677_G_A_b38                  0                      0   \n",
       "chr1_16841_G_T_b38                  0                      0   \n",
       "chr1_16856_A_G_b38                  0                      0   \n",
       "chr1_17005_A_G_b38                  0                      0   \n",
       "\n",
       "                    5_prime_UTR_variant_d  frameshift_variant_d  \\\n",
       "SNP                                                               \n",
       "chr1_13550_G_A_b38                      0                     0   \n",
       "chr1_14677_G_A_b38                      0                     0   \n",
       "chr1_16841_G_T_b38                      0                     0   \n",
       "chr1_16856_A_G_b38                      0                     0   \n",
       "chr1_17005_A_G_b38                      0                     0   \n",
       "\n",
       "                    intron_variant_d  missense_variant_d  \\\n",
       "SNP                                                        \n",
       "chr1_13550_G_A_b38                 0                   0   \n",
       "chr1_14677_G_A_b38                 1                   0   \n",
       "chr1_16841_G_T_b38                 1                   0   \n",
       "chr1_16856_A_G_b38                 0                   0   \n",
       "chr1_17005_A_G_b38                 0                   0   \n",
       "\n",
       "                    non_coding_transcript_exon_variant_d  \\\n",
       "SNP                                                        \n",
       "chr1_13550_G_A_b38                                     1   \n",
       "chr1_14677_G_A_b38                                     0   \n",
       "chr1_16841_G_T_b38                                     0   \n",
       "chr1_16856_A_G_b38                                     0   \n",
       "chr1_17005_A_G_b38                                     1   \n",
       "\n",
       "                    splice_acceptor_variant_d  splice_donor_variant_d  \\\n",
       "SNP                                                                     \n",
       "chr1_13550_G_A_b38                          0                       0   \n",
       "chr1_14677_G_A_b38                          0                       0   \n",
       "chr1_16841_G_T_b38                          0                       0   \n",
       "chr1_16856_A_G_b38                          0                       1   \n",
       "chr1_17005_A_G_b38                          0                       0   \n",
       "\n",
       "                    splice_region_variant_d  stop_gained_d  \\\n",
       "SNP                                                          \n",
       "chr1_13550_G_A_b38                        0              0   \n",
       "chr1_14677_G_A_b38                        0              0   \n",
       "chr1_16841_G_T_b38                        0              0   \n",
       "chr1_16856_A_G_b38                        0              0   \n",
       "chr1_17005_A_G_b38                        0              0   \n",
       "\n",
       "                    synonymous_variant_d  \n",
       "SNP                                       \n",
       "chr1_13550_G_A_b38                     0  \n",
       "chr1_14677_G_A_b38                     0  \n",
       "chr1_16841_G_T_b38                     0  \n",
       "chr1_16856_A_G_b38                     0  \n",
       "chr1_17005_A_G_b38                     0  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8048655, 17)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "enhancer_d\t63300\t0.00786467\n",
      "promoter_d\t62692\t0.00778913\n",
      "open_chromatin_region_d\t190956\t0.0237252\n",
      "promoter_flanking_region_d\t179666\t0.0223225\n",
      "CTCF_binding_site_d\t217021\t0.0269636\n",
      "TF_binding_site_d\t34386\t0.00427227\n",
      "3_prime_UTR_variant_d\t83841\t0.0104168\n",
      "5_prime_UTR_variant_d\t21382\t0.00265659\n",
      "frameshift_variant_d\t0\t0\n",
      "intron_variant_d\t4012743\t0.498561\n",
      "missense_variant_d\t28796\t0.00357774\n",
      "non_coding_transcript_exon_variant_d\t138288\t0.0171815\n",
      "splice_acceptor_variant_d\t371\t4.60947e-05\n",
      "splice_donor_variant_d\t696\t8.64741e-05\n",
      "splice_region_variant_d\t8801\t0.00109347\n",
      "stop_gained_d\t367\t4.55977e-05\n",
      "synonymous_variant_d\t30129\t0.00374336\n",
      "not_annot\t3430119\t0.426173\n"
     ]
    }
   ],
   "source": [
    "## Background frequencies\n",
    "import numpy as np\n",
    "tot = df.shape[0]\n",
    "bg  = dict()\n",
    "for typ, n in zip(df.columns, np.sum(df.values, axis=0)):\n",
    "    bg[typ] = n/tot\n",
    "    print(\"{:s}\\t{:d}\\t{:g}\".format(typ, n, n/tot))\n",
    "    \n",
    "not_annot = np.sum(np.sum(df.values, axis=1) == 0)\n",
    "bg['not_annot'] = not_annot/tot\n",
    "print(\"{:s}\\t{:d}\\t{:g}\".format('not_annot', not_annot, not_annot/tot))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/analysis/plots/tissue_table.txt\"\n",
    "tissues, descriptions, tstrings = utils.read_tissues_str(tissue_file)\n",
    "\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "tissue_colors = dict()\n",
    "tissue_names = dict()\n",
    "tissue_samples = dict()\n",
    "\n",
    "tshorst_dict = dict(zip(tstrings, tissues))\n",
    "\n",
    "for tshort, tfull in zip(tissues, descriptions):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]\n",
    "\n",
    "special_tissues = ['ag', 'haa', 'liv', 'msg', 'pan', 'pit', 'si', 'spl', 'va', 'wb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mpmath\n",
    "import collections\n",
    "mpmath.mp.dps = 50\n",
    "def pvalue(x): return float(mpmath.log10(1 - 0.5 * (1 + mpmath.erf(x/mpmath.sqrt(2)))))\n",
    "\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "def tejaas(filepath, mafcutoff=0.01):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            pos   = int(arr[2])\n",
    "            p     = float(arr[7])\n",
    "            chrom = int(arr[1])\n",
    "            q     = float(arr[4])\n",
    "            mu    = float(arr[5])\n",
    "            sigma = float(arr[6])\n",
    "            maf   = float(arr[3])\n",
    "            if maf < mafcutoff or maf > (1-mafcutoff):\n",
    "                continue\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pvalue( (q - mu) / sigma)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf))\n",
    "    return res\n",
    "\n",
    "alltranseqtls = list()\n",
    "transeqtl_dict = dict()\n",
    "tejaas_file = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma01_knn30_cut5e-8/{:s}/trans_eqtls.txt\"\n",
    "for tissue in tissues:\n",
    "    if tissue in special_tissues:\n",
    "        tejaas_file = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_optim_gamma_knn30_cut5e-8/{:s}/trans_eqtls_ldpruned.txt\"\n",
    "    else:\n",
    "        tejaas_file = \"/cbscratch/franco/trans-eqtl/protein_coding_lncRNA_gamma01_knn30_cut5e-8/{:s}/trans_eqtls_ldpruned.txt\"\n",
    "    tissue_tejaas_file = tejaas_file.format(tissue)\n",
    "    tissue_trans_eqtls = tejaas(tissue_tejaas_file)\n",
    "    transeqtl_dict[tissue] = tissue_trans_eqtls\n",
    "    alltranseqtls += tissue_trans_eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/users/fsimone/opt/miniconda/3/envs/env3.6/lib/python3.6/site-packages/ipykernel_launcher.py:19: RuntimeWarning: invalid value encountered in double_scalars\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as ss\n",
    "\n",
    "## Count ocurrence of trans-eQTL in different tissues\n",
    "teqtl_ntissues = collections.defaultdict(int)\n",
    "for snp in alltranseqtls:\n",
    "    teqtl_ntissues[snp.rsid] += 1\n",
    "    \n",
    "teqtl_cats = dict()\n",
    "for tissue in tissues:\n",
    "    teqtl_cats[tissue] = dict()\n",
    "    ids = [snp.rsid for snp in transeqtl_dict[tissue]]\n",
    "    ntot = len(set(ids))\n",
    "    df_slice = df.loc[ids]\n",
    "    cat_counts = np.sum(df_slice, axis=0)\n",
    "    for c in cat_counts.index:\n",
    "        # teqtl_cats[tissue][c] = cat_counts[c] \n",
    "        teqtl_cats[tissue][c] = dict(counts = int(cat_counts[c]), \n",
    "                                      frac = cat_counts[c]/ntot, \n",
    "                                      enrichment = cat_counts[c]/ntot / bg[c],\n",
    "                                      pval = ss.binom_test(cat_counts[c], ntot, bg[c], alternative='greater') )\n",
    "    not_annot = np.sum(np.sum(df_slice, axis=1) == 0)\n",
    "    # teqtl_cats[tissue]['no_annot'] = not_annot\n",
    "    teqtl_cats[tissue]['not_annot'] = dict(counts = int(not_annot), \n",
    "                                     frac = not_annot/ntot, \n",
    "                                     enrichment = not_annot/ntot / bg['not_annot'],\n",
    "                                     pval = ss.binom_test(not_annot, ntot, bg['not_annot'], alternative='greater') )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/users/fsimone/opt/miniconda/3/envs/env3.6/lib/python3.6/site-packages/ipykernel_launcher.py:12: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  if sys.path[0] == '':\n"
     ]
    }
   ],
   "source": [
    "## Get category counts across all trans-eqtls in all tissues\n",
    "teqtl_cats['all'] = dict()\n",
    "all_rsids = list(teqtl_ntissues.keys())\n",
    "ntot = len(set(all_rsids))\n",
    "df_slice = df.loc[all_rsids]\n",
    "cat_counts = np.sum(df_slice, axis=0)\n",
    "for c in cat_counts.index:\n",
    "    #print(c, cat_counts[c])\n",
    "#     teqtl_cats['all'][c] = cat_counts[c] \n",
    "    teqtl_cats['all'][c] = dict( counts = int(cat_counts[c]), \n",
    "                           frac = cat_counts[c]/ntot, \n",
    "                           enrichment = cat_counts[c]/ntot / bg[c],\n",
    "                           pval = ss.binom_test(cat_counts[c], ntot, bg[c], alternative='greater') )\n",
    "not_annot = np.sum(np.sum(df_slice, axis=1) == 0)\n",
    "# teqtl_cats['all']['no_annot'] = not_annot\n",
    "teqtl_cats['all']['not_annot'] = dict( counts = int(not_annot), \n",
    "                                frac = not_annot/ntot, \n",
    "                                enrichment = not_annot/ntot / bg['not_annot'],\n",
    "                                pval = ss.binom_test(not_annot, ntot, bg['not_annot'], alternative='greater') )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20811 4031 1596 778 1211\n",
      "0.73208569317902 0.14180180814014845 0.05614380694410244 0.027368346994054948 0.04260034474267422\n"
     ]
    }
   ],
   "source": [
    "## Get category counts across trans-eqtls in 1, 2, 3, 4, 5+ tissues\n",
    "n1_rsids = [k for k in teqtl_ntissues.keys() if teqtl_ntissues[k] == 1]\n",
    "n2_rsids = [k for k in teqtl_ntissues.keys() if teqtl_ntissues[k] == 2]\n",
    "n3_rsids = [k for k in teqtl_ntissues.keys() if teqtl_ntissues[k] == 3]\n",
    "n4_rsids = [k for k in teqtl_ntissues.keys() if teqtl_ntissues[k] == 4]\n",
    "n5_rsids = [k for k in teqtl_ntissues.keys() if teqtl_ntissues[k] >= 5]\n",
    "\n",
    "ntot = len(all_rsids)\n",
    "print(len(n1_rsids), len(n2_rsids), len(n3_rsids), len(n4_rsids), len(n5_rsids))\n",
    "print(len(n1_rsids)/ntot, len(n2_rsids)/ntot, len(n3_rsids)/ntot, len(n4_rsids)/ntot, len(n5_rsids)/ntot)\n",
    "\n",
    "ncombinations = dict(zip(['n1', 'n2', 'n3', 'n4', 'n5'], [n1_rsids, n2_rsids, n3_rsids, n4_rsids, n5_rsids]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n1 20811\n",
      "n2 4031\n",
      "n3 1596\n",
      "n4 778\n",
      "n5 1211\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/users/fsimone/opt/miniconda/3/envs/env3.6/lib/python3.6/site-packages/ipykernel_launcher.py:12: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  if sys.path[0] == '':\n"
     ]
    }
   ],
   "source": [
    "for k in ncombinations:\n",
    "    ids = ncombinations[k]\n",
    "    ntot = len(set(ids))\n",
    "    print(k, ntot)\n",
    "    teqtl_cats[k] = dict()\n",
    "    df_slice = df.loc[ids]\n",
    "    cat_counts = np.sum(df_slice, axis=0)\n",
    "    for c in cat_counts.index:\n",
    "        # teqtl_cats[k][c] = cat_counts[c]\n",
    "        teqtl_cats[k][c] = dict(counts = int(cat_counts[c]), \n",
    "                           frac = cat_counts[c]/ntot, \n",
    "                           enrichment = cat_counts[c]/ntot / bg[c],\n",
    "                           pval = ss.binom_test(cat_counts[c], ntot, bg[c], alternative='greater') )\n",
    "    not_annot = np.sum(np.sum(df_slice, axis=1) == 0)\n",
    "    # teqtl_cats[k]['no_annot'] = not_annot\n",
    "    teqtl_cats[k]['not_annot'] = dict(counts = int(not_annot), \n",
    "                                frac = not_annot/ntot, \n",
    "                                enrichment = not_annot/ntot / bg['not_annot'],\n",
    "                                pval = ss.binom_test(not_annot, ntot, bg['not_annot'], alternative='greater') )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "json_dict = json.dumps(teqtl_cats)\n",
    "with open(\"gtex_v8_feature_enrichments.txt\", 'w') as outstream:\n",
    "    outstream.write(json_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
