{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "sys.path.append('gwas-trans-eqtl-analysis/analysis/utils/')\n",
    "import myutils\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "loading_file = \"datasets/hore_2016/gene_loadings.txt\"\n",
    "df = pd.read_csv(loading_file, header=0, index_col=0, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dbsnp_catalog_path = \"datasets/gtex_v8/metadata/variant_b37_rs_id_dbSNP151_GRCh38p7_map.tsv\"\n",
    "# maps variant id to rsid sbSNP151 GRCh38p7\n",
    "dbsnp_dict, variantid_dict = myutils.get_id_dictionaries(dbsnp_catalog_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "def load_snpinfo(snpinfofile):\n",
    "    snp_list = list()\n",
    "    with gzip.open(snpinfofile) as inpt:\n",
    "        for line in inpt:\n",
    "            arr = line.decode().strip().split(\" \")\n",
    "            var_id = arr[1]\n",
    "            snp_list.append(var_id)\n",
    "    return snp_list\n",
    "\n",
    "snpinfofile = \"datasets/gtex_v8/gtex_v8_snpinfo_SHAPEIT2.txt.gz\"\n",
    "snps_list = load_snpinfo(snpinfofile)\n",
    "rsid_list = [dbsnp_dict[x] for x in snps_list]\n",
    "tejaas_rsids = dict(zip(rsid_list, snps_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['rs17611866', 'rs9924520', 'rs194862', 'rs6882516', 'rs12630796'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-a8999edb8344>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mpaper_rsids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"rs194862\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"rs9924520\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"rs289749\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"rs6882516\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"rs2401506\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"rs17611866\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"rs12630796\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpaper_rsids\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2804\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2805\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2806\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2807\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2808\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[0;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1551\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1552\u001b[0m         self._validate_read_indexer(\n\u001b[0;32m-> 1553\u001b[0;31m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis_number\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraise_missing\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1554\u001b[0m         )\n\u001b[1;32m   1555\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[0;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1644\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"loc\"\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1645\u001b[0m                 \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1646\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{not_found} not in index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1647\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1648\u001b[0m             \u001b[0;31m# we skip the warning on Categorical/Interval\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['rs17611866', 'rs9924520', 'rs194862', 'rs6882516', 'rs12630796'] not in index\""
     ]
    }
   ],
   "source": [
    "paper_rsids = [\"rs194862\",\"rs9924520\",\"rs289749\",\"rs6882516\",\"rs2401506\",\"rs17611866\",\"rs12630796\"]\n",
    "df[paper_rsids]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs2401506 chr22_18144963_C_G_b38\n",
      "rs289749 chr16_56992382_G_A_b38\n",
      "in paper  rs289749\n",
      "rs194862 chr7_104145832_T_C_b38\n",
      "in paper  rs194862\n",
      "rs9924520 chr16_10880498_G_A_b38\n",
      "in paper  rs9924520\n",
      "rs289749 chr16_56992382_G_A_b38\n",
      "in paper  rs289749\n",
      "rs6882516 chr5_157755889_C_A_b38\n",
      "in paper  rs6882516\n",
      "rs2401506 chr22_18144963_C_G_b38\n",
      "rs17611866 chr16_3317229_T_C_b38\n",
      "in paper  rs17611866\n",
      "rs12630796 chr3_101512090_C_G_b38\n",
      "analyzed 715 snps of the 944 SNPs in Hore\n",
      "6 from main Hore paper\n"
     ]
    }
   ],
   "source": [
    "paper_rsids = [\"rs194862\",\"rs9924520\",\"rs289749\",\"rs6882516\",\"rs2401506\",\"rs17611866\",\"rs12630796\"]\n",
    "#in ld rs13320918 rs2553419 rs2682386 rs9859077 rs2141180\n",
    "\n",
    "# for rsid in paper_rsids:\n",
    "counter = 0\n",
    "counter_paper = 0\n",
    "for rsid in list(df.columns)+paper_rsids:\n",
    "    if rsid.startswith(\"rs\"):\n",
    "        if rsid in paper_rsids:\n",
    "            print(rsid, variantid_dict[rsid])\n",
    "        if rsid in tejaas_rsids:\n",
    "            counter += 1\n",
    "            if rsid in paper_rsids:\n",
    "                counter_paper += 1\n",
    "                print(\"in paper \", rsid)\n",
    "            \n",
    "print(f\"analyzed {counter} snps of the {df.shape[1]} SNPs in Hore\")\n",
    "print(f\"{counter_paper} from main Hore paper\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as has 2011 trans-eqtls with 586 lead variants\n",
      "av has 1664 trans-eqtls with 463 lead variants\n",
      "ag has 822 trans-eqtls with 184 lead variants\n",
      "aa has 5760 trans-eqtls with 1298 lead variants\n",
      "ac has 7551 trans-eqtls with 1539 lead variants\n",
      "at has 1463 trans-eqtls with 391 lead variants\n",
      "bam has 4780 trans-eqtls with 1182 lead variants\n",
      "ban has 3890 trans-eqtls with 954 lead variants\n",
      "bca has 27 trans-eqtls with 19 lead variants\n",
      "bceh has 270 trans-eqtls with 128 lead variants\n",
      "bce has 96 trans-eqtls with 40 lead variants\n",
      "bco has 113 trans-eqtls with 64 lead variants\n",
      "bfr has 58 trans-eqtls with 32 lead variants\n",
      "bhi has 170 trans-eqtls with 51 lead variants\n",
      "bhy has 325 trans-eqtls with 83 lead variants\n",
      "bnu has 1451 trans-eqtls with 379 lead variants\n",
      "bpu has 14 trans-eqtls with 10 lead variants\n",
      "bsp has 186 trans-eqtls with 61 lead variants\n",
      "bsu has 202 trans-eqtls with 28 lead variants\n",
      "br has 1764 trans-eqtls with 505 lead variants\n",
      "ebv has 1241 trans-eqtls with 297 lead variants\n",
      "fib has 767 trans-eqtls with 209 lead variants\n",
      "cols has 320 trans-eqtls with 75 lead variants\n",
      "colt has 5282 trans-eqtls with 1174 lead variants\n",
      "esog has 71 trans-eqtls with 17 lead variants\n",
      "esom has 343 trans-eqtls with 9 lead variants\n",
      "esomu has 68 trans-eqtls with 20 lead variants\n",
      "haa has 0 trans-eqtls with 0 lead variants\n",
      "hlv has 953 trans-eqtls with 247 lead variants\n",
      "kc has 39 trans-eqtls with 14 lead variants\n",
      "liv has 189 trans-eqtls with 37 lead variants\n",
      "lu has 1975 trans-eqtls with 477 lead variants\n",
      "msg has 17 trans-eqtls with 3 lead variants\n",
      "ms has 11347 trans-eqtls with 2253 lead variants\n",
      "nt has 2457 trans-eqtls with 720 lead variants\n",
      "ov has 321 trans-eqtls with 80 lead variants\n",
      "pan has 0 trans-eqtls with 0 lead variants\n",
      "pit has 39 trans-eqtls with 10 lead variants\n",
      "pro has 4193 trans-eqtls with 886 lead variants\n",
      "snse has 3010 trans-eqtls with 725 lead variants\n",
      "sse has 1548 trans-eqtls with 400 lead variants\n",
      "si has 286 trans-eqtls with 82 lead variants\n",
      "spl has 9 trans-eqtls with 7 lead variants\n",
      "sto has 905 trans-eqtls with 252 lead variants\n",
      "tes has 5452 trans-eqtls with 1286 lead variants\n",
      "thy has 4097 trans-eqtls with 998 lead variants\n",
      "ut has 8563 trans-eqtls with 1683 lead variants\n",
      "va has 0 trans-eqtls with 0 lead variants\n",
      "wb has 0 trans-eqtls with 0 lead variants\n",
      "as has 2021 trans-eqtls with 587 lead variants\n",
      "av has 16356 trans-eqtls with 3534 lead variants\n",
      "ag has 32131 trans-eqtls with 5942 lead variants\n",
      "aa has 5777 trans-eqtls with 1299 lead variants\n",
      "ac has 7600 trans-eqtls with 1549 lead variants\n",
      "at has 1463 trans-eqtls with 391 lead variants\n",
      "bam has 4899 trans-eqtls with 1193 lead variants\n",
      "ban has 3907 trans-eqtls with 959 lead variants\n",
      "bca has 29 trans-eqtls with 19 lead variants\n",
      "bceh has 271 trans-eqtls with 129 lead variants\n",
      "bce has 99 trans-eqtls with 41 lead variants\n",
      "bco has 115 trans-eqtls with 64 lead variants\n",
      "bfr has 63 trans-eqtls with 32 lead variants\n",
      "bhi has 171 trans-eqtls with 51 lead variants\n",
      "bhy has 967 trans-eqtls with 268 lead variants\n",
      "bnu has 1461 trans-eqtls with 379 lead variants\n",
      "bpu has 57 trans-eqtls with 22 lead variants\n",
      "bsp has 188 trans-eqtls with 63 lead variants\n",
      "bsu has 279 trans-eqtls with 51 lead variants\n",
      "br has 1792 trans-eqtls with 510 lead variants\n",
      "ebv has 1259 trans-eqtls with 306 lead variants\n",
      "fib has 768 trans-eqtls with 210 lead variants\n",
      "cols has 463 trans-eqtls with 113 lead variants\n",
      "colt has 30588 trans-eqtls with 5687 lead variants\n",
      "esog has 72 trans-eqtls with 18 lead variants\n",
      "esom has 342 trans-eqtls with 8 lead variants\n",
      "esomu has 68 trans-eqtls with 20 lead variants\n",
      "haa has 655 trans-eqtls with 151 lead variants\n",
      "hlv has 953 trans-eqtls with 247 lead variants\n",
      "kc has 39 trans-eqtls with 14 lead variants\n",
      "liv has 9553 trans-eqtls with 2006 lead variants\n",
      "lu has 1979 trans-eqtls with 477 lead variants\n",
      "msg has 1763 trans-eqtls with 427 lead variants\n",
      "ms has 11227 trans-eqtls with 2232 lead variants\n",
      "nt has 2459 trans-eqtls with 721 lead variants\n",
      "ov has 516 trans-eqtls with 124 lead variants\n",
      "pan has 784 trans-eqtls with 215 lead variants\n",
      "pit has 2089 trans-eqtls with 479 lead variants\n",
      "pro has 20796 trans-eqtls with 4027 lead variants\n",
      "snse has 3010 trans-eqtls with 725 lead variants\n",
      "sse has 1545 trans-eqtls with 399 lead variants\n",
      "si has 35668 trans-eqtls with 6611 lead variants\n",
      "spl has 19876 trans-eqtls with 4069 lead variants\n",
      "sto has 902 trans-eqtls with 252 lead variants\n",
      "tes has 5523 trans-eqtls with 1300 lead variants\n",
      "thy has 4108 trans-eqtls with 999 lead variants\n",
      "ut has 8671 trans-eqtls with 1696 lead variants\n",
      "va has 326 trans-eqtls with 94 lead variants\n",
      "wb has 31546 trans-eqtls with 5890 lead variants\n"
     ]
    }
   ],
   "source": [
    "### 1st, check how many snps from Hore we replicate (in the same 3 tissues)\n",
    "### If we replicate some, then\n",
    "### 2nd, run Tejaas with beta values only for them\n",
    "import collections\n",
    "\n",
    "basepath = \"/home/franco/soedinglab\"\n",
    "basepath_data = os.path.join(basepath, \"datasets\")\n",
    "basepath_results = os.path.join(\"gwas-trans-eqtl-analysis\", \"results\")\n",
    "basepath_gtex = os.path.join(basepath_data, \"gtex_v8\")\n",
    "\n",
    "gamma1_file = \"protein_coding_lncRNA_gamma01_knn30_cut5e-8\"\n",
    "optim_file = \"protein_coding_lncRNA_optim_gamma_knn30_cut5e-8\"\n",
    "\n",
    "gamma1 = \"gamma01_knn30_cut5e-8\"\n",
    "optim = \"optim_gamma_knn30_cut5e-8\"\n",
    "basepath_gamma1_trans_eqtls = os.path.join(basepath_gtex, gamma1_file)\n",
    "basepath_optim_trans_eqtls = os.path.join(basepath_gtex, optim_file)\n",
    "\n",
    "chromlist = range(1,23)\n",
    "\n",
    "tissue_file    = os.path.join(\"gwas-trans-eqtl-analysis/analysis/tissue_table.txt\")\n",
    "gtex_meta_file = os.path.join(\"../../../../trans-eqtl-manuscript/figures/results/gtex/gtex_v8_metadata.json\")\n",
    "optim_tissues = ['ag', 'haa', 'liv', 'msg', 'pan', 'pit', 'si', 'spl', 'va', 'wb']\n",
    "\n",
    "tshorts, tfulls, tstrings = myutils.read_tissues(tissue_file)\n",
    "tissues_string = dict(zip(tshorts, tstrings))\n",
    "tissues_full = dict(zip(tshorts, tfulls))\n",
    "tissues_abbr = dict(zip(tfulls, tshorts))\n",
    "\n",
    "tissue_eqtls_gamma1, tissue_eqtls_ld_gamma1 = myutils.read_trans_eqtls(basepath_gamma1_trans_eqtls, \n",
    "                                                                        tshorts, \n",
    "                                                                        dbsnp_dict)\n",
    "tissue_eqtls_optim, tissue_eqtls_ld_optim = myutils.read_trans_eqtls(basepath_optim_trans_eqtls, \n",
    "                                                                        tshorts, \n",
    "                                                                        dbsnp_dict)\n",
    "\n",
    "tissue_eqtls_ld = collections.defaultdict(list)\n",
    "tissue_eqtls_all = collections.defaultdict(list)\n",
    "for tissue in tshorts:\n",
    "    if tissue in optim_tissues:\n",
    "        tissue_eqtls_ld[tissue] = tissue_eqtls_ld_optim[tissue]\n",
    "        tissue_eqtls_all[tissue] = tissue_eqtls_optim[tissue]\n",
    "    else:\n",
    "        tissue_eqtls_ld[tissue] = tissue_eqtls_ld_gamma1[tissue]\n",
    "        tissue_eqtls_all[tissue] = tissue_eqtls_gamma1[tissue]\n",
    "        \n",
    "rsids_ld = collections.defaultdict(set)\n",
    "for tissue in tshorts:\n",
    "    for entry in tissue_eqtls_ld[tissue]:\n",
    "        rsids_ld[tissue].add(entry.dbsnp)\n",
    "        \n",
    "rsids_all = collections.defaultdict(set)\n",
    "for tissue in tshorts:\n",
    "    for entry in tissue_eqtls_all[tissue]:\n",
    "        rsids_all[tissue].add(entry.dbsnp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "av rs1932201 replicated!\n",
      "ag rs3817362 replicated!\n",
      "ag rs1932201 replicated!\n",
      "ag rs1792218 replicated!\n",
      "ag rs2405143 replicated!\n",
      "ag rs4742478 replicated!\n",
      "ag rs11641185 replicated!\n",
      "aa rs11641185 replicated!\n",
      "liv rs3817362 replicated!\n",
      "liv rs740000 replicated!\n",
      "ms rs3817362 replicated!\n",
      "ms rs11641185 replicated!\n",
      "ms rs1851555 replicated!\n",
      "ms rs740000 replicated!\n",
      "si rs3817362 replicated!\n",
      "si rs1932201 replicated!\n",
      "si rs2217285 replicated!\n",
      "si rs2405143 replicated!\n",
      "si rs10050872 replicated!\n",
      "si rs1028695 replicated!\n",
      "si rs4742478 replicated!\n",
      "spl rs3817362 replicated!\n",
      "spl rs12531170 replicated!\n",
      "spl rs1474404 replicated!\n",
      "tes rs3817362 replicated!\n",
      "tes rs740000 replicated!\n",
      "ut rs2134180 replicated!\n",
      "wb rs12941904 replicated!\n",
      "wb rs9915548 replicated!\n",
      "wb rs11641185 replicated!\n"
     ]
    }
   ],
   "source": [
    "hore_tissues = [\"ebv\", \"as\", \"sse\", \"snse\"]\n",
    "\n",
    "for tissue in tshorts:\n",
    "    myrsids = list(rsids_all[tissue])\n",
    "    for rsid in df.columns:\n",
    "        # for rsid in paper_rsids:\n",
    "        if rsid.startswith(\"rs\"):\n",
    "            if rsid in myrsids:\n",
    "                print(f\"{tissue} {rsid} replicated!\")\n",
    "                if rsid in paper_rsids:\n",
    "                    print(\"in paper!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import readgtf\n",
    "\n",
    "gencode_file = os.path.join(basepath_data, \"gencode.v26.annotation.gtf.gz\")\n",
    "geneinfo = readgtf.gencode_v12(gencode_file)\n",
    "\n",
    "ensembl_gene_symbol_map = dict()\n",
    "for gene in geneinfo:\n",
    "    ensembl_id = gene.ensembl_id.split(\".\")[0]\n",
    "    ensembl_gene_symbol_map[ensembl_id] = gene.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Trans-eQTLs Variant ids for each tissue\n",
    "# For each trans-eQTL the ensemblIDs of target genes as list\n",
    "target_genes = collections.defaultdict(lambda: collections.defaultdict(list))\n",
    "for tissue in tshorts:\n",
    "    if tissue in optim_tissues:\n",
    "        basepath_trans_eqtls = basepath_optim_trans_eqtls\n",
    "    else:\n",
    "        basepath_trans_eqtls = basepath_gamma1_trans_eqtls\n",
    "    \n",
    "    # Getting the tejaas file belongig to the tissue with the calculated target genes for all the trans-eQTLs found\n",
    "    target_gene_file = os.path.join(basepath_trans_eqtls, tissue, \"target_genes.txt\")\n",
    "    if os.path.exists(target_gene_file):\n",
    "        with open(target_gene_file) as fin:\n",
    "            next(fin)\n",
    "            for line in fin:\n",
    "                arr = line.split(\"\\t\")\n",
    "                variant_id = arr[0]\n",
    "                # getting rid of version number eg. ENSG00000009780.15 becomes ENSG00000009780\n",
    "                ensembl_id = arr[1].split(\".\")[0]\n",
    "                target_genes[tissue][variant_id].append(ensembl_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSG00000117616',\n",
       " 'ENSG00000076003',\n",
       " 'ENSG00000188282',\n",
       " 'ENSG00000206561',\n",
       " 'ENSG00000248019',\n",
       " 'ENSG00000138640',\n",
       " 'ENSG00000181381',\n",
       " 'ENSG00000106483',\n",
       " 'ENSG00000189233',\n",
       " 'ENSG00000108091',\n",
       " 'ENSG00000119705',\n",
       " 'ENSG00000100603',\n",
       " 'ENSG00000270240',\n",
       " 'ENSG00000271503',\n",
       " 'ENSG00000161381',\n",
       " 'ENSG00000274173']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_genes['wb'][variantid_dict['rs11641185']]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
