{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "sys.path.append('../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "from utils import readgtf\n",
    "from utils import utils\n",
    "import mpmath\n",
    "import collections\n",
    "from operator import attrgetter\n",
    "import gzip\n",
    "\n",
    "# gene_info = readgtf.gencode_v12(\"/cbscratch/franco/datasets/GENCODE/gencode.v26.annotation.gtf.gz\", trim=True)\n",
    "# gene_info_dict = collections.defaultdict(dict)\n",
    "# for gene in gene_info:\n",
    "#     gene_info_dict[gene.chrom][gene.ensembl_id] = gene.typ\n",
    "\n",
    "mpmath.mp.dps = 50\n",
    "def pvalue(x): return float(mpmath.log10(1 - 0.5 * (1 + mpmath.erf(x/mpmath.sqrt(2)))))\n",
    "\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'target', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "CT_FIELDS = ['tissue', 'ncis', 'ntrans', 'ncistrans', 'randtrans', 'enrichment', 'pval']\n",
    "class CisTrans(collections.namedtuple('_CisTrans', CT_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "CT_TYPE_FIELDS = ['tissue', 'genetype', 'hits', 'enrichment', 'pval', 'frac']\n",
    "class CisTrans_type(collections.namedtuple('_CisTransType', CT_TYPE_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "def load_snp_maf(filepath, tissue):\n",
    "    snp_maf_dict = collections.defaultdict(lambda:False)\n",
    "    for chrm in np.arange(1,23):\n",
    "        with open(filepath.format(tissue, chrm)) as instream:\n",
    "            for line in instream:\n",
    "                arr = line.strip().split(\"\\t\")\n",
    "                snp_maf_dict[arr[0]] = float(arr[1])\n",
    "    return snp_maf_dict\n",
    "\n",
    "def tejaas(filepath, mafcutoff=0.01):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = int(arr[1])\n",
    "            pos   = int(arr[2])\n",
    "            maf   = float(arr[3])\n",
    "            if maf < mafcutoff or maf > (1-mafcutoff):\n",
    "                continue\n",
    "            q     = float(arr[4])\n",
    "            mu    = float(arr[5])\n",
    "            sigma = float(arr[6])\n",
    "            p     = float(arr[7])\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pvalue( (q - mu) / sigma)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf, target=None))\n",
    "    return res\n",
    "\n",
    "def read_cis(filepath):\n",
    "    res = list()\n",
    "    if not os.path.exists(filepath) or os.stat(filepath).st_size == 0:\n",
    "        print(\"File empty or does not exist\")\n",
    "        return res\n",
    "    with gzip.open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr  = line.decode().strip().split(\"\\t\")\n",
    "            rsid = arr[0]\n",
    "            if rsid.startswith(\"chrX\"):\n",
    "                continue\n",
    "            pos = int(rsid.split(\"_\")[1])\n",
    "            chrom = int(rsid.split(\"_\")[0][3:])\n",
    "            gene = arr[1].split(\":\")[-1].split(\".\")[0]\n",
    "            maf  = float(arr[5])\n",
    "            logp = np.log10(float(arr[6]))\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, target=gene, maf=maf))\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "maf1 /cbscratch/franco/datasets/gtex_v8/genotypes/snpMAFs/{:s}_snp_CHR{:d}.maf /usr/users/fsimone/vcfs_0.01/\n"
     ]
    }
   ],
   "source": [
    "# Filter by allowed snps according to MAF\n",
    "basepath = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_lncRNA_freeze/\"\n",
    "baseoutdir = os.path.join(basepath, \"GTExPortal_sqtl_analysis\")\n",
    "if not os.path.exists(baseoutdir): os.makedirs(baseoutdir)\n",
    "\n",
    "MAF = 0.01\n",
    "title = \"maf{:g}\".format(MAF*100)\n",
    "maffile = \"/cbscratch/franco/datasets/gtex_v8/genotypes/snpMAFs/{:s}_snp_CHR{:d}.maf\"\n",
    "randompath = \"/usr/users/fsimone/vcfs_{:.2f}/\".format(MAF)\n",
    "\n",
    "print(title, maffile, randompath)\n",
    "\n",
    "outdir = os.path.join(baseoutdir, title)\n",
    "if not os.path.exists(outdir): os.makedirs(outdir)\n",
    "\n",
    "# NTOT_SNPS = 0\n",
    "# snp_maf_dict = collections.defaultdict(lambda:False)\n",
    "# with open(maffile) as instream:\n",
    "#     for line in instream:\n",
    "#         snp_maf_dict[line.rstrip().split()[1]] = True\n",
    "#         NTOT_SNPS += 1\n",
    "# print(\"Total Number of SNPs at MAF {:g}%: {:d}\".format(MAF*100, NTOT_SNPS))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for a set of eQTLs, get me those targeting a specific type of targets\n",
    "def cis_typespecific_eqtls(cistrans_target_eqtls, genetype_dict):\n",
    "    cis_typespecific_eqtls = [x for x in cistrans_target_eqtls if genetype_dict[x.target]]\n",
    "    \n",
    "    uniq_cis_snps = list(set([x.rsid for x in cis_typespecific_eqtls]))\n",
    "    unique_targets = list(set([x.target for x in cis_typespecific_eqtls]))\n",
    "    return cis_typespecific_eqtls, uniq_cis_snps, unique_targets\n",
    "\n",
    "# for a set of cis and trans-eQTLs, return the cis-trans ids and the cis-eQTLs with its targets\n",
    "def cross_ref_cis_trans(trans_ids, cis_eqtls):\n",
    "    cis_ids = list(set([x.rsid for x in cis_eqtls]))\n",
    "    \n",
    "    #Intersection between cis-eqtls (MatrixEQTL) and trans-eqtls (TEJAAS)\n",
    "    cis_trans_eqtls_ids = list(set.intersection(set(trans_ids), set(cis_ids)))\n",
    "    \n",
    "    #set up a dict for fast look up later\n",
    "    cis_trans_dict = dict()\n",
    "    for x in cis_trans_eqtls_ids:\n",
    "        cis_trans_dict[x] = True\n",
    "    \n",
    "    # List of cis-trans-eqtls with its target gene\n",
    "    cis_target_eqtls = [x for x in cis_eqtls if cis_trans_dict.get(x.rsid, False)]\n",
    "\n",
    "    return cis_trans_eqtls_ids, cis_target_eqtls\n",
    "\n",
    "def crossref_trans_tejaas(transeqtls, cis_eqtls):\n",
    "    trans_ids = [x.rsid for x in transeqtls]\n",
    "    a, b = cross_ref_cis_trans(trans_ids, cis_eqtls)\n",
    "    return a, b\n",
    "\n",
    "def get_cistype_fractions(ciseqtls, valid_types, alltypes_dict):\n",
    "    cistype_frac_dict = dict()\n",
    "    NCIS = len(list(set([x.rsid for x in ciseqtls])))\n",
    "    for gtype in valid_types:\n",
    "        cishits = list()\n",
    "        for ciseqtl in ciseqtls:\n",
    "            if alltypes_dict[gtype][ciseqtl.target]:\n",
    "                cishits.append(ciseqtl.rsid)\n",
    "        NCIS_TYPE = len(list(set(cishits)))\n",
    "        cistype_frac_dict[gtype] = NCIS_TYPE / NCIS\n",
    "        # print(\"CIS_frac:\", gtype, NCIS_TYPE, NCIS)\n",
    "    return cistype_frac_dict\n",
    "\n",
    "def sample_background_50000_simple(ciseqtls, randompath):\n",
    "    randtrans = list()\n",
    "    chroms    = [str(x) for x in np.arange(1,23)]\n",
    "    for nid in [\"{:03d}\".format(x) for x in np.arange(1, 11)]:\n",
    "        Nrand=\"50000\"\n",
    "        randomfile = randompath+\"random_\"+Nrand+\"_\"+nid\n",
    "\n",
    "        rand_ids = list()\n",
    "        for chrm in chroms:\n",
    "            with open(os.path.join(randomfile, \"chr{:s}.txt\".format(chrm))) as ins:\n",
    "                rand_ids += [line.rstrip() for line in ins]\n",
    "\n",
    "        a, b = cross_ref_cis_trans(rand_ids, ciseqtls)\n",
    "        randtrans.append( len(a) )\n",
    "    return np.mean(randtrans)\n",
    "\n",
    "def sample_background_50000(ciseqtls, randompath, valid_types = [], alltypes_dict = dict()):\n",
    "    randtrans = list()\n",
    "    res_dict  = collections.defaultdict(list)\n",
    "    chroms    = [str(x) for x in np.arange(1,23)]\n",
    "    for nid in [\"{:03d}\".format(x) for x in np.arange(1, 11)]:\n",
    "        Nrand=\"50000\"\n",
    "        randomfile = randompath+\"random_\"+Nrand+\"_\"+nid\n",
    "\n",
    "        rand_ids = list()\n",
    "        for chrm in chroms:\n",
    "            with open(os.path.join(randomfile, \"chr{:s}.txt\".format(chrm))) as ins:\n",
    "                rand_ids += [line.rstrip() for line in ins]\n",
    "\n",
    "        a, b = cross_ref_cis_trans(rand_ids, ciseqtls)\n",
    "        randtrans.append( len(a) )\n",
    "        if len(valid_types) > 0:\n",
    "            for gtype in valid_types:\n",
    "                cis_types_eqtls, uniq_cis_snps, uniq_targets = cis_typespecific_eqtls(b, alltypes_dict[gtype])\n",
    "                # save only the number of uniq cistrans snps found for that gene type\n",
    "                res_dict[gtype].append(len(uniq_cis_snps)) \n",
    "    return np.mean(randtrans), res_dict\n",
    "\n",
    "\n",
    "def sample_binomial(n, p, NTIMES):\n",
    "    array_n = list()\n",
    "    for i in range(NTIMES):\n",
    "        n_success = np.random.binomial(n, p)\n",
    "        array_n.append(n_success)\n",
    "    return array_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENEINFO_FIELDS = ['name', 'ensembl_id', 'chrom', 'start', 'end', 'typ']\n",
    "# class GeneInfo(collections.namedtuple('_GeneInfo', GENEINFO_FIELDS)):\n",
    "#     __slots__ = ()\n",
    "    \n",
    "# def read_TFannot(infile):\n",
    "#     TF_list = list()\n",
    "#     with open(infile) as instream:\n",
    "#         next(instream)\n",
    "#         for line in instream:\n",
    "#             arr = line.rstrip().split()\n",
    "#             TF_list.append(GeneInfo(ensembl_id=arr[0], chrom=int(arr[1]), start=int(arr[2]), end=int(arr[3]), name=arr[4], typ=\"TF\"))\n",
    "#     return TF_list\n",
    "\n",
    "# base_dir = \"/cbscratch/franco/datasets\"\n",
    "# TF_annot = read_TFannot(\"../TF_annotation.txt\")\n",
    "\n",
    "# TF_dict = collections.defaultdict(dict)\n",
    "# for g in TF_annot:\n",
    "#     TF_dict[g.chrom][g.ensembl_id] = \"TF\"\n",
    "\n",
    "# # Reformat genetype dict, we can add as many gene annotations as we want here\n",
    "# alltypes_dict = collections.defaultdict(dict)\n",
    "# genetypes = []\n",
    "# for chrm in range(1,23):\n",
    "#     gene_info_dict[chrm]\n",
    "#     for k in gene_info_dict[chrm].keys():\n",
    "#         genetype = gene_info_dict[chrm][k]\n",
    "#         if genetype not in alltypes_dict:\n",
    "#             alltypes_dict[genetype] = collections.defaultdict(lambda:False)\n",
    "#             genetypes.append(genetype)\n",
    "#         alltypes_dict[genetype][k] = True\n",
    "#     # Add TF dictionary\n",
    "#     for k in TF_dict[chrm].keys():\n",
    "#         genetype = \"TF\"\n",
    "#         if genetype not in alltypes_dict:\n",
    "#             alltypes_dict[genetype] = collections.defaultdict(lambda:False)\n",
    "#             genetypes.append(genetype)\n",
    "#         alltypes_dict[genetype][k] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "json_file = \"../gtex_v8_metadata.json\"\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "    \n",
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/main/tissues.txt\"\n",
    "tissues, descriptions = utils.read_tissues(tissue_file)\n",
    "tissue_names   = dict()\n",
    "tissue_colors  = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tissues, descriptions):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gtexportal_dir = \"/cbscratch/franco/datasets/gtex_v8/expression/gtex_portal/eQTLs/GTEx_Analysis_v8_sQTL/\"\n",
    "dataset = \"gtex_v8\"\n",
    "# meqtl_expr  = \"tmm_cclm\"\n",
    "tejaas_expr = \"raw\"\n",
    "K = 30\n",
    "pcutoff = 5e-8\n",
    "MIN_TRANS = 1\n",
    "MIN_CIS   = 1\n",
    "\n",
    "res_dict = dict()\n",
    "res_dict_cistype = collections.defaultdict(dict)\n",
    "res_dict_randomtype = collections.defaultdict(dict)\n",
    "for tissue in tissues:\n",
    "    gtex_t = \"-\".join([dataset, tissue])\n",
    "    \n",
    "    snp_maf_dict = load_snp_maf(maffile, tissue)\n",
    "    \n",
    "    tejaas_file = os.path.join(basepath, tejaas_expr, gtex_t, \"tejaas\", \"permnull_sb0.1_knn{:d}\".format(K), \"trans_eqtls_{:g}.txt\".format(pcutoff))\n",
    "    if not os.path.exists(tejaas_file):\n",
    "        print(\"{:s} has no trans-eqtl results\".format(tissue))\n",
    "        continue\n",
    "    transeqtls = tejaas(tejaas_file)\n",
    "    # transeqtls = [x for x in transeqtls if snp_maf_dict[x.rsid]]\n",
    "    \n",
    "    if len(transeqtls) < MIN_TRANS:\n",
    "        print(\"{:s} has less than {:d} trans-eqtls\".format(tissue, MIN_TRANS))\n",
    "        continue\n",
    "    \n",
    "    signif_cisfile = os.path.join(gtexportal_dir, \"{:s}.v8.sqtl_signifpairs.txt.gz\".format(tissue_names[tissue].replace(\" \", \"_\")))\n",
    "    if not os.path.exists(signif_cisfile) or os.stat(signif_cisfile).st_size == 0:\n",
    "        print(\"{:s} has no cis-file in GTEx!\".format(tissue_names[tissue]))\n",
    "        continue\n",
    "    ciseqtls = read_cis(signif_cisfile)\n",
    "    cis_ids = list(set([x.rsid for x in ciseqtls]))\n",
    "    \n",
    "    if len(ciseqtls) < MIN_CIS:\n",
    "        print(\"{:s} has less than {:d} cis-eqtls\".format(tissue, MIN_CIS))\n",
    "        continue\n",
    "    \n",
    "    cis_trans_eqtls_ids, cistrans_target_eqtls = crossref_trans_tejaas(transeqtls, ciseqtls)\n",
    "\n",
    "    \n",
    "    valid_types = []\n",
    "    genetarget_counts = dict()\n",
    "    for genetype in genetypes:\n",
    "        cis_types_eqtls, uniq_cis_snps, unique_targets = cis_typespecific_eqtls(cistrans_target_eqtls, alltypes_dict[genetype])\n",
    "        if len(uniq_cis_snps) > 0:\n",
    "            # print(genetype, len(uniq_cis_snps), len(unique_targets))\n",
    "            valid_types.append(genetype)\n",
    "            genetarget_counts[genetype] = [cis_types_eqtls, uniq_cis_snps, unique_targets]\n",
    "    \n",
    "    # randtrans: contains avg number of random snps that are cis-trans-eQTLs\n",
    "    # rand_res_dict: contains, for each genetype, a list with the number of those random cis-trans-eQTLs targeting that genetype\n",
    "    randtrans, rand_res_dict = sample_background_50000(ciseqtls, randompath, valid_types, alltypes_dict)\n",
    "    FRAC_CISTRANS = len(cis_trans_eqtls_ids) / len(transeqtls)\n",
    "    FRAC_RANDOM_GWCISTRANS = randtrans / 50000 \n",
    "    \n",
    "    enrichment = FRAC_CISTRANS / FRAC_RANDOM_GWCISTRANS\n",
    "   \n",
    "    ncis = len(cis_ids)\n",
    "    ntrans = len(transeqtls)\n",
    "    ncistrans = len(cis_trans_eqtls_ids)\n",
    "    \n",
    "    randtrans1k_bin = sample_binomial(ntrans, FRAC_RANDOM_GWCISTRANS, 10000000)\n",
    "    num_null = np.array(randtrans1k_bin) /  ntrans\n",
    "\n",
    "    null_enrichments = num_null / FRAC_RANDOM_GWCISTRANS\n",
    "    ecdf = ECDF(null_enrichments)\n",
    "    pval = 1 - ecdf(enrichment)\n",
    "    \n",
    "    \n",
    "    res_dict[tissue] = CisTrans(tissue=tissue, ncis=ncis, ntrans=ntrans, \n",
    "                                ncistrans=ncistrans, randtrans=FRAC_RANDOM_GWCISTRANS,\n",
    "                                enrichment=enrichment, pval=pval)\n",
    "    \n",
    "    print(f\"########## Tissue: {tissue} - {ntrans} trans-eqtls - {ncistrans} cis-trans-eqtls #########\")\n",
    "    print(f\"{tissue:>20}        Enrichment: {enrichment:>g} - pval: {pval:>g}\")\n",
    "    \n",
    "    cistype_fracs_dict = get_cistype_fractions(ciseqtls, valid_types, alltypes_dict)\n",
    "    \n",
    "    for vt in valid_types:\n",
    "        # from the 50000 random SNPs, randtrans is the nÂº that are cistrans\n",
    "        FRAC_RANDOM_TYPE_NULL = np.mean(np.array(rand_res_dict[vt])) / 50000 #randtrans\n",
    "        FRAC_CIS_TYPE_NULL    = cistype_fracs_dict[vt]\n",
    "        uniq_snps_targettype  = len(genetarget_counts[vt][1]) # contains --> genetarget_counts[genetype] = [cis_types_eqtls, uniq_cis_snps, unique_targets]\n",
    "        FRAC_CISTYPE =  uniq_snps_targettype / ntrans\n",
    "        \n",
    "        if FRAC_CIS_TYPE_NULL != 1.0:        \n",
    "            cis_type_enrichment    = FRAC_CISTYPE / FRAC_CIS_TYPE_NULL\n",
    "            # Calculate pvalue\n",
    "            randtrans1k_bin = sample_binomial(ntrans, FRAC_CIS_TYPE_NULL, 10000000)\n",
    "            num_null = np.array(randtrans1k_bin) /  ntrans\n",
    "            null_enrichments = num_null / FRAC_CIS_TYPE_NULL\n",
    "            ecdf = ECDF(null_enrichments)\n",
    "            cis_pval = 1 - ecdf(cis_type_enrichment)\n",
    "        else:\n",
    "            cis_pval = 1.0\n",
    "            cis_type_enrichment = 1.0\n",
    "          \n",
    "        random_type_enrichment = FRAC_CISTYPE / FRAC_RANDOM_TYPE_NULL\n",
    "        # Calculate pvalue\n",
    "        randtrans1k_bin = sample_binomial(ntrans, FRAC_RANDOM_TYPE_NULL, 10000000)\n",
    "        num_null = np.array(randtrans1k_bin) /  ntrans\n",
    "        null_enrichments = num_null / FRAC_RANDOM_TYPE_NULL\n",
    "        ecdf = ECDF(null_enrichments)\n",
    "        random_pval = 1 - ecdf(random_type_enrichment)\n",
    "                  \n",
    "        print(f\"{vt:>20}    CIS Enrichment: {cis_type_enrichment:>g} - pval: {cis_pval:>g}\")\n",
    "        print(f\"{vt:>20} RANDOM Enrichment: {random_type_enrichment:>g} - pval: {random_pval:>g}\")\n",
    "    \n",
    "        res_dict_cistype[tissue][vt]    = CisTrans_type(tissue=tissue, genetype=vt,\n",
    "                                                        hits=uniq_snps_targettype,\n",
    "                                                        frac=FRAC_CIS_TYPE_NULL,\n",
    "                                                        enrichment=cis_type_enrichment, pval=cis_pval)\n",
    "        res_dict_randomtype[tissue][vt] = CisTrans_type(tissue=tissue, genetype=vt, \n",
    "                                                        hits=uniq_snps_targettype,\n",
    "                                                        frac=FRAC_RANDOM_TYPE_NULL,\n",
    "                                                        enrichment=random_type_enrichment, pval=random_pval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/users/fsimone/opt/miniconda/3/envs/env3.6/lib/python3.6/site-packages/ipykernel_launcher.py:83: RuntimeWarning: divide by zero encountered in log10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########## Tissue: as - 1280 trans-eqtls - 26 cis-trans-eqtls #########\n",
      "                  as        Enrichment: 0.40798 - pval: 1\n",
      "########## Tissue: av - 671 trans-eqtls - 29 cis-trans-eqtls #########\n",
      "                  av        Enrichment: 1.14276 - pval: 0.199263\n",
      "########## Tissue: ag - 349 trans-eqtls - 99 cis-trans-eqtls #########\n",
      "                  ag        Enrichment: 14.4272 - pval: 0\n",
      "########## Tissue: aa - 2615 trans-eqtls - 182 cis-trans-eqtls #########\n",
      "                  aa        Enrichment: 2.06536 - pval: 0\n",
      "########## Tissue: ac - 3479 trans-eqtls - 90 cis-trans-eqtls #########\n",
      "                  ac        Enrichment: 1.54426 - pval: 3.93e-05\n",
      "########## Tissue: at - 920 trans-eqtls - 14 cis-trans-eqtls #########\n",
      "                  at        Enrichment: 0.324506 - pval: 1\n",
      "########## Tissue: bam - 2598 trans-eqtls - 59 cis-trans-eqtls #########\n",
      "                 bam        Enrichment: 3.93857 - pval: 0\n",
      "########## Tissue: ban - 2064 trans-eqtls - 4 cis-trans-eqtls #########\n",
      "                 ban        Enrichment: 0.231043 - pval: 0.999858\n",
      "########## Tissue: bca - 18 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bca        Enrichment: 0 - pval: 0.212491\n",
      "########## Tissue: bceh - 157 trans-eqtls - 2 cis-trans-eqtls #########\n",
      "                bceh        Enrichment: 0.733382 - pval: 0.514373\n",
      "########## Tissue: bce - 49 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bce        Enrichment: 0 - pval: 0.643715\n",
      "########## Tissue: bco - 76 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bco        Enrichment: 0 - pval: 0.686293\n",
      "########## Tissue: bfr - 33 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bfr        Enrichment: 0 - pval: 0.325287\n",
      "########## Tissue: bhi - 108 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bhi        Enrichment: 0 - pval: 0.616491\n",
      "########## Tissue: bhy - 207 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bhy        Enrichment: 0 - pval: 0.881645\n",
      "########## Tissue: bnu - 663 trans-eqtls - 8 cis-trans-eqtls #########\n",
      "                 bnu        Enrichment: 0.861145 - pval: 0.582832\n",
      "########## Tissue: bpu - 7 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bpu        Enrichment: 0 - pval: 0.0649446\n",
      "########## Tissue: bsp - 86 trans-eqtls - 1 cis-trans-eqtls #########\n",
      "                 bsp        Enrichment: 1.49229 - pval: 0.145167\n",
      "########## Tissue: bsu - 40 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bsu        Enrichment: 0 - pval: 0.188277\n",
      "########## Tissue: br - 841 trans-eqtls - 40 cis-trans-eqtls #########\n",
      "                  br        Enrichment: 1.33881 - pval: 0.0282662\n",
      "########## Tissue: ebv - 514 trans-eqtls - 21 cis-trans-eqtls #########\n",
      "                 ebv        Enrichment: 2.61998 - pval: 2.83e-05\n",
      "########## Tissue: fib - 390 trans-eqtls - 27 cis-trans-eqtls #########\n",
      "                 fib        Enrichment: 1.50646 - pval: 0.0143354\n",
      "########## Tissue: cols - 138 trans-eqtls - 18 cis-trans-eqtls #########\n",
      "                cols        Enrichment: 4.788 - pval: 0\n",
      "########## Tissue: colt - 2198 trans-eqtls - 241 cis-trans-eqtls #########\n",
      "                colt        Enrichment: 3.72765 - pval: 0\n",
      "########## Tissue: esog - 30 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                esog        Enrichment: 0 - pval: 0.57707\n",
      "########## Tissue: esom - 90 trans-eqtls - 85 cis-trans-eqtls #########\n",
      "                esom        Enrichment: 25.6587 - pval: 0\n",
      "########## Tissue: esomu - 35 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "               esomu        Enrichment: 0 - pval: 0.741573\n",
      "haa has less than 1 trans-eqtls\n",
      "########## Tissue: hlv - 437 trans-eqtls - 4 cis-trans-eqtls #########\n",
      "                 hlv        Enrichment: 0.426848 - pval: 0.95793\n",
      "########## Tissue: kc - 17 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                  kc        Enrichment: 0 - pval: 0.0404668\n",
      "########## Tissue: liv - 83 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 liv        Enrichment: 0 - pval: 0.616022\n",
      "########## Tissue: lu - 907 trans-eqtls - 90 cis-trans-eqtls #########\n",
      "                  lu        Enrichment: 2.29186 - pval: 0\n",
      "########## Tissue: msg - 4 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 msg        Enrichment: 0 - pval: 0.0408659\n",
      "########## Tissue: ms - 4827 trans-eqtls - 394 cis-trans-eqtls #########\n",
      "                  ms        Enrichment: 1.81807 - pval: 0\n",
      "########## Tissue: nt - 1573 trans-eqtls - 27 cis-trans-eqtls #########\n",
      "                  nt        Enrichment: 0.33854 - pval: 1\n",
      "pan has less than 1 trans-eqtls\n",
      "########## Tissue: pit - 11 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 pit        Enrichment: 0 - pval: 0.23156\n",
      "########## Tissue: snse - 1411 trans-eqtls - 121 cis-trans-eqtls #########\n",
      "                snse        Enrichment: 1.95707 - pval: 0\n",
      "########## Tissue: sse - 642 trans-eqtls - 111 cis-trans-eqtls #########\n",
      "                 sse        Enrichment: 3.4498 - pval: 0\n",
      "########## Tissue: si - 133 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                  si        Enrichment: 0 - pval: 0.853303\n",
      "########## Tissue: spl - 5 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 spl        Enrichment: 0 - pval: 0.0996644\n",
      "########## Tissue: sto - 453 trans-eqtls - 30 cis-trans-eqtls #########\n",
      "                 sto        Enrichment: 3.02619 - pval: 0\n",
      "########## Tissue: thy - 2658 trans-eqtls - 127 cis-trans-eqtls #########\n",
      "                 thy        Enrichment: 0.898295 - pval: 0.886162\n",
      "wb has less than 1 trans-eqtls\n",
      "########## Tissue: ov - 178 trans-eqtls - 14 cis-trans-eqtls #########\n",
      "                  ov        Enrichment: 5.46344 - pval: 1e-07\n",
      "########## Tissue: pro - 2101 trans-eqtls - 36 cis-trans-eqtls #########\n",
      "                 pro        Enrichment: 0.893922 - pval: 0.720304\n",
      "########## Tissue: tes - 2344 trans-eqtls - 288 cis-trans-eqtls #########\n",
      "                 tes        Enrichment: 1.76042 - pval: 0\n",
      "########## Tissue: ut - 3506 trans-eqtls - 27 cis-trans-eqtls #########\n",
      "                  ut        Enrichment: 0.759626 - pval: 0.916655\n",
      "va has less than 1 trans-eqtls\n",
      "Bladder has no cis-file in GTEx!\n"
     ]
    }
   ],
   "source": [
    "gtexportal_dir = \"/cbscratch/franco/datasets/gtex_v8/expression/gtex_portal/eQTLs/GTEx_Analysis_v8_sQTL/\"\n",
    "dataset = \"gtex_v8\"\n",
    "# meqtl_expr  = \"tmm_cclm\"\n",
    "tejaas_expr = \"raw\"\n",
    "K = 30\n",
    "pcutoff = 5e-8\n",
    "MIN_TRANS = 1\n",
    "MIN_CIS   = 1\n",
    "\n",
    "res_dict = dict()\n",
    "res_target_dict = dict()\n",
    "for tissue in tissues:\n",
    "    gtex_t = \"-\".join([dataset, tissue])\n",
    "    \n",
    "    snp_maf_dict = load_snp_maf(maffile, tissue)\n",
    "    \n",
    "    tejaas_file = os.path.join(basepath, tejaas_expr, gtex_t, \"tejaas\", \"permnull_sb0.1_knn{:d}\".format(K), \"trans_eqtls_{:g}.txt\".format(pcutoff))\n",
    "    if not os.path.exists(tejaas_file):\n",
    "        print(\"{:s} has no trans-eqtl results\".format(tissue))\n",
    "        continue\n",
    "    transeqtls = tejaas(tejaas_file)\n",
    "    # transeqtls = [x for x in transeqtls if snp_maf_dict[x.rsid]]\n",
    "    \n",
    "    if len(transeqtls) < MIN_TRANS:\n",
    "        print(\"{:s} has less than {:d} trans-eqtls\".format(tissue, MIN_TRANS))\n",
    "        continue\n",
    "    \n",
    "    signif_cisfile = os.path.join(gtexportal_dir, \"{:s}.v8.sqtl_signifpairs.txt.gz\".format(tissue_names[tissue].replace(\" \", \"_\")))\n",
    "    if not os.path.exists(signif_cisfile) or os.stat(signif_cisfile).st_size == 0:\n",
    "        print(\"{:s} has no cis-file in GTEx!\".format(tissue_names[tissue]))\n",
    "        continue\n",
    "    ciseqtls = read_cis(signif_cisfile)\n",
    "    cis_ids = list(set([x.rsid for x in ciseqtls]))\n",
    "    \n",
    "    if len(ciseqtls) < MIN_CIS:\n",
    "        print(\"{:s} has less than {:d} cis-eqtls\".format(tissue, MIN_CIS))\n",
    "        continue\n",
    "    \n",
    "    cis_trans_eqtls_ids, cistrans_target_eqtls = crossref_trans_tejaas(transeqtls, ciseqtls)\n",
    "    \n",
    "    randtrans = sample_background_50000_simple(ciseqtls, randompath)\n",
    "    FRAC_CISTRANS = len(cis_trans_eqtls_ids) / len(transeqtls)\n",
    "    FRAC_RANDOM_GWCISTRANS = randtrans / 50000 \n",
    "    \n",
    "    enrichment = FRAC_CISTRANS / FRAC_RANDOM_GWCISTRANS\n",
    "   \n",
    "    ncis = len(cis_ids)\n",
    "    ntrans = len(transeqtls)\n",
    "    ncistrans = len(cis_trans_eqtls_ids)\n",
    "    \n",
    "    randtrans1k_bin = sample_binomial(ntrans, FRAC_RANDOM_GWCISTRANS, 10000000)\n",
    "    num_null = np.array(randtrans1k_bin) /  ntrans\n",
    "\n",
    "    null_enrichments = num_null / FRAC_RANDOM_GWCISTRANS\n",
    "    ecdf = ECDF(null_enrichments)\n",
    "    pval = 1 - ecdf(enrichment)\n",
    "    \n",
    "    res_dict[tissue] = CisTrans(tissue=tissue, ncis=ncis, ntrans=ntrans, \n",
    "                                ncistrans=ncistrans, randtrans=FRAC_RANDOM_GWCISTRANS,\n",
    "                                enrichment=enrichment, pval=pval)\n",
    "    \n",
    "    res_target_dict[tissue] = cistrans_target_eqtls\n",
    "    \n",
    "    print(f\"########## Tissue: {tissue} - {ntrans} trans-eqtls - {ncistrans} cis-trans-eqtls #########\")\n",
    "    print(f\"{tissue:>20}        Enrichment: {enrichment:>g} - pval: {pval:>g}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "outcisfilename = os.path.join(outdir,\"CisSQTL_enrichment_results_\"+title+\".txt\")\n",
    "# if os.path.exists(outcisfilename):\n",
    "#     print(\"Warning! File exists\")\n",
    "#     raise\n",
    "with open(outcisfilename, 'w') as outstream:\n",
    "    for tissue in tissues:\n",
    "        if tissue in res_dict:\n",
    "            line = f\"{tissue}\\t{res_dict[tissue].ncis}\\t{res_dict[tissue].ntrans}\\t{res_dict[tissue].ncistrans}\\t{res_dict[tissue].randtrans}\\t{res_dict[tissue].enrichment}\\t{res_dict[tissue].pval}\\n\"\n",
    "            outstream.write(line)\n",
    "\n",
    "targets_outfile = os.path.join(outdir,\"CistransSQTL_targets.txt\")\n",
    "# if os.path.exists(targets_outfile):\n",
    "#     print(\"Warning! File exists\")\n",
    "#     raise\n",
    "with open(targets_outfile, 'w') as outstream:\n",
    "    for tissue in tissues:\n",
    "        if tissue in res_target_dict:\n",
    "            for snp in res_target_dict[tissue]:\n",
    "                line = f\"{tissue}\\t{snp.rsid}\\t{snp.logp}\\t{snp.target}\\t{snp.maf}\\n\"\n",
    "                outstream.write(line)\n",
    "            \n",
    "# outcistypefilename = os.path.join(outdir,\"CisSQTL_target_enrichment_results_\"+title+\".txt\")\n",
    "# if os.path.exists(outcistypefilename):\n",
    "#     print(\"Warning! File exists\")\n",
    "#     raise\n",
    "# with open(outcistypefilename, 'w') as outstream:\n",
    "#     for tissue in tissues:\n",
    "#         if tissue in res_dict_cistype:\n",
    "#             for genetype in res_dict_cistype[tissue].keys():\n",
    "#                 cis_hits = res_dict_cistype[tissue][genetype].hits\n",
    "#                 cis_frac = res_dict_cistype[tissue][genetype].frac\n",
    "#                 cis_e    = res_dict_cistype[tissue][genetype].enrichment\n",
    "#                 cis_pval = res_dict_cistype[tissue][genetype].pval\n",
    "#                 random_hits = res_dict_randomtype[tissue][genetype].hits\n",
    "#                 random_frac = res_dict_randomtype[tissue][genetype].frac\n",
    "#                 random_e    = res_dict_randomtype[tissue][genetype].enrichment\n",
    "#                 random_pval = res_dict_randomtype[tissue][genetype].pval\n",
    "#                 cisline    = f\"{tissue}\\t{genetype}\\tCIS\\t{cis_hits}\\t{cis_frac}\\t{cis_e}\\t{cis_pval}\\n\"\n",
    "#                 randomline = f\"{tissue}\\t{genetype}\\tRANDOM\\t{random_hits}\\t{random_frac}\\t{random_e}\\t{random_pval}\\n\"\n",
    "#                 outstream.write(cisline)\n",
    "#                 outstream.write(randomline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_data_dict = dict()\n",
    "targets_outfile = os.path.join(outdir,\"CistransSQTL_targets.txt\")\n",
    "with open(targets_outfile) as instream:"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
