{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "from statsmodels.distributions.empirical_distribution import ECDF\n",
    "sys.path.append('../../')\n",
    "sys.path.append('/usr/users/fsimone/tejaas')\n",
    "from utils import readgtf\n",
    "from utils import utils\n",
    "import mpmath\n",
    "import collections\n",
    "from operator import attrgetter\n",
    "import gzip\n",
    "\n",
    "mpmath.mp.dps = 50\n",
    "def pvalue(x): return float(mpmath.log10(1 - 0.5 * (1 + mpmath.erf(x/mpmath.sqrt(2)))))\n",
    "\n",
    "SNPRES_FIELDS = ['rsid', 'chrom', 'pos', 'logp', 'target', 'maf']\n",
    "class SNPRes(collections.namedtuple('_SNPRes', SNPRES_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "CT_FIELDS = ['tissue', 'ncis', 'ntrans', 'ncistrans', 'randtrans', 'enrichment', 'pval', 'binom']\n",
    "class CisTrans(collections.namedtuple('_CisTrans', CT_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "def load_snp_maf(filepath, tissue):\n",
    "    snp_maf_dict = collections.defaultdict(lambda:False)\n",
    "    with open(filepath) as instream:\n",
    "        next(instream)\n",
    "        for line in instream:\n",
    "            arr = line.strip().split(\"\\t\")\n",
    "            snp_maf_dict[arr[0]] = float(arr[1])\n",
    "    return snp_maf_dict\n",
    "\n",
    "def tejaas(filepath, mafcutoff=0.01):\n",
    "    res = list()\n",
    "    with open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr   = line.strip().split(\"\\t\")\n",
    "            rsid  = arr[0]\n",
    "            chrom = int(arr[1])\n",
    "            pos   = int(arr[2])\n",
    "            maf   = float(arr[3])\n",
    "            if maf < mafcutoff or maf > (1-mafcutoff):\n",
    "                continue\n",
    "            q     = float(arr[4])\n",
    "            mu    = float(arr[5])\n",
    "            sigma = float(arr[6])\n",
    "            p     = float(arr[7])\n",
    "            if sigma == 0:\n",
    "                continue\n",
    "            logp  = np.log10(p) if p != 0 else pvalue( (q - mu) / sigma)\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf, target=None))\n",
    "    return res\n",
    "\n",
    "def read_cis(filepath):\n",
    "    res = list()\n",
    "    if not os.path.exists(filepath) or os.stat(filepath).st_size == 0:\n",
    "        print(\"File empty or does not exist\")\n",
    "        return res\n",
    "    with gzip.open(filepath, 'r') as mfile:\n",
    "        next(mfile)\n",
    "        for line in mfile:\n",
    "            arr  = line.decode().strip().split(\"\\t\")\n",
    "            rsid = arr[0]\n",
    "            if rsid.startswith(\"chrX\"):\n",
    "                continue\n",
    "            pos = int(rsid.split(\"_\")[1])\n",
    "            chrom = int(rsid.split(\"_\")[0][3:])\n",
    "            gene = arr[1].split(\":\")[-1].split(\".\")[0]\n",
    "            maf  = float(arr[5])\n",
    "            logp = np.log10(float(arr[6]))\n",
    "            res.append(SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, target=gene, maf=maf))\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mode v26\n"
     ]
    }
   ],
   "source": [
    "gene_info = readgtf.gencode_v12(\"/cbscratch/franco/datasets/GENCODE/gencode.v26.annotation.gtf.gz\", trim=True)\n",
    "gene_info_dict = collections.defaultdict(dict)\n",
    "for gene in gene_info:\n",
    "    gene_info_dict[gene.chrom][gene.ensembl_id] = gene.typ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter by allowed snps according to MAF\n",
    "tejaas_expr = \"raw\"\n",
    "K = 30 #not used at the moment\n",
    "pcutoff = 5e-8\n",
    "preproc = \"permnull_sb0.1_knn30\"\n",
    "MIN_TRANS = 1\n",
    "MIN_CIS   = 1\n",
    "use_LD    = True\n",
    "\n",
    "gtexportal_dir = \"/cbscratch/franco/datasets/gtex_v8/expression/gtex_portal/eQTLs/GTEx_Analysis_v8_eQTL/\"\n",
    "summary_dir = \"{:s}/summary_{:g}\".format(tejaas_expr,pcutoff)\n",
    "basepath = \"/cbscratch/franco/trans-eqtl/dev-pipeline/gtex_v8_lncRNA_freeze/\"\n",
    "baseoutdir = os.path.join(basepath, summary_dir, \"GTExPortal_eqtl_analysis\", preproc)\n",
    "if not os.path.exists(baseoutdir): os.makedirs(baseoutdir)\n",
    "\n",
    "# MAF = 0.01\n",
    "# title = \"maf{:g}\".format(MAF*100)\n",
    "# randompath = \"/usr/users/fsimone/vcfs_{:.2f}/\".format(MAF)\n",
    "# print(title, randompath)\n",
    "\n",
    "# outdir = os.path.join(baseoutdir, title)\n",
    "# if not os.path.exists(outdir): os.makedirs(outdir)\n",
    "\n",
    "# read all variants ids\n",
    "all_snp_ids = pd.read_csv(os.path.join(basepath, tejaas_expr, \"all_variants_pvalues_tejaas.txt\"), usecols=[0], header=0, sep=\"\\t\")\n",
    "# all_snp_ids = list(all_snp_ids.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_list = list(all_snp_ids.values.reshape(-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENEINFO_FIELDS = ['name', 'ensembl_id', 'chrom', 'start', 'end', 'typ']\n",
    "class GeneInfo(collections.namedtuple('_GeneInfo', GENEINFO_FIELDS)):\n",
    "    __slots__ = ()\n",
    "    \n",
    "def read_TFannot(infile):\n",
    "    TF_list = list()\n",
    "    with open(infile) as instream:\n",
    "        next(instream)\n",
    "        for line in instream:\n",
    "            arr = line.rstrip().split()\n",
    "            TF_list.append(GeneInfo(ensembl_id=arr[0], chrom=int(arr[1]), start=int(arr[2]), end=int(arr[3]), name=arr[4], typ=\"TF\"))\n",
    "    return TF_list\n",
    "\n",
    "base_dir = \"/cbscratch/franco/datasets\"\n",
    "\n",
    "TFs_file = \"../../external/TF_annotation.txt\"\n",
    "if not os.path.exists(TFs_file):\n",
    "    lambert_file = \"../../external/TFs_Lambert_2018.csv\"\n",
    "    TFs = pd.read_csv(lambert_file, header = 0)\n",
    "    TrueTFs = list(TFs[ TFs[\"isTF?\"] == \"Yes\" ].ID)\n",
    "    TF_annot = [x for x in gene_info if x.ensembl_id in TrueTFs]\n",
    "    ## Generate annotation file for future analysis\n",
    "    with open(TFs_file, 'w') as outstream:\n",
    "        outstream.write(\"ensembl_id\\tchrom\\tstart\\tend\\tname\\n\")\n",
    "        for e in TF_annot:\n",
    "            outstream.write(\"\\t\".join([e.ensembl_id, str(e.chrom), str(e.start), str(e.end), e.name])+\"\\n\")\n",
    "\n",
    "TF_annot = read_TFannot(TFs_file)\n",
    "TF_dict = collections.defaultdict(dict)\n",
    "for g in TF_annot:\n",
    "    TF_dict[g.chrom][g.ensembl_id] = \"TF\"\n",
    "\n",
    "# Reformat genetype dict, we can add as many gene annotations as we want here\n",
    "alltypes_dict = collections.defaultdict(dict)\n",
    "genetypes = []\n",
    "for chrm in range(1,23):\n",
    "    gene_info_dict[chrm]\n",
    "    for k in gene_info_dict[chrm].keys():\n",
    "        genetype = gene_info_dict[chrm][k]\n",
    "        if genetype not in alltypes_dict:\n",
    "            alltypes_dict[genetype] = collections.defaultdict(lambda:False)\n",
    "            genetypes.append(genetype)\n",
    "        alltypes_dict[genetype][k] = True\n",
    "    # Add TF dictionary\n",
    "    for k in TF_dict[chrm].keys():\n",
    "        genetype = \"TF\"\n",
    "        if genetype not in alltypes_dict:\n",
    "            alltypes_dict[genetype] = collections.defaultdict(lambda:False)\n",
    "            genetypes.append(genetype)\n",
    "        alltypes_dict[genetype][k] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "json_file = \"../../gtex_v8_metadata.json\"\n",
    "with open(json_file) as instream:\n",
    "    gtex_meta = json.load(instream)\n",
    "    \n",
    "tissue_file = \"/usr/users/fsimone/trans-eqtl-pipeline/main/tissues.txt\"\n",
    "tissues, descriptions = utils.read_tissues(tissue_file)\n",
    "tissue_names   = dict()\n",
    "tissue_colors  = dict()\n",
    "tissue_samples = dict()\n",
    "for tshort, tfull in zip(tissues, descriptions):\n",
    "    tissue_names[tshort] = tfull\n",
    "    tissue_colors[tshort] = \"#\" + gtex_meta[tfull.replace(\" \", \"_\")][\"colorHex\"]\n",
    "    tissue_samples[tshort] = gtex_meta[tfull.replace(\" \", \"_\")][\"rnaSeqAndGenotypeSampleCount\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as ss\n",
    "\n",
    "# for a set of cis and trans-eQTLs, return the cis-trans ids and the cis-eQTLs with its targets\n",
    "def cross_ref_cis_trans(trans_ids, cis_eqtls):\n",
    "    cis_ids = list(set([x.rsid for x in cis_eqtls]))\n",
    "    \n",
    "    #Intersection between cis-eqtls (MatrixEQTL) and trans-eqtls (TEJAAS)\n",
    "    cis_trans_eqtls_ids = list(set.intersection(set(trans_ids), set(cis_ids)))\n",
    "    \n",
    "    #set up a dict for fast look up later\n",
    "    cis_trans_dict = dict()\n",
    "    for x in cis_trans_eqtls_ids:\n",
    "        cis_trans_dict[x] = True\n",
    "    \n",
    "    # List of cis-trans-eqtls with its target gene\n",
    "    cis_target_eqtls = [x for x in cis_eqtls if cis_trans_dict.get(x.rsid, False)]\n",
    "\n",
    "    return cis_trans_eqtls_ids, cis_target_eqtls\n",
    "\n",
    "def crossref_trans_tejaas(transeqtls, cis_eqtls):\n",
    "    trans_ids = [x.rsid for x in transeqtls]\n",
    "    a, b = cross_ref_cis_trans(trans_ids, cis_eqtls)\n",
    "    return a, b\n",
    "\n",
    "def sample_background_50000_simple(ciseqtls, randompath):\n",
    "    randtrans = list()\n",
    "    chroms    = [str(x) for x in np.arange(1,23)]\n",
    "    for nid in [\"{:03d}\".format(x) for x in np.arange(1, 11)]:\n",
    "        Nrand=\"50000\"\n",
    "        randomfile = randompath+\"random_\"+Nrand+\"_\"+nid\n",
    "\n",
    "        rand_ids = list()\n",
    "        for chrm in chroms:\n",
    "            with open(os.path.join(randomfile, \"chr{:s}.txt\".format(chrm))) as ins:\n",
    "                rand_ids += [line.rstrip() for line in ins]\n",
    "\n",
    "        a, b = cross_ref_cis_trans(rand_ids, ciseqtls)\n",
    "        randtrans.append( len(a) )\n",
    "    return np.mean(randtrans)\n",
    "\n",
    "def sample_rand_bg(ciseqtls, snps_list, nchoose = 20000, niter = 20):\n",
    "    nchoose = 20000 # len(trans_eqtls_signif) \n",
    "    niter = 20\n",
    "    randtrans = list()\n",
    "    print(f'Iteration', end=\"\")\n",
    "    for k in range(niter):\n",
    "        chooseidx = np.sort(np.random.choice(len(snps_list), nchoose, replace = False))\n",
    "        rand_ids = [snps_list[i] for i in chooseidx]\n",
    "        # nannot_k = annotated_random(snps_list, nchoose, dhs_file)\n",
    "        print(f' {k}', end=\"\")\n",
    "        a, b = cross_ref_cis_trans(rand_ids, ciseqtls)\n",
    "        randtrans.append( len(a) )\n",
    "    print(\"\")\n",
    "    return np.mean(randtrans)\n",
    "\n",
    "def sample_binomial(n, p, NTIMES):\n",
    "    array_n = list()\n",
    "    for i in range(NTIMES):\n",
    "        n_success = np.random.binomial(n, p)\n",
    "        array_n.append(n_success)\n",
    "    return array_n\n",
    "\n",
    "def smart_LD_filter(trans_eqtls, trans_eqtls_ld_regions_file):\n",
    "    pass_snps = list()\n",
    "    with open(trans_eqtls_ld_regions_file) as instream:\n",
    "        for line in instream:\n",
    "            arr = line.strip(\"\\n\").split(\"\\t\")\n",
    "            if len(arr[4]) > 0:\n",
    "                pass_snps.append(arr[0])\n",
    "    return [x for x in trans_eqtls if x.rsid in pass_snps]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "av Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ag Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "aa Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ac Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "at Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bam Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ban Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bca Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bceh Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bce Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bco Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bfr Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bhi Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bhy Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bnu Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bpu Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bsp Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bsu Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "br Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ebv Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "fib Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "cols Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "colt Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "esog Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "esom Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "esomu Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "haa Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "hlv Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "kc Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "liv Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "lu Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "msg Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ms Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "nt Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "pan Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "pit Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "snse Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "sse Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "si Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "spl Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "sto Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "thy Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "wb Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ov Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "pro Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "tes Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "ut Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "va Iteration 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\n",
      "bl Bladder has no cis-file in GTEx!\n"
     ]
    }
   ],
   "source": [
    "cis_bg = dict()\n",
    "for tissue in tissues:\n",
    "    print(tissue, end=\" \")\n",
    "    signif_cisfile = os.path.join(gtexportal_dir, \"{:s}.v8.signif_variant_gene_pairs.txt.gz\".format(tissue_names[tissue].replace(\" \", \"_\")))\n",
    "    if not os.path.exists(signif_cisfile) or os.stat(signif_cisfile).st_size == 0:\n",
    "        print(\"{:s} has no cis-file in GTEx!\".format(tissue_names[tissue]))\n",
    "        continue\n",
    "    ciseqtls = read_cis(signif_cisfile)\n",
    "    cis_bg[tissue] = sample_rand_bg(ciseqtls, snps_list, nchoose = 20000, niter = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_bg_file = os.path.join(gtexportal_dir, \"gtex_background_freqs_ciseqtls.txt\")\n",
    "if not os.path.exists(cis_bg_file):\n",
    "    with open(cis_bg_file, 'w') as outstream:\n",
    "        outstream.write(\"tissue\\trand_count\\tfreq\\n\")\n",
    "        for t in cis_bg:\n",
    "            outstream.write(f\"{t}\\t{cis_bg[t]}\\t{cis_bg[t]/20000} \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########## Tissue: as - 199 trans-eqtls - 17 cis-trans-eqtls #########\n",
      "                  as        Enrichment: 0.729601 - binom 0.93844\n",
      "########## Tissue: av - 113 trans-eqtls - 17 cis-trans-eqtls #########\n",
      "                  av        Enrichment: 1.79258 - binom 0.0133264\n",
      "########## Tissue: ag - 51 trans-eqtls - 9 cis-trans-eqtls #########\n",
      "                  ag        Enrichment: 4.13038 - binom 0.000282267\n",
      "########## Tissue: aa - 420 trans-eqtls - 61 cis-trans-eqtls #########\n",
      "                  aa        Enrichment: 1.71155 - binom 2.96248e-05\n",
      "########## Tissue: ac - 529 trans-eqtls - 19 cis-trans-eqtls #########\n",
      "                  ac        Enrichment: 1.11103 - binom 0.352905\n",
      "########## Tissue: at - 139 trans-eqtls - 11 cis-trans-eqtls #########\n",
      "                  at        Enrichment: 0.661899 - binom 0.952041\n",
      "########## Tissue: bam - 411 trans-eqtls - 3 cis-trans-eqtls #########\n",
      "                 bam        Enrichment: 0.468277 - binom 0.955057\n",
      "########## Tissue: ban - 313 trans-eqtls - 8 cis-trans-eqtls #########\n",
      "                 ban        Enrichment: 1.024 - binom 0.522038\n",
      "########## Tissue: bca - 3 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bca        Enrichment: 0 - binom 1\n",
      "########## Tissue: bceh - 24 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                bceh        Enrichment: 0 - binom 1\n",
      "########## Tissue: bce - 7 trans-eqtls - 1 cis-trans-eqtls #########\n",
      "                 bce        Enrichment: 2.2859 - binom 0.363475\n",
      "########## Tissue: bco - 18 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bco        Enrichment: 0 - binom 1\n",
      "########## Tissue: bfr - 8 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bfr        Enrichment: 0 - binom 1\n",
      "########## Tissue: bhi - 16 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bhi        Enrichment: 0 - binom 1\n",
      "########## Tissue: bhy - 24 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bhy        Enrichment: 0 - binom 1\n",
      "########## Tissue: bnu - 99 trans-eqtls - 2 cis-trans-eqtls #########\n",
      "                 bnu        Enrichment: 0.492312 - binom 0.917302\n",
      "bpu has less than 1 trans-eqtls\n",
      "########## Tissue: bsp - 18 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bsp        Enrichment: 0 - binom 1\n",
      "########## Tissue: bsu - 5 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 bsu        Enrichment: 0 - binom 1\n",
      "########## Tissue: br - 143 trans-eqtls - 13 cis-trans-eqtls #########\n",
      "                  br        Enrichment: 1.35837 - binom 0.162223\n",
      "########## Tissue: ebv - 75 trans-eqtls - 3 cis-trans-eqtls #########\n",
      "                 ebv        Enrichment: 1.87948 - binom 0.214517\n",
      "########## Tissue: fib - 64 trans-eqtls - 11 cis-trans-eqtls #########\n",
      "                 fib        Enrichment: 1.40009 - binom 0.156338\n",
      "########## Tissue: cols - 27 trans-eqtls - 7 cis-trans-eqtls #########\n",
      "                cols        Enrichment: 4.11719 - binom 0.00113454\n",
      "########## Tissue: colt - 363 trans-eqtls - 50 cis-trans-eqtls #########\n",
      "                colt        Enrichment: 1.95184 - binom 5.29276e-06\n",
      "########## Tissue: esog - 6 trans-eqtls - 1 cis-trans-eqtls #########\n",
      "                esog        Enrichment: 2.55683 - binom 0.332649\n",
      "########## Tissue: esom - 4 trans-eqtls - 3 cis-trans-eqtls #########\n",
      "                esom        Enrichment: 7.05384 - binom 0.00442461\n",
      "########## Tissue: esomu - 4 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "               esomu        Enrichment: 0 - binom 1\n",
      "haa has less than 1 trans-eqtls\n",
      "########## Tissue: hlv - 72 trans-eqtls - 6 cis-trans-eqtls #########\n",
      "                 hlv        Enrichment: 1.32459 - binom 0.299424\n",
      "########## Tissue: kc - 4 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                  kc        Enrichment: 0 - binom 1\n",
      "########## Tissue: liv - 13 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 liv        Enrichment: 0 - binom 1\n",
      "########## Tissue: lu - 154 trans-eqtls - 29 cis-trans-eqtls #########\n",
      "                  lu        Enrichment: 1.97097 - binom 0.000305814\n",
      "########## Tissue: msg - 1 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 msg        Enrichment: 0 - binom 1\n",
      "########## Tissue: ms - 741 trans-eqtls - 117 cis-trans-eqtls #########\n",
      "                  ms        Enrichment: 1.42785 - binom 5.89906e-05\n",
      "########## Tissue: nt - 247 trans-eqtls - 24 cis-trans-eqtls #########\n",
      "                  nt        Enrichment: 0.720936 - binom 0.970411\n",
      "pan has less than 1 trans-eqtls\n",
      "########## Tissue: pit - 2 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 pit        Enrichment: 0 - binom 1\n",
      "########## Tissue: snse - 235 trans-eqtls - 43 cis-trans-eqtls #########\n",
      "                snse        Enrichment: 1.66378 - binom 0.000601113\n",
      "########## Tissue: sse - 105 trans-eqtls - 27 cis-trans-eqtls #########\n",
      "                 sse        Enrichment: 1.99948 - binom 0.000282543\n",
      "########## Tissue: si - 19 trans-eqtls - 2 cis-trans-eqtls #########\n",
      "                  si        Enrichment: 3.70156 - binom 0.100525\n",
      "########## Tissue: spl - 1 trans-eqtls - 0 cis-trans-eqtls #########\n",
      "                 spl        Enrichment: 0 - binom 1\n",
      "########## Tissue: sto - 77 trans-eqtls - 5 cis-trans-eqtls #########\n",
      "                 sto        Enrichment: 1.27043 - binom 0.358216\n",
      "########## Tissue: thy - 393 trans-eqtls - 34 cis-trans-eqtls #########\n",
      "                 thy        Enrichment: 0.622179 - binom 0.999457\n",
      "wb has less than 1 trans-eqtls\n",
      "########## Tissue: ov - 25 trans-eqtls - 2 cis-trans-eqtls #########\n",
      "                  ov        Enrichment: 3.14589 - binom 0.132189\n",
      "########## Tissue: pro - 309 trans-eqtls - 17 cis-trans-eqtls #########\n",
      "                 pro        Enrichment: 1.52982 - binom 0.0567278\n",
      "########## Tissue: tes - 412 trans-eqtls - 63 cis-trans-eqtls #########\n",
      "                 tes        Enrichment: 1.33312 - binom 0.011358\n",
      "########## Tissue: ut - 550 trans-eqtls - 11 cis-trans-eqtls #########\n",
      "                  ut        Enrichment: 1.40056 - binom 0.168412\n",
      "va has less than 1 trans-eqtls\n",
      "Bladder has no cis-file in GTEx!\n"
     ]
    }
   ],
   "source": [
    "res_dict = dict()\n",
    "res_target_dict = dict()\n",
    "   \n",
    "cis_bg_freq = dict()\n",
    "with open(cis_bg_file) as instream:\n",
    "    next(instream)\n",
    "    for line in instream:\n",
    "        arr = line.strip().split(\"\\t\")\n",
    "        cis_bg_freq[arr[0]] = float(arr[2])\n",
    "\n",
    "for tissue in tissues:   \n",
    "    tejaas_file = os.path.join(basepath, summary_dir, tissue, \"tejaas\", preproc, \"trans_eqtls.txt\")\n",
    "    if use_LD:\n",
    "        tejaas_regions_file = tejaas_file+\".ld_regions\"\n",
    "        tejaas_file = tejaas_file+\".ld_prune\"\n",
    "    if not os.path.exists(tejaas_file):\n",
    "        print(\"{:s} has no trans-eqtl results\".format(tissue))\n",
    "        continue\n",
    "    transeqtls = tejaas(tejaas_file)\n",
    "    if use_LD:\n",
    "        transeqtls = smart_LD_filter(transeqtls, tejaas_regions_file)\n",
    "    \n",
    "    if len(transeqtls) < MIN_TRANS:\n",
    "        print(\"{:s} has less than {:d} trans-eqtls\".format(tissue, MIN_TRANS))\n",
    "        continue\n",
    "    \n",
    "    signif_cisfile = os.path.join(gtexportal_dir, \"{:s}.v8.signif_variant_gene_pairs.txt.gz\".format(tissue_names[tissue].replace(\" \", \"_\")))\n",
    "    if not os.path.exists(signif_cisfile) or os.stat(signif_cisfile).st_size == 0:\n",
    "        print(\"{:s} has no cis-file in GTEx!\".format(tissue_names[tissue]))\n",
    "        continue\n",
    "    ciseqtls = read_cis(signif_cisfile)\n",
    "    cis_ids = list(set([x.rsid for x in ciseqtls]))\n",
    "    \n",
    "    if len(ciseqtls) < MIN_CIS:\n",
    "        print(\"{:s} has less than {:d} cis-eqtls\".format(tissue, MIN_CIS))\n",
    "        continue\n",
    "    \n",
    "    cis_trans_eqtls_ids, cistrans_target_eqtls = crossref_trans_tejaas(transeqtls, ciseqtls)\n",
    "    \n",
    "    FRAC_CISTRANS = len(cis_trans_eqtls_ids) / len(transeqtls)\n",
    "    FRAC_RANDOM_GWCISTRANS = cis_bg_freq[tissue] # randtrans / 50000 \n",
    "    \n",
    "    enrichment = FRAC_CISTRANS / FRAC_RANDOM_GWCISTRANS\n",
    "   \n",
    "    ncis = len(cis_ids)\n",
    "    ntrans = len(transeqtls)\n",
    "    ncistrans = len(cis_trans_eqtls_ids)\n",
    "    pval_binom = ss.binom_test(ncistrans, ntrans, FRAC_RANDOM_GWCISTRANS, alternative='greater')\n",
    "    \n",
    "    res_dict[tissue] = CisTrans(tissue=tissue, ncis=ncis, ntrans=ntrans, \n",
    "                                ncistrans=ncistrans, randtrans=FRAC_RANDOM_GWCISTRANS,\n",
    "                                enrichment=enrichment, pval=1, binom=pval_binom)\n",
    "    \n",
    "    res_target_dict[tissue] = cistrans_target_eqtls\n",
    "    \n",
    "    print(f\"########## Tissue: {tissue} - {ntrans} trans-eqtls - {ncistrans} cis-trans-eqtls #########\")\n",
    "    print(f\"{tissue:>20}        Enrichment: {enrichment:>g} - binom {pval_binom:>g}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "outcisfilename = os.path.join(baseoutdir,\"CisEQTL_enrichment.txt\")\n",
    "if use_LD:\n",
    "    outcisfilename = outcisfilename + \".ld_prune\"\n",
    "with open(outcisfilename, 'w') as outstream:\n",
    "    for tissue in tissues:\n",
    "        if tissue in res_dict:\n",
    "            line = f\"{tissue}\\t{res_dict[tissue].ncis}\\t{res_dict[tissue].ntrans}\\t{res_dict[tissue].ncistrans}\\t{res_dict[tissue].randtrans}\\t{res_dict[tissue].enrichment}\\t{res_dict[tissue].binom}\\n\"\n",
    "            outstream.write(line)\n",
    "\n",
    "targets_outfile = os.path.join(baseoutdir,\"CistransEQTL_targets.txt\")\n",
    "if use_LD:\n",
    "    targets_outfile = targets_outfile + \".ld_prune\"\n",
    "with open(targets_outfile, 'w') as outstream:\n",
    "    for tissue in tissues:\n",
    "        if tissue in res_target_dict:\n",
    "            for snp in res_target_dict[tissue]:\n",
    "                line = f\"{tissue}\\t{snp.rsid}\\t{snp.logp}\\t{snp.target}\\t{snp.maf}\\n\"\n",
    "                outstream.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "targets_outfile = os.path.join(baseoutdir,\"CistransEQTL_targets.txt\")\n",
    "if use_LD:\n",
    "    targets_outfile = targets_outfile + \".ld_prune\"\n",
    "res_target_dict = collections.defaultdict(list)\n",
    "with open(targets_outfile) as instream:\n",
    "    for line in instream:\n",
    "        arr = line.strip().split(\"\\t\")\n",
    "        t    = arr[0]\n",
    "        rsid = arr[1]\n",
    "        logp = -float(arr[2])\n",
    "        targ = arr[3]\n",
    "        maf  = float(arr[4])\n",
    "        chrom = int(rsid.split(\"_\")[0][3:])\n",
    "        pos   = int(rsid.split(\"_\")[1])\n",
    "        res_target_dict[t].append( SNPRes(rsid=rsid, chrom=chrom, pos=pos, logp=-logp, maf=maf, target=targ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as - lincRNA (1) - enrichment: 0.4586206742143724 - pval_binom 0.8974431321708211\n",
      "as - protein_coding (15) - enrichment: 0.8968100738425279 - pval_binom 0.8197933631376447\n",
      "as - antisense (1) - enrichment: 0.5088753438026795 - pval_binom 0.8704060000766479\n",
      "as - processed_transcript (1) - enrichment: 3.1129693714504056 - pval_binom 0.2761981973397053\n",
      "as - sense_intronic (1) - enrichment: 3.124875506961803 - pval_binom 0.2753006839607016\n",
      "as - TEC (3) - enrichment: 7.671176360442618 - pval_binom 0.006830898904564\n",
      "av - lincRNA (3) - enrichment: 1.1118352631214032 - pval_binom 0.5137429682087641\n",
      "av - protein_coding (17) - enrichment: 0.8804707048865426 - pval_binom 0.8507884487748867\n",
      "av - antisense (2) - enrichment: 0.9208947081287507 - pval_binom 0.6490216417203969\n",
      "av - processed_transcript (2) - enrichment: 5.22260218205504 - pval_binom 0.056028046646193624\n",
      "av - TEC (1) - enrichment: 2.0989528795811516 - pval_binom 0.38129519592470207\n",
      "ag - protein_coding (11) - enrichment: 1.5333721856553308 - pval_binom 0.018902902615159225\n",
      "aa - lincRNA (10) - enrichment: 1.3389882462157423 - pval_binom 0.2124941531822667\n",
      "aa - protein_coding (61) - enrichment: 0.9914707747682431 - pval_binom 0.5931360105852528\n",
      "aa - antisense (6) - enrichment: 0.9787319746221589 - pval_binom 0.5811054104862237\n",
      "aa - processed_transcript (5) - enrichment: 4.224970622696794 - pval_binom 0.006877720328817586\n",
      "aa - TEC (2) - enrichment: 1.4351055195977347 - pval_binom 0.40699081634889883\n",
      "ac - lincRNA (2) - enrichment: 0.7606331211142656 - pval_binom 0.753912039227438\n",
      "ac - protein_coding (18) - enrichment: 1.1632317783262203 - pval_binom 0.2166786863728597\n",
      "ac - antisense (1) - enrichment: 0.49485598650172813 - pval_binom 0.8775938406438143\n",
      "ac - processed_transcript (1) - enrichment: 2.6276728859185 - pval_binom 0.3183727844635278\n",
      "at - lincRNA (4) - enrichment: 2.621174219310177 - pval_binom 0.0608851488616298\n",
      "at - protein_coding (9) - enrichment: 0.7019970297351107 - pval_binom 0.9800642289194078\n",
      "at - antisense (2) - enrichment: 1.5675102944810055 - pval_binom 0.3679746977222736\n",
      "at - 3prime_overlapping_ncRNA (1) - enrichment: 187.5848676078329 - pval_binom 0.005317480296906362\n",
      "bam - protein_coding (2) - enrichment: 1.2798479787140156 - pval_binom 0.5313245513743654\n",
      "ban - lincRNA (1) - enrichment: 1.1358405153651954 - pval_binom 0.6065203495485801\n",
      "ban - protein_coding (3) - enrichment: 0.6957351733552315 - pval_binom 0.9007977319960807\n",
      "ban - antisense (1) - enrichment: 1.404184334574571 - pval_binom 0.5256793104315164\n",
      "bce - protein_coding (2) - enrichment: 1.6131192589294603 - pval_binom 0.3842970472005106\n",
      "bnu - protein_coding (1) - enrichment: 0.8587899230403874 - pval_binom 0.8254553288722446\n",
      "br - lincRNA (4) - enrichment: 1.9235455564669168 - pval_binom 0.1485567345270723\n",
      "br - protein_coding (16) - enrichment: 1.2268954452972092 - pval_binom 0.14242878261794972\n",
      "ebv - lincRNA (1) - enrichment: 2.6834635767537955 - pval_binom 0.32109274458131853\n",
      "ebv - protein_coding (3) - enrichment: 1.0170597469407827 - pval_binom 0.6650277998379903\n",
      "ebv - TF (2) - enrichment: 11.76335477096732 - pval_binom 0.010796177173331978\n",
      "fib - lincRNA (1) - enrichment: 0.5443124098998963 - pval_binom 0.8508306304240103\n",
      "fib - protein_coding (25) - enrichment: 1.3049573590618273 - pval_binom 0.006713043963160117\n",
      "fib - TF (4) - enrichment: 2.4296459530743917 - pval_binom 0.07912431898769678\n",
      "cols - protein_coding (9) - enrichment: 1.3378871628994167 - pval_binom 0.1350288950517107\n",
      "cols - antisense (2) - enrichment: 2.40145031619686 - pval_binom 0.20039599235975708\n",
      "colt - lincRNA (7) - enrichment: 0.8379921733326456 - pval_binom 0.7404535206046003\n",
      "colt - protein_coding (59) - enrichment: 1.074938452000999 - pval_binom 0.22038037623224346\n",
      "colt - antisense (5) - enrichment: 0.7523667450484711 - pval_binom 0.8026916834335163\n",
      "colt - processed_transcript (3) - enrichment: 2.81197755358271 - pval_binom 0.09191916855642829\n",
      "colt - TEC (3) - enrichment: 1.9017374736632278 - pval_binom 0.20988333274360624\n",
      "colt - TF (5) - enrichment: 1.0268247457619482 - pval_binom 0.5406052512096876\n",
      "esom - protein_coding (15) - enrichment: 1.1072768792978327 - pval_binom 0.3332864038547369\n",
      "hlv - lincRNA (1) - enrichment: 1.351543940669802 - pval_binom 0.5379513152775977\n",
      "hlv - protein_coding (6) - enrichment: 0.9976947862316722 - pval_binom 0.654089397890306\n",
      "hlv - antisense (1) - enrichment: 1.862491879813148 - pval_binom 0.42512198885250885\n",
      "lu - lincRNA (3) - enrichment: 0.5250024775721795 - pval_binom 0.9332733912246978\n",
      "lu - protein_coding (40) - enrichment: 1.0587172581687367 - pval_binom 0.33028420395340685\n",
      "lu - antisense (10) - enrichment: 2.2307385248556675 - pval_binom 0.013055064599805885\n",
      "lu - processed_transcript (2) - enrichment: 2.6136823575768333 - pval_binom 0.17823152343791152\n",
      "lu - TEC (2) - enrichment: 2.038342143016324 - pval_binom 0.25725819291984797\n",
      "lu - TF (2) - enrichment: 0.6528388875583445 - pval_binom 0.8175583563934481\n",
      "ms - lincRNA (15) - enrichment: 1.0927931988204234 - pval_binom 0.39941681205769813\n",
      "ms - protein_coding (142) - enrichment: 1.0593376316164573 - pval_binom 0.11519612759769843\n",
      "ms - antisense (15) - enrichment: 1.2404898080092837 - pval_binom 0.23045911467740185\n",
      "ms - processed_transcript (1) - enrichment: 0.5085156219537329 - pval_binom 0.8614896201780161\n",
      "ms - TEC (3) - enrichment: 1.150432245787809 - pval_binom 0.48456334473362817\n",
      "ms - sense_overlapping (1) - enrichment: 2.6245840251273975 - pval_binom 0.3170935002631834\n",
      "ms - TF (6) - enrichment: 0.5652787065933521 - pval_binom 0.9571993155642521\n",
      "nt - lincRNA (2) - enrichment: 0.6188502731880569 - pval_binom 0.8456558211365899\n",
      "nt - protein_coding (29) - enrichment: 1.2074824321333462 - pval_binom 0.05748266633160516\n",
      "nt - antisense (1) - enrichment: 0.3784045529423622 - pval_binom 0.9355456526840131\n",
      "snse - lincRNA (5) - enrichment: 0.6573699115178065 - pval_binom 0.8860466055565458\n",
      "snse - protein_coding (59) - enrichment: 0.9848144861572701 - pval_binom 0.6248118448382235\n",
      "snse - antisense (3) - enrichment: 0.45143816983307095 - pval_binom 0.966122744942302\n",
      "snse - processed_transcript (1) - enrichment: 0.8977790096233406 - pval_binom 0.67393470557391\n",
      "snse - sense_intronic (2) - enrichment: 1.5629325474353455 - pval_binom 0.3666435079588738\n",
      "snse - TEC (1) - enrichment: 0.7430832084151908 - pval_binom 0.7422290554861626\n",
      "snse - TF (4) - enrichment: 0.814949610306963 - pval_binom 0.7294662058444331\n",
      "sse - lincRNA (6) - enrichment: 0.7300807320419523 - pval_binom 0.8402004448712626\n",
      "sse - protein_coding (59) - enrichment: 0.9593232894258212 - pval_binom 0.7444444718692889\n",
      "sse - antisense (8) - enrichment: 1.1773377973630923 - pval_binom 0.3695341081622529\n",
      "sse - sense_intronic (2) - enrichment: 1.5969797882386032 - pval_binom 0.35681396739226307\n",
      "sse - TEC (4) - enrichment: 3.0093893780767487 - pval_binom 0.04482882638316217\n",
      "sse - 3prime_overlapping_ncRNA (1) - enrichment: 20.712818074031606 - pval_binom 0.04714418752006393\n",
      "sse - TF (2) - enrichment: 0.3954708888252877 - pval_binom 0.9649419924237297\n",
      "si - protein_coding (1) - enrichment: 0.9185357762239388 - pval_binom 0.7923781666754683\n",
      "sto - lincRNA (1) - enrichment: 1.0466226943580343 - pval_binom 0.6319440673466075\n",
      "sto - protein_coding (4) - enrichment: 0.5983666008134315 - pval_binom 0.974152438660908\n",
      "sto - antisense (2) - enrichment: 2.4964401443954403 - pval_binom 0.18864538328879354\n",
      "sto - processed_transcript (1) - enrichment: 6.6425779421402495 - pval_binom 0.14065397646892885\n",
      "thy - lincRNA (7) - enrichment: 1.3917740347013048 - pval_binom 0.23502635623484547\n",
      "thy - protein_coding (33) - enrichment: 0.9526070481193158 - pval_binom 0.7275333226387402\n",
      "thy - antisense (4) - enrichment: 0.9651666921001073 - pval_binom 0.6026726289421104\n",
      "thy - processed_transcript (4) - enrichment: 6.139726543085854 - pval_binom 0.004152495939513479\n",
      "thy - sense_intronic (1) - enrichment: 1.3442887545819433 - pval_binom 0.527143940177722\n",
      "thy - TEC (1) - enrichment: 1.1594390110644988 - pval_binom 0.5807635420989223\n",
      "thy - TF (2) - enrichment: 0.6487390931839746 - pval_binom 0.8212929421628758\n",
      "ov - lincRNA (2) - enrichment: 2.56642060109327 - pval_binom 0.17844471297037906\n",
      "ov - protein_coding (2) - enrichment: 0.5374618828726644 - pval_binom 0.9557535911943199\n",
      "ov - antisense (1) - enrichment: 1.9667770022055455 - pval_binom 0.4101325727578238\n",
      "ov - processed_transcript (1) - enrichment: 8.900912729170951 - pval_binom 0.10708096659028843\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pro - lincRNA (2) - enrichment: 0.5922808872830353 - pval_binom 0.8663768955589742\n",
      "pro - protein_coding (17) - enrichment: 1.0667217513827587 - pval_binom 0.4200299657532772\n",
      "pro - antisense (4) - enrichment: 1.554668825661192 - pval_binom 0.25370256093007365\n",
      "pro - processed_transcript (1) - enrichment: 2.0730516142298994 - pval_binom 0.38510325678939716\n",
      "pro - TEC (2) - enrichment: 4.108716694686433 - pval_binom 0.08494486154805331\n",
      "tes - lincRNA (10) - enrichment: 0.5998103865828157 - pval_binom 0.9772876592407453\n",
      "tes - protein_coding (73) - enrichment: 1.177484457077194 - pval_binom 0.02651596113976638\n",
      "tes - antisense (12) - enrichment: 1.0034237359194478 - pval_binom 0.5395819708669676\n",
      "tes - processed_transcript (5) - enrichment: 3.766131915363014 - pval_binom 0.011074660541587312\n",
      "tes - TEC (1) - enrichment: 0.6902591988417153 - pval_binom 0.7672108004918619\n",
      "tes - TF (3) - enrichment: 0.7294907261492378 - pval_binom 0.783038410169045\n",
      "ut - lincRNA (2) - enrichment: 1.4867926426604958 - pval_binom 0.3951164390060632\n",
      "ut - protein_coding (4) - enrichment: 0.6092590753293091 - pval_binom 0.9524787292298699\n",
      "ut - antisense (4) - enrichment: 3.3536304838925965 - pval_binom 0.026306517242633443\n",
      "Bladder has no cis-file in GTEx!\n"
     ]
    }
   ],
   "source": [
    "CisTrans_Type_FIELDS = ['tissue', 'genetype', 'ncistrans', 'ntype', 'frac_cis', 'frac_cistrans', 'enrichment', 'binom']\n",
    "class CisTrans_Type(collections.namedtuple('_CisTrans_Type', CisTrans_Type_FIELDS)):\n",
    "    __slots__ = ()\n",
    "\n",
    "gtexportal_dir = \"/cbscratch/franco/datasets/gtex_v8/expression/gtex_portal/eQTLs/GTEx_Analysis_v8_eQTL/\"\n",
    "\n",
    "enrichment_genetypes = list()\n",
    "\n",
    "for tissue in tissues:\n",
    "    signif_cisfile = os.path.join(gtexportal_dir, \"{:s}.v8.signif_variant_gene_pairs.txt.gz\".format(tissue_names[tissue].replace(\" \", \"_\")))\n",
    "    if not os.path.exists(signif_cisfile) or os.stat(signif_cisfile).st_size == 0:\n",
    "        print(\"{:s} has no cis-file in GTEx!\".format(tissue_names[tissue]))\n",
    "        continue\n",
    "    ciseqtls = read_cis(signif_cisfile)\n",
    "    \n",
    "    TOTAL_CIS = len(ciseqtls)\n",
    "    cis_counts_dict = collections.defaultdict(int)\n",
    "    for eqtl in ciseqtls:\n",
    "        for genetype in genetypes:\n",
    "            if alltypes_dict[genetype][eqtl.target]:\n",
    "                cis_counts_dict[genetype] += 1\n",
    "    \n",
    "    if tissue in res_target_dict:\n",
    "        TOTAL_CIS_TRANS = len(res_target_dict[tissue])\n",
    "        cistrans_counts_dict = collections.defaultdict(int)\n",
    "        for eqtl in res_target_dict[tissue]:\n",
    "            for genetype in genetypes:\n",
    "                if alltypes_dict[genetype][eqtl.target]:\n",
    "                    cistrans_counts_dict[genetype] += 1\n",
    "                \n",
    "        for genetype in genetypes:\n",
    "            if TOTAL_CIS > 0 and TOTAL_CIS_TRANS > 0:\n",
    "                frac_cis      = cis_counts_dict[genetype] / TOTAL_CIS\n",
    "                frac_cistrans = cistrans_counts_dict[genetype] / TOTAL_CIS_TRANS\n",
    "                if frac_cis > 0: \n",
    "                    if frac_cistrans > 0:\n",
    "                        enrichment = frac_cistrans / frac_cis\n",
    "                        pval_binom = ss.binom_test(cistrans_counts_dict[genetype], TOTAL_CIS_TRANS, frac_cis, alternative='greater')\n",
    "                        print(f\"{tissue} - {genetype} ({cistrans_counts_dict[genetype]}) - enrichment: {enrichment} - pval_binom {pval_binom}\")\n",
    "                        enrichment_genetypes.append(CisTrans_Type(tissue=tissue, genetype=genetype, \\\n",
    "                                                                  ncistrans=TOTAL_CIS_TRANS, \\\n",
    "                                                                  ntype=cistrans_counts_dict[genetype], \\\n",
    "                                                                  frac_cis=frac_cis, \\\n",
    "                                                                  frac_cistrans=frac_cistrans, \\\n",
    "                                                                  enrichment=enrichment, \\\n",
    "                                                                  binom=pval_binom))\n",
    "        #         else:\n",
    "        #             print(f\"{tissue} - Error on genetype {genetype}: {cis_counts_dict[genetype]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "targets_enrichment_outfile = os.path.join(baseoutdir, \"CisTrans_EQTL_target_enrichments.txt\")\n",
    "if use_LD:\n",
    "    targets_enrichment_outfile = targets_enrichment_outfile + \".ld_prune\"\n",
    "with open(targets_enrichment_outfile, 'w') as outstream:\n",
    "    outstream.write(\"tissue\\tgenetype\\tncistrans\\tntype\\tfrac_cis\\tfrac_cistrans\\tenrichment\\tpval_binom\\n\")\n",
    "    for e in enrichment_genetypes:\n",
    "        outstream.write(f\"{e.tissue}\\t{e.genetype}\\t{e.ncistrans}\\t{e.ntype}\\t{e.frac_cis}\\t{e.frac_cistrans}\\t{e.enrichment}\\t{e.binom}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List of Unique cis-eQTLs across all tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cis_list = list()\n",
    "for tissue in tissues:\n",
    "    signif_cisfile = os.path.join(gtexportal_dir, \"{:s}.v8.signif_variant_gene_pairs.txt.gz\".format(tissue_names[tissue].replace(\" \", \"_\")))\n",
    "    if not os.path.exists(signif_cisfile) or os.stat(signif_cisfile).st_size == 0:\n",
    "        print(\"{:s} has no cis-file in GTEx!\".format(tissue_names[tissue]))\n",
    "        continue\n",
    "    ciseqtls = read_cis(signif_cisfile)\n",
    "    print(tissue, len(ciseqtls))\n",
    "    all_cis_list += [(e.rsid, e.target) for e in ciseqtls]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_cis_list = set(all_cis_list)\n",
    "len(set_cis_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"unique_snp_gene_target_list_all_tissues.txt\", 'w') as outstream:\n",
    "    for pair in set_cis_list:\n",
    "        outstream.write(\"{:s}\\t{:s}\\n\".format(pair[0], pair[1]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
